Index: DEPS
===================================================================
--- DEPS	(revision 204265)
+++ DEPS	(working copy)
@@ -86,7 +86,7 @@
   'sctp-refimpl':
     'https://sctp-refimpl.googlecode.com/svn',
   'jsoncpp':
-    'http://jsoncpp.svn.sourceforge.net/svnroot/jsoncpp',
+    'http://svn.code.sf.net/p/jsoncpp/code',
   'webrtc':
     'http://webrtc.googlecode.com/svn',
   'web-page-replay':
