Index: features.gypi
===================================================================
--- features.gypi	(revision 145278)
+++ features.gypi	(working copy)
@@ -193,7 +193,7 @@
           'ENABLE_CALENDAR_PICKER=1',
           'ENABLE_DATALIST_ELEMENT=1',
           'ENABLE_INPUT_SPEECH=1',
-          'ENABLE_INPUT_TYPE_COLOR=1',
+          'ENABLE_INPUT_TYPE_COLOR=0',
           'ENABLE_INPUT_TYPE_WEEK=1',
           'ENABLE_INPUT_MULTIPLE_FIELDS_UI=1',
           'ENABLE_LEGACY_NOTIFICATIONS=1',
