diff --git base/task/thread_pool/thread_pool_impl.cc base/task/thread_pool/thread_pool_impl.cc
index cc1579306df670f66fbdf8336d796e73d203d3b8..16758431c299bfe49dece53c8b0b66b64a11d13a 100644
--- base/task/thread_pool/thread_pool_impl.cc
+++ base/task/thread_pool/thread_pool_impl.cc
@@ -127,6 +127,10 @@ ThreadPoolImpl::ThreadPoolImpl(
         tracked_ref_factory_.GetTrackedRef(), monitor_worker_thread_priorities,
         record_lock_contention);
   }
+
+  // After https://crrev.com/e2e090c363 the ThreadPool is being created earlier
+  // during initialization. Detach now so we can rebind later on the UI thread.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
 }
 
 ThreadPoolImpl::~ThreadPoolImpl() {
