diff --git base/command_line.cc base/command_line.cc
index 69e4303165259..682e8d2bfa6ce 100644
--- base/command_line.cc
+++ base/command_line.cc
@@ -355,11 +355,10 @@ void CommandLine::AppendSwitchNative(StringPiece switch_string,
 #if BUILDFLAG(ENABLE_COMMANDLINE_SEQUENCE_CHECKS)
   sequence_checker_.Check();
 #endif
-#if BUILDFLAG(IS_WIN)
   const std::string switch_key = ToLowerASCII(switch_string);
+#if BUILDFLAG(IS_WIN)
   StringType combined_switch_string(UTF8ToWide(switch_key));
 #elif BUILDFLAG(IS_POSIX) || BUILDFLAG(IS_FUCHSIA)
-  StringPiece switch_key = switch_string;
   StringType combined_switch_string(switch_key);
 #endif
   size_t prefix_length = GetSwitchPrefixLength(combined_switch_string);
