diff --git base/time/time_win.cc base/time/time_win.cc
index eabcc10ad2c31..014dd169992f0 100644
--- base/time/time_win.cc
+++ base/time/time_win.cc
@@ -406,7 +406,7 @@ DWORD (*g_tick_function)(void) = &timeGetTimeWrapper;
 // "rollover" counter.
 union LastTimeAndRolloversState {
   // The state as a single 32-bit opaque value.
-  std::atomic<int32_t> as_opaque_32;
+  std::atomic<int32_t> as_opaque_32{0};
 
   // The state as usable values.
   struct {
