diff --git base/allocator/partition_alloc_features.cc base/allocator/partition_alloc_features.cc
index ef25b0dc2ef44..012e64d349a55 100644
--- base/allocator/partition_alloc_features.cc
+++ base/allocator/partition_alloc_features.cc
@@ -36,7 +36,7 @@ const base::FeatureParam<UnretainedDanglingPtrMode>
     kUnretainedDanglingPtrModeParam = {
         &kPartitionAllocUnretainedDanglingPtr,
         "mode",
-        UnretainedDanglingPtrMode::kCrash,
+        UnretainedDanglingPtrMode::kDumpWithoutCrashing,
         &kUnretainedDanglingPtrModeOption,
 };
 
