diff --git base/debug/dump_without_crashing.cc base/debug/dump_without_crashing.cc
index b9858dc0a7a33..302774bfd773c 100644
--- base/debug/dump_without_crashing.cc
+++ base/debug/dump_without_crashing.cc
@@ -7,6 +7,7 @@
 #include "base/check.h"
 #include "base/metrics/histogram_functions.h"
 #include "base/no_destructor.h"
+#include "base/synchronization/lock.h"
 #include "base/trace_event/base_tracing.h"
 
 namespace {
