diff --git base/task/thread_pool/thread_pool_impl.cc base/task/thread_pool/thread_pool_impl.cc
index d7f9c33417f6403ea1705a4fd7c0beafb0473b11..a6561d3dad060ae5b3b77c4dfb651dc033f7469b 100644
--- base/task/thread_pool/thread_pool_impl.cc
+++ base/task/thread_pool/thread_pool_impl.cc
@@ -108,6 +108,10 @@ ThreadPoolImpl::ThreadPoolImpl(std::string_view histogram_label,
         ThreadGroupType::BACKGROUND, task_tracker_->GetTrackedRef(),
         tracked_ref_factory_.GetTrackedRef());
   }
+
+  // After https://crrev.com/e2e090c363 the ThreadPool is being created earlier
+  // during initialization. Detach now so we can rebind later on the UI thread.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
 }
 
 ThreadPoolImpl::~ThreadPoolImpl() {
