diff --git base/allocator/partition_allocator/partition_alloc_base/time/time_win.cc base/allocator/partition_allocator/partition_alloc_base/time/time_win.cc
index e36aeb64abdb7..7deae93b86f03 100644
--- base/allocator/partition_allocator/partition_alloc_base/time/time_win.cc
+++ base/allocator/partition_allocator/partition_alloc_base/time/time_win.cc
@@ -327,7 +327,7 @@ DWORD (*g_tick_function)(void) = &timeGetTimeWrapper;
 // "rollover" counter.
 union LastTimeAndRolloversState {
   // The state as a single 32-bit opaque value.
-  std::atomic<int32_t> as_opaque_32;
+  std::atomic<int32_t> as_opaque_32{0};
 
   // The state as usable values.
   struct {
