diff --git base/logging.cc base/logging.cc
index 496e68dfcd33a..b1b0af0710f28 100644
--- base/logging.cc
+++ base/logging.cc
@@ -20,6 +20,7 @@
 #include "base/debug/crash_logging.h"
 #include "base/immediate_crash.h"
 #include "base/pending_task.h"
+#include "base/process/process_handle.h"
 #include "base/strings/string_piece.h"
 #include "base/task/common/task_annotator.h"
 #include "base/trace_event/base_tracing.h"
@@ -72,7 +73,6 @@ typedef HANDLE FileHandle;
 #include <stdlib.h>
 #include <string.h>
 #include <sys/stat.h>
-#include "base/process/process_handle.h"
 #define MAX_PATH PATH_MAX
 typedef FILE* FileHandle;
 #endif
