diff --git base/test/BUILD.gn base/test/BUILD.gn
index 99324469362e14c0168cb15d123c9484b3674359..94edc8164ad47991a60c04af02d8a104e5ff169b 100644
--- base/test/BUILD.gn
+++ base/test/BUILD.gn
@@ -219,11 +219,6 @@ static_library("test_support") {
   # that would require special support in gn2bp (AOSP Cronet).
   if (!is_cronet_build) {
     public_deps += [ "//third_party/perfetto:perfetto_test_support" ]
-    public_deps += [ ":test_trace_processor" ]
-    sources += [
-      "test_trace_processor.cc",
-      "test_trace_processor.h",
-    ]
     deps += [
       ":amalgamated_perfetto_sql_stdlib",
       ":gen_cc_chrome_track_event_descriptor",
@@ -618,7 +613,7 @@ action("amalgamated_perfetto_sql_stdlib") {
 # processor depends on dev_sqlite. The two share the same symbols but have
 # different implementations, so we need to hide dev_sqlite in this shared
 # library even in non-component builds to prevent duplicate symbols.
-_target_type = "shared_library"
+_target_type = "component"
 if (is_ios) {
   _target_type = "ios_framework_bundle"
 }
@@ -627,6 +622,8 @@ target(_target_type, "test_trace_processor") {
   defines = [ "TEST_TRACE_PROCESSOR_IMPL" ]
   testonly = true
   sources = [
+    "test_trace_processor.cc",
+    "test_trace_processor.h",
     "test_trace_processor_export.h",
     "test_trace_processor_impl.cc",
     "test_trace_processor_impl.h",
@@ -644,33 +641,6 @@ target(_target_type, "test_trace_processor") {
     output_name = "TestTraceProcessor"
     bundle_deps_filter = [ "//third_party/icu:icudata" ]
   }
-
-  # Set rpath on dependent tests so that they can find the shared library
-  # in a non-component build.
-  if (!is_component_build) {
-    all_dependent_configs = []
-    if (is_linux || is_chromeos) {
-      all_dependent_configs +=
-          [ "//build/config/gcc:rpath_for_built_shared_libraries" ]
-    }
-    if (is_mac) {
-      # We need the lib to be in $root_out_dir/Libraries so the library is where
-      # the linker expects it, since we set the install_name flag. We need to
-      # set this flag so that the library can be found when it is bundled in the
-      # Content Shell Framework (see test_trace_processor_bundle_data).
-      output_dir = "$root_out_dir/Libraries"
-      ldflags = [ "-Wl,-install_name,@loader_path/Libraries/libtest_trace_processor.dylib" ]
-    }
-  }
-}
-
-if (!is_component_build && is_mac) {
-  bundle_data("test_trace_processor_bundle_data") {
-    testonly = true
-    sources = [ "$root_out_dir/Libraries/libtest_trace_processor.dylib" ]
-    outputs = [ "{{bundle_contents_dir}}/Libraries/{{source_file_part}}" ]
-    public_deps = [ ":test_trace_processor" ]
-  }
 }
 
 if (is_android) {
diff --git base/test/test_trace_processor_export.h base/test/test_trace_processor_export.h
index 94b5a2ab1e7518f0e4db9f0b58d392459441659f..6dacffae228efc88e84b8ba14c400566ad8c3ca5 100644
--- base/test/test_trace_processor_export.h
+++ base/test/test_trace_processor_export.h
@@ -5,6 +5,7 @@
 #ifndef BASE_TEST_TEST_TRACE_PROCESSOR_EXPORT_H_
 #define BASE_TEST_TEST_TRACE_PROCESSOR_EXPORT_H_
 
+#if defined(COMPONENT_BUILD)
 #if defined(WIN32)
 
 #if defined(TEST_TRACE_PROCESSOR_IMPL)
@@ -17,4 +18,8 @@
 #define TEST_TRACE_PROCESSOR_EXPORT __attribute__((visibility("default")))
 #endif  // defined(WIN32)
 
+#else  // defined(COMPONENT_BUILD)
+#define TEST_TRACE_PROCESSOR_EXPORT
+#endif
+
 #endif  // BASE_TEST_TEST_TRACE_PROCESSOR_EXPORT_H_
diff --git content/shell/BUILD.gn content/shell/BUILD.gn
index a15a8d75913adffc2fed5deda418b399d3989535..0d58ba7950962fdaa310ce8e6e37c3cc6043f280 100644
--- content/shell/BUILD.gn
+++ content/shell/BUILD.gn
@@ -925,7 +925,6 @@ if (is_mac) {
       # Specify a sensible install_name for static builds. The library is
       # dlopen()ed so this is not used to resolve the module.
       ldflags = [ "-Wl,-install_name,@executable_path/../Frameworks/$output_name.framework/$output_name" ]
-      deps += [ "//base/test:test_trace_processor_bundle_data" ]
     } else {
       # Both the main :content_shell and :content_shell_helper_app executables
       # need to link the framework. Because they are at different directory
diff --git ui/views/examples/BUILD.gn ui/views/examples/BUILD.gn
index 500dd0d66f6c7d4f1a4b4cb479c3adf9f4749a27..6272c444d27209ba76174216fae92f2a0e9f7e1d 100644
--- ui/views/examples/BUILD.gn
+++ ui/views/examples/BUILD.gn
@@ -356,7 +356,6 @@ if (is_mac) {
       # Specify a sensible install_name for static builds. The library is
       # dlopen()ed so this is not used to resolve the module.
       ldflags = [ "-Wl,-install_name,@executable_path/../Frameworks/$output_name.framework/$output_name.framework" ]
-      deps += [ "//base/test:test_trace_processor_bundle_data" ]
     }
 
     info_plist_target = ":views_examples_framework_plist"
