diff --git base/command_line.cc base/command_line.cc
index 735a1e8bcbe6f..d284d2c86b11b 100644
--- base/command_line.cc
+++ base/command_line.cc
@@ -338,11 +338,10 @@ void CommandLine::AppendSwitchPath(StringPiece switch_string,
 
 void CommandLine::AppendSwitchNative(StringPiece switch_string,
                                      CommandLine::StringPieceType value) {
-#if defined(OS_WIN)
   const std::string switch_key = ToLowerASCII(switch_string);
+#if defined(OS_WIN)
   StringType combined_switch_string(UTF8ToWide(switch_key));
 #elif defined(OS_POSIX) || defined(OS_FUCHSIA)
-  StringPiece switch_key = switch_string;
   StringType combined_switch_string(switch_key);
 #endif
   size_t prefix_length = GetSwitchPrefixLength(combined_switch_string);
