diff --git base/task/thread_pool/thread_pool_impl.cc base/task/thread_pool/thread_pool_impl.cc
index 4a934c05c7bb03646b8636c7ad12ebd76fe37791..2176dba3bc01a99a9dd7eff87d2ff6b60ffece7f 100644
--- base/task/thread_pool/thread_pool_impl.cc
+++ base/task/thread_pool/thread_pool_impl.cc
@@ -122,6 +122,10 @@ ThreadPoolImpl::ThreadPoolImpl(std::string_view histogram_label,
         ThreadGroupType::BACKGROUND, task_tracker_->GetTrackedRef(),
         tracked_ref_factory_.GetTrackedRef(), monitor_worker_thread_priorities);
   }
+
+  // After https://crrev.com/e2e090c363 the ThreadPool is being created earlier
+  // during initialization. Detach now so we can rebind later on the UI thread.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
 }
 
 ThreadPoolImpl::~ThreadPoolImpl() {
