diff --git base/values.cc base/values.cc
index 7d042cdcd459..09eb01af5745 100644
--- base/values.cc
+++ base/values.cc
@@ -23,20 +23,6 @@
 
 namespace base {
 
-// base::Value must be standard layout to guarantee that writing to
-// |bool_type_| then reading |type_| is defined behaviour. See:
-//
-// [class.union]:
-//   If a standard-layout union contains several standard-layout structs that
-//   share a common initial sequence (9.2), and if an object of this
-//   standard-layout union type contains one of the standard-layout structs,
-//   it is permitted to inspect the common initial sequence of any of
-//   standard-layout struct members;
-//
-static_assert(std::is_standard_layout<Value>::value,
-              "base::Value should be a standard-layout C++ class in order "
-              "to avoid undefined behaviour in its implementation!");
-
 static_assert(sizeof(Value::DoubleStorage) == sizeof(double),
               "The double and DoubleStorage types should have the same size");
 
