diff --git base/task/thread_pool/thread_pool_impl.cc base/task/thread_pool/thread_pool_impl.cc
index 92300920227dd8aa9d1ab3aa25d0b435f70ca024..df5f3d1fc7e1f5e4ef98aa152322de3eed6b9afe 100644
--- base/task/thread_pool/thread_pool_impl.cc
+++ base/task/thread_pool/thread_pool_impl.cc
@@ -110,6 +110,10 @@ ThreadPoolImpl::ThreadPoolImpl(std::string_view histogram_label,
         ThreadGroupType::BACKGROUND, task_tracker_->GetTrackedRef(),
         tracked_ref_factory_.GetTrackedRef(), monitor_worker_thread_priorities);
   }
+
+  // After https://crrev.com/e2e090c363 the ThreadPool is being created earlier
+  // during initialization. Detach now so we can rebind later on the UI thread.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
 }
 
 ThreadPoolImpl::~ThreadPoolImpl() {
