diff --git base/metrics/statistics_recorder.cc base/metrics/statistics_recorder.cc
index 1f808cac46bc0..8ca2df9cfc5b9 100644
--- base/metrics/statistics_recorder.cc
+++ base/metrics/statistics_recorder.cc
@@ -32,7 +32,7 @@ bool HistogramNameLesser(const base::HistogramBase* a,
 }  // namespace
 
 // static
-LazyInstance<Lock>::Leaky StatisticsRecorder::lock_;
+LazyInstance<Lock>::Leaky StatisticsRecorder::lock_ = LAZY_INSTANCE_INITIALIZER;
 
 // static
 StatisticsRecorder* StatisticsRecorder::top_ = nullptr;
