diff --git base/threading/hang_watcher.cc base/threading/hang_watcher.cc
index 5eff8563b1309..f62a711b57ca5 100644
--- base/threading/hang_watcher.cc
+++ base/threading/hang_watcher.cc
@@ -719,6 +719,7 @@ void HangWatcher::WatchStateSnapShot::Init(
         any_hung_thread_has_dumping_enabled = true;
       }
 
+#if BUILDFLAG(ENABLE_BASE_TRACING)
       // Emit trace events for monitored threads.
       if (ThreadTypeLoggingLevelGreaterOrEqual(watch_state.get()->thread_type(),
                                                LoggingLevel::kUmaOnly)) {
@@ -728,6 +729,7 @@ void HangWatcher::WatchStateSnapShot::Init(
         TRACE_EVENT_BEGIN("base", "HangWatcher::ThreadHung", track, deadline);
         TRACE_EVENT_END("base", track, now);
       }
+#endif
 
       // Attempt to mark the thread as needing to stay within its current
       // WatchHangsInScope until capture is complete.
