diff --git third_party/blink/public/web/web_element.h third_party/blink/public/web/web_element.h
index 015e6dce8b7871e780fdf2d282033474099c7280..b513194c0545017c5a690d039691e6ebaa608b3e 100644
--- third_party/blink/public/web/web_element.h
+++ third_party/blink/public/web/web_element.h
@@ -86,6 +86,9 @@ class BLINK_EXPORT WebElement : public WebNode {
   WebString TextContent() const;
   WebString TextContentAbridged(unsigned int max_length) const;
   WebString InnerHTML() const;
+  WebString AttributeLocalName(unsigned index) const;
+  WebString AttributeValue(unsigned index) const;
+  unsigned AttributeCount() const;
 
   void Focus();
 
diff --git third_party/blink/renderer/core/exported/web_element.cc third_party/blink/renderer/core/exported/web_element.cc
index 7a172a04538bc81ff707c312d6f99e10335c783d..94c57c0be1ebf650553e7ccfd8838bfc19b03bdf 100644
--- third_party/blink/renderer/core/exported/web_element.cc
+++ third_party/blink/renderer/core/exported/web_element.cc
@@ -187,6 +187,24 @@ void WebElement::SetAttribute(const WebString& attr_name,
                                   IGNORE_EXCEPTION_FOR_TESTING);
 }
 
+unsigned WebElement::AttributeCount() const {
+  if (!ConstUnwrap<Element>()->hasAttributes())
+    return 0;
+  return ConstUnwrap<Element>()->Attributes().size();
+}
+
+WebString WebElement::AttributeLocalName(unsigned index) const {
+  if (index >= AttributeCount())
+    return WebString();
+  return ConstUnwrap<Element>()->Attributes().at(index).LocalName();
+}
+
+WebString WebElement::AttributeValue(unsigned index) const {
+  if (index >= AttributeCount())
+    return WebString();
+  return ConstUnwrap<Element>()->Attributes().at(index).Value();
+}
+
 WebString WebElement::TextContent() const {
   return ConstUnwrap<Element>()->textContent();
 }
