diff --git third_party/blink/common/widget/constants.cc third_party/blink/common/widget/constants.cc
index c9b9b1e2a11fa1d11e115651be07851cd371f00a..650669ac0eb7c6d3be432dcd7d6860272a8d6e0d 100644
--- third_party/blink/common/widget/constants.cc
+++ third_party/blink/common/widget/constants.cc
@@ -6,7 +6,9 @@
 
 namespace blink {
 
-const int kMinimumWindowSize = 100;
+// Allow JavaScript resize down to kMinWindowSize which applies for Document
+// Picture-in-Picture windows.
+const int kMinimumWindowSize = 52;
 
 // TODO(b/307160156, b/307182741); Investigate where else is the window size
 // limited to be able to drop this even more until 9 instead 29.
