diff --git content/browser/renderer_host/render_widget_host_view_child_frame.cc content/browser/renderer_host/render_widget_host_view_child_frame.cc
index ef607a087cdf..1320a4d274f6 100644
--- content/browser/renderer_host/render_widget_host_view_child_frame.cc
+++ content/browser/renderer_host/render_widget_host_view_child_frame.cc
@@ -662,6 +662,7 @@ void RenderWidgetHostViewChildFrame::SubmitCompositorFrame(
                "RenderWidgetHostViewChildFrame::OnSwapCompositorFrame");
   support_->SubmitCompositorFrame(local_surface_id, std::move(frame),
                                   std::move(hit_test_region_list));
+  ProcessFrameSwappedCallbacks();
 }
 
 void RenderWidgetHostViewChildFrame::OnDidNotProduceFrame(
@@ -945,7 +946,6 @@ void RenderWidgetHostViewChildFrame::OnFirstSurfaceActivation(
   last_activated_surface_info_ = surface_info;
   has_frame_ = true;
   FirstSurfaceActivation(surface_info);
-  ProcessFrameSwappedCallbacks();
 }
 
 void RenderWidgetHostViewChildFrame::OnFrameTokenChanged(uint32_t frame_token) {
