diff --git content/browser/frame_host/render_widget_host_view_guest.cc content/browser/frame_host/render_widget_host_view_guest.cc
index fc720a7..1c30365 100644
--- content/browser/frame_host/render_widget_host_view_guest.cc
+++ content/browser/frame_host/render_widget_host_view_guest.cc
@@ -247,13 +247,14 @@ void RenderWidgetHostViewGuest::Destroy() {
 }
 
 gfx::Size RenderWidgetHostViewGuest::GetPhysicalBackingSize() const {
+  RenderWidgetHostViewBase* rwhv = GetOwnerRenderWidgetHostView();
+  if (rwhv)
+    return rwhv->GetPhysicalBackingSize();
   // We obtain the reference to native view from the owner RenderWidgetHostView.
   // If the guest is embedded inside a cross-process frame, it is possible to
   // reach here after the frame is detached in which case there will be no owner
   // view.
-  if (!GetOwnerRenderWidgetHostView())
-    return gfx::Size();
-  return RenderWidgetHostViewBase::GetPhysicalBackingSize();
+  return gfx::Size();
 }
 
 base::string16 RenderWidgetHostViewGuest::GetSelectedText() {
