diff --git content/browser/scheduler/browser_task_executor.cc content/browser/scheduler/browser_task_executor.cc
index 332ef34e24dc..ee866ca5f3b6 100644
--- content/browser/scheduler/browser_task_executor.cc
+++ content/browser/scheduler/browser_task_executor.cc
@@ -244,7 +244,7 @@ void BrowserTaskExecutor::PostFeatureListSetup() {
 
 // static
 void BrowserTaskExecutor::Shutdown() {
-  if (!g_browser_task_executor)
+  if (!g_browser_task_executor || !g_browser_task_executor->ui_thread_executor_)
     return;
 
   DCHECK(g_browser_task_executor->ui_thread_executor_);
