diff --git content/browser/scheduler/browser_task_executor.cc content/browser/scheduler/browser_task_executor.cc
index e06e26f80a5e0..2da99c8f97990 100644
--- content/browser/scheduler/browser_task_executor.cc
+++ content/browser/scheduler/browser_task_executor.cc
@@ -285,7 +285,7 @@ BrowserTaskExecutor::OnUserInputStart() {
 
 // static
 void BrowserTaskExecutor::Shutdown() {
-  if (!g_browser_task_executor)
+  if (!g_browser_task_executor || !g_browser_task_executor->ui_thread_executor_)
     return;
 
   DCHECK(Get()->ui_thread_executor_);
