diff --git content/public/browser/web_ui_controller_factory.h content/public/browser/web_ui_controller_factory.h
index 9d9c17ffd6474..4eb79c65369af 100644
--- content/public/browser/web_ui_controller_factory.h
+++ content/public/browser/web_ui_controller_factory.h
@@ -47,9 +47,6 @@ class CONTENT_EXPORT WebUIControllerFactory {
   virtual bool UseWebUIForURL(BrowserContext* browser_context,
                               const GURL& url) = 0;
 
- private:
-  friend class ScopedWebUIControllerFactoryRegistration;
-
   static void UnregisterFactoryForTesting(WebUIControllerFactory* factory);
 };
 
diff --git content/public/browser/webui_config_map.h content/public/browser/webui_config_map.h
index e7b6a9b841abb..1c98be86c0db1 100644
--- content/public/browser/webui_config_map.h
+++ content/public/browser/webui_config_map.h
@@ -68,6 +68,10 @@ class CONTENT_EXPORT WebUIConfigMap {
   std::vector<WebUIConfigInfo> GetWebUIConfigList(
       BrowserContext* browser_context);
 
+  WebUIControllerFactory* controller_factory() const {
+    return webui_controller_factory_.get();
+  }
+
  private:
   void AddWebUIConfigImpl(std::unique_ptr<WebUIConfig> config);
 
