diff --git content/browser/scheduler/browser_task_executor.cc content/browser/scheduler/browser_task_executor.cc
index eb9f505243bf..191368199d44 100644
--- content/browser/scheduler/browser_task_executor.cc
+++ content/browser/scheduler/browser_task_executor.cc
@@ -245,7 +245,7 @@ void BrowserTaskExecutor::PostFeatureListSetup() {
 
 // static
 void BrowserTaskExecutor::Shutdown() {
-  if (!g_browser_task_executor)
+  if (!g_browser_task_executor || !g_browser_task_executor->ui_thread_executor_)
     return;
 
   DCHECK(g_browser_task_executor->ui_thread_executor_);
