diff --git common.gypi common.gypi
index f97bf66..116c7bc 100644
--- common.gypi
+++ common.gypi
@@ -9,6 +9,9 @@
   # Variables expected to be overriden on the GYP command line (-D) or by
   # ~/.gyp/include.gypi.
   'variables': {
+    # Directory for CEF source files. This will be set by cef.gypi.
+    'cef_directory%' : '',
+
     # Putting a variables dict inside another variables dict looks kind of
     # weird.  This is done so that 'host_arch', 'chromeos', etc are defined as
     # variables within the outer variables dict here.  This is necessary
@@ -5486,6 +5489,9 @@
           4510, # Default constructor could not be generated
           4512, # Assignment operator could not be generated
           4610, # Object can never be instantiated
+          # Disable Warning 4702 ("Unreachable code")
+          # https://code.google.com/p/chromium/issues/detail?id=346399#c55
+          4702,
         ],
         'msvs_settings': {
           'VCCLCompilerTool': {
diff --git mac/strip_save_dsym mac/strip_save_dsym
index c9cf226..0dedbe3 100755
--- mac/strip_save_dsym
+++ mac/strip_save_dsym
@@ -48,7 +48,7 @@ def macho_archs(macho):
                  "bundle"]
   macho_types_re = "Mach-O (?:64-bit )?(?:" + "|".join(macho_types) + ")"
 
-  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "--", macho],
+  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "-L", "--", macho],
                               stdout=subprocess.PIPE)
 
   archs = []
diff --git toolchain/win/setup_toolchain.py toolchain/win/setup_toolchain.py
index 5e292ab..70ad093 100644
--- toolchain/win/setup_toolchain.py
+++ toolchain/win/setup_toolchain.py
@@ -25,6 +25,8 @@ def ExtractImportantEnvironment():
   envvars_to_save = (
       'goma_.*',  # TODO(scottmg): This is ugly, but needed for goma.
       'include',  # Needed by midl compiler.
+      'lib',
+      'libpath',
       'path',
       'pathext',
       'systemroot',
