diff --git common.gypi common.gypi
index cff7376..d16e12d 100644
--- common.gypi
+++ common.gypi
@@ -9,6 +9,9 @@
   # Variables expected to be overriden on the GYP command line (-D) or by
   # ~/.gyp/include.gypi.
   'variables': {
+    # Directory for CEF source files. This will be set by cef.gypi.
+    'cef_directory%' : '',
+
     # Putting a variables dict inside another variables dict looks kind of
     # weird.  This is done so that 'host_arch', 'chromeos', etc are defined as
     # variables within the outer variables dict here.  This is necessary
diff --git mac/strip_save_dsym mac/strip_save_dsym
index c9cf226..0dedbe3 100755
--- mac/strip_save_dsym
+++ mac/strip_save_dsym
@@ -48,7 +48,7 @@ def macho_archs(macho):
                  "bundle"]
   macho_types_re = "Mach-O (?:64-bit )?(?:" + "|".join(macho_types) + ")"
 
-  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "--", macho],
+  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "-L", "--", macho],
                               stdout=subprocess.PIPE)
 
   archs = []
diff --git vs_toolchain.py vs_toolchain.py
index cea0d6d..158d0a8 100755
--- vs_toolchain.py
+++ vs_toolchain.py
@@ -32,8 +32,8 @@ def SetEnvironmentAndGetRuntimeDllDirs():
       bool(int(os.environ.get('DEPOT_TOOLS_WIN_TOOLCHAIN', '1')))
   # When running on a non-Windows host, only do this if the SDK has explicitly
   # been downloaded before (in which case json_data_file will exist).
-  if ((sys.platform in ('win32', 'cygwin') and depot_tools_win_toolchain) or
-      os.path.exists(json_data_file)):
+  if ((sys.platform in ('win32', 'cygwin') or os.path.exists(json_data_file)) and
+      depot_tools_win_toolchain):
     if not os.path.exists(json_data_file):
       Update()
     with open(json_data_file, 'r') as tempf:
