diff --git build/config/compiler/BUILD.gn build/config/compiler/BUILD.gn
index 23a8da453c3d2..66ad9b2ab0ed6 100644
--- build/config/compiler/BUILD.gn
+++ build/config/compiler/BUILD.gn
@@ -2015,8 +2015,6 @@ config("thin_archive") {
   # confuses lldb.
   if ((is_posix && !is_nacl && !is_apple) || is_fuchsia) {
     arflags = [ "-T" ]
-  } else if (is_win && use_lld) {
-    arflags = [ "/llvmlibthin" ]
   }
 }
 
