diff --git build/config/compiler/BUILD.gn build/config/compiler/BUILD.gn
index b37e8d1f2ab6..f42ea327a9e8 100644
--- build/config/compiler/BUILD.gn
+++ build/config/compiler/BUILD.gn
@@ -157,7 +157,7 @@ declare_args() {
             !use_clang_coverage && !(is_android && use_order_profiling) &&
             (use_lld ||
              (use_gold &&
-              ((!is_android && linux_use_bundled_binutils) || is_chromeos ||
+              ((!is_android && linux_use_bundled_binutils && current_cpu != "x86") || is_chromeos ||
                !(current_cpu == "x86" || current_cpu == "x64"))))
 }
 
@@ -1696,8 +1696,6 @@ config("thin_archive") {
   # archive names to 16 characters, which is not what we want).
   if ((is_posix && !is_nacl && !is_mac && !is_ios) || is_fuchsia) {
     arflags = [ "-T" ]
-  } else if (is_win && use_lld) {
-    arflags = [ "/llvmlibthin" ]
   }
 }
 
