diff --git build/config/compiler/BUILD.gn build/config/compiler/BUILD.gn
index 72ea590d6968..0c34d20addfa 100644
--- build/config/compiler/BUILD.gn
+++ build/config/compiler/BUILD.gn
@@ -1834,8 +1834,6 @@ config("thin_archive") {
   # archives.
   if ((is_posix && !is_nacl && (!is_apple || use_lld)) || is_fuchsia) {
     arflags = [ "-T" ]
-  } else if (is_win && use_lld) {
-    arflags = [ "/llvmlibthin" ]
   }
 }
 
