diff --git build/config/compiler/BUILD.gn build/config/compiler/BUILD.gn
index 3c7b4d82d0b46..7f890ab3aa546 100644
--- build/config/compiler/BUILD.gn
+++ build/config/compiler/BUILD.gn
@@ -2195,8 +2195,6 @@ config("thin_archive") {
   # confuses lldb.
   if ((is_posix && !is_nacl && !is_apple) || is_fuchsia) {
     arflags = [ "-T" ]
-  } else if (is_win && use_lld) {
-    arflags = [ "/llvmlibthin" ]
   }
 }
 
