diff --git common.gypi common.gypi
index f3f28c7..2d2f0f7 100644
--- common.gypi
+++ common.gypi
@@ -9,6 +9,9 @@
   # Variables expected to be overriden on the GYP command line (-D) or by
   # ~/.gyp/include.gypi.
   'variables': {
+    # Directory for CEF source files. This will be set by cef.gypi.
+    'cef_directory%' : '',
+
     # Putting a variables dict inside another variables dict looks kind of
     # weird.  This is done so that 'host_arch', 'chromeos', etc are defined as
     # variables within the outer variables dict here.  This is necessary
@@ -1567,7 +1570,7 @@
       ['OS=="win" and "<!pymod_do_main(dir_exists <(windows_sdk_default_path))"=="True"', {
         'windows_sdk_path%': '<(windows_sdk_default_path)',
       }, {
-        'windows_sdk_path%': 'C:/Program Files (x86)/Windows Kits/8.0',
+        'windows_sdk_path%': 'C:/Program Files (x86)/Windows Kits/8.1',
       }],
       ['OS=="win" and "<!pymod_do_main(dir_exists <(directx_sdk_default_path))"=="True"', {
         'directx_sdk_path%': '<(directx_sdk_default_path)',
@@ -5394,6 +5397,10 @@
           4510, # Default constructor could not be generated
           4512, # Assignment operator could not be generated
           4610, # Object can never be instantiated
+          4996, # 'X': was declared deprecated (for GetVersionEx)
+          # Disable Warning 4702 ("Unreachable code")
+          # https://code.google.com/p/chromium/issues/detail?id=346399#c55
+          4702,
         ],
         'msvs_settings': {
           'VCCLCompilerTool': {
diff --git gyp_chromium gyp_chromium
index 326919c0..67d0a00 100755
--- gyp_chromium
+++ gyp_chromium
@@ -263,12 +263,6 @@ if __name__ == '__main__':
   if sys.platform not in ('darwin',):
     args.append('--no-circular-check')
 
-  # We explicitly don't support the make gyp generator (crbug.com/348686). Be
-  # nice and fail here, rather than choking in gyp.
-  if re.search(r'(^|,|\s)make($|,|\s)', os.environ.get('GYP_GENERATORS', '')):
-    print 'Error: make gyp generator not supported (check GYP_GENERATORS).'
-    sys.exit(1)
-
   # If CHROMIUM_GYP_SYNTAX_CHECK is set to 1, it will invoke gyp with --check
   # to enfore syntax checking.
   syntax_check = os.environ.get('CHROMIUM_GYP_SYNTAX_CHECK')
diff --git mac/strip_save_dsym mac/strip_save_dsym
index c9cf226..0dedbe3 100755
--- mac/strip_save_dsym
+++ mac/strip_save_dsym
@@ -48,7 +48,7 @@ def macho_archs(macho):
                  "bundle"]
   macho_types_re = "Mach-O (?:64-bit )?(?:" + "|".join(macho_types) + ")"
 
-  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "--", macho],
+  file_cmd = subprocess.Popen(["/usr/bin/file", "-b", "-L", "--", macho],
                               stdout=subprocess.PIPE)
 
   archs = []
diff --git toolchain/win/setup_toolchain.py toolchain/win/setup_toolchain.py
index 5e292ab..9e0bb8d 100644
--- toolchain/win/setup_toolchain.py
+++ toolchain/win/setup_toolchain.py
@@ -22,9 +22,13 @@ and the files will be written to the current directory.
 def ExtractImportantEnvironment():
   """Extracts environment variables required for the toolchain from the
   current environment."""
+  # This list should be kept synchronized with _ExtractImportantEnvironment from
+  # tools/gyp/pylib/gyp/msvs_emulation.py.
   envvars_to_save = (
       'goma_.*',  # TODO(scottmg): This is ugly, but needed for goma.
       'include',  # Needed by midl compiler.
+      'lib',
+      'libpath',
       'path',
       'pathext',
       'systemroot',
@@ -90,7 +94,7 @@ important_env_vars = ExtractImportantEnvironment()
 path = important_env_vars["PATH"].split(";")
 
 # Add 32-bit compiler path to the beginning and write the block.
-path32 = [os.path.join(vs_path, "VC\\BIN")] + \
+path32 = [os.path.join(vs_path, "VC\\BIN\\amd64_x86")] + \
          [os.path.join(win_sdk_path, "bin\\x86")] + \
          path
 important_env_vars["PATH"] = ";".join(path32)
