diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index 8b2f8a9424a95..a87abd0508f51 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -29,6 +29,7 @@ static_library("safe_browsing") {
     "//components/browser_sync",
     "//components/enterprise:enterprise",
     "//components/enterprise/common:strings",
+    "//components/gcm_driver:gcm_buildflags",
     "//components/keyed_service/content",
     "//components/language/core/browser",
     "//components/no_state_prefetch/browser",
