diff --git chrome/browser/ui/extensions/extension_dialog_utils_views.cc chrome/browser/ui/extensions/extension_dialog_utils_views.cc
index 12c3bc947ddb641a538f18bb742a69ca00ef6ec4..e28ef5223c966f24d8745dac756ea646f685553e 100644
--- chrome/browser/ui/extensions/extension_dialog_utils_views.cc
+++ chrome/browser/ui/extensions/extension_dialog_utils_views.cc
@@ -88,10 +88,10 @@ void ShowDialog(BrowserWindowInterface* browser,
                 std::unique_ptr<ui::DialogModel> dialog_model) {
   ToolbarButtonProvider* toolbar_button_provider =
       BrowserView::GetBrowserViewForBrowser(browser)->toolbar_button_provider();
-  CHECK(toolbar_button_provider);
 
-  views::BubbleAnchor anchor =
-      toolbar_button_provider->GetDefaultExtensionDialogAnchor();
+  views::BubbleAnchor anchor = toolbar_button_provider
+      ? toolbar_button_provider->GetDefaultExtensionDialogAnchor()
+      : views::BubbleAnchor();
   auto bubble = std::make_unique<views::BubbleDialogModelHost>(
       std::move(dialog_model), std::move(anchor),
       views::BubbleBorder::TOP_RIGHT);
diff --git chrome/browser/ui/views/certificate_selector.cc chrome/browser/ui/views/certificate_selector.cc
index 98e63b81b595e73a4b030910efbde85a685dbdad..e9f5306f711530bbe6b1db2f9cabfd31d4d81646 100644
--- chrome/browser/ui/views/certificate_selector.cc
+++ chrome/browser/ui/views/certificate_selector.cc
@@ -337,8 +337,13 @@ void CertificateSelector::ViewCertButtonPressed() {
   if (!cert) {
     return;
   }
+  auto native_window = web_contents_->GetTopLevelNativeWindow();
+  if (!native_window) {
+    // Not supported with CEF windowless rendering.
+    return;
+  }
   ShowCertificateViewerForClientAuth(web_contents_,
-                                     web_contents_->GetTopLevelNativeWindow(),
+                                     native_window,
                                      cert->certificate());
 }
 
