diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 6cb9ee1f0764..8109a134907e 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -12,6 +12,7 @@ import("//build/config/crypto.gni")
 import("//build/config/features.gni")
 import("//build/config/linux/gtk/gtk.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -1870,6 +1871,7 @@ static_library("browser") {
     "//build:branding_buildflags",
     "//build:lacros_buildflags",
     "//cc",
+    "//cef/libcef/features",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -2221,6 +2223,10 @@ static_library("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     sources += [
       "after_startup_task_utils_android.cc",
@@ -3422,8 +3428,6 @@ static_library("browser") {
       "nearby_sharing/outgoing_share_target_info.h",
       "nearby_sharing/paired_key_verification_runner.cc",
       "nearby_sharing/paired_key_verification_runner.h",
-      "nearby_sharing/share_target.cc",
-      "nearby_sharing/share_target.h",
       "nearby_sharing/share_target_discovered_callback.h",
       "nearby_sharing/share_target_info.cc",
       "nearby_sharing/share_target_info.h",
@@ -3617,13 +3621,6 @@ static_library("browser") {
       "serial/serial_chooser_context_factory.cc",
       "serial/serial_chooser_context_factory.h",
       "serial/serial_chooser_histograms.h",
-      "sharesheet/sharesheet_controller.h",
-      "sharesheet/sharesheet_service.cc",
-      "sharesheet/sharesheet_service.h",
-      "sharesheet/sharesheet_service_delegate.cc",
-      "sharesheet/sharesheet_service_delegate.h",
-      "sharesheet/sharesheet_service_factory.cc",
-      "sharesheet/sharesheet_service_factory.h",
       "sharing/click_to_call/click_to_call_context_menu_observer.cc",
       "sharing/click_to_call/click_to_call_context_menu_observer.h",
       "sharing/click_to_call/click_to_call_metrics.cc",
diff --git chrome/browser/ui/views/toolbar/toolbar_view.cc chrome/browser/ui/views/toolbar/toolbar_view.cc
index 7ccd06a9b0fe..183da2c7238d 100644
--- chrome/browser/ui/views/toolbar/toolbar_view.cc
+++ chrome/browser/ui/views/toolbar/toolbar_view.cc
@@ -232,8 +232,8 @@ void ToolbarView::Init() {
     media_button = std::make_unique<MediaToolbarButtonView>(browser_);
   }
 
-  std::unique_ptr<SharesheetButton> sharesheet_button;
 #if defined(OS_CHROMEOS)
+  std::unique_ptr<SharesheetButton> sharesheet_button;
   if (base::FeatureList::IsEnabled(features::kSharesheet)) {
     sharesheet_button = std::make_unique<SharesheetButton>(browser_);
   }
@@ -276,8 +276,10 @@ void ToolbarView::Init() {
   if (media_button)
     media_button_ = AddChildView(std::move(media_button));
 
+#if defined(OS_CHROMEOS)
   if (sharesheet_button)
     sharesheet_button_ = AddChildView(std::move(sharesheet_button));
+#endif
 
   if (toolbar_account_icon_container) {
     toolbar_account_icon_container_ =
@@ -966,3 +968,4 @@ void ToolbarView::OnTouchUiChanged() {
     PreferredSizeChanged();
   }
 }
+
diff --git chrome/browser/ui/views/toolbar/toolbar_view.h chrome/browser/ui/views/toolbar/toolbar_view.h
index f4af66fa98b3..3c4d171c091e 100644
--- chrome/browser/ui/views/toolbar/toolbar_view.h
+++ chrome/browser/ui/views/toolbar/toolbar_view.h
@@ -273,7 +273,9 @@ class ToolbarView : public views::AccessiblePaneView,
   media_router::CastToolbarButton* cast_ = nullptr;
   ToolbarAccountIconContainerView* toolbar_account_icon_container_ = nullptr;
   AvatarToolbarButton* avatar_ = nullptr;
+#if defined(OS_CHROMEOS)
   SharesheetButton* sharesheet_button_ = nullptr;
+#endif
   MediaToolbarButtonView* media_button_ = nullptr;
   BrowserAppMenuButton* app_menu_button_ = nullptr;
 
@@ -302,3 +304,4 @@ class ToolbarView : public views::AccessiblePaneView,
 };
 
 #endif  // CHROME_BROWSER_UI_VIEWS_TOOLBAR_TOOLBAR_VIEW_H_
+
