diff --git chrome/browser/ui/autofill/address_bubbles_controller.cc chrome/browser/ui/autofill/address_bubbles_controller.cc
index cfeffcb14b60abf722edf0fdd85886d2e63d5b86..56684d6e5b6b91be4ae9eb7428503b1ea997caee 100644
--- chrome/browser/ui/autofill/address_bubbles_controller.cc
+++ chrome/browser/ui/autofill/address_bubbles_controller.cc
@@ -56,11 +56,15 @@ AutofillBubbleBase* ShowSaveBubble(
     content::WebContents* web_contents,
     bool shown_by_user_gesture,
     base::WeakPtr<AddressBubbleControllerDelegate> delegate) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   auto controller = std::make_unique<SaveAddressBubbleController>(
       delegate, web_contents, profile, save_address_bubble_type);
 
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
-      ->GetAutofillBubbleHandler()
+  return browser_window->GetAutofillBubbleHandler()
       ->ShowSaveAddressProfileBubble(web_contents, std::move(controller),
                                      shown_by_user_gesture);
 }
@@ -71,10 +75,14 @@ AutofillBubbleBase* ShowUpdateBubble(
     content::WebContents* web_contents,
     bool shown_by_user_gesture,
     base::WeakPtr<AddressBubbleControllerDelegate> delegate) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   auto update_controller = std::make_unique<UpdateAddressBubbleController>(
       delegate, web_contents, profile, original_profile);
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
-      ->GetAutofillBubbleHandler()
+  return browser_window->GetAutofillBubbleHandler()
       ->ShowUpdateAddressProfileBubble(
           web_contents, std::move(update_controller), shown_by_user_gesture);
 }
@@ -82,11 +90,15 @@ AutofillBubbleBase* ShowUpdateBubble(
 #if BUILDFLAG(ENABLE_DICE_SUPPORT)
 AutofillBubbleBase* ShowSignInPromo(content::WebContents* web_contents,
                                     const AutofillProfile& autofill_profile) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   // TODO(crbug.com/381390420): Expose the `AutofillBubbleHandler` in
   // `BrowserWindowInterface` and use that instead.
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
-      ->GetAutofillBubbleHandler()
-      ->ShowAddressSignInPromo(web_contents, autofill_profile);
+  return browser_window->GetAutofillBubbleHandler()->ShowAddressSignInPromo(
+      web_contents, autofill_profile);
 }
 #endif  // BUILDFLAG(ENABLE_DICE_SUPPORT)
 
@@ -244,8 +256,6 @@ void AddressBubblesController::DoShowBubble() {
 
   SetBubbleView(*show_bubble_view_callback_.Run(
       web_contents(), shown_by_user_gesture_, GetWeakPtr()));
-
-  CHECK(bubble_view());
 }
 
 BubbleType AddressBubblesController::GetBubbleType() const {
@@ -312,10 +322,12 @@ void AddressBubblesController::SetUpBubble(
 }
 
 void AddressBubblesController::MaybeShowIOSDektopAddressPromo() {
-  Browser* browser =
-      BrowserWindow::FindBrowserWindowWithWebContents(web_contents())
-          ->AsBrowserView()
-          ->browser();
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents());
+  if (!browser_window) {
+    return;
+  }
+  Browser* browser = browser_window->AsBrowserView()->browser();
 
   // Verify if user is eligible for iOS promo, and attempt showing if they are.
   ios_promos_utils::VerifyIOSPromoEligibility(
diff --git chrome/browser/ui/autofill/payments/save_card_bubble_controller_impl.cc chrome/browser/ui/autofill/payments/save_card_bubble_controller_impl.cc
index 11ae2d1d559b5cd4ea9e9911697255e4fae51cba..9c67ff132d1cc110dc77698abdfe4cff7b647aae 100644
--- chrome/browser/ui/autofill/payments/save_card_bubble_controller_impl.cc
+++ chrome/browser/ui/autofill/payments/save_card_bubble_controller_impl.cc
@@ -816,6 +816,9 @@ void SaveCardBubbleControllerImpl::DoShowBubble() {
   }
 
   Browser* browser = chrome::FindBrowserWithTab(web_contents());
+  if (!browser) {
+    return;
+  }
   if (current_bubble_type_ == PaymentsBubbleType::kUploadComplete) {
     SetBubbleView(*browser->window()
                        ->GetAutofillBubbleHandler()
