diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index 2d7bb6e888f5890744b0ab6c3c8403115d32fb7c..19e5fd39101d89e0b73cff98a201226ee2698850 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -3,6 +3,7 @@
 # found in the LICENSE file.
 
 import("//build/config/buildflags_paint_preview.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//chrome/common/request_header_integrity/buildflags.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -134,6 +135,7 @@ static_library("renderer") {
     ":process_state",
     "//base/allocator:buildflags",
     "//build:chromeos_buildflags",
+    "//cef/libcef/features:buildflags",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -253,6 +255,10 @@ static_library("renderer") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_plugins) {
     sources += [
       "plugins/chrome_plugin_placeholder.cc",
