diff --git chrome/browser/extensions/BUILD.gn chrome/browser/extensions/BUILD.gn
index 45ffea4f8d3b9c1b906dbfb2af90650461e3a7e9..9d90bd3f5e8d2f3b9295339e532003b30da591c1 100644
--- chrome/browser/extensions/BUILD.gn
+++ chrome/browser/extensions/BUILD.gn
@@ -3170,10 +3170,10 @@ source_set("browser_tests") {
 
     if (!is_chromeos) {
       # Chrome OS doesn't use background mode or download bubbles.
-      deps += [
-        "//chrome/browser/background/extensions",
-        "//chrome/browser/download/bubble",
-      ]
+      if (enable_background_mode) {
+        deps += [ "//chrome/browser/background/extensions" ]
+      }
+      deps += [ "//chrome/browser/download/bubble" ]
     }
 
     if (enable_background_mode) {
