diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index ae8aef3a076c..d344ddcf3953 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -23,6 +23,7 @@ static_library("safe_browsing") {
     "//chrome/common:constants",
     "//components/browser_sync",
     "//components/enterprise/common:strings",
+    "//components/gcm_driver:gcm_buildflags",
     "//components/keyed_service/content",
     "//components/language/core/browser",
     "//components/page_info",
@@ -66,7 +67,10 @@ static_library("safe_browsing") {
         "settings_reset_prompt/settings_reset_prompt_util_win.h",
       ]
     }
-    deps += [ "//extensions/browser" ]
+    deps += [
+      "//chrome/browser:ntp_background_proto",
+      "//extensions/browser",
+    ]
   }
 
   if (safe_browsing_mode != 0) {
