diff --git chrome/browser/chrome_content_browser_client.cc chrome/browser/chrome_content_browser_client.cc
index cf4250c7ba26..b8a05db7cced 100644
--- chrome/browser/chrome_content_browser_client.cc
+++ chrome/browser/chrome_content_browser_client.cc
@@ -1049,12 +1049,16 @@ void LaunchURL(
   }
 }
 
+}  // namespace
+
 std::string GetProduct() {
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(switches::kProductVersion))
+    return command_line->GetSwitchValueASCII(switches::kProductVersion);
+
   return version_info::GetProductNameAndVersionForUserAgent();
 }
 
-}  // namespace
-
 std::string GetUserAgent() {
   base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
   if (command_line->HasSwitch(switches::kUserAgent)) {
diff --git chrome/browser/chrome_content_browser_client.h chrome/browser/chrome_content_browser_client.h
index 07e0e051ff1d..41337a1719a1 100644
--- chrome/browser/chrome_content_browser_client.h
+++ chrome/browser/chrome_content_browser_client.h
@@ -77,7 +77,8 @@ class Origin;
 
 class ChromeSerialDelegate;
 
-// Returns the user agent of Chrome.
+// Returns the product and user agent of Chrome.
+std::string GetProduct();
 std::string GetUserAgent();
 
 blink::UserAgentMetadata GetUserAgentMetadata();
