diff --git chrome/browser/extensions/api/chrome_extensions_api_client.cc chrome/browser/extensions/api/chrome_extensions_api_client.cc
index e875b597230e0626c036797971fd06b3b6dc8061..567580698f9ce55f094ff1967ef6df14e21cd3a6 100644
--- chrome/browser/extensions/api/chrome_extensions_api_client.cc
+++ chrome/browser/extensions/api/chrome_extensions_api_client.cc
@@ -16,6 +16,7 @@
 #include "base/task/single_thread_task_runner.h"
 #include "build/build_config.h"
 #include "build/chromeos_buildflags.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/extensions/api/automation_internal/chrome_automation_internal_api_delegate.h"
 #include "chrome/browser/extensions/api/declarative_content/chrome_content_rules_registry.h"
 #include "chrome/browser/extensions/api/declarative_content/default_content_predicate_evaluators.h"
@@ -102,6 +103,10 @@
 #include "chromeos/ash/components/settings/cros_settings.h"
 #endif
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/extensions/chrome_mime_handler_view_guest_delegate_cef.h"
+#endif
+
 #if BUILDFLAG(ENABLE_PRINTING)
 #include "chrome/browser/printing/printing_init.h"
 #endif
@@ -368,7 +373,11 @@ ChromeExtensionsAPIClient::CreateGuestViewManagerDelegate() const {
 std::unique_ptr<MimeHandlerViewGuestDelegate>
 ChromeExtensionsAPIClient::CreateMimeHandlerViewGuestDelegate(
     MimeHandlerViewGuest* guest) const {
+#if BUILDFLAG(ENABLE_CEF)
+  return std::make_unique<ChromeMimeHandlerViewGuestDelegateCef>(guest);
+#else
   return std::make_unique<ChromeMimeHandlerViewGuestDelegate>();
+#endif
 }
 #endif
 
diff --git chrome/browser/extensions/api/tabs/tabs_api.cc chrome/browser/extensions/api/tabs/tabs_api.cc
index 8599fed2dfbbc43a1f222c45143133020bf7e337..5a4fcd344a8488584419583a2a5a3673267fb0d7 100644
--- chrome/browser/extensions/api/tabs/tabs_api.cc
+++ chrome/browser/extensions/api/tabs/tabs_api.cc
@@ -2661,31 +2661,31 @@ ExtensionFunction::ResponseAction TabsUpdateFunction::Run() {
   }
 
   // tabs_internal::GetTabById may return a null window for prerender tabs.
-  if (!window || !ExtensionTabUtil::BrowserSupportsTabs(
-                     window->GetBrowserWindowInterface())) {
-    return RespondNow(Error(ExtensionTabUtil::kNoCurrentWindowError));
-  }
+  TabListInterface* tab_list = nullptr;
+  ::tabs::TabInterface* target_tab = nullptr;
+  if (window && ExtensionTabUtil::BrowserSupportsTabs(
+                    window->GetBrowserWindowInterface())) {
+    tab_list = TabListInterface::From(window->GetBrowserWindowInterface());
+    CHECK(tab_list);
 
-  // Cache the original web contents.
-  content::WebContents* original_contents = contents;
+    // Update the active (aka selected) tab.
+    if (!UpdateActiveTab(*params, *window->profile(),
+                         *window->GetBrowserWindowInterface(), *tab_list,
+                         tab_index, error)) {
+      return RespondNow(Error(std::move(error)));
+    }
 
-  // Update the active (aka selected) tab.
-  TabListInterface* tab_list =
-      TabListInterface::From(window->GetBrowserWindowInterface());
-  CHECK(tab_list);
-  if (!UpdateActiveTab(*params, *window->profile(),
-                       *window->GetBrowserWindowInterface(), *tab_list,
-                       tab_index, error)) {
-    return RespondNow(Error(std::move(error)));
+    // Update the highlighted tab.
+    target_tab = tab_list->GetTab(tab_index);
+    CHECK(target_tab);
+    if (!UpdateHighlightedTab(*params, *window->profile(), *tab_list,
+                              *target_tab, error)) {
+      return RespondNow(Error(std::move(error)));
+    }
   }
 
-  // Update the highlighted tab.
-  ::tabs::TabInterface* target_tab = tab_list->GetTab(tab_index);
-  CHECK(target_tab);
-  if (!UpdateHighlightedTab(*params, *window->profile(), *tab_list, *target_tab,
-                            error)) {
-    return RespondNow(Error(std::move(error)));
-  }
+  // Cache the original web contents.
+  content::WebContents* original_contents = contents;
 
   if (params->update_properties.muted &&
       !SetTabAudioMuted(contents, *params->update_properties.muted,
@@ -2695,6 +2695,10 @@ ExtensionFunction::ResponseAction TabsUpdateFunction::Run() {
   }
 
   if (params->update_properties.opener_tab_id) {
+    if (!target_tab) {
+      return RespondNow(Error(ExtensionTabUtil::kNoCurrentWindowError));
+    }
+
     int opener_id = *params->update_properties.opener_tab_id;
     content::WebContents* opener_contents = nullptr;
     if (opener_id == tab_id) {
@@ -2709,8 +2713,13 @@ ExtensionFunction::ResponseAction TabsUpdateFunction::Run() {
 
     ::tabs::TabInterface* opener_tab =
         ::tabs::TabInterface::GetFromContents(opener_contents);
-    CHECK(opener_tab);
-    if (!SetOpenerOfTab(*window->profile(), *target_tab, *opener_tab, error)) {
+    if (!opener_tab) {
+      return RespondNow(
+          Error("Tab opener must be in the same window as the updated tab."));
+    }
+
+    auto* profile = Profile::FromBrowserContext(browser_context());
+    if (!SetOpenerOfTab(*profile, *target_tab, *opener_tab, error)) {
       return RespondNow(Error(std::move(error)));
     }
   }
@@ -2718,14 +2727,19 @@ ExtensionFunction::ResponseAction TabsUpdateFunction::Run() {
   // TODO(https://crbug.com/505306735): Support on desktop android.
 #if !BUILDFLAG(IS_ANDROID)
   if (params->update_properties.auto_discardable) {
-    bool state = *params->update_properties.auto_discardable;
-    resource_coordinator::TabLifecycleUnitExternal::FromWebContents(
-        original_contents)
-        ->SetAutoDiscardable(state);
+    auto* lifecycle_unit =
+        resource_coordinator::TabLifecycleUnitExternal::FromWebContents(
+            original_contents);
+    if (!lifecycle_unit) {
+      return RespondNow(Error(ExtensionTabUtil::kNoCurrentWindowError));
+    }
+
+    lifecycle_unit->SetAutoDiscardable(
+        *params->update_properties.auto_discardable);
   }
 #endif
 
-  if (params->update_properties.pinned) {
+  if (target_tab && params->update_properties.pinned) {
     bool pinned = *params->update_properties.pinned;
 
     if (target_tab->IsPinned() != pinned) {
@@ -2755,7 +2769,8 @@ ExtensionFunction::ResponseAction TabsUpdateFunction::Run() {
   // Navigate the tab to a new location if the url is different.
   if (params->update_properties.url) {
     std::string updated_url = *params->update_properties.url;
-    if (window->profile()->IsIncognitoProfile() &&
+    auto* profile = Profile::FromBrowserContext(browser_context());
+    if (profile->IsIncognitoProfile() &&
         !IsURLAllowedInIncognito(GURL(updated_url))) {
       return RespondNow(Error(ErrorUtils::FormatErrorMessage(
           tabs_constants::kURLsNotAllowedInIncognitoError, updated_url)));
diff --git chrome/browser/extensions/extension_tab_util.cc chrome/browser/extensions/extension_tab_util.cc
index dd919b9675f64712572b162ee15e23680edbbee2..fc2af59c4529bd6cfbd2505c317a6790fa440957 100644
--- chrome/browser/extensions/extension_tab_util.cc
+++ chrome/browser/extensions/extension_tab_util.cc
@@ -73,6 +73,7 @@
 #include "chrome/browser/ui/android/tab_model/tab_model.h"
 #include "chrome/browser/ui/android/tab_model/tab_model_list.h"
 #else
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/resource_coordinator/tab_lifecycle_unit_external.h"
 // nogncheck
 #include "chrome/browser/ui/navigator/browser_navigator_params.h"  // nogncheck
@@ -93,6 +94,10 @@
 
 static_assert(BUILDFLAG(ENABLE_EXTENSIONS_CORE));
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/extensions/chrome_extension_util.h"
+#endif
+
 using content::NavigationEntry;
 using content::WebContents;
 using extensions::mojom::APIPermissionID;
@@ -672,6 +677,14 @@ bool ExtensionTabUtil::GetTabById(int tab_id,
     }
   }
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (cef::GetAlloyTabById(tab_id, profile, include_incognito, out_contents)) {
+    // |out_window| and |out_tab_index| are tied to a specific Browser window,
+    // which doesn't exist for an Alloy style browser.
+    return true;
+  }
+#endif
+
   // Prerendering tab is not visible and it cannot be in `TabStripModel`, if the
   // tab id exists as a prerendering tab, and the API will returns
   // `api::tabs::TAB_INDEX_NONE` for `out_tab_index` and a valid `WebContents`.
diff --git chrome/browser/ui/tab_helpers.h chrome/browser/ui/tab_helpers.h
index c6e05eb620dd216529c32d6b3660fc44c8b47baf..99b20faea5b8b70d0720ccca1a6aaf92bb3637c5 100644
--- chrome/browser/ui/tab_helpers.h
+++ chrome/browser/ui/tab_helpers.h
@@ -6,6 +6,7 @@
 #define CHROME_BROWSER_UI_TAB_HELPERS_H_
 
 #include "build/build_config.h"
+#include "cef/libcef/features/features.h"
 
 #if BUILDFLAG(IS_ANDROID)
 
@@ -41,6 +42,10 @@ namespace tabs {
 class TabModel;
 }  // namespace tabs
 
+#if BUILDFLAG(ENABLE_CEF)
+class CefBrowserPlatformDelegateAlloy;
+#endif
+
 // A "tab contents" is a WebContents that is used as a tab in a browser window
 // (or the equivalent on Android). The TabHelpers class allows specific classes
 // to attach the set of tab helpers that is used for tab contents.
@@ -79,6 +84,10 @@ class TabHelpers {
   // Link Preview shows a preview of a page, then promote it as a new tab.
   friend class PreviewTab;
 
+#if BUILDFLAG(ENABLE_CEF)
+  friend class CefBrowserPlatformDelegateAlloy;
+#endif
+
   // FYI: Do NOT add any more friends here. The functions above are the ONLY
   // ones that need to call AttachTabHelpers; if you think you do, re-read the
   // design document linked above, especially the section "Reusing tab helpers".
