diff --git chrome/browser/file_select_helper.cc chrome/browser/file_select_helper.cc
index 94f809da0d7d553591766ee42655f573eb92b1b7..c32e1b1be25f13c8c8ced4a2b55c48bcb5f33f61 100644
--- chrome/browser/file_select_helper.cc
+++ chrome/browser/file_select_helper.cc
@@ -67,30 +67,6 @@ using content::WebContents;
 
 DEFINE_LOCAL_ELEMENT_IDENTIFIER_VALUE(kCancelButtonId);
 
-WEB_CONTENTS_USER_DATA_KEY_IMPL(FileSelectHelper::ActiveHelpers);
-
-FileSelectHelper::ActiveHelpers::ActiveHelpers(
-    content::WebContents* web_contents)
-    : content::WebContentsUserData<ActiveHelpers>(*web_contents) {}
-
-FileSelectHelper::ActiveHelpers::~ActiveHelpers() = default;
-
-// static
-void FileSelectHelper::ActiveHelpers::Add(
-    content::WebContents* tab,
-    scoped_refptr<FileSelectHelper> helper) {
-  GetOrCreateForWebContents(tab)->helpers_.insert(std::move(helper));
-}
-
-// static
-void FileSelectHelper::ActiveHelpers::Remove(FileSelectHelper* helper) {
-  if (helper->web_contents_) {
-    if (auto* active_helpers = FromWebContents(helper->web_contents_)) {
-      active_helpers->helpers_.erase(helper);
-    }
-  }
-}
-
 namespace {
 
 void DeleteFiles(std::vector<base::FilePath> paths) {
@@ -142,13 +118,8 @@ FileSelectHelper::FileSelectHelper(Profile* profile)
 FileSelectHelper::~FileSelectHelper() {
   // There may be pending file dialogs, we need to tell them that we've gone
   // away so they don't try and call back to us.
-  if (select_file_dialog_) {
+  if (select_file_dialog_)
     select_file_dialog_->ListenerDestroyed();
-  }
-
-  if (!temporary_files_.empty()) {
-    DeleteTemporaryFiles();
-  }
 }
 
 void FileSelectHelper::FileSelected(const ui::SelectedFileInfo& file,
@@ -288,6 +259,12 @@ void FileSelectHelper::OnListDone(int error) {
   }
 
   if (dialog_type_ == ui::SelectFileDialog::SELECT_UPLOAD_FOLDER) {
+    if (run_from_cef_) {
+      // Don't show the upload confirmation dialog when triggered via CEF
+      // (initially or recursively).
+      PerformContentAnalysisIfNeeded(std::move(chooser_files));
+      return;
+    }
     auto model = CreateConfirmationDialog(
         entry->display_name_, std::move(chooser_files),
         base::BindOnce(&FileSelectHelper::PerformContentAnalysisIfNeeded,
@@ -440,6 +417,16 @@ void FileSelectHelper::DeleteTemporaryFiles() {
       base::BindOnce(&DeleteFiles, std::move(temporary_files_)));
 }
 
+void FileSelectHelper::CleanUp() {
+  if (!temporary_files_.empty()) {
+    DeleteTemporaryFiles();
+
+    // Now that the temporary files have been scheduled for deletion, there
+    // is no longer any reason to keep this instance around.
+    Release();
+  }
+}
+
 bool FileSelectHelper::AbortIfWebContentsDestroyed() {
   if (abort_on_missing_web_contents_in_tests_ &&
       (render_frame_host_ == nullptr || web_contents_ == nullptr)) {
@@ -476,31 +463,51 @@ FileSelectHelper::GetFileTypesFromAcceptType(
   std::vector<base::FilePath::StringType>* extensions =
       &file_type->extensions.back();
 
+  // Create individual filters for each accept type.
+  std::vector<std::vector<base::FilePath::StringType>> all_extensions;
+  std::vector<std::u16string> all_overrides;
+  std::vector<std::u16string> all_mimetypes;
+
   // Find the corresponding extensions.
   int valid_type_count = 0;
   int description_id = 0;
+  std::string ascii_type;
   for (const auto& accept_type : accept_types) {
+    std::vector<base::FilePath::StringType> current_extensions;
+    description_id = 0;
+    ascii_type.clear();
+
     size_t old_extension_size = extensions->size();
     if (accept_type[0] == '.') {
       // If the type starts with a period it is assumed to be a file extension
       // so we just have to add it to the list.
       base::FilePath::StringType ext =
           base::FilePath::FromUTF16Unsafe(accept_type).value();
-      extensions->push_back(ext.substr(1));
+      current_extensions.push_back(ext.substr(1));
     } else {
       if (!base::IsStringASCII(accept_type))
         continue;
-      std::string ascii_type = base::UTF16ToASCII(accept_type);
+      ascii_type = base::UTF16ToASCII(accept_type);
       if (ascii_type == "image/*")
         description_id = IDS_IMAGE_FILES;
       else if (ascii_type == "audio/*")
         description_id = IDS_AUDIO_FILES;
       else if (ascii_type == "video/*")
         description_id = IDS_VIDEO_FILES;
-
-      net::GetExtensionsForMimeType(ascii_type, extensions);
+      net::GetExtensionsForMimeType(ascii_type, &current_extensions);
     }
 
+    if (!current_extensions.empty()) {
+      all_extensions.push_back(current_extensions);
+      all_overrides.push_back(description_id != 0 ?
+                              l10n_util::GetStringUTF16(description_id) :
+                              std::u16string());
+      all_mimetypes.push_back(ascii_type.empty() ?
+                              std::u16string() : accept_type);
+
+      extensions->insert(extensions->end(), current_extensions.begin(),
+                         current_extensions.end());
+    }
     if (extensions->size() > old_extension_size)
       valid_type_count++;
   }
@@ -517,12 +524,28 @@ FileSelectHelper::GetFileTypesFromAcceptType(
   //    dialog uses the first extension in the list to form the description,
   //    like "EHTML Files". This is not what we want.
   if (valid_type_count > 1 ||
-      (valid_type_count == 1 && description_id == 0 && extensions->size() > 1))
+      (valid_type_count == 1 && description_id == 0 && extensions->size() > 1)) {
     description_id = IDS_CUSTOM_FILES;
+    ascii_type.clear();
+  }
 
-  if (description_id) {
-    file_type->extension_description_overrides.push_back(
-        l10n_util::GetStringUTF16(description_id));
+  file_type->extension_description_overrides.push_back(
+      description_id != 0 ?
+          l10n_util::GetStringUTF16(description_id) :
+          std::u16string());
+  file_type->extension_mimetypes.push_back(
+      ascii_type.empty() ? std::u16string() : base::ASCIIToUTF16(ascii_type));
+
+  if (all_extensions.size() > 1) {
+    // Insert filters for the specific accept types at the beginning.
+    file_type->extensions.insert(file_type->extensions.begin(),
+        all_extensions.begin(), all_extensions.end());
+    file_type->extension_description_overrides.insert(
+        file_type->extension_description_overrides.begin(),
+        all_overrides.begin(), all_overrides.end());
+    file_type->extension_mimetypes.insert(
+        file_type->extension_mimetypes.begin(),
+        all_mimetypes.begin(), all_mimetypes.end());
   }
 
   return file_type;
@@ -532,19 +555,16 @@ FileSelectHelper::GetFileTypesFromAcceptType(
 void FileSelectHelper::RunFileChooser(
     content::RenderFrameHost* render_frame_host,
     scoped_refptr<content::FileSelectListener> listener,
-    const FileChooserParams& params) {
+    const FileChooserParams& params,
+    bool run_from_cef) {
   Profile* profile = Profile::FromBrowserContext(
       render_frame_host->GetProcess()->GetBrowserContext());
 
-  // The `FileSelectHelper::ActiveHelpers` will keep this instance alive for the
-  // duration of the operation on this tab. Asynchronous tasks (like zipping
-  // or content analysis) may also hold a reference.
+  // FileSelectHelper will keep itself alive until it sends the result
+  // message.
   scoped_refptr<FileSelectHelper> file_select_helper(
       new FileSelectHelper(profile));
-  auto* web_contents = WebContents::FromRenderFrameHost(render_frame_host);
-  if (web_contents) {
-    ActiveHelpers::Add(web_contents, file_select_helper);
-  }
+  file_select_helper->run_from_cef_ = run_from_cef;
   file_select_helper->RunFileChooser(render_frame_host, std::move(listener),
                                      params.Clone());
 }
@@ -555,14 +575,10 @@ void FileSelectHelper::EnumerateDirectory(
     scoped_refptr<content::FileSelectListener> listener,
     const base::FilePath& path) {
   Profile* profile = Profile::FromBrowserContext(tab->GetBrowserContext());
-  // The `FileSelectHelper::ActiveHelpers` will keep this instance alive for the
-  // duration of the operation on this tab. Asynchronous tasks (like zipping
-  // or content analysis) may also hold a reference.
+  // FileSelectHelper will keep itself alive until it sends the result
+  // message.
   scoped_refptr<FileSelectHelper> file_select_helper(
       new FileSelectHelper(profile));
-  if (tab) {
-    ActiveHelpers::Add(tab, file_select_helper);
-  }
   file_select_helper->EnumerateDirectoryImpl(tab, std::move(listener), path);
 }
 
@@ -610,6 +626,13 @@ void FileSelectHelper::RunFileChooser(
       FROM_HERE, {base::MayBlock()},
       base::BindOnce(&FileSelectHelper::GetFileTypesInThreadPool, this,
                      std::move(params)));
+
+  // Because this class returns notifications to the RenderViewHost, it is
+  // difficult for callers to know how long to keep a reference to this
+  // instance. We AddRef() here to keep the instance alive after we return
+  // to the caller, until the last callback is received from the file dialog.
+  // At that point, we must call RunFileChooserEnd().
+  AddRef();
 }
 
 void FileSelectHelper::GetFileTypesInThreadPool(FileChooserParamsPtr params) {
@@ -696,8 +719,8 @@ void FileSelectHelper::RunFileChooserOnUIThread(
 }
 
 // This method is called when we receive the last callback from the file chooser
-// dialog or if the renderer was destroyed. Perform any cleanup and unregister
-// from the tab-scoped manager.
+// dialog or if the renderer was destroyed. Perform any cleanup and release the
+// reference we added in RunFileChooser().
 void FileSelectHelper::RunFileChooserEnd() {
 #if !BUILDFLAG(IS_ANDROID)
   // Ensure picture-in-picture occlusion mitigation stops, even if we need to
@@ -711,11 +734,20 @@ void FileSelectHelper::RunFileChooserEnd() {
   // If there are temporary files, then this instance needs to stick around
   // until web_contents_ is destroyed, so that this instance can delete the
   // temporary files.
-  if (!temporary_files_.empty()) {
+  if (!temporary_files_.empty())
     return;
+
+  if (listener_)
+    listener_->FileSelectionCanceled();
+  render_frame_host_ = nullptr;
+  web_contents_ = nullptr;
+  // If the dialog was actually opened, dispose of our reference.
+  if (select_file_dialog_) {
+    select_file_dialog_->ListenerDestroyed();
+    select_file_dialog_.reset();
   }
 
-  ActiveHelpers::Remove(this);
+  Release();
 }
 
 void FileSelectHelper::EnumerateDirectoryImpl(
@@ -727,6 +759,12 @@ void FileSelectHelper::EnumerateDirectoryImpl(
   dialog_type_ = ui::SelectFileDialog::SELECT_NONE;
   web_contents_ = tab;
   listener_ = std::move(listener);
+  // Because this class returns notifications to the RenderViewHost, it is
+  // difficult for callers to know how long to keep a reference to this
+  // instance. We AddRef() here to keep the instance alive after we return
+  // to the caller, until the last callback is received from the enumeration
+  // code. At that point, we must call EnumerateDirectoryEnd().
+  AddRef();
 #if BUILDFLAG(IS_ANDROID)
   if (path.IsContentUri()) {
     base::ThreadPool::PostTaskAndReplyWithResult(
@@ -739,9 +777,10 @@ void FileSelectHelper::EnumerateDirectoryImpl(
 }
 
 // This method is called when we receive the last callback from the enumeration
-// code. Perform any cleanup and unregister from the tab-scoped manager.
+// code. Perform any cleanup and release the reference we added in
+// EnumerateDirectoryImpl().
 void FileSelectHelper::EnumerateDirectoryEnd() {
-  ActiveHelpers::Remove(this);
+  Release();
 }
 
 void FileSelectHelper::RenderFrameHostChanged(
@@ -768,6 +807,7 @@ void FileSelectHelper::WebContentsDestroyed() {
   render_frame_host_ = nullptr;
   web_contents_ = nullptr;
   profile_ = nullptr;
+  CleanUp();
 }
 
 void FileSelectHelper::OnTabDeactivated(tabs::TabInterface* tab) {
diff --git chrome/browser/file_select_helper.h chrome/browser/file_select_helper.h
index e2599cf85006370a4a152e95e394f5aa81af30f2..ba572985c16f306e37d37f2948d8041bbeb6a635 100644
--- chrome/browser/file_select_helper.h
+++ chrome/browser/file_select_helper.h
@@ -18,7 +18,6 @@
 #include "components/enterprise/common/files_scan_data.h"
 #include "content/public/browser/browser_thread.h"
 #include "content/public/browser/web_contents_observer.h"
-#include "content/public/browser/web_contents_user_data.h"
 #include "net/base/directory_lister.h"
 #include "third_party/blink/public/mojom/choosers/file_chooser.mojom.h"
 #include "ui/shell_dialogs/select_file_dialog.h"
@@ -63,22 +62,6 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
                          public content::WebContentsObserver,
                          private net::DirectoryLister::DirectoryListerDelegate {
  public:
-  // Tab-scoped container that anchors the lifetime of active operations.
-  class ActiveHelpers : public content::WebContentsUserData<ActiveHelpers> {
-   public:
-    ~ActiveHelpers() override;
-
-    static void Add(content::WebContents* tab,
-                    scoped_refptr<FileSelectHelper> helper);
-    static void Remove(FileSelectHelper* helper);
-
-   private:
-    friend WebContentsUserData;
-    explicit ActiveHelpers(content::WebContents* web_contents);
-    std::set<scoped_refptr<FileSelectHelper>> helpers_;
-    WEB_CONTENTS_USER_DATA_KEY_DECL();
-  };
-
   FileSelectHelper(const FileSelectHelper&) = delete;
   FileSelectHelper& operator=(const FileSelectHelper&) = delete;
 
@@ -86,7 +69,8 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
   static void RunFileChooser(
       content::RenderFrameHost* render_frame_host,
       scoped_refptr<content::FileSelectListener> listener,
-      const blink::mojom::FileChooserParams& params);
+      const blink::mojom::FileChooserParams& params,
+      bool run_from_cef = false);
 
   // Enumerates all the files in directory.
   static void EnumerateDirectory(
@@ -94,10 +78,6 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
       scoped_refptr<content::FileSelectListener> listener,
       const base::FilePath& path);
 
-  base::WeakPtr<FileSelectHelper> GetWeakPtr() {
-    return weak_ptr_factory_.GetWeakPtr();
-  }
-
  private:
   friend class base::RefCountedThreadSafe<FileSelectHelper>;
   friend class base::DeleteHelper<FileSelectHelper>;
@@ -138,7 +118,6 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
   FRIEND_TEST_ALL_PREFIXES(FileSelectHelperTest, GetFileTypesFromAcceptType);
   FRIEND_TEST_ALL_PREFIXES(FileSelectHelperTest, MultipleFileExtensionsForMime);
   FRIEND_TEST_ALL_PREFIXES(FileSelectHelperTest, ConfirmationDialog);
-  FRIEND_TEST_ALL_PREFIXES(FileSelectHelperTest, TaskKeepsAlive);
   FRIEND_TEST_ALL_PREFIXES(policy::DlpFilesControllerAshBrowserTest,
                            FilesUploadCallerPassed);
 
@@ -270,6 +249,9 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
   // vector.
   void DeleteTemporaryFiles();
 
+  // Cleans up when the initiator of the file chooser is no longer valid.
+  void CleanUp();
+
   // Calls RunFileChooserEnd() if the webcontents was destroyed. Returns true
   // if the file chooser operation shouldn't proceed.
   bool AbortIfWebContentsDestroyed();
@@ -357,7 +339,12 @@ class FileSelectHelper : public base::RefCountedThreadSafe<
   std::unique_ptr<ScopedTuckPictureInPicture> scoped_tuck_picture_in_picture_;
 #endif  // !BUILDFLAG(IS_ANDROID)
 
+  // Set to true if this dialog was triggered via CEF.
+  bool run_from_cef_ = false;
+
+#if BUILDFLAG(IS_CHROMEOS)
   base::WeakPtrFactory<FileSelectHelper> weak_ptr_factory_{this};
+#endif  // BUILDFLAG(IS_CHROMEOS)
 };
 
 #endif  // CHROME_BROWSER_FILE_SELECT_HELPER_H_
diff --git chrome/browser/ui/select_file_policy/chrome_select_file_policy.h chrome/browser/ui/select_file_policy/chrome_select_file_policy.h
index 993b89dca34875d458370dc0730a98fb36bf7bff..d82beccf7014fe79c84b82d549d4015ee5c23b95 100644
--- chrome/browser/ui/select_file_policy/chrome_select_file_policy.h
+++ chrome/browser/ui/select_file_policy/chrome_select_file_policy.h
@@ -30,6 +30,8 @@ class ChromeSelectFilePolicy : public ui::SelectFilePolicy {
   // Returns true if local state allows showing file pickers.
   static bool FileSelectDialogsAllowed();
 
+  content::WebContents* source_contents() const { return source_contents_; }
+
  private:
   raw_ptr<content::WebContents, AcrossTasksDanglingUntriaged> source_contents_;
 };
diff --git ui/shell_dialogs/execute_select_file_win.cc ui/shell_dialogs/execute_select_file_win.cc
index b9f46bd6ebf8da7c668a67a7a3aea0e35910cbe3..ca2dceef5e0a077a2394f528e0156e6fef0c7293 100644
--- ui/shell_dialogs/execute_select_file_win.cc
+++ ui/shell_dialogs/execute_select_file_win.cc
@@ -31,7 +31,7 @@ namespace ui {
 namespace {
 
 // Stop switch for the AutoCloseDialogEventHandler.
-BASE_FEATURE(kAutoCloseFileDialogs, base::FEATURE_ENABLED_BY_DEFAULT);
+BASE_FEATURE(kAutoCloseFileDialogs, base::FEATURE_DISABLED_BY_DEFAULT);
 
 // RAII wrapper around AutoCloseDialogEventHandler.
 class ScopedAutoCloseDialogEventHandler {
diff --git ui/shell_dialogs/select_file_dialog.cc ui/shell_dialogs/select_file_dialog.cc
index d44aa9ea9f3f81e2d6dd3f28f67980ded747f7d8..60f42e1299d5cd909974de91cd48c4b79b4c1590 100644
--- ui/shell_dialogs/select_file_dialog.cc
+++ ui/shell_dialogs/select_file_dialog.cc
@@ -90,8 +90,10 @@ void SelectFileDialog::SetFactory(
 // static
 scoped_refptr<SelectFileDialog> SelectFileDialog::Create(
     Listener* listener,
-    std::unique_ptr<ui::SelectFilePolicy> policy) {
-  if (dialog_factory_)
+    std::unique_ptr<ui::SelectFilePolicy> policy,
+    bool run_from_cef) {
+  // Avoid reentrancy of the CEF factory.
+  if (dialog_factory_ && (!run_from_cef || !dialog_factory_->IsCefFactory()))
     return dialog_factory_->Create(listener, std::move(policy));
   return CreateSelectFileDialog(listener, std::move(policy));
 }
diff --git ui/shell_dialogs/select_file_dialog.h ui/shell_dialogs/select_file_dialog.h
index b0f4b3df0a4f10b306ba2d25e5fd51440a25bcd4..7fbb2be86bce8cf7676c0166a5510335ab94e0b5 100644
--- ui/shell_dialogs/select_file_dialog.h
+++ ui/shell_dialogs/select_file_dialog.h
@@ -96,7 +96,8 @@ class SHELL_DIALOGS_EXPORT SelectFileDialog
   // listener when there are no calls to SelectFile() outstanding.
   static scoped_refptr<SelectFileDialog> Create(
       Listener* listener,
-      std::unique_ptr<SelectFilePolicy> policy);
+      std::unique_ptr<SelectFilePolicy> policy,
+      bool run_from_cef = false);
 
   SelectFileDialog(const SelectFileDialog&) = delete;
   SelectFileDialog& operator=(const SelectFileDialog&) = delete;
@@ -130,6 +131,10 @@ class SHELL_DIALOGS_EXPORT SelectFileDialog
     // FileTypeExtensions instead of this and the above vector?
     std::vector<std::u16string> extension_description_overrides;
 
+    // Original mime types for the specified extensions. Entries correspond to
+    // |extensions|; if left blank then there was no mime type.
+    std::vector<std::u16string> extension_mimetypes;
+
     // Specifies whether there will be a filter added for all files (i.e. *.*).
     bool include_all_files = false;
 
@@ -224,6 +229,19 @@ class SHELL_DIALOGS_EXPORT SelectFileDialog
                   const GURL* caller = nullptr);
   bool HasMultipleFileTypeChoices();
 
+  // Match the types used by CefWindowHandle.
+#if BUILDFLAG(IS_MAC)
+  using WidgetType = void*;
+  static constexpr WidgetType kNullWidget = nullptr;
+#else
+  using WidgetType = gfx::AcceleratedWidget;
+  static constexpr WidgetType kNullWidget = gfx::kNullAcceleratedWidget;
+#endif
+
+  void set_owning_widget(WidgetType widget) {
+    owning_widget_ = widget;
+  }
+
  protected:
   friend class base::RefCountedThreadSafe<SelectFileDialog>;
 
@@ -248,6 +266,11 @@ class SHELL_DIALOGS_EXPORT SelectFileDialog
   // The listener to be notified of selection completion.
   raw_ptr<Listener> listener_;
 
+  std::unique_ptr<SelectFilePolicy> select_file_policy_;
+
+  // Support override of the |owning_window| value.
+  WidgetType owning_widget_ = kNullWidget;
+
  private:
   // Tests if the file selection dialog can be displayed by
   // testing if the AllowFileSelectionDialogs-Policy is
@@ -260,8 +283,6 @@ class SHELL_DIALOGS_EXPORT SelectFileDialog
 
   // Returns true if the dialog has multiple file type choices.
   virtual bool HasMultipleFileTypeChoicesImpl() = 0;
-
-  std::unique_ptr<SelectFilePolicy> select_file_policy_;
 };
 
 SelectFileDialog* CreateSelectFileDialog(
diff --git ui/shell_dialogs/select_file_dialog_factory.h ui/shell_dialogs/select_file_dialog_factory.h
index c7acd9b05fbb8c94c37a03c9b7e170548cb3dfbb..3e95e4125fa245e9671437e847058a865a52876a 100644
--- ui/shell_dialogs/select_file_dialog_factory.h
+++ ui/shell_dialogs/select_file_dialog_factory.h
@@ -24,6 +24,8 @@ class SHELL_DIALOGS_EXPORT SelectFileDialogFactory {
   virtual SelectFileDialog* Create(
       ui::SelectFileDialog::Listener* listener,
       std::unique_ptr<ui::SelectFilePolicy> policy) = 0;
+
+  virtual bool IsCefFactory() const { return false; }
 };
 
 }  // namespace ui
diff --git ui/shell_dialogs/select_file_dialog_mac.mm ui/shell_dialogs/select_file_dialog_mac.mm
index 5da67cfb7362d01fac8d0be4565d319087ec3cf9..6e618083bc851ba2d9a11adebfeb8ac1afffa89c 100644
--- ui/shell_dialogs/select_file_dialog_mac.mm
+++ ui/shell_dialogs/select_file_dialog_mac.mm
@@ -105,6 +105,10 @@ void SelectFileDialogImpl::SelectFileImpl(
     mojo_window->CreateSelectFileDialog(std::move(receiver));
   } else {
     NSWindow* ns_window = gfx_window.GetNativeNSWindow();
+    if (!ns_window && owning_widget_) {
+      NSView* view = ((__bridge NSView*)owning_widget_);
+      ns_window = [view window];
+    }
     mojo::MakeSelfOwnedReceiver(
         std::make_unique<remote_cocoa::SelectFileDialogBridge>(ns_window),
         std::move(receiver));
diff --git ui/shell_dialogs/select_file_dialog_win.cc ui/shell_dialogs/select_file_dialog_win.cc
index e3edd353d533ad35749cca0e0a248e47d371f821..863f04f0a5d29f2ec6dc1d7cd31b41badcee7488 100644
--- ui/shell_dialogs/select_file_dialog_win.cc
+++ ui/shell_dialogs/select_file_dialog_win.cc
@@ -250,6 +250,8 @@ void SelectFileDialogImpl::SelectFileImpl(
   HWND owner = owning_window && owning_window->GetRootWindow()
                    ? owning_window->GetHost()->GetAcceleratedWidget()
                    : nullptr;
+  if (!owner)
+    owner = owning_widget_;
 
   std::unique_ptr<RunState> run_state = BeginRun(owner);
 
