diff --git components/content_settings/renderer/content_settings_agent_impl.cc components/content_settings/renderer/content_settings_agent_impl.cc
index 2394aeb4fb33ed4c99fa2be3892a7ad6db5ecfaa..7f27b04a0dc50a33f56cc4bb3eae3b132c5b051c 100644
--- components/content_settings/renderer/content_settings_agent_impl.cc
+++ components/content_settings/renderer/content_settings_agent_impl.cc
@@ -146,7 +146,7 @@ ContentSetting GetContentSettingFromRules(
       return rule.GetContentSetting();
     }
   }
-  NOTREACHED();
+  return CONTENT_SETTING_DEFAULT;
 }
 }  // namespace
 
