diff --git chrome/utility/chrome_content_utility_client.cc chrome/utility/chrome_content_utility_client.cc
index f779813dd5333..547ba5ebfc161 100644
--- chrome/utility/chrome_content_utility_client.cc
+++ chrome/utility/chrome_content_utility_client.cc
@@ -12,6 +12,7 @@
 #include "base/files/file_path.h"
 #include "base/path_service.h"
 #include "build/build_config.h"
+#include "cef/libcef/features/runtime.h"
 #include "chrome/common/chrome_paths.h"
 #include "chrome/common/profiler/thread_profiler.h"
 #include "chrome/common/profiler/thread_profiler_configuration.h"
@@ -56,7 +57,8 @@ void ChromeContentUtilityClient::UtilityThreadStarted() {
       command_line->GetSwitchValueASCII(switches::kProcessType);
   // An in-process utility thread may run in other processes, only set up
   // collector in a utility process.
-  if (process_type == switches::kUtilityProcess) {
+  if (process_type == switches::kUtilityProcess &&
+      !cef::IsAlloyRuntimeEnabled()) {
     // The HeapProfilerController should have been created in
     // ChromeMainDelegate::PostEarlyInitialization.
     using HeapProfilerController = heap_profiling::HeapProfilerController;
