diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index 44798c35c1e7..88a7639a96d9 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -4,6 +4,7 @@
 
 import("//build/config/features.gni")
 import("//build/config/jumbo.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//components/nacl/features.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -121,6 +122,7 @@ jumbo_static_library("renderer") {
   defines = []
 
   deps = [
+    "//cef/libcef/features",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -187,6 +189,10 @@ jumbo_static_library("renderer") {
 
   configs += [ "//build/config/compiler:wexit_time_destructors" ]
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_nacl) {
     deps += [
       "//components/nacl/loader",
