diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index 283e1c70773a3403f8ca94cecda63a7018bdedac..5afef920ae4328dcb553c14d59235f71b4b6850e 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -4,6 +4,7 @@
 
 import("//build/config/buildflags_paint_preview.gni")
 import("//build/config/features.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//chrome/common/request_header_integrity/buildflags.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -136,6 +137,7 @@ static_library("renderer") {
     ":process_state",
     "//base/allocator:buildflags",
     "//build:chromeos_buildflags",
+    "//cef/libcef/features:buildflags",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -250,6 +252,10 @@ static_library("renderer") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_plugins) {
     sources += [
       "plugins/chrome_plugin_placeholder.cc",
