diff --git chrome/browser/extensions/BUILD.gn chrome/browser/extensions/BUILD.gn
index 7f94b49042c431f867493942fafb855ece96f14a..a8b4c85f373534b678cbabe653320e93e37acc5f 100644
--- chrome/browser/extensions/BUILD.gn
+++ chrome/browser/extensions/BUILD.gn
@@ -3111,10 +3111,10 @@ source_set("browser_tests") {
 
     if (!is_chromeos) {
       # Chrome OS doesn't use background mode or download bubbles.
-      deps += [
-        "//chrome/browser/background/extensions",
-        "//chrome/browser/download/bubble",
-      ]
+      if (enable_background_mode) {
+        deps += [ "//chrome/browser/background/extensions" ]
+      }
+      deps += [ "//chrome/browser/download/bubble" ]
     }
 
     if (use_aura) {
