diff --git chrome/browser/printing/print_backend_service_manager.cc chrome/browser/printing/print_backend_service_manager.cc
index 4a0f4f0cc8ffb2c1c64fd6af7042609354636c83..6e795f30d2e2d8b5daedd470d48a859b44ecdb4c 100644
--- chrome/browser/printing/print_backend_service_manager.cc
+++ chrome/browser/printing/print_backend_service_manager.cc
@@ -97,7 +97,15 @@ void MaybeCloneOrCopy(
 // to support modal dialogs from OOP.
 uint32_t NativeViewToUint(gfx::NativeView view) {
 #if BUILDFLAG(IS_WIN)
-  return base::win::HandleToUint32(views::HWNDForNativeView(view));
+  // |view| will be nullptr with CEF OSR. In that case return the desktop
+  // window handle (same behavior as PrintingContextWin::GetRootWindow) to
+  // avoid a DCHECK in PrintingContextSystemDialogWin::AskUserForSettings
+  // with OOP printing.
+  HWND hwnd = views::HWNDForNativeView(view);
+  if (!hwnd) {
+    hwnd = ::GetDesktopWindow();
+  }
+  return base::win::HandleToUint32(hwnd);
 #else
   NOTREACHED();
 #endif
