diff --git chrome/browser/ui/startup/startup_browser_creator.cc chrome/browser/ui/startup/startup_browser_creator.cc
index e1d02996005979b33cca70ed68cd1c0fc622fb80..9bcf0360f4161d03c67f1be25788d9b7e72ebf9b 100644
--- chrome/browser/ui/startup/startup_browser_creator.cc
+++ chrome/browser/ui/startup/startup_browser_creator.cc
@@ -773,6 +773,13 @@ std::optional<bool> MaybeHandleValidateCrx(
   return false;
 }
 
+StartupBrowserCreator::ProcessCommandLineCallback*
+GetProcessCommandLineCallback() {
+  static base::NoDestructor<StartupBrowserCreator::ProcessCommandLineCallback>
+      callback;
+  return callback.get();
+}
+
 }  // namespace
 
 StartupBrowserCreator::StartupBrowserCreator() = default;
@@ -1614,6 +1621,12 @@ void StartupBrowserCreator::ProcessCommandLineWithProfile(
       {profile, mode}, last_opened_profiles);
 }
 
+// static
+void StartupBrowserCreator::RegisterProcessCommandLineCallback(
+    ProcessCommandLineCallback cb) {
+  *GetProcessCommandLineCallback() = cb;
+}
+
 // static
 void StartupBrowserCreator::ProcessCommandLineAlreadyRunning(
     const base::CommandLine& command_line,
@@ -1623,6 +1636,11 @@ void StartupBrowserCreator::ProcessCommandLineAlreadyRunning(
     return;
   }
 
+  auto* cb = GetProcessCommandLineCallback();
+  if (!cb->is_null() && cb->Run(command_line, cur_dir)) {
+    return;
+  }
+
   Profile* profile = nullptr;
   if (profile_path_info.mode == StartupProfileMode::kBrowserWindow) {
     ProfileManager* profile_manager = g_browser_process->profile_manager();
diff --git chrome/browser/ui/startup/startup_browser_creator.h chrome/browser/ui/startup/startup_browser_creator.h
index 9e6c961694cc51b6ca8e738b2cbb6704468dc3ba..1c6aec7b6da3d61ebf27f258ef7d75b12713e35b 100644
--- chrome/browser/ui/startup/startup_browser_creator.h
+++ chrome/browser/ui/startup/startup_browser_creator.h
@@ -9,6 +9,7 @@
 #include <vector>
 
 #include "base/files/file_path.h"
+#include "base/functional/callback.h"
 #include "base/gtest_prod_util.h"
 #include "base/memory/raw_ptr.h"
 #include "build/build_config.h"
@@ -93,6 +94,13 @@ class StartupBrowserCreator {
              StartupProfileInfo profile_info,
              const Profiles& last_opened_profiles);
 
+  // Registers a callback that will be executed each time
+  // ProcessCommandLineAlreadyRunning is called.
+  using ProcessCommandLineCallback = base::RepeatingCallback<bool(
+      const base::CommandLine& command_line,
+      const base::FilePath& cur_dir)>;
+  static void RegisterProcessCommandLineCallback(ProcessCommandLineCallback cb);
+
   // This function performs command-line handling and is invoked only after
   // start up (for example when we get a start request for another process).
   // |command_line| holds the command line we need to process.
