diff --git chrome/browser/ui/startup/startup_browser_creator.cc chrome/browser/ui/startup/startup_browser_creator.cc
index 26e845c75d014f75e4e6ef4212282531c64d185f..344b9c1cb1e0e1c181b1b311e96677f07abdfd6f 100644
--- chrome/browser/ui/startup/startup_browser_creator.cc
+++ chrome/browser/ui/startup/startup_browser_creator.cc
@@ -649,6 +649,13 @@ std::optional<bool> MaybeHandleValidateCrx(
   return false;
 }
 
+StartupBrowserCreator::ProcessCommandLineCallback*
+GetProcessCommandLineCallback() {
+  static base::NoDestructor<StartupBrowserCreator::ProcessCommandLineCallback>
+      callback;
+  return callback.get();
+}
+
 }  // namespace
 
 StartupBrowserCreator::StartupBrowserCreator() = default;
@@ -1490,6 +1497,12 @@ void StartupBrowserCreator::ProcessCommandLineWithProfile(
       {profile, mode}, last_opened_profiles);
 }
 
+// static
+void StartupBrowserCreator::RegisterProcessCommandLineCallback(
+    ProcessCommandLineCallback cb) {
+  *GetProcessCommandLineCallback() = cb;
+}
+
 // static
 void StartupBrowserCreator::ProcessCommandLineAlreadyRunning(
     const base::CommandLine& command_line,
@@ -1499,6 +1512,11 @@ void StartupBrowserCreator::ProcessCommandLineAlreadyRunning(
     return;
   }
 
+  auto* cb = GetProcessCommandLineCallback();
+  if (!cb->is_null() && cb->Run(command_line, cur_dir)) {
+    return;
+  }
+
   Profile* profile = nullptr;
   if (profile_path_info.mode == StartupProfileMode::kBrowserWindow) {
     ProfileManager* profile_manager = g_browser_process->profile_manager();
diff --git chrome/browser/ui/startup/startup_browser_creator.h chrome/browser/ui/startup/startup_browser_creator.h
index c384e930cb0481ef2b3417f8f6a811cfc31f33b6..b541ad3adddfe2674be0730615f7143253b51d0e 100644
--- chrome/browser/ui/startup/startup_browser_creator.h
+++ chrome/browser/ui/startup/startup_browser_creator.h
@@ -9,6 +9,7 @@
 #include <vector>
 
 #include "base/files/file_path.h"
+#include "base/functional/callback.h"
 #include "base/gtest_prod_util.h"
 #include "base/memory/raw_ptr.h"
 #include "build/build_config.h"
@@ -93,6 +94,13 @@ class StartupBrowserCreator {
              StartupProfileInfo profile_info,
              const Profiles& last_opened_profiles);
 
+  // Registers a callback that will be executed each time
+  // ProcessCommandLineAlreadyRunning is called.
+  using ProcessCommandLineCallback = base::RepeatingCallback<bool(
+      const base::CommandLine& command_line,
+      const base::FilePath& cur_dir)>;
+  static void RegisterProcessCommandLineCallback(ProcessCommandLineCallback cb);
+
   // This function performs command-line handling and is invoked only after
   // start up (for example when we get a start request for another process).
   // |command_line| holds the command line we need to process.
