diff --git chrome/browser/ui/bookmarks/bookmark_stats.cc chrome/browser/ui/bookmarks/bookmark_stats.cc
index 19d75a725dbcfaf100cfe2d1ec2fc551ffecde65..c325a7116c45ad68f9bd54d3e7760dd7b388b0d8 100644
--- chrome/browser/ui/bookmarks/bookmark_stats.cc
+++ chrome/browser/ui/bookmarks/bookmark_stats.cc
@@ -29,10 +29,14 @@ bool IsBookmarkBarLocation(BookmarkLaunchLocation location) {
 auto GetMetricProfile(const Profile* profile) {
   DCHECK(profile);
   DCHECK(profile->IsRegularProfile() ||
-         profile->IsPrimaryOTRProfileWithRegularParent());
+         profile->IsPrimaryOTRProfileWithRegularParent() ||
+         (profile->IsOffTheRecord() &&
+          profile->GetOTRProfileID().IsUniqueForCEF()));
   if (profile->IsRegularProfile()) {
     return profile_metrics::BrowserProfileType::kRegular;
-  } else if (profile->IsIncognitoProfile()) {
+  } else if (profile->IsIncognitoProfile() ||
+             (profile->IsOffTheRecord() &&
+              profile->GetOTRProfileID().IsUniqueForCEF())) {
     return profile_metrics::BrowserProfileType::kIncognito;
   }
   return profile_metrics::BrowserProfileType::kEnterpriseIsolated;
diff --git chrome/browser/ui/views/incognito_clear_browsing_data_dialog.cc chrome/browser/ui/views/incognito_clear_browsing_data_dialog.cc
index f25b0d6283336b899d58cc83879820d2b46e9cfe..5aedb7c65fe40e40dd061156fb38625fbed24c58 100644
--- chrome/browser/ui/views/incognito_clear_browsing_data_dialog.cc
+++ chrome/browser/ui/views/incognito_clear_browsing_data_dialog.cc
@@ -30,7 +30,9 @@ IncognitoClearBrowsingDataDialog::IncognitoClearBrowsingDataDialog(
       dialog_type_(type),
       incognito_profile_(incognito_profile) {
   DCHECK(incognito_profile_);
-  DCHECK(incognito_profile_->IsPrimaryOTRProfileWithRegularParent());
+  DCHECK(incognito_profile_->IsPrimaryOTRProfileWithRegularParent() ||
+         (incognito_profile_->IsOffTheRecord() &&
+          incognito_profile_->GetOTRProfileID().IsUniqueForCEF()));
   SetButtons(static_cast<int>(ui::mojom::DialogButton::kNone));
   SetShowCloseButton(true);
 
diff --git chrome/browser/ui/views/incognito_clear_browsing_data_dialog_coordinator.cc chrome/browser/ui/views/incognito_clear_browsing_data_dialog_coordinator.cc
index e87fb564bcd043a0010788eedbc61825d005160d..7dbd5ef335945c06ab25a1400ab021080fce564e 100644
--- chrome/browser/ui/views/incognito_clear_browsing_data_dialog_coordinator.cc
+++ chrome/browser/ui/views/incognito_clear_browsing_data_dialog_coordinator.cc
@@ -23,6 +23,10 @@ IncognitoClearBrowsingDataDialogCoordinator::
 void IncognitoClearBrowsingDataDialogCoordinator::Show(
     IncognitoClearBrowsingDataDialogInterface::Type type,
     views::BubbleAnchor anchor) {
+  // The full toolbar may not be visible.
+  if (anchor.IsNull())
+    return;
+
   if (bubble_tracker_.view() && bubble_tracker_.view()->GetWidget()) {
     // Ensure the previous bubble is closed before creating and showing the new
     // one.
diff --git chrome/browser/ui/views/profiles/incognito_menu_view.cc chrome/browser/ui/views/profiles/incognito_menu_view.cc
index 5e34acec16e1f581d7842e4210f00fd9b107ad70..bbe3cd63f1a921d0e62ad91a60af4ee009837fb9 100644
--- chrome/browser/ui/views/profiles/incognito_menu_view.cc
+++ chrome/browser/ui/views/profiles/incognito_menu_view.cc
@@ -26,7 +26,9 @@
 IncognitoMenuView::IncognitoMenuView(views::BubbleAnchor anchor_element,
                                      BrowserWindowInterface* browser)
     : ProfileMenuViewBase(anchor_element, browser) {
-  CHECK(profile().IsIncognitoProfile());
+  CHECK(profile().IsIncognitoProfile() ||
+        (profile().IsOffTheRecord() &&
+         profile().GetOTRProfileID().IsUniqueForCEF()));
   GetViewAccessibility().SetName(GetAccessibleWindowTitle(),
                                  ax::mojom::NameFrom::kAttribute);
 
diff --git chrome/browser/ui/views/profiles/profile_menu_coordinator.cc chrome/browser/ui/views/profiles/profile_menu_coordinator.cc
index 74ccee8d69fbf8ad8c621bcd53d8aff02e6c0671..7dcd7b25770852088c8c674982a8a739a45e6364 100644
--- chrome/browser/ui/views/profiles/profile_menu_coordinator.cc
+++ chrome/browser/ui/views/profiles/profile_menu_coordinator.cc
@@ -99,7 +99,9 @@ void ProfileMenuCoordinator::ShowWithPromoResults(
   auto avatar_toolbar_button = GetAvatarToolbarButton();
   std::unique_ptr<ProfileMenuViewBase> bubble;
   const bool is_incognito = GetProfile()->IsIncognitoProfile();
-  if (is_incognito) {
+  if (is_incognito ||
+         (GetProfile()->IsOffTheRecord() &&
+          GetProfile()->GetOTRProfileID().IsUniqueForCEF())) {
     bubble = std::make_unique<IncognitoMenuView>(avatar_toolbar_button,
                                                  &browser_.get());
   } else if (GetProfile()->IsEnterpriseIsolatedModeProfile()) {
