diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 9fe789540c05..a2646c13e5d5 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -12,6 +12,7 @@ import("//build/config/crypto.gni")
 import("//build/config/features.gni")
 import("//build/config/linux/gtk/gtk.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -1824,6 +1825,7 @@ static_library("browser") {
     "//base/util/values:values_util",
     "//build:branding_buildflags",
     "//cc",
+    "//cef/libcef/features",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -2156,6 +2158,10 @@ static_library("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     sources += [
       "after_startup_task_utils_android.cc",
