diff --git chrome/browser/task_manager/providers/render_process_host_task_provider.cc chrome/browser/task_manager/providers/render_process_host_task_provider.cc
index 11b05cbfc6077..45a5b11158a21 100644
--- chrome/browser/task_manager/providers/render_process_host_task_provider.cc
+++ chrome/browser/task_manager/providers/render_process_host_task_provider.cc
@@ -110,18 +110,14 @@ void RenderProcessHostTaskProvider::OnRenderProcessHostCreated(
 void RenderProcessHostTaskProvider::RenderProcessExited(
     content::RenderProcessHost* host,
     const content::ChildProcessTerminationInfo& info) {
-  if (is_updating_) {
-    DeleteTask(host->GetID());
-    host_observation_.RemoveObservation(host);
-  }
+  DeleteTask(host->GetID());
+  host_observation_.RemoveObservation(host);
 }
 
 void RenderProcessHostTaskProvider::RenderProcessHostDestroyed(
     content::RenderProcessHost* host) {
-  if (is_updating_) {
-    DeleteTask(host->GetID());
-    host_observation_.RemoveObservation(host);
-  }
+  DeleteTask(host->GetID());
+  host_observation_.RemoveObservation(host);
 }
 
 }  // namespace task_manager
