diff --git chrome/browser/devtools/chrome_devtools_manager_delegate.cc chrome/browser/devtools/chrome_devtools_manager_delegate.cc
index 692e883ad3b60..4b45234cea0bc 100644
--- chrome/browser/devtools/chrome_devtools_manager_delegate.cc
+++ chrome/browser/devtools/chrome_devtools_manager_delegate.cc
@@ -14,6 +14,7 @@
 #include "base/strings/strcat.h"
 #include "base/strings/utf_string_conversions.h"
 #include "build/build_config.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/browser_features.h"
 #include "chrome/browser/devtools/chrome_devtools_session.h"
 #include "chrome/browser/devtools/device/android_device_manager.h"
@@ -66,6 +67,10 @@
 #include "components/prefs/pref_service.h"
 #endif
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/extensions/chrome_extension_util.h"
+#endif
+
 using content::DevToolsAgentHost;
 
 const char ChromeDevToolsManagerDelegate::kTypeApp[] = "app";
@@ -280,6 +285,12 @@ std::string ChromeDevToolsManagerDelegate::GetTargetType(
     return DevToolsAgentHost::kTypePage;
   }
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (cef::IsAlloyContents(web_contents, /*primary_only=*/true)) {
+    return DevToolsAgentHost::kTypePage;
+  }
+#endif
+
   return DevToolsAgentHost::kTypeOther;
 }
 
