diff --git chrome/browser/ui/browser_command_controller.cc chrome/browser/ui/browser_command_controller.cc
index d637eecb9ee1cf7e3de4f720ee29599abe1d562c..14a3dae9eeee3f3f68f74413be42238a829b7aa1 100644
--- chrome/browser/ui/browser_command_controller.cc
+++ chrome/browser/ui/browser_command_controller.cc
@@ -531,6 +531,7 @@ bool BrowserCommandController::ExecuteCommandWithDisposition(
   // choose to not implement CommandUpdaterDelegate inside this class and
   // therefore command_updater_ doesn't have the delegate set).
   if (!SupportsCommand(id) || !IsCommandEnabled(id)) {
+    LOG(WARNING) << "Invalid/disabled command " << id;
     return false;
   }
 
@@ -548,6 +549,13 @@ bool BrowserCommandController::ExecuteCommandWithDisposition(
   DCHECK(command_updater_.IsCommandEnabled(id))
       << "Invalid/disabled command " << id;
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (browser_->cef_delegate() &&
+      browser_->cef_delegate()->HandleCommand(id, disposition)) {
+    return true;
+  }
+#endif
+
   // The order of commands in this switch statement must match the function
   // declaration order in browser.h!
   switch (id) {
@@ -1431,12 +1439,14 @@ void BrowserCommandController::TabRestoreServiceLoaded(
 
 bool BrowserCommandController::IsShowingMainUI() {
   return browser_->SupportsWindowFeature(
-      Browser::WindowFeature::kFeatureTabStrip);
+             Browser::WindowFeature::kFeatureTabStrip) ||
+         browser_->toolbar_overridden();
 }
 
 bool BrowserCommandController::IsShowingLocationBar() {
   return browser_->SupportsWindowFeature(
-      Browser::WindowFeature::kFeatureLocationBar);
+             Browser::WindowFeature::kFeatureLocationBar) ||
+         browser_->toolbar_overridden();
 }
 
 void BrowserCommandController::InitCommandState() {
diff --git chrome/browser/ui/toolbar/app_menu_model.cc chrome/browser/ui/toolbar/app_menu_model.cc
index a3fe5c31ca3ab269a5368bde5181e760a056b4c2..21c7986504f3cf240fea302b50fc83be65873631 100644
--- chrome/browser/ui/toolbar/app_menu_model.cc
+++ chrome/browser/ui/toolbar/app_menu_model.cc
@@ -814,10 +814,12 @@ FindAndEditSubMenuModel::FindAndEditSubMenuModel(
     ui::SimpleMenuModel::Delegate* delegate)
     : SimpleMenuModel(delegate) {
   AddItemWithStringIdAndVectorIcon(this, IDC_FIND, IDS_FIND, kSearchMenuIcon);
+  if (delegate->IsCommandIdVisible(IDC_EDIT_MENU)) {
   AddSeparator(ui::NORMAL_SEPARATOR);
   AddItemWithStringIdAndVectorIcon(this, IDC_CUT, IDS_CUT, kCutMenuIcon);
   AddItemWithStringIdAndVectorIcon(this, IDC_COPY, IDS_COPY, kCopyMenuIcon);
   AddItemWithStringIdAndVectorIcon(this, IDC_PASTE, IDS_PASTE, kPasteMenuIcon);
+  }
 }
 
 class SaveAndShareSubMenuModel : public ui::SimpleMenuModel {
@@ -887,6 +889,57 @@ bool ArePromotionsEnabled() {
   return local_state && local_state->GetBoolean(prefs::kPromotionsEnabled);
 }
 
+#if BUILDFLAG(ENABLE_CEF)
+using IsVisibleCallback = base::RepeatingCallback<bool(int)>;
+
+void FilterMenuModel(ui::SimpleMenuModel* model,
+                     const IsVisibleCallback& is_visible) {
+  std::optional<size_t> last_separator;
+  size_t visible_ct = 0;
+  for (size_t i = 0; i < model->GetItemCount(); ++i) {
+    const auto type = model->GetTypeAt(i);
+    if (type == ui::MenuModel::TYPE_SEPARATOR) {
+      if (last_separator) {
+        // Remove multiple separators in a row. Prefer to remove a NORMAL
+        // separator if possible (as compared to zoom/edit controls which use
+        // UPPER/LOWER separators).
+        if (model->GetSeparatorTypeAt(*last_separator) ==
+            ui::NORMAL_SEPARATOR) {
+          model->RemoveItemAt(*last_separator);
+          i--;
+          last_separator = i;
+        } else {
+          model->RemoveItemAt(i);
+          i--;
+        }
+      } else if (visible_ct == 0) {
+        // Remove leading separator.
+        model->RemoveItemAt(i);
+        i--;
+      } else {
+        last_separator = i;
+      }
+      visible_ct = 0;
+    } else if (is_visible.Run(model->GetCommandIdAt(i))) {
+      last_separator = std::nullopt;
+      visible_ct++;
+
+      if (type == ui::MenuModel::TYPE_SUBMENU) {
+        // Filter sub-menu.
+        auto sub_model =
+            static_cast<ui::SimpleMenuModel*>(model->GetSubmenuModelAt(i));
+        FilterMenuModel(sub_model, is_visible);
+      }
+    }
+  }
+
+  if (last_separator) {
+    // Remove trailing separator.
+    model->RemoveItemAt(*last_separator);
+  }
+}
+#endif  // BUILDFLAG(ENABLE_CEF)
+
 }  // namespace
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -1866,7 +1919,7 @@ bool AppMenuModel::IsCommandIdChecked(int command_id) const {
   return false;
 }
 
-bool AppMenuModel::IsCommandIdEnabled(int command_id) const {
+bool AppMenuModel::IsCommandIdEnabledInternal(int command_id) const {
   GlobalError* error =
       GlobalErrorServiceFactory::GetForProfile(browser_->profile())
           ->GetGlobalErrorByMenuItemCommandID(command_id);
@@ -1882,6 +1935,30 @@ bool AppMenuModel::IsCommandIdEnabled(int command_id) const {
   }
 }
 
+bool AppMenuModel::IsCommandIdEnabled(int command_id) const {
+  if (!IsCommandIdEnabledInternal(command_id)) {
+    return false;
+  }
+
+#if BUILDFLAG(ENABLE_CEF)
+  if (browser_->cef_delegate()) {
+    return browser_->cef_delegate()->IsAppMenuItemEnabled(command_id);
+  }
+#endif
+
+  return true;
+}
+
+bool AppMenuModel::IsCommandIdVisible(int command_id) const {
+#if BUILDFLAG(ENABLE_CEF)
+  if (browser_->cef_delegate()) {
+    return browser_->cef_delegate()->IsAppMenuItemVisible(command_id);
+  }
+#endif
+
+  return true;
+}
+
 bool AppMenuModel::IsCommandIdAlerted(int command_id) const {
   if (command_id == IDC_VIEW_PASSWORDS ||
       command_id == IDC_SHOW_PASSWORD_MANAGER) {
@@ -2063,8 +2140,10 @@ void AppMenuModel::Build() {
                                    IDS_CLEAR_BROWSING_DATA,
                                    kTrashCanRefreshIcon);
 
+  if (IsCommandIdVisible(IDC_ZOOM_MENU)) {
   AddSeparator(ui::NORMAL_SEPARATOR);
   CreateZoomMenu();
+  }
   AddSeparator(ui::NORMAL_SEPARATOR);
 
   AddItemWithStringIdAndVectorIcon(this, IDC_PRINT, IDS_PRINT, kPrintMenuIcon);
@@ -2178,6 +2257,11 @@ void AppMenuModel::Build() {
   }
 #endif  // !BUILDFLAG(IS_CHROMEOS)
 
+#if BUILDFLAG(ENABLE_CEF)
+  FilterMenuModel(this, base::BindRepeating(&AppMenuModel::IsCommandIdVisible,
+                                            base::Unretained(this)));
+#endif
+
   uma_action_recorded_ = false;
 }
 
diff --git chrome/browser/ui/toolbar/app_menu_model.h chrome/browser/ui/toolbar/app_menu_model.h
index c961aef70c8211542ec5fa135917fadc4e2bf9ff..480998af8972d96324a374f08baf854d2c3ab9cb 100644
--- chrome/browser/ui/toolbar/app_menu_model.h
+++ chrome/browser/ui/toolbar/app_menu_model.h
@@ -238,6 +238,7 @@ class AppMenuModel : public ui::SimpleMenuModel,
   void ExecuteCommand(int command_id, int event_flags) override;
   bool IsCommandIdChecked(int command_id) const override;
   bool IsCommandIdEnabled(int command_id) const override;
+  bool IsCommandIdVisible(int command_id) const override;
   bool IsCommandIdAlerted(int command_id) const override;
   bool GetAcceleratorForCommandId(int command_id,
                                   ui::Accelerator* accelerator) const override;
@@ -277,6 +278,8 @@ class AppMenuModel : public ui::SimpleMenuModel,
   void LogSafetyHubInteractionMetrics(safety_hub::SafetyHubModuleType sh_module,
                                       int event_flags);
 
+  bool IsCommandIdEnabledInternal(int command_id) const;
+
  private:
   // Adds actionable global error menu items to the menu.
   // Examples: Extension permissions and sign in errors.
diff --git chrome/browser/ui/views/frame/browser_view.cc chrome/browser/ui/views/frame/browser_view.cc
index f55a24fe7762649f6eaf9b7d88e633f1218da72e..eed69148dcce7333769806ea0cda7de3baa12d07 100644
--- chrome/browser/ui/views/frame/browser_view.cc
+++ chrome/browser/ui/views/frame/browser_view.cc
@@ -843,13 +843,23 @@ class BrowserView::AccessibilityModeObserver : public ui::AXModeObserver {
 ///////////////////////////////////////////////////////////////////////////////
 // BrowserView, public:
 
+BrowserView::BrowserView() : BrowserView(nullptr) {}
+
 BrowserView::BrowserView(Browser* browser)
     : views::ClientView(nullptr, nullptr),
       exclusive_access_context_(
           std::make_unique<ExclusiveAccessContextImpl>(*this)),
-      browser_(browser),
       accessibility_mode_observer_(
           std::make_unique<AccessibilityModeObserver>(this)) {
+  if (browser) {
+    InitBrowser(browser);
+  }
+}
+
+void BrowserView::InitBrowser(Browser* browser) {
+  DCHECK(!browser_);
+  browser_ = browser;
+
   SetShowIcon(::ShouldShowWindowIcon(
       browser_.get(), AppUsesWindowControlsOverlay(), AppUsesTabbed()));
 
@@ -919,8 +929,15 @@ BrowserView::BrowserView(Browser* browser)
   contents_container->SetLayoutManager(std::make_unique<ContentsLayoutManager>(
       contents_view, lens_overlay_view_));
 
-  toolbar_ = top_container_->AddChildView(
-      std::make_unique<ToolbarView>(browser_.get(), this));
+  toolbar_ = OverrideCreateToolbar();
+  if (!toolbar_) {
+    toolbar_ = new ToolbarView(browser_.get(), this, std::nullopt);
+  } else {
+    browser_->set_toolbar_overridden(true);
+    // Update state that depends on the above flag.
+    browser_->command_controller()->FullscreenStateChanged();
+  }
+  top_container_->AddChildView(base::WrapUnique(toolbar_.get()));
 
   top_container_separator_ = top_container_->AddChildView(
       ContentsSeparator::CreateContentsSeparator());
@@ -1055,6 +1072,10 @@ BrowserView::BrowserView(Browser* browser)
 }
 
 BrowserView::~BrowserView() {
+  // If the Toolbar is not overloaded it will be destroyed via
+  // RemoveAllChildViews().
+  WillDestroyToolbar();
+
   // Remove the layout manager to avoid dangling. This needs to be earlier than
   // other cleanups that destroy views referenced in the layout manager.
   SetLayoutManager(nullptr);
@@ -1067,7 +1088,9 @@ BrowserView::~BrowserView() {
 
   // All the tabs should have been destroyed already. If we were closed by the
   // OS with some tabs than the BrowserNativeWidget should have destroyed them.
-  DCHECK_EQ(0, browser_->tab_strip_model()->count());
+  if (browser_) {
+    DCHECK_EQ(0, browser_->tab_strip_model()->count());
+  }
 
   // Stop the animation timer explicitly here to avoid running it in a nested
   // message loop, which may run by Browser destructor.
@@ -1165,7 +1188,7 @@ void BrowserView::SetDisableRevealerDelayForTesting(bool disable) {
   g_disable_revealer_delay_for_testing = disable;
 }
 
-gfx::Rect BrowserView::GetFindBarBoundingBox() const {
+gfx::Rect BrowserView::GetFindBarBoundingBoxImpl() const {
   gfx::Rect contents_bounds = contents_container_->ConvertRectToWidget(
       contents_container_->GetLocalBounds());
 
@@ -1187,6 +1210,16 @@ gfx::Rect BrowserView::GetFindBarBoundingBox() const {
   return contents_container_->GetMirroredRect(contents_bounds);
 }
 
+gfx::Rect BrowserView::GetFindBarBoundingBox() const {
+  auto bounds = GetFindBarBoundingBoxImpl();
+#if BUILDFLAG(ENABLE_CEF)
+  if (browser() && browser()->cef_delegate()) {
+    browser()->cef_delegate()->UpdateFindBarBoundingBox(&bounds);
+  }
+#endif
+  return bounds;
+}
+
 int BrowserView::GetTabStripHeight() const {
   // We want to return tabstrip_->height(), but we might be called in the midst
   // of layout, when that hasn't yet been updated to reflect the current state.
@@ -1743,6 +1776,28 @@ gfx::Point BrowserView::GetThemeOffsetFromBrowserView() const {
       ThemeProperties::kFrameHeightAboveTabs - browser_view_origin.y());
 }
 
+void BrowserView::WillDestroyToolbar() {
+  // Reset autofill bubble handler to make sure it does not out-live toolbar,
+  // since it is responsible for showing autofill related bubbles from toolbar's
+  // child views and it is an observer for avatar toolbar button if any.
+  autofill_bubble_handler_.reset();
+
+  toolbar_button_provider_ = nullptr;
+  if (GetBrowserViewLayout()) {
+    GetBrowserViewLayout()->reset_toolbar();
+  }
+
+  if (toolbar_ && toolbar_->parent()) {
+    // Remove now instead of waiting for RemoveAllChildViews(), as there is
+    // otherwise no guarantee that the Toolbar will be removed before the
+    // BrowserView is removed (and destroyed).
+    toolbar_->parent()->RemoveChildView(toolbar_);
+    toolbar_.ClearAndDelete();
+  } else {
+    toolbar_ = nullptr;
+  }
+}
+
 bool BrowserView::IsLoadingAnimationRunning() const {
   return loading_animation_timer_.IsRunning();
 }
@@ -2073,7 +2128,11 @@ void BrowserView::Restore() {
 
 bool BrowserView::ShouldHideUIForFullscreen() const {
   // Immersive mode needs UI for the slide-down top panel.
-  if (ImmersiveModeController::From(browser())->IsEnabled()) {
+  // Avoid callback into |immersive_mode_controller| during construction.
+  // See CEF issue #3527.
+  auto* const immersive_mode_controller =
+      ImmersiveModeController::From(browser());
+  if (immersive_mode_controller && immersive_mode_controller->IsEnabled()) {
     return false;
   }
 
@@ -3253,6 +3312,9 @@ void BrowserView::ShowOneClickSigninConfirmation(
 
 DownloadBubbleUIController* BrowserView::GetDownloadBubbleUIController() {
 #if !BUILDFLAG(IS_CHROMEOS)
+  if (!browser_) {
+    return nullptr;
+  }
   if (auto* download_controller =
           browser_->GetFeatures().download_toolbar_ui_controller()) {
     return download_controller->bubble_controller();
@@ -4367,9 +4429,10 @@ bool BrowserView::GetSavedWindowPlacement(
 views::View* BrowserView::GetContentsView() {
   if (multi_contents_view_) {
     return multi_contents_view_->GetActiveContentsView();
-  } else {
+  } else if (contents_container_view_) {
     return contents_container_view_->contents_view();
   }
+  return nullptr;
 }
 
 views::ClientView* BrowserView::CreateClientView(views::Widget* widget) {
@@ -4548,9 +4611,22 @@ void BrowserView::DeleteBrowserWindow() {
   // before the frame is destroyed to mitigate UAF risk.
   browser_widget_->SaveWindowPlacementIfNeeded();
 
-  browser_widget_.reset();
-  // BrowserWidget owns BrowserView in its views::View hierarchy and `this` will
-  // not be valid after this returns.
+  const bool views_hosted =
+#if BUILDFLAG(ENABLE_CEF)
+      browser() && browser()->cef_delegate() &&
+      browser()->cef_delegate()->IsViewsHosted();
+#else
+      false;
+#endif
+
+  if (views_hosted) {
+    // The BrowserWidget is owned by CefWindowWidgetDelegate.
+    browser_widget_.release();
+  } else {
+    browser_widget_.reset();
+    // BrowserWidget owns BrowserView in its views::View hierarchy and `this`
+    // will not be valid after this returns.
+  }
 }
 
 void BrowserView::SetForceShowBookmarkBarFlag(
@@ -4865,11 +4941,36 @@ void BrowserView::GetAccessiblePanes(std::vector<views::View*>* panes) {
 bool BrowserView::ShouldDescendIntoChildForEventHandling(
     gfx::NativeView child,
     const gfx::Point& location) {
+#if BUILDFLAG(ENABLE_CEF)
+  const bool frameless_pip =
+      GetIsPictureInPictureType() &&
+      !browser_->SupportsWindowFeature(Browser::WindowFeature::kFeatureTitleBar);
+  if (frameless_pip) {
+    int result = GetFrameView()->NonClientHitTest(location);
+    if (result == HTTOP || result == HTTOPLEFT || result == HTTOPRIGHT) {
+      // Allow resize from the top of a frameless window.
+      return false;
+    }
+  }
+#endif
+
+  std::optional<SkRegion> draggable_region;
+
   // Window for PWAs with window-controls-overlay display override should claim
   // mouse events that fall within the draggable region.
   web_app::AppBrowserController* controller = browser()->app_controller();
-  if (AreDraggableRegionsEnabled() && controller &&
-      controller->draggable_region().has_value()) {
+  if (AreDraggableRegionsEnabled() && controller) {
+    draggable_region = controller->draggable_region();
+  }
+
+#if BUILDFLAG(ENABLE_CEF)
+  // Match logic in PictureInPictureBrowserFrameView::NonClientHitTest.
+  if (!draggable_region.has_value() && frameless_pip) {
+    draggable_region = browser_->cef_delegate()->GetDraggableRegion();
+  }
+#endif
+
+  if (draggable_region.has_value()) {
     // Draggable regions are defined relative to the web contents.
     gfx::Point point_in_contents_web_view_coords(location);
     views::View::ConvertPointToTarget(GetWidget()->GetRootView(),
@@ -4878,9 +4979,8 @@ bool BrowserView::ShouldDescendIntoChildForEventHandling(
 
     // Draggable regions should be ignored for clicks into any browser view's
     // owned widgets, for example alerts, permission prompts or find bar.
-    return !controller->draggable_region()->contains(
-               point_in_contents_web_view_coords.x(),
-               point_in_contents_web_view_coords.y()) ||
+    return !draggable_region->contains(point_in_contents_web_view_coords.x(),
+                                       point_in_contents_web_view_coords.y()) ||
            WidgetOwnedByAnchorContainsPoint(point_in_contents_web_view_coords);
   }
 
@@ -5134,7 +5234,9 @@ void BrowserView::Layout(PassKey) {
 
   // TODO(jamescook): Why was this in the middle of layout code?
   toolbar_->location_bar()->omnibox_view()->SetFocusBehavior(
-      IsToolbarVisible() ? FocusBehavior::ALWAYS : FocusBehavior::NEVER);
+      (IsToolbarVisible() || browser_->toolbar_overridden())
+          ? FocusBehavior::ALWAYS
+          : FocusBehavior::NEVER);
   GetFrameView()->UpdateMinimumSize();
 
   // Some of the situations when the BrowserView is laid out are:
@@ -5203,6 +5305,12 @@ void BrowserView::AddedToWidget() {
   SetThemeProfileForWindow(GetNativeWindow(), browser_->GetProfile());
 #endif
 
+  // This browser view may already have a custom button provider set (e.g the
+  // hosted app frame).
+  if (!toolbar_button_provider_) {
+    SetToolbarButtonProvider(toolbar_);
+  }
+
   toolbar_->Init();
 
   UpdateTabSearchBubbleHost();
@@ -5301,12 +5409,6 @@ void BrowserView::AddedToWidget() {
 
   EnsureFocusOrder();
 
-  // This browser view may already have a custom button provider set (e.g the
-  // hosted app frame).
-  if (!toolbar_button_provider_) {
-    SetToolbarButtonProvider(toolbar_);
-  }
-
 #if !BUILDFLAG(IS_CHROMEOS)
   browser_->GetFeatures().download_toolbar_ui_controller()->Init();
 #endif
@@ -5340,6 +5442,13 @@ void BrowserView::AddedToWidget() {
                 weak_ptr_factory_.GetWeakPtr()));
   }
 
+  if (browser_->window()) {
+    // Initialize the browser features that rely on the browser window now that
+    // it is initialized. This will instead be called from the Browser
+    // constructor for default Chrome UI windows.
+    browser_->GetFeatures().InitPostWindowConstruction(browser_.get());
+  }
+
   initialized_ = true;
 }
 
diff --git chrome/browser/ui/views/frame/browser_view.h chrome/browser/ui/views/frame/browser_view.h
index d8334d0355229d3ba5cb798053e74f65c00627aa..b2ac175f5e2b0a14dc7c751ab850a070bde217cc 100644
--- chrome/browser/ui/views/frame/browser_view.h
+++ chrome/browser/ui/views/frame/browser_view.h
@@ -141,7 +141,9 @@ class BrowserView : public BrowserWindow,
   // locate this object using just the handle.
   static constexpr char kBrowserViewKey[] = "__BROWSER_VIEW__";
 
+  BrowserView();
   explicit BrowserView(Browser* browser);
+  void InitBrowser(Browser* browser);
   BrowserView(const BrowserView&) = delete;
   BrowserView& operator=(const BrowserView&) = delete;
   ~BrowserView() override;
@@ -858,10 +860,16 @@ class BrowserView : public BrowserWindow,
   bool IsLockedFullscreen() const;
 #endif
 
- protected:
+  // Called during Toolbar destruction to remove dependent objects that have
+  // dangling references.
+  virtual void WillDestroyToolbar();
+
   // BrowserWindow:
   void DeleteBrowserWindow() final;
 
+ protected:
+  virtual ToolbarView* OverrideCreateToolbar() { return nullptr; }
+
  private:
   // Do not friend BrowserViewLayout. Use the BrowserViewLayoutDelegate
   // interface to keep these two classes decoupled and testable.
@@ -1137,11 +1145,13 @@ class BrowserView : public BrowserWindow,
   class ExclusiveAccessContextImpl;
   std::unique_ptr<ExclusiveAccessContextImpl> exclusive_access_context_;
 
+  gfx::Rect GetFindBarBoundingBoxImpl() const;
+
   // The BrowserWidget that owns this view.
   std::unique_ptr<BrowserWidget> browser_widget_;
 
   // The owning Browser object. `browser_` will outlive this.
-  const raw_ptr<Browser> browser_;
+  raw_ptr<Browser> browser_;
 
   base::CallbackListSubscription chip_visibility_subscription_;
 
diff --git chrome/browser/ui/views/frame/browser_widget.cc chrome/browser/ui/views/frame/browser_widget.cc
index 67e932d5bb4fe0a9eeac01bf84e57ab24027f65c..db60982ca0fbdf19b4bb2a0e191be94666837bbe 100644
--- chrome/browser/ui/views/frame/browser_widget.cc
+++ chrome/browser/ui/views/frame/browser_widget.cc
@@ -112,14 +112,27 @@ ui::ColorProviderKey::SchemeVariant GetSchemeVariant(
 ////////////////////////////////////////////////////////////////////////////////
 // BrowserWidget, public:
 
+BrowserWidget::BrowserWidget() : BrowserWidget(nullptr) {}
+
 BrowserWidget::BrowserWidget(BrowserView* browser_view)
     : browser_native_widget_(nullptr),
       root_view_(nullptr),
       browser_frame_view_(nullptr),
-      browser_view_(browser_view) {
+      browser_view_(nullptr) {
   set_is_secondary_widget(false);
   // Don't focus anything on creation, selecting a tab will set the focus.
   set_focus_on_creation(false);
+  if (browser_view) {
+    SetBrowserView(browser_view);
+  }
+}
+
+void BrowserWidget::SetBrowserFrameView(BrowserFrameView* browser_frame_view) {
+  browser_frame_view_ = browser_frame_view;
+}
+
+void BrowserWidget::SetBrowserView(BrowserView* browser_view) {
+  browser_view_ = browser_view;
 }
 
 BrowserWidget::~BrowserWidget() {
@@ -140,7 +153,11 @@ BrowserWidget::~BrowserWidget() {
   // Do this here and not in ~BrowserView() as BrowserWindowFeatures may attempt
   // to read state on the BrowserWidget as they undergo destruction, and
   // BrowserWidget state is destroyed at the end of this scope.
-  browser_view_->browser()->GetFeatures().TearDownPreBrowserWindowDestruction();
+  if (browser_view_ && browser_view_->browser()) {
+    browser_view_->browser()
+        ->GetFeatures()
+        .TearDownPreBrowserWindowDestruction();
+  }
 }
 
 void BrowserWidget::InitBrowserWidget() {
@@ -229,6 +246,8 @@ void BrowserWidget::GetWindowPlacement(
     ui::mojom::WindowShowState* show_state) const {
   if (browser_native_widget_) {
     browser_native_widget_->GetWindowPlacement(bounds, show_state);
+  } else {
+    *show_state = ui::mojom::WindowShowState::kDefault;
   }
 }
 
@@ -339,6 +358,9 @@ ui::ColorProviderKey::ThemeInitializerSupplier* BrowserWidget::GetCustomTheme()
 }
 
 void BrowserWidget::OnNativeWidgetWorkspaceChanged() {
+  if (!browser_view_) {
+    return;
+  }
   chrome::SaveWindowWorkspace(browser_view_->browser(), GetWorkspace());
   chrome::SaveWindowVisibleOnAllWorkspaces(browser_view_->browser(),
                                            IsVisibleOnAllWorkspaces());
@@ -544,6 +566,13 @@ void BrowserWidget::SelectNativeTheme() {
     return;
   }
 
+  // Always use the NativeTheme for forced color modes.
+  if (ui::NativeTheme::IsForcedDarkMode() ||
+      ui::NativeTheme::IsForcedLightMode()) {
+    SetNativeTheme(native_theme);
+    return;
+  }
+
   // Ignore the system theme for web apps with window-controls-overlay as the
   // display_override so the web contents can blend with the overlay by using
   // the developer-provided theme color for a better experience. Context:
@@ -560,15 +589,17 @@ void BrowserWidget::SelectNativeTheme() {
 void BrowserWidget::OnTouchUiChanged() {
   client_view()->InvalidateLayout();
 
-  // For standard browser frame, if we do not invalidate the FrameView
-  // the client window bounds will not be properly updated which could cause
-  // visual artifacts. See crbug.com/1035959 for details.
-  if (non_client_view()->frame_view()) {
-    // Note that invalidating a view invalidates all of its ancestors, so it is
-    // not necessary to also invalidate the NonClientView or RootView here.
-    non_client_view()->frame_view()->InvalidateLayout();
-  } else {
-    non_client_view()->InvalidateLayout();
+  if (non_client_view()) {
+    // For standard browser frame, if we do not invalidate the FrameView
+    // the client window bounds will not be properly updated which could cause
+    // visual artifacts. See crbug.com/1035959 for details.
+    if (non_client_view()->frame_view()) {
+      // Note that invalidating a view invalidates all of its ancestors, so it
+      // is not necessary to also invalidate the NonClientView or RootView here.
+      non_client_view()->frame_view()->InvalidateLayout();
+    } else {
+      non_client_view()->InvalidateLayout();
+    }
   }
   GetRootView()->InvalidateLayout();
 }
@@ -609,5 +640,8 @@ bool BrowserWidget::RegenerateFrameOnThemeChange(
 }
 
 bool BrowserWidget::IsIncognitoBrowser() const {
+  if (!browser_view_) {
+    return true;
+  }
   return browser_view_->browser()->profile()->IsIncognitoProfile();
 }
diff --git chrome/browser/ui/views/frame/browser_widget.h chrome/browser/ui/views/frame/browser_widget.h
index 470e63b9352a1641cab22f416810eec7cbc03ad9..9c22ace412038b437160bd691c395decba9b94e0 100644
--- chrome/browser/ui/views/frame/browser_widget.h
+++ chrome/browser/ui/views/frame/browser_widget.h
@@ -58,6 +58,7 @@ enum class TabDragKind {
 class BrowserWidget : public views::Widget,
                       public views::ContextMenuController {
  public:
+  BrowserWidget();
   explicit BrowserWidget(BrowserView* browser_view);
 
   BrowserWidget(const BrowserWidget&) = delete;
@@ -98,7 +99,7 @@ class BrowserWidget : public views::Widget,
 
   // ThemeService calls this when a user has changed their theme, indicating
   // that it's time to redraw everything.
-  void UserChangedTheme(BrowserThemeChangeType theme_change_type);
+  virtual void UserChangedTheme(BrowserThemeChangeType theme_change_type);
 
   // views::Widget:
   views::internal::RootView* CreateRootView() override;
@@ -132,22 +133,27 @@ class BrowserWidget : public views::Widget,
   void SetTabDragKind(TabDragKind tab_drag_kind);
   TabDragKind tab_drag_kind() const { return tab_drag_kind_; }
 
+  BrowserView* browser_view() const { return browser_view_.get(); }
+
  protected:
+  void SetBrowserFrameView(BrowserFrameView* browser_frame_view);
+  void SetBrowserView(BrowserView* browser_view);
+
   // views::Widget:
   void OnNativeThemeUpdated(ui::NativeTheme* observed_theme) override;
   ui::ColorProviderKey GetColorProviderKey() const override;
 
+  // Select a native theme that is appropriate for the current context. This is
+  // currently only needed for Linux to switch between the regular NativeTheme
+  // and the GTK NativeTheme instance.
+  void SelectNativeTheme();
+
  private:
   void OnTouchUiChanged();
 
   // Callback for MenuRunner.
   void OnMenuClosed();
 
-  // Select a native theme that is appropriate for the current context. This is
-  // currently only needed for Linux to switch between the regular NativeTheme
-  // and the GTK NativeTheme instance.
-  void SelectNativeTheme();
-
   // Regenerate the frame on theme change if necessary. Returns true if
   // regenerated.
   bool RegenerateFrameOnThemeChange(BrowserThemeChangeType theme_change_type);
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout.cc chrome/browser/ui/views/frame/layout/browser_view_layout.cc
index 01f6c81406ad2c742258e75004662734e55cad95..5c6d1fdad834c3fee5a5d8dd8173b1850fef9611 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout.cc
+++ chrome/browser/ui/views/frame/layout/browser_view_layout.cc
@@ -23,6 +23,10 @@
 #include "ui/views/widget/widget.h"
 #include "ui/views/widget/widget_observer.h"
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/views/chrome_views_util.h"
+#endif
+
 using web_modal::ModalDialogHostObserver;
 using web_modal::WebContentsModalDialogHost;
 
@@ -56,6 +60,8 @@ class BrowserViewLayout::BrowserModalDialogHostViews
     observer_list_.Notify(&ModalDialogHostObserver::OnHostDestroying);
   }
 
+  bool HasObservers() const { return !observer_list_.empty(); }
+
   void NotifyPositionRequiresUpdate() {
     observer_list_.Notify(&ModalDialogHostObserver::OnPositionRequiresUpdate);
   }
@@ -200,18 +206,21 @@ void BrowserViewLayout::UpdateBubbles() {
     exclusive_access_bubble->RepositionIfVisible();
   }
 
-  // Adjust any hosted dialogs if the browser's dialog hosting bounds changed.
-  const gfx::Rect dialog_bounds(dialog_host_->GetDialogPosition(gfx::Size()),
-                                dialog_host_->GetMaximumDialogSize());
-  const gfx::Rect host_widget_bounds =
-      dialog_host_->GetHostWidget()
-          ? dialog_host_->GetHostWidget()->GetClientAreaBoundsInScreen()
-          : gfx::Rect();
-  const gfx::Rect dialog_bounds_in_screen =
-      dialog_bounds + host_widget_bounds.OffsetFromOrigin();
-  if (latest_dialog_bounds_in_screen_ != dialog_bounds_in_screen) {
-    latest_dialog_bounds_in_screen_ = dialog_bounds_in_screen;
-    dialog_host_->NotifyPositionRequiresUpdate();
+  // Avoid unnecessary calls to UpdateDialogTopInsetInBrowserView().
+  if (dialog_host_->HasObservers()) {
+    // Adjust any hosted dialogs if the browser's dialog hosting bounds changed.
+    const gfx::Rect dialog_bounds(dialog_host_->GetDialogPosition(gfx::Size()),
+                                  dialog_host_->GetMaximumDialogSize());
+    const gfx::Rect host_widget_bounds =
+        dialog_host_->GetHostWidget()
+            ? dialog_host_->GetHostWidget()->GetClientAreaBoundsInScreen()
+            : gfx::Rect();
+    const gfx::Rect dialog_bounds_in_screen =
+        dialog_bounds + host_widget_bounds.OffsetFromOrigin();
+    if (latest_dialog_bounds_in_screen_ != dialog_bounds_in_screen) {
+      latest_dialog_bounds_in_screen_ = dialog_bounds_in_screen;
+      dialog_host_->NotifyPositionRequiresUpdate();
+    }
   }
 }
 
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout.h chrome/browser/ui/views/frame/layout/browser_view_layout.h
index c475f825745d9b2364acf1ee934b61547e7b7009..2c6f4eed484d5988f4c7dcd5c36430f71a9bcd4b 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout.h
+++ chrome/browser/ui/views/frame/layout/browser_view_layout.h
@@ -134,6 +134,8 @@ class BrowserViewLayout : public views::LayoutManager {
     return views_.side_panel_animation_content;
   }
 
+  void reset_toolbar() { views_.toolbar = nullptr; }
+
   // views::LayoutManager overrides:
   gfx::Size GetPreferredSize(
       const views::View* host,
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout_delegate.h chrome/browser/ui/views/frame/layout/browser_view_layout_delegate.h
index fb4278538ec18077f6738e37c847605d0ecbf4c9..ee44a7dcd49952c602b9df4374c2a22f59582224 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout_delegate.h
+++ chrome/browser/ui/views/frame/layout/browser_view_layout_delegate.h
@@ -40,6 +40,7 @@ class BrowserViewLayoutDelegate {
       const gfx::Rect& available_space,
       views::Label& window_title_label) const = 0;
   virtual int GetTopInsetInBrowserView() const = 0;
+  virtual void UpdateDialogTopInsetInBrowserView(int* dialog_top_y) const = 0;
   virtual bool IsToolbarVisible() const = 0;
   virtual bool IsBookmarkBarVisible() const = 0;
   virtual bool IsInfobarVisible() const = 0;
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.cc chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.cc
index 5fdc6824d1dfe8b1d8c3fe13dd4b825067045f75..7bc3b123b22c52bfd2e5183d13885d99ca961a1a 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.cc
+++ chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.cc
@@ -90,6 +90,15 @@ int BrowserViewLayoutDelegateImpl::GetTopInsetInBrowserView() const {
          browser_view_->y();
 }
 
+void BrowserViewLayoutDelegateImpl::UpdateDialogTopInsetInBrowserView(
+    int* dialog_top_y) const {
+#if BUILDFLAG(ENABLE_CEF)
+  if (auto cef_delegate = browser_view_->browser()->cef_delegate()) {
+    cef_delegate->UpdateDialogTopInset(dialog_top_y);
+  }
+#endif
+}
+
 void BrowserViewLayoutDelegateImpl::LayoutWebAppWindowTitle(
     const gfx::Rect& available_space,
     views::Label& window_title_label) const {
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.h chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.h
index 346b09fb010fc354e03facfb3cb4dc749036a85f..e50686d427571e48bf32f7bb16f2a48b91a82c76 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.h
+++ chrome/browser/ui/views/frame/layout/browser_view_layout_delegate_impl.h
@@ -31,6 +31,7 @@ class BrowserViewLayoutDelegateImpl : public BrowserViewLayoutDelegate {
   BrowserLayoutParams GetBrowserLayoutParams(
       bool use_browser_bounds) const override;
   int GetTopInsetInBrowserView() const override;
+  void UpdateDialogTopInsetInBrowserView(int* dialog_top_y) const override;
   void LayoutWebAppWindowTitle(const gfx::Rect& available_space,
                                views::Label& window_title_label) const override;
   bool IsToolbarVisible() const override;
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.cc chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.cc
index ad51c0f5dfcd94ed5e7039271f15575739c3ae01..242b84dd67f73aa50c21f34634137266510fa3fb 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.cc
+++ chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.cc
@@ -29,6 +29,10 @@
 #include "chrome/browser/ui/fullscreen_util_mac.h"
 #endif
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/views/chrome_views_util.h"
+#endif
+
 namespace {
 
 // The number of pixels the constrained window should overlap the bottom
@@ -439,6 +443,15 @@ void BrowserViewLayoutImplOld::LayoutWebUITabStrip(
 
 void BrowserViewLayoutImplOld::LayoutToolbar(gfx::Rect& available_bounds) {
   TRACE_EVENT0("ui", "BrowserViewLayout::LayoutToolbar");
+
+#if BUILDFLAG(ENABLE_CEF)
+  if (cef::IsCefView(views().toolbar)) {
+    // CEF may take ownership of the toolbar. Early exit to avoid the DCHECK
+    // in LayoutManager::SetViewVisibility().
+    return;
+  }
+#endif
+
   bool toolbar_visible = delegate().IsToolbarVisible();
   SetViewVisibility(views().toolbar, toolbar_visible);
 
@@ -737,7 +750,7 @@ gfx::Point BrowserViewLayoutImplOld::GetDialogPosition(
   }
   const int middle_x =
       leading_x + layout_result.contents_container_bounds.width() / 2;
-  return gfx::Point(middle_x - dialog_size.width() / 2, dialog_top_y_);
+  return gfx::Point(middle_x - dialog_size.width() / 2, GetDialogTopY());
 }
 
 gfx::Size BrowserViewLayoutImplOld::GetMaximumDialogSize() const {
@@ -751,6 +764,12 @@ gfx::Size BrowserViewLayoutImplOld::GetMaximumDialogSize() const {
   // universally.
   views::View* view = views().contents_container;
   gfx::Rect content_area = view->ConvertRectToWidget(view->GetLocalBounds());
-  const int top = dialog_top_y_;
+  const int top = GetDialogTopY();
   return gfx::Size(content_area.width(), content_area.bottom() - top);
 }
+
+int BrowserViewLayoutImplOld::GetDialogTopY() const {
+  int dialog_top_y = dialog_top_y_;
+  delegate().UpdateDialogTopInsetInBrowserView(&dialog_top_y);
+  return dialog_top_y;
+}
diff --git chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.h chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.h
index d9cfc2e6d87ef4547884cceedfd1a1014a1aac0f..3189d0fc4d8497a58dacea1fca6de9383e56962d 100644
--- chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.h
+++ chrome/browser/ui/views/frame/layout/browser_view_layout_impl_old.h
@@ -71,6 +71,8 @@ class BrowserViewLayoutImplOld : public BrowserViewLayout {
 
   bool IsImmersiveModeEnabledWithoutToolbar() const;
 
+  int GetDialogTopY() const;
+
   // Whether or not to use the browser based content minimum size.
   const bool use_browser_content_minimum_size_ = false;
 
diff --git chrome/browser/ui/views/frame/picture_in_picture_browser_frame_view.cc chrome/browser/ui/views/frame/picture_in_picture_browser_frame_view.cc
index e82beb006a06eef0e454f8cf4d8c974c87d01428..3c4e212eaf3faca9e13b2abeccc0c6eec5dbef7f 100644
--- chrome/browser/ui/views/frame/picture_in_picture_browser_frame_view.cc
+++ chrome/browser/ui/views/frame/picture_in_picture_browser_frame_view.cc
@@ -711,6 +711,11 @@ PictureInPictureBrowserFrameView::PictureInPictureBrowserFrameView(
       IsOverlayViewVisible()) {
     PictureInPictureWindowManager::GetInstance()->ClearCachedBounds();
   }
+
+  if (!browser_view->browser()->SupportsWindowFeature(
+          Browser::WindowFeature::kFeatureTitleBar)) {
+    top_bar_container_view_->SetVisible(false);
+  }
 }
 
 PictureInPictureBrowserFrameView::~PictureInPictureBrowserFrameView() {
@@ -826,18 +831,42 @@ gfx::Rect PictureInPictureBrowserFrameView::GetWindowBoundsForClientBounds(
 
 int PictureInPictureBrowserFrameView::NonClientHitTest(
     const gfx::Point& point) {
-  // Allow interacting with the buttons.
-  if (GetLocationIconViewBounds().Contains(point) ||
-      GetBackToTabControlsBounds().Contains(point) ||
-      GetCloseControlsBounds().Contains(point)) {
-    return HTCLIENT;
+  const bool frameless = !top_bar_container_view_->GetVisible();
+  if (!frameless) {
+    // Allow interacting with the buttons.
+    if (GetLocationIconViewBounds().Contains(point) ||
+        GetBackToTabControlsBounds().Contains(point) ||
+        GetCloseControlsBounds().Contains(point)) {
+      return HTCLIENT;
+    }
+
+    for (size_t i = 0; i < content_setting_views_.size(); i++) {
+      if (GetContentSettingViewBounds(i).Contains(point)) {
+        return HTCLIENT;
+      }
+    }
   }
 
-  for (size_t i = 0; i < content_setting_views_.size(); i++) {
-    if (GetContentSettingViewBounds(i).Contains(point)) {
-      return HTCLIENT;
+#if BUILDFLAG(ENABLE_CEF)
+  if (frameless) {
+    // Match logic in BrowserView::ShouldDescendIntoChildForEventHandling.
+    const auto draggable_region =
+        browser_view()->browser()->cef_delegate()->GetDraggableRegion();
+    if (draggable_region.has_value()) {
+      // Draggable regions are defined relative to the web contents.
+      gfx::Point point_in_contents_web_view_coords(point);
+      views::View::ConvertPointToTarget(GetWidget()->GetRootView(),
+                                        browser_view()->contents_web_view(),
+                                        &point_in_contents_web_view_coords);
+
+      if (draggable_region->contains(
+              point_in_contents_web_view_coords.x(),
+              point_in_contents_web_view_coords.y())) {
+        return HTCAPTION;
+      }
     }
   }
+#endif  // BUILDFLAG(ENABLE_CEF)
 
   // Allow dragging and resizing the window.
   int window_component = GetHTComponentForFrame(
@@ -902,7 +931,8 @@ void PictureInPictureBrowserFrameView::Layout(PassKey) {
   gfx::Rect content_area = GetLocalBounds();
   content_area.Inset(FrameBorderInsets());
   gfx::Rect top_bar = content_area;
-  top_bar.set_height(kTopControlsHeight);
+  top_bar.set_height(
+      top_bar_container_view_->GetVisible() ? kTopControlsHeight : 0);
   top_bar_container_view_->SetBoundsRect(top_bar);
 #if !BUILDFLAG(IS_ANDROID)
   if (auto_pip_setting_overlay_) {
@@ -1482,7 +1512,8 @@ gfx::Insets PictureInPictureBrowserFrameView::FrameBorderInsets() const {
 }
 
 int PictureInPictureBrowserFrameView::GetTopAreaHeight() const {
-  return FrameBorderInsets().top() + kTopControlsHeight;
+  return FrameBorderInsets().top() +
+         (top_bar_container_view_->GetVisible() ? kTopControlsHeight : 0);
 }
 
 gfx::Size PictureInPictureBrowserFrameView::GetNonClientViewAreaSize() const {
diff --git chrome/browser/ui/views/frame/shadow_overlay_view.cc chrome/browser/ui/views/frame/shadow_overlay_view.cc
index 1713dcf489e01f35143ff4cc0d574cd64612fd97..b7cac9c6b3429fe2da329e481a1a2b0fce9b6aaf 100644
--- chrome/browser/ui/views/frame/shadow_overlay_view.cc
+++ chrome/browser/ui/views/frame/shadow_overlay_view.cc
@@ -220,6 +220,9 @@ void ShadowOverlayView::VisibilityChanged(View* starting_from, bool visible) {
 }
 
 void ShadowOverlayView::AddedToWidget() {
+  if (!browser_view_->toolbar_height_side_panel()) {
+    return;
+  }
   side_panel_observer_.Observe(browser_view_->toolbar_height_side_panel());
   side_panel_observer_.GetSource()->animation_coordinator()->AddObserver(
       kShadowOverlayOpacityAnimation, this);
diff --git chrome/browser/ui/views/omnibox/omnibox_popup_closer.cc chrome/browser/ui/views/omnibox/omnibox_popup_closer.cc
index a87e6825164ed9a0a3737913b7d58366e26733a2..81d394d519556d3add7c649b16cb536ed3c318d8 100644
--- chrome/browser/ui/views/omnibox/omnibox_popup_closer.cc
+++ chrome/browser/ui/views/omnibox/omnibox_popup_closer.cc
@@ -50,7 +50,8 @@ OmniboxPopupCloser::~OmniboxPopupCloser() = default;
 
 void OmniboxPopupCloser::OnMouseEvent(ui::MouseEvent* event) {
   // Close the omnibox popup if the click is outside the omnibox view.
-  if (!browser_view_->browser()->is_delete_scheduled() &&
+  if (browser_view_->browser() &&
+      !browser_view_->browser()->is_delete_scheduled() &&
       event->type() == ui::EventType::kMousePressed) {
     LocationBarView* location_bar_view = browser_view_->GetLocationBarView();
     CHECK(location_bar_view);
diff --git chrome/browser/ui/views/page_action/page_action_icon_controller.cc chrome/browser/ui/views/page_action/page_action_icon_controller.cc
index 796b619f0412cdc713ab33fed97af751e3390d7a..00fa5326bfc57e2feb5eb826484f5fcec2d7e197 100644
--- chrome/browser/ui/views/page_action/page_action_icon_controller.cc
+++ chrome/browser/ui/views/page_action/page_action_icon_controller.cc
@@ -109,6 +109,14 @@ void PageActionIconController::Init(const PageActionIconParams& params,
     if (IsPageActionMigrated(type)) {
       continue;
     }
+
+#if BUILDFLAG(ENABLE_CEF)
+    if (params.browser && params.browser->cef_delegate() &&
+        !params.browser->cef_delegate()->IsPageActionIconVisible(type)) {
+      continue;
+    }
+#endif
+
     switch (type) {
       case PageActionIconType::kPaymentsOfferNotification:
         add_page_action_icon(
diff --git chrome/browser/ui/views/toolbar/toolbar_view.cc chrome/browser/ui/views/toolbar/toolbar_view.cc
index 78b7dfb77ab164f035ede4152ebc2a818a30df94..1f72691fe8779ef0c21b7aa6245a8353198ca175 100644
--- chrome/browser/ui/views/toolbar/toolbar_view.cc
+++ chrome/browser/ui/views/toolbar/toolbar_view.cc
@@ -216,12 +216,14 @@ END_METADATA
 DEFINE_CLASS_ELEMENT_IDENTIFIER_VALUE(ToolbarView, kToolbarElementId);
 DEFINE_CLASS_ELEMENT_IDENTIFIER_VALUE(ToolbarView, kToolbarContainerElementId);
 
-ToolbarView::ToolbarView(Browser* browser, BrowserView* browser_view)
+ToolbarView::ToolbarView(Browser* browser,
+                         BrowserView* browser_view,
+                         std::optional<DisplayMode> display_mode)
     : AnimationDelegateViews(this),
       browser_(browser),
       browser_view_(browser_view),
       app_menu_icon_controller_(browser->profile(), this),
-      display_mode_(GetDisplayMode(browser)) {
+      display_mode_(display_mode ? *display_mode : GetDisplayMode(browser)) {
   SetID(VIEW_ID_TOOLBAR);
   SetProperty(views::kElementIdentifierKey, kToolbarElementId);
 
@@ -256,9 +258,24 @@ ToolbarView::~ToolbarView() {
   if (browser_view_->GetSupportsTabStrip()) {
     browser()->GetTabStripModel()->RemoveObserver(this);
   }
+
+  browser_view_->WillDestroyToolbar();
 }
 
 void ToolbarView::Init() {
+#if BUILDFLAG(ENABLE_CEF)
+  using ToolbarButtonType = cef::BrowserDelegate::ToolbarButtonType;
+  auto button_visible = [this](ToolbarButtonType type) {
+    if (this->browser_->cef_delegate()) {
+      return this->browser_->cef_delegate()->IsToolbarButtonVisible(type);
+    }
+    return true;
+  };
+#define BUTTON_VISIBLE(type) button_visible(ToolbarButtonType::type)
+#else
+#define BUTTON_VISIBLE(type) true
+#endif
+
 #if defined(USE_AURA)
   // Avoid generating too many occlusion tracking calculation events before this
   // function returns. The occlusion status will be computed only once once this
@@ -281,7 +298,7 @@ void ToolbarView::Init() {
 
   auto location_bar = std::make_unique<LocationBarView>(
       browser_, browser_->profile(), browser_->command_controller(), this,
-      display_mode_ != DisplayMode::kNormal);
+      display_mode_ != DisplayMode::kNormal && !browser_->toolbar_overridden());
   // Make sure the toolbar shows by default.
   size_animation_.Reset(1);
 
@@ -351,7 +368,8 @@ void ToolbarView::Init() {
   }
 
   std::unique_ptr<MediaToolbarButtonView> media_button;
-  if (base::FeatureList::IsEnabled(media::kGlobalMediaControls)) {
+  if (base::FeatureList::IsEnabled(media::kGlobalMediaControls) &&
+      BUTTON_VISIBLE(kMedia)) {
     media_button = std::make_unique<MediaToolbarButtonView>(
         browser_view_,
         std::make_unique<MediaToolbarButtonContextualMenu>(browser_));
@@ -409,7 +427,7 @@ void ToolbarView::Init() {
   pinned_toolbar_actions_container_ = container_view_->AddChildView(
       std::make_unique<PinnedToolbarActionsContainer>(browser_view_, this));
 
-  if (features::HasTabSearchToolbarButton()) {
+  if (features::HasTabSearchToolbarButton() && BUTTON_VISIBLE(kTabSearch)) {
     tab_search_button_ =
         pinned_toolbar_actions_container()->CreatePermanentButtonFor(
             kActionTabSearch);
@@ -438,7 +456,8 @@ void ToolbarView::Init() {
 
   // Only show the Battery Saver button when it is not controlled by the OS. On
   // ChromeOS the battery icon in the shelf shows the same information.
-  if (!performance_manager::user_tuning::IsBatterySaverModeManagedByOS()) {
+  if (!performance_manager::user_tuning::IsBatterySaverModeManagedByOS() &&
+      BUTTON_VISIBLE(kBatterySaver)) {
     battery_saver_button_ = container_view_->AddChildView(
         std::make_unique<BatterySaverButton>(browser_view_));
   }
@@ -467,7 +486,7 @@ void ToolbarView::Init() {
                                browser_->profile()->IsGuestSession() ||
                                browser_->profile()->IsRegularProfile();
 #endif
-  avatar_->SetVisible(show_avatar_toolbar_button);
+  avatar_->SetVisible(show_avatar_toolbar_button && BUTTON_VISIBLE(kAvatar));
 
 #if BUILDFLAG(ENABLE_WEBUI_TAB_STRIP)
   auto new_tab_button = std::make_unique<ToolbarButton>(base::BindRepeating(
@@ -824,7 +843,9 @@ void ToolbarView::Layout(PassKey) {
 
   if (display_mode_ == DisplayMode::kNormal) {
     LayoutCommon();
-    UpdateClipPath();
+    if (!browser_->toolbar_overridden()) {
+      UpdateClipPath();
+    }
   }
 
   if (toolbar_controller_) {
diff --git chrome/browser/ui/views/toolbar/toolbar_view.h chrome/browser/ui/views/toolbar/toolbar_view.h
index 415970aadb0ed48099ae622fb6c3c1feaffc83a1..d5e0b4e8bebccf0aa511e602ea870c9c04f781b4 100644
--- chrome/browser/ui/views/toolbar/toolbar_view.h
+++ chrome/browser/ui/views/toolbar/toolbar_view.h
@@ -90,7 +90,8 @@ class ToolbarView : public views::AccessiblePaneView,
   DECLARE_CLASS_ELEMENT_IDENTIFIER_VALUE(kToolbarElementId);
   DECLARE_CLASS_ELEMENT_IDENTIFIER_VALUE(kToolbarContainerElementId);
 
-  ToolbarView(Browser* browser, BrowserView* browser_view);
+  ToolbarView(Browser* browser, BrowserView* browser_view,
+              std::optional<DisplayMode> display_mode);
   ToolbarView(const ToolbarView&) = delete;
   ToolbarView& operator=(const ToolbarView&) = delete;
   ~ToolbarView() override;
