diff --git chrome/browser/printing/print_backend_service_manager.cc chrome/browser/printing/print_backend_service_manager.cc
index e74fac03876879d46c1f939ddecccc9270f8adf6..a2846756b36855b8634988862e2e791fe95fe164 100644
--- chrome/browser/printing/print_backend_service_manager.cc
+++ chrome/browser/printing/print_backend_service_manager.cc
@@ -97,7 +97,15 @@ void MaybeCloneOrCopy(
 // to support modal dialogs from OOP.
 uint32_t NativeViewToUint(gfx::NativeView view) {
 #if BUILDFLAG(IS_WIN)
-  return base::win::HandleToUint32(views::HWNDForNativeView(view));
+  // |view| will be nullptr with CEF OSR. In that case return the desktop
+  // window handle (same behavior as PrintingContextWin::GetRootWindow) to
+  // avoid a DCHECK in PrintingContextSystemDialogWin::AskUserForSettings
+  // with OOP printing.
+  HWND hwnd = views::HWNDForNativeView(view);
+  if (!hwnd) {
+    hwnd = ::GetDesktopWindow();
+  }
+  return base::win::HandleToUint32(hwnd);
 #else
   NOTREACHED();
 #endif
