diff --git chrome/browser/renderer_context_menu/render_view_context_menu.cc chrome/browser/renderer_context_menu/render_view_context_menu.cc
index 2ae51b7a08eeb91a07d93b78d2ef5fbfbaf995e7..3cc39c4bf5285bed17557253fd51d19360e66c14 100644
--- chrome/browser/renderer_context_menu/render_view_context_menu.cc
+++ chrome/browser/renderer_context_menu/render_view_context_menu.cc
@@ -382,6 +382,18 @@ base::OnceCallback<void(RenderViewContextMenu*)>* GetMenuShownCallback() {
   return callback.get();
 }
 
+RenderViewContextMenu::MenuCreatedCallback* GetMenuCreatedCallback() {
+  static base::NoDestructor<RenderViewContextMenu::MenuCreatedCallback>
+      callback;
+  return callback.get();
+}
+
+RenderViewContextMenu::MenuShowHandlerCallback* GetMenuShowHandlerCallback() {
+  static base::NoDestructor<RenderViewContextMenu::MenuShowHandlerCallback>
+      callback;
+  return callback.get();
+}
+
 enum class UmaEnumIdLookupType {
   GeneralEnumId,
   ContextSpecificEnumId,
@@ -662,6 +674,10 @@ int FindUMAEnumValueForCommand(int id, UmaEnumIdLookupType type) {
     return 1;
   }
 
+  // Match the MENU_ID_USER_FIRST to MENU_ID_USER_LAST range from cef_types.h.
+  if (id >= 26500 && id <= 28500)
+    return 1;
+
   id = CollapseCommandsForUMA(id);
   const auto& map = GetIdcToUmaMap(type);
   auto it = map.find(id);
@@ -934,6 +950,14 @@ RenderViewContextMenu::RenderViewContextMenu(
                     : nullptr;
 #endif  // BUILDFLAG(IS_CHROMEOS)
 
+  auto* cb = GetMenuCreatedCallback();
+  if (!cb->is_null()) {
+    first_observer_ = cb->Run(this);
+    if (first_observer_) {
+      observers_.AddObserver(first_observer_.get());
+    }
+  }
+
   observers_.AddObserver(&autofill_context_menu_manager_);
 }
 
@@ -1320,6 +1344,12 @@ void RenderViewContextMenu::InitMenu() {
       content_type_->SupportsGroup(ContextMenuContentType::ITEM_GROUP_PAGE)) {
     AppendReadAnythingItem();
   }
+
+  if (first_observer_) {
+    // Do this last so that the observer can optionally modify previously
+    // created items.
+    first_observer_->InitMenu(params_);
+  }
 }
 
 Profile* RenderViewContextMenu::GetProfile() const {
@@ -3732,6 +3762,26 @@ void RenderViewContextMenu::RemoveObserverForTesting(
   observers_.RemoveObserver(observer);
 }
 
+// static
+void RenderViewContextMenu::RegisterMenuCreatedCallback(
+    MenuCreatedCallback cb) {
+  *GetMenuCreatedCallback() = cb;
+}
+
+// static
+void RenderViewContextMenu::RegisterMenuShowHandlerCallback(
+    MenuShowHandlerCallback cb) {
+  *GetMenuShowHandlerCallback() = cb;
+}
+
+bool RenderViewContextMenu::UseShowHandler() {
+  auto* cb = GetMenuShowHandlerCallback();
+  if (!cb->is_null() && cb->Run(this)) {
+    return true;
+  }
+  return false;
+}
+
 custom_handlers::ProtocolHandlerRegistry::ProtocolHandlerList
 RenderViewContextMenu::GetHandlersForLinkUrl() {
   custom_handlers::ProtocolHandlerRegistry::ProtocolHandlerList handlers =
diff --git chrome/browser/renderer_context_menu/render_view_context_menu.h chrome/browser/renderer_context_menu/render_view_context_menu.h
index fa21659ea613bb9210154525564662d3a266c57e..79059a98ddacc05f1f990e76096b7e6c59ac62b9 100644
--- chrome/browser/renderer_context_menu/render_view_context_menu.h
+++ chrome/browser/renderer_context_menu/render_view_context_menu.h
@@ -153,7 +153,21 @@ class RenderViewContextMenu
   void AddObserverForTesting(RenderViewContextMenuObserver* observer);
   void RemoveObserverForTesting(RenderViewContextMenuObserver* observer);
 
+  // Registers a callback that will be called each time a context menu is
+  // created.
+  using MenuCreatedCallback = base::RepeatingCallback<
+      std::unique_ptr<RenderViewContextMenuObserver>(RenderViewContextMenu*)>;
+  static void RegisterMenuCreatedCallback(MenuCreatedCallback cb);
+
+  // Registers a callback that will be called each time before context menu is
+  // shown. The callback should return true if the show has handled.
+  using MenuShowHandlerCallback = base::RepeatingCallback<
+      bool(RenderViewContextMenu*)>;
+  static void RegisterMenuShowHandlerCallback(MenuShowHandlerCallback cb);
+
  protected:
+  bool UseShowHandler();
+
   Profile* GetProfile() const;
 
   // This may return nullptr (e.g. for WebUI dialogs). Virtual to allow tests to
@@ -500,6 +514,9 @@ class RenderViewContextMenu
   //   built.
   bool is_protocol_submenu_valid_ = false;
 
+  // An observer returned via MenuCreatedCallback that will be called first.
+  std::unique_ptr<RenderViewContextMenuObserver> first_observer_;
+
   // An observer that handles spelling suggestions, "Add to dictionary", and
   // "Use enhanced spell check" items.
   std::unique_ptr<SpellingMenuObserver> spelling_suggestions_menu_observer_;
diff --git chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.h chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.h
index cb51224f9892c26fba26a059798242e319b57d38..b5a3946999d8f4e06443312441f70f93a488d706 100644
--- chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.h
+++ chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.h
@@ -29,6 +29,7 @@ class RenderViewContextMenuMacCocoa : public RenderViewContextMenuMac {
 
   // RenderViewContextMenu:
   void Show() override;
+  bool IsRunning() override;
 
  private:
   // RenderViewContextMenuViewsMac:
diff --git chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.mm chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.mm
index 2e34cf104fbe68248486c33c99a831ff046d42d2..850489d12a4582923c113f4a4ec6e1fe42d4291c 100644
--- chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.mm
+++ chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_cocoa.mm
@@ -70,6 +70,10 @@ RenderViewContextMenuMacCocoa::~RenderViewContextMenuMacCocoa() {
 }
 
 void RenderViewContextMenuMacCocoa::Show() {
+  if (UseShowHandler()) {
+    return;
+  }
+
   views::Widget* widget = views::Widget::GetTopLevelWidgetForNativeView(
       source_web_contents_->GetNativeView());
 
@@ -94,6 +98,10 @@ void RenderViewContextMenuMacCocoa::Show() {
                       views::ElementTrackerViews::GetContextForWidget(widget));
 }
 
+bool RenderViewContextMenuMacCocoa::IsRunning() {
+  return menu_controller_ && [menu_controller_ isMenuOpen];
+}
+
 void RenderViewContextMenuMacCocoa::CancelToolkitMenu() {
   [menu_controller_ cancel];
 }
diff --git chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.h chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.h
index d3bdffdbf66145c888741c3d060b9a9210a8424b..58760dc4d6287d74651475e339cca3c99c335382 100644
--- chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.h
+++ chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.h
@@ -33,6 +33,7 @@ class RenderViewContextMenuMacRemoteCocoa : public RenderViewContextMenuMac {
 
   // RenderViewContextMenu:
   void Show() override;
+  bool IsRunning() override;
 
  private:
   // RenderViewContextMenuViewsMac:
diff --git chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.mm chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.mm
index 208dc7254b3408f0e4f71ebcb118028173a3489a..12bcf8c171c20d255417f7cd6155817c7efd70f1 100644
--- chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.mm
+++ chrome/browser/ui/cocoa/renderer_context_menu/render_view_context_menu_mac_remote_cocoa.mm
@@ -43,6 +43,10 @@ void RenderViewContextMenuMacRemoteCocoa::Show() {
       target_view_id_);
 }
 
+bool RenderViewContextMenuMacRemoteCocoa::IsRunning() {
+  return runner_ && runner_->IsRunning();
+}
+
 void RenderViewContextMenuMacRemoteCocoa::CancelToolkitMenu() {
   runner_->Cancel();
 }
diff --git chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.cc chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.cc
index 995a9c9c8ca222481dfad5e7e1007ab609c40338..b743c01bc6ebb61e58ad108904ffb0dc17ef6083 100644
--- chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.cc
+++ chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.cc
@@ -150,6 +150,9 @@ void RenderViewContextMenuViews::RunMenuAt(views::Widget* parent,
 bool RenderViewContextMenuViews::GetAcceleratorForCommandId(
     int command_id,
     ui::Accelerator* accel) const {
+  if (RenderViewContextMenu::GetAcceleratorForCommandId(command_id, accel))
+    return true;
+
   // There are no formally defined accelerators we can query so we assume
   // that Ctrl+C, Ctrl+V, Ctrl+X, Ctrl-A, etc do what they normally do.
   switch (command_id) {
@@ -366,6 +369,10 @@ void RenderViewContextMenuViews::AppendPlatformEditableItems() {
 }
 
 void RenderViewContextMenuViews::Show() {
+  if (UseShowHandler()) {
+    return;
+  }
+
   if (base::CommandLine::ForCurrentProcess()->HasSwitch(switches::kKioskMode)) {
     return;
   }
@@ -412,6 +419,11 @@ void RenderViewContextMenuViews::Show() {
   }
 }
 
+bool RenderViewContextMenuViews::IsRunning() {
+  return static_cast<ToolkitDelegateViews*>(toolkit_delegate())
+             ->IsMenuRunning();
+}
+
 views::Widget* RenderViewContextMenuViews::GetTopLevelWidget() {
   return views::Widget::GetTopLevelWidgetForNativeView(GetActiveNativeView());
 }
diff --git chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.h chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.h
index 8433cdc9c88b716170df529d6a88eeaacf786b29..aea62cda5cd22458afed447389305593d7f204bf 100644
--- chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.h
+++ chrome/browser/ui/views/renderer_context_menu/render_view_context_menu_views.h
@@ -45,6 +45,7 @@ class RenderViewContextMenuViews : public RenderViewContextMenu {
 
   // RenderViewContextMenuBase implementation.
   void Show() override;
+  bool IsRunning() override;
 
  protected:
   RenderViewContextMenuViews(content::RenderFrameHost& render_frame_host,
diff --git chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.cc chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.cc
index d46f3ec0eec9cc0e3d4b2ff49142133f69cff9ba..c1e166725f526d1c10b9d0e53ae214a05a04d9fb 100644
--- chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.cc
+++ chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.cc
@@ -92,6 +92,10 @@ void ChromeWebContentsViewDelegateViews::ShowMenu(
   context_menu_->Show();
 }
 
+bool ChromeWebContentsViewDelegateViews::IsMenuRunning() {
+  return context_menu_ && context_menu_->IsRunning();
+}
+
 void ChromeWebContentsViewDelegateViews::ShowContextMenu(
     content::RenderFrameHost& render_frame_host,
     const content::ContextMenuParams& params) {
diff --git chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.h chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.h
index 427c68bd899104d779a3d32933df5b6b2fae00c9..1632f7bbfd0c2a5804ad3831776822001d395fd9 100644
--- chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.h
+++ chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views.h
@@ -55,6 +55,7 @@ class ChromeWebContentsViewDelegateViews
       content::RenderFrameHost& render_frame_host,
       const content::ContextMenuParams& params) override;
   void ShowMenu(std::unique_ptr<RenderViewContextMenuBase> menu) override;
+  bool IsMenuRunning() override;
 
  private:
   // The context menu is reset every time we show it, but we keep a pointer to
diff --git chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.h chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.h
index 07e5b3613a2df3a7117dcc1c2947dca70fff204f..293282690bd5a29a8c5ebbf8b5fa3db63900ece0 100644
--- chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.h
+++ chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.h
@@ -55,6 +55,7 @@ class ChromeWebContentsViewDelegateViewsMac
       content::RenderFrameHost& render_frame_host,
       const content::ContextMenuParams& params) override;
   void ShowMenu(std::unique_ptr<RenderViewContextMenuBase> menu) override;
+  bool IsMenuRunning() override;
 
  private:
   content::RenderWidgetHostView* GetActiveRenderWidgetHostView() const;
diff --git chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.mm chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.mm
index 13ce1cf5f40411623ea55e63e139c83e4f76c9af..e722d85d50a0309ac709f502ecb6958d566f85ad 100644
--- chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.mm
+++ chrome/browser/ui/views/tab_contents/chrome_web_contents_view_delegate_views_mac.mm
@@ -147,6 +147,10 @@ void ChromeWebContentsViewDelegateViewsMac::ShowMenu(
   context_menu_->Show();
 }
 
+bool ChromeWebContentsViewDelegateViewsMac::IsMenuRunning() {
+  return context_menu_ && context_menu_->IsRunning();
+}
+
 content::RenderWidgetHostView*
 ChromeWebContentsViewDelegateViewsMac::GetActiveRenderWidgetHostView() const {
   return web_contents_->GetTopLevelRenderWidgetHostView();
diff --git components/renderer_context_menu/context_menu_delegate.h components/renderer_context_menu/context_menu_delegate.h
index 042428f77f4ad021d82db228772d82dc9cd9216c..e4efd98ca45d5be49b17d75533719b3ac157767d 100644
--- components/renderer_context_menu/context_menu_delegate.h
+++ components/renderer_context_menu/context_menu_delegate.h
@@ -45,6 +45,8 @@ class ContextMenuDelegate {
   // Displays the context menu.
   virtual void ShowMenu(std::unique_ptr<RenderViewContextMenuBase> menu) = 0;
 
+  virtual bool IsMenuRunning() = 0;
+
  private:
   raw_ptr<content::WebContents> web_contents_ = nullptr;
 };
diff --git components/renderer_context_menu/render_view_context_menu_base.cc components/renderer_context_menu/render_view_context_menu_base.cc
index ad4a62e7b52962fd2cd7babba369d1a8ac474a39..3a8b295fa8fa4951aeabc9d19d98d08fd10ae18b 100644
--- components/renderer_context_menu/render_view_context_menu_base.cc
+++ components/renderer_context_menu/render_view_context_menu_base.cc
@@ -411,6 +411,17 @@ bool RenderViewContextMenuBase::IsCommandIdChecked(int id) const {
   return false;
 }
 
+bool RenderViewContextMenuBase::GetAcceleratorForCommandId(
+    int id,
+    ui::Accelerator* accelerator) const {
+  for (auto& observer : observers_) {
+    if (observer.IsCommandIdSupported(id))
+      return observer.GetAccelerator(id, accelerator);
+  }
+
+  return false;
+}
+
 void RenderViewContextMenuBase::ExecuteCommand(int id, int event_flags) {
   command_executed_ = true;
   RecordUsedItem(id);
diff --git components/renderer_context_menu/render_view_context_menu_base.h components/renderer_context_menu/render_view_context_menu_base.h
index 3129962e3160982670615c1ae0ad59f876048afb..b13856169889e875a3944ffc0e6382c6d179552e 100644
--- components/renderer_context_menu/render_view_context_menu_base.h
+++ components/renderer_context_menu/render_view_context_menu_base.h
@@ -85,8 +85,13 @@ class RenderViewContextMenuBase : public ui::SimpleMenuModel::Delegate,
   // Programmatically closes the context menu.
   void Cancel();
 
+  virtual bool IsRunning() = 0;
+
   const ui::SimpleMenuModel& menu_model() const { return menu_model_; }
   const content::ContextMenuParams& params() const { return params_; }
+  content::WebContents* source_web_contents() const {
+    return source_web_contents_;
+  }
 
   // Returns true if the specified command id is known and valid for
   // this menu. If the command is known |enabled| is set to indicate
@@ -95,6 +100,9 @@ class RenderViewContextMenuBase : public ui::SimpleMenuModel::Delegate,
 
   // SimpleMenuModel::Delegate implementation.
   bool IsCommandIdChecked(int command_id) const override;
+  bool GetAcceleratorForCommandId(
+      int command_id,
+      ui::Accelerator* accelerator) const override;
   void ExecuteCommand(int command_id, int event_flags) override;
   void OnMenuWillShow(ui::SimpleMenuModel* source) override;
   void MenuClosed(ui::SimpleMenuModel* source) override;
diff --git components/renderer_context_menu/render_view_context_menu_observer.cc components/renderer_context_menu/render_view_context_menu_observer.cc
index 9fdda1636003df5f0981f6073bc459fa5b5056ae..538bd05a412961e54214833fad915de7104b5154 100644
--- components/renderer_context_menu/render_view_context_menu_observer.cc
+++ components/renderer_context_menu/render_view_context_menu_observer.cc
@@ -15,3 +15,8 @@ bool RenderViewContextMenuObserver::IsCommandIdChecked(int command_id) {
 bool RenderViewContextMenuObserver::IsCommandIdEnabled(int command_id) {
   return false;
 }
+
+bool RenderViewContextMenuObserver::GetAccelerator(int command_id,
+                                                   ui::Accelerator* accel) {
+  return false;
+}
diff --git components/renderer_context_menu/render_view_context_menu_observer.h components/renderer_context_menu/render_view_context_menu_observer.h
index b2aa338c76cd36031abe49190c1f43db4e2b6f03..233c34003bbc5a3e3588bdd60b3ba6ea4c426ea9 100644
--- components/renderer_context_menu/render_view_context_menu_observer.h
+++ components/renderer_context_menu/render_view_context_menu_observer.h
@@ -11,6 +11,10 @@ namespace content {
 struct ContextMenuParams;
 }
 
+namespace ui {
+class Accelerator;
+}
+
 // The interface used for implementing context-menu items. The following
 // instruction describe how to implement a context-menu item with this
 // interface.
@@ -100,6 +104,8 @@ class RenderViewContextMenuObserver {
   virtual bool IsCommandIdChecked(int command_id);
   virtual bool IsCommandIdEnabled(int command_id);
 
+  virtual bool GetAccelerator(int command_id, ui::Accelerator* accel);
+
   // Called when a user selects the specified context-menu item. This is
   // only called when the observer returns true for IsCommandIdSupported()
   // for that |command_id|.
diff --git components/renderer_context_menu/views/toolkit_delegate_views.cc components/renderer_context_menu/views/toolkit_delegate_views.cc
index 0be2640c8173c86720b44a8ad4ab10fdf9e6686e..8bfdc6185690fbe01ba0aaab8c7bc6973a78a37c 100644
--- components/renderer_context_menu/views/toolkit_delegate_views.cc
+++ components/renderer_context_menu/views/toolkit_delegate_views.cc
@@ -31,6 +31,10 @@ void ToolkitDelegateViews::RunMenuAt(views::Widget* parent,
                           anchor_position, type);
 }
 
+bool ToolkitDelegateViews::IsMenuRunning() const {
+  return menu_runner_ && menu_runner_->IsRunning();
+}
+
 void ToolkitDelegateViews::Init(ui::SimpleMenuModel* menu_model) {
   menu_adapter_ = std::make_unique<views::MenuModelAdapter>(menu_model);
   std::unique_ptr<views::MenuItemView> menu_view = menu_adapter_->CreateMenu();
diff --git components/renderer_context_menu/views/toolkit_delegate_views.h components/renderer_context_menu/views/toolkit_delegate_views.h
index 7762b3e6673d91edd01858bfd325ee57e3d9f766..5bd1544aaaa6229bd20955e0e0ded1360d1395b2 100644
--- components/renderer_context_menu/views/toolkit_delegate_views.h
+++ components/renderer_context_menu/views/toolkit_delegate_views.h
@@ -39,6 +39,7 @@ class ToolkitDelegateViews : public RenderViewContextMenuBase::ToolkitDelegate {
                  const gfx::Point& point,
                  ui::mojom::MenuSourceType type);
   views::MenuItemView* menu_view() { return menu_view_; }
+  bool IsMenuRunning() const;
 
  protected:
   // ToolkitDelegate:
