diff --git chrome/browser/profiles/profile_window.cc chrome/browser/profiles/profile_window.cc
index c00884e06f960..858fdb75529f8 100644
--- chrome/browser/profiles/profile_window.cc
+++ chrome/browser/profiles/profile_window.cc
@@ -272,7 +272,9 @@ void BubbleViewModeFromAvatarBubbleMode(BrowserWindow::AvatarBubbleMode mode,
       *bubble_view_mode = BUBBLE_VIEW_MODE_PROFILE_CHOOSER;
       return;
     case BrowserWindow::AVATAR_BUBBLE_MODE_DEFAULT:
-      *bubble_view_mode = profile->IsIncognitoProfile()
+      *bubble_view_mode = profile->IsIncognitoProfile() ||
+                          (profile->IsOffTheRecord() &&
+                           profile->GetOTRProfileID().IsUniqueForCEF())
                               ? profiles::BUBBLE_VIEW_MODE_INCOGNITO
                               : profiles::BUBBLE_VIEW_MODE_PROFILE_CHOOSER;
   }
diff --git chrome/browser/ui/views/profiles/incognito_menu_view.cc chrome/browser/ui/views/profiles/incognito_menu_view.cc
index 38970b9869c05..35dd0be0030ed 100644
--- chrome/browser/ui/views/profiles/incognito_menu_view.cc
+++ chrome/browser/ui/views/profiles/incognito_menu_view.cc
@@ -38,7 +38,9 @@
 IncognitoMenuView::IncognitoMenuView(views::Button* anchor_button,
                                      Browser* browser)
     : ProfileMenuViewBase(anchor_button, browser) {
-  DCHECK(browser->profile()->IsIncognitoProfile());
+  DCHECK(browser->profile()->IsIncognitoProfile() ||
+         (browser->profile()->IsOffTheRecord() &&
+          browser->profile()->GetOTRProfileID().IsUniqueForCEF()));
   GetViewAccessibility().OverrideName(GetAccessibleWindowTitle());
 
   chrome::RecordDialogCreation(
diff --git chrome/browser/ui/views/profiles/profile_menu_view_base.cc chrome/browser/ui/views/profiles/profile_menu_view_base.cc
index 7f8f15b406de5..d0cf255156980 100644
--- chrome/browser/ui/views/profiles/profile_menu_view_base.cc
+++ chrome/browser/ui/views/profiles/profile_menu_view_base.cc
@@ -512,7 +512,9 @@ void ProfileMenuViewBase::ShowBubble(profiles::BubbleViewMode view_mode,
 
   ProfileMenuViewBase* bubble = nullptr;
   if (view_mode == profiles::BUBBLE_VIEW_MODE_INCOGNITO) {
-    DCHECK(browser->profile()->IsIncognitoProfile());
+    DCHECK(browser->profile()->IsIncognitoProfile() ||
+           (browser->profile()->IsOffTheRecord() &&
+            browser->profile()->GetOTRProfileID().IsUniqueForCEF()));
     bubble = new IncognitoMenuView(anchor_button, browser);
   } else {
 #if BUILDFLAG(IS_CHROMEOS_ASH)
