diff --git chrome/utility/chrome_content_utility_client.cc chrome/utility/chrome_content_utility_client.cc
index 8a498ddd2c563..1a96939d54079 100644
--- chrome/utility/chrome_content_utility_client.cc
+++ chrome/utility/chrome_content_utility_client.cc
@@ -14,6 +14,7 @@
 #include "base/path_service.h"
 #include "base/task/single_thread_task_runner.h"
 #include "build/build_config.h"
+#include "cef/libcef/features/runtime.h"
 #include "chrome/common/chrome_paths.h"
 #include "chrome/common/profiler/thread_profiler.h"
 #include "chrome/common/profiler/thread_profiler_configuration.h"
@@ -55,7 +56,11 @@ void ChromeContentUtilityClient::UtilityThreadStarted() {
       command_line->GetSwitchValueASCII(switches::kProcessType);
   // An in-process utility thread may run in other processes, only set up
   // collector in a utility process.
-  if (process_type == switches::kUtilityProcess) {
+  if (process_type == switches::kUtilityProcess
+#if BUILDFLAG(ENABLE_ALLOY_BOOTSTRAP)
+      && !cef::IsAlloyRuntimeEnabled()
+#endif
+      ) {
     const auto* heap_profiler_controller =
         heap_profiling::HeapProfilerController::GetInstance();
     // The HeapProfilerController should have been created in
