diff --git chrome/browser/metrics/chrome_browser_main_extra_parts_metrics.cc chrome/browser/metrics/chrome_browser_main_extra_parts_metrics.cc
index 0c12da9dfa8f7333cbbbe629da2fd575e9f09cc1..419c69006451d6a3ebde868fafa3fd2828d7bed0 100644
--- chrome/browser/metrics/chrome_browser_main_extra_parts_metrics.cc
+++ chrome/browser/metrics/chrome_browser_main_extra_parts_metrics.cc
@@ -34,6 +34,7 @@
 #include "base/version.h"
 #include "build/build_config.h"
 #include "build/config/compiler/compiler_buildflags.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/about_flags.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/buildflags.h"
@@ -898,9 +899,11 @@ void RecordStartupMetrics() {
   key_credential_manager_support::ReportKeyCredentialManagerSupport();
 #endif  // BUILDFLAG(IS_WIN)
 
+#if !BUILDFLAG(ENABLE_CEF)
 #if BUILDFLAG(IS_WIN) || BUILDFLAG(IS_MAC)
   crypto::MaybeMeasureTpmOperations(unexportable_keys::GetDefaultConfig());
 #endif  // BUILDFLAG(IS_WIN) || BUILDFLAG(IS_MAC)
+#endif  // !BUILDFLAG(ENABLE_CEF)
 
   // Record whether Chrome is the default browser or not.
   // Disabled on Linux due to hanging browser tests, see crbug.com/1216328.
