diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index e4537f6dcdfe70125649c1a83a9f7eef5dcb4d90..70d48bd8603895e5d4a9932a44ac105da99336bc 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -4,6 +4,7 @@
 
 import("//build/config/buildflags_paint_preview.gni")
 import("//build/config/features.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//chrome/common/request_header_integrity/buildflags.gni")
 import("//components/nacl/features.gni")
@@ -134,6 +135,7 @@ static_library("renderer") {
     ":process_state",
     "//base/allocator:buildflags",
     "//build:chromeos_buildflags",
+    "//cef/libcef/features:buildflags",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -248,6 +250,10 @@ static_library("renderer") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_nacl) {
     deps += [ "//components/nacl/renderer" ]
 
