diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 8ce5e842bbb2..95b0356cab27 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -9,6 +9,7 @@ import("//build/config/jumbo.gni")
 import("//build/config/linux/gtk/gtk.gni")
 import("//build/config/ui.gni")
 import("//build/split_static_library.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//components/feature_engagement/features.gni")
 import("//components/feed/features.gni")
@@ -1889,6 +1890,7 @@ jumbo_split_static_library("browser") {
     "//base/util/values:values_util",
     "//build:branding_buildflags",
     "//cc",
+    "//cef/libcef/features",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -2192,6 +2194,10 @@ jumbo_split_static_library("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     sources += [
       "after_startup_task_utils_android.cc",
