diff --git chrome/browser/devtools/protocol/page_handler.cc chrome/browser/devtools/protocol/page_handler.cc
index 663a27ff332f2..5c8222267f860 100644
--- chrome/browser/devtools/protocol/page_handler.cc
+++ chrome/browser/devtools/protocol/page_handler.cc
@@ -14,7 +14,6 @@
 #if BUILDFLAG(ENABLE_PRINTING)
 #include "components/printing/browser/print_to_pdf/pdf_print_utils.h"
 #if BUILDFLAG(ENABLE_PRINT_PREVIEW)
-#include "chrome/browser/printing/pdf_nup_converter_client.h"
 #include "chrome/browser/printing/print_view_manager.h"
 #else
 #include "chrome/browser/printing/print_view_manager_basic.h"
