diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index ddfdfd49ad2b2..1b53c56bb677f 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -5,6 +5,7 @@
 import("//build/config/buildflags_paint_preview.gni")
 import("//build/config/chromeos/ui_mode.gni")
 import("//build/config/features.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//components/nacl/features.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -136,6 +137,7 @@ static_library("renderer") {
   deps = [
     "//base/allocator:buildflags",
     "//build:chromeos_buildflags",
+    "//cef/libcef/features",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -243,6 +245,10 @@ static_library("renderer") {
 
   configs += [ "//build/config/compiler:wexit_time_destructors" ]
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_nacl) {
     deps += [ "//components/nacl/renderer" ]
 
