diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index b2eaaf2b29fd..842e99c37e95 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -5,6 +5,7 @@
 import("//build/config/buildflags_paint_preview.gni")
 import("//build/config/features.gni")
 import("//build/config/jumbo.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//components/nacl/features.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -129,6 +130,7 @@ jumbo_static_library("renderer") {
   defines = []
 
   deps = [
+    "//cef/libcef/features",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -194,6 +196,10 @@ jumbo_static_library("renderer") {
 
   configs += [ "//build/config/compiler:wexit_time_destructors" ]
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_nacl) {
     deps += [
       "//components/nacl/loader",
