diff --git pepper_flash.cc pepper_flash.cc
index 16b4328..63b21c1 100644
--- pepper_flash.cc
+++ pepper_flash.cc
@@ -119,8 +119,15 @@ bool CheckPepperFlashManifest(const base::DictionaryValue& manifest,
 
   std::string arch;
   manifest.GetStringASCII("x-ppapi-arch", &arch);
+#if defined(OS_MACOSX)
+  // On Mac newer versions of the plugin are a universal binary and use "mac"
+  // as the value.
+  if (arch != kPepperFlashArch && arch != kPepperFlashOperatingSystem)
+    return false;
+#else
   if (arch != kPepperFlashArch)
     return false;
+#endif
 
   *version_out = version;
   return true;
