diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index d06407d4ca6a18e66c49ea725e6e9c40c8042c7a..c62e844edb75561e97a165789b24597a6f22f28d 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -11,6 +11,7 @@ import("//build/config/compiler/pgo/pgo.gni")
 import("//build/config/features.gni")
 import("//build/config/python.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -2039,6 +2040,7 @@ source_set("browser") {
     "//build/config/compiler:compiler_buildflags",
     "//build/config/linux/dbus:buildflags",
     "//cc",
+    "//cef/libcef/features:buildflags",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -3051,6 +3053,10 @@ source_set("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     deps += [
       "//chrome/browser/android",
