diff --git chrome/browser/ui/autofill/address_bubbles_controller.cc chrome/browser/ui/autofill/address_bubbles_controller.cc
index 9a959e457fd148bcde186fe01b67ea19dddaf683..8421a07350b8385dc659890eae0f701b88c38820 100644
--- chrome/browser/ui/autofill/address_bubbles_controller.cc
+++ chrome/browser/ui/autofill/address_bubbles_controller.cc
@@ -56,10 +56,15 @@ AutofillBubbleBase* ShowSaveBubble(
     content::WebContents* web_contents,
     bool shown_by_user_gesture,
     base::WeakPtr<AddressBubbleControllerDelegate> delegate) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   auto controller = std::make_unique<SaveAddressBubbleController>(
       delegate, web_contents, profile, is_migration_to_account);
 
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
+  return browser_window
       ->GetAutofillBubbleHandler()
       ->ShowSaveAddressProfileBubble(web_contents, std::move(controller),
                                      shown_by_user_gesture);
@@ -71,9 +76,14 @@ AutofillBubbleBase* ShowUpdateBubble(
     content::WebContents* web_contents,
     bool shown_by_user_gesture,
     base::WeakPtr<AddressBubbleControllerDelegate> delegate) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   auto update_controller = std::make_unique<UpdateAddressBubbleController>(
       delegate, web_contents, profile, original_profile);
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
+  return browser_window
       ->GetAutofillBubbleHandler()
       ->ShowUpdateAddressProfileBubble(
           web_contents, std::move(update_controller), shown_by_user_gesture);
@@ -82,9 +92,14 @@ AutofillBubbleBase* ShowUpdateBubble(
 #if BUILDFLAG(ENABLE_DICE_SUPPORT)
 AutofillBubbleBase* ShowSignInPromo(content::WebContents* web_contents,
                                     const AutofillProfile& autofill_profile) {
+  auto browser_window =
+      BrowserWindow::FindBrowserWindowWithWebContents(web_contents);
+  if (!browser_window) {
+    return nullptr;
+  }
   // TODO(crbug.com/381390420): Expose the `AutofillBubbleHandler` in
   // `BrowserWindowInterface` and use that instead.
-  return BrowserWindow::FindBrowserWindowWithWebContents(web_contents)
+  return browser_window
       ->GetAutofillBubbleHandler()
       ->ShowAddressSignInPromo(web_contents, autofill_profile);
 }
@@ -232,8 +247,6 @@ void AddressBubblesController::DoShowBubble() {
 
   set_bubble_view(show_bubble_view_callback_.Run(
       web_contents(), shown_by_user_gesture_, GetWeakPtr()));
-
-  CHECK(bubble_view());
 }
 
 BubbleType AddressBubblesController::GetBubbleType() const {
