diff --git chrome/browser/printing/print_backend_service_manager.cc chrome/browser/printing/print_backend_service_manager.cc
index e5c205c1889571bc7bc6bf0ed2f5d8521827c273..8d12a5cdb730eab49f4f7973ed887e9cb77cf8f4 100644
--- chrome/browser/printing/print_backend_service_manager.cc
+++ chrome/browser/printing/print_backend_service_manager.cc
@@ -96,7 +96,15 @@ void MaybeCloneOrCopy(
 // to support modal dialogs from OOP.
 uint32_t NativeViewToUint(gfx::NativeView view) {
 #if BUILDFLAG(IS_WIN)
-  return base::win::HandleToUint32(views::HWNDForNativeView(view));
+  // |view| will be nullptr with CEF OSR. In that case return the desktop
+  // window handle (same behavior as PrintingContextWin::GetRootWindow) to
+  // avoid a DCHECK in PrintingContextSystemDialogWin::AskUserForSettings
+  // with OOP printing.
+  HWND hwnd = views::HWNDForNativeView(view);
+  if (!hwnd) {
+    hwnd = ::GetDesktopWindow();
+  }
+  return base::win::HandleToUint32(hwnd);
 #else
   NOTREACHED();
 #endif
