diff --git chrome/browser/chrome_content_browser_client.cc chrome/browser/chrome_content_browser_client.cc
index b4087b7737c2..ff13279d154e 100644
--- chrome/browser/chrome_content_browser_client.cc
+++ chrome/browser/chrome_content_browser_client.cc
@@ -1003,10 +1003,6 @@ void LaunchURL(const GURL& url,
   }
 }
 
-std::string GetProduct() {
-  return version_info::GetProductNameAndVersionForUserAgent();
-}
-
 void MaybeAppendSecureOriginsAllowlistSwitch(base::CommandLine* cmdline) {
   // |allowlist| combines pref/policy + cmdline switch in the browser process.
   // For renderer and utility (e.g. NetworkService) processes the switch is the
@@ -1173,6 +1169,14 @@ const blink::UserAgentBrandList& GetBrandVersionList() {
   return *greased_brand_version_list;
 }
 
+std::string GetProduct() {
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(switches::kProductVersion))
+    return command_line->GetSwitchValueASCII(switches::kProductVersion);
+
+  return version_info::GetProductNameAndVersionForUserAgent();
+}
+
 std::string GetUserAgent() {
   base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
   if (command_line->HasSwitch(switches::kUserAgent)) {
diff --git chrome/browser/chrome_content_browser_client.h chrome/browser/chrome_content_browser_client.h
index 219eac813d9b..f933fc41ab85 100644
--- chrome/browser/chrome_content_browser_client.h
+++ chrome/browser/chrome_content_browser_client.h
@@ -90,7 +90,8 @@ class ChromeXrIntegrationClient;
 }
 #endif
 
-// Returns the user agent of Chrome.
+// Returns the product and user agent of Chrome.
+std::string GetProduct();
 std::string GetUserAgent();
 
 blink::UserAgentMetadata GetUserAgentMetadata();
