diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index 14bef91c5370..cb1faba7c23e 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -246,6 +246,7 @@ jumbo_static_library("safe_browsing") {
         "//chrome/common/safe_browsing:download_type_util",
         "//chrome/services/file_util/public/cpp",
         "//components/content_settings/core/browser",
+        "//components/gcm_driver:gcm_buildflags",
         "//components/language/core/common",
         "//components/prefs",
         "//components/safe_browsing/core:webprotect_proto",
