diff --git chrome/browser/ui/prefs/pref_watcher.h chrome/browser/ui/prefs/pref_watcher.h
index dd06c6d22b16..1f3a5aa7ac2b 100644
--- chrome/browser/ui/prefs/pref_watcher.h
+++ chrome/browser/ui/prefs/pref_watcher.h
@@ -29,10 +29,10 @@ class PrefWatcher : public KeyedService {
   void RegisterRendererPreferenceWatcher(
       mojo::PendingRemote<blink::mojom::RendererPreferenceWatcher> watcher);
 
- private:
   // KeyedService overrides:
   void Shutdown() override;
 
+ private:
   void UpdateRendererPreferences();
   void OnWebPrefChanged(const std::string& pref_name);
 
