diff --git chrome/browser/ui/views/frame/browser_native_widget_mac.h chrome/browser/ui/views/frame/browser_native_widget_mac.h
index 7dd0ba6851c17db45689ef3aee4440a25e8a4552..70be9d606c59e69c321566cade2beda44edb1030 100644
--- chrome/browser/ui/views/frame/browser_native_widget_mac.h
+++ chrome/browser/ui/views/frame/browser_native_widget_mac.h
@@ -11,6 +11,7 @@
 #include "ui/base/mojom/window_show_state.mojom-forward.h"
 #include "ui/views/widget/native_widget_mac.h"
 
+class Browser;
 class BrowserWidget;
 class BrowserView;
 @class BrowserWindowTouchBarController;
@@ -57,6 +58,21 @@ class BrowserNativeWidgetMac : public views::NativeWidgetMac,
   bool ShouldUseInitialVisibleOnAllWorkspaces() const override;
   void AnnounceTextInInProcessWindow(const std::u16string& text) override;
 
+  static void ValidateUserInterfaceItem(
+      Browser* browser,
+      int32_t command,
+      remote_cocoa::mojom::ValidateUserInterfaceItemResult* result);
+  static bool WillExecuteCommand(
+      Browser* browser,
+      int32_t command,
+      WindowOpenDisposition window_open_disposition,
+      bool is_before_first_responder);
+  static bool ExecuteCommand(
+      Browser* browser,
+      int32_t command,
+      WindowOpenDisposition window_open_disposition,
+      bool is_before_first_responder);
+
  protected:
   ~BrowserNativeWidgetMac() override;
 
diff --git chrome/browser/ui/views/frame/browser_native_widget_mac.mm chrome/browser/ui/views/frame/browser_native_widget_mac.mm
index f45be88706afc7c6ad4ecbdd9f4c621137bc065d..52c4b9d4b3c3510467b289eaae8368ab2a9579dd 100644
--- chrome/browser/ui/views/frame/browser_native_widget_mac.mm
+++ chrome/browser/ui/views/frame/browser_native_widget_mac.mm
@@ -217,7 +217,15 @@ void BrowserNativeWidgetMac::OnWidgetDestroyed(views::Widget* widget) {
 void BrowserNativeWidgetMac::ValidateUserInterfaceItem(
     int32_t tag,
     remote_cocoa::mojom::ValidateUserInterfaceItemResult* result) {
-  Browser* const browser = browser_view_ ? browser_view_->browser() : nullptr;
+  return ValidateUserInterfaceItem(
+      browser_view_ ? browser_view_->browser() : nullptr, tag, result);
+}
+
+// static
+void BrowserNativeWidgetMac::ValidateUserInterfaceItem(
+    Browser* browser,
+    int32_t tag,
+    remote_cocoa::mojom::ValidateUserInterfaceItemResult* result) {
   if (!browser || !chrome::SupportsCommand(browser, tag)) {
     result->enable = false;
     return;
@@ -412,9 +420,16 @@ bool BrowserNativeWidgetMac::WillExecuteCommand(
   if (!browser_view_) {
     return false;
   }
+  return WillExecuteCommand(browser_view_->browser(), command,
+                            window_open_disposition, is_before_first_responder);
+}
 
-  Browser* const browser = browser_view_->browser();
-
+// static
+bool BrowserNativeWidgetMac::WillExecuteCommand(
+    Browser* browser,
+    int32_t command,
+    WindowOpenDisposition window_open_disposition,
+    bool is_before_first_responder) {
   if (is_before_first_responder) {
     // The specification for this private extensions API is incredibly vague.
     // For now, we avoid triggering chrome commands prior to giving the
@@ -446,12 +461,23 @@ bool BrowserNativeWidgetMac::ExecuteCommand(
     int32_t command,
     WindowOpenDisposition window_open_disposition,
     bool is_before_first_responder) {
-  if (!WillExecuteCommand(command, window_open_disposition,
-                          is_before_first_responder)) {
+  if (!browser_view_) {
     return false;
   }
+  return ExecuteCommand(browser_view_->browser(), command,
+                        window_open_disposition, is_before_first_responder);
+}
 
-  Browser* browser = browser_view_->browser();
+// static
+bool BrowserNativeWidgetMac::ExecuteCommand(
+    Browser* browser,
+    int32_t command,
+    WindowOpenDisposition window_open_disposition,
+    bool is_before_first_responder) {
+  if (!WillExecuteCommand(browser, command, window_open_disposition,
+                          is_before_first_responder)) {
+    return false;
+  }
 
   if (command == IDC_TOGGLE_VERTICAL_TABS) {
     if (auto* controller =
@@ -461,10 +487,13 @@ bool BrowserNativeWidgetMac::ExecuteCommand(
           is_vertical, tabs::VerticalTabStripEntryPoint::kMacViewMenu);
     }
   } else if (command == IDC_CLEAR_BROWSING_DATA) {
-    views::ElementTrackerViews::GetInstance()->NotifyCustomEvent(
-        browsing_data_important_sites_util::
-            kOpenClearBrowsingDataDialogViaAcceleratorEventId,
-        browser_view_);
+    BrowserView* browser_view = BrowserView::GetBrowserViewForBrowser(browser);
+    if (browser_view) {
+      views::ElementTrackerViews::GetInstance()->NotifyCustomEvent(
+          browsing_data_important_sites_util::
+              kOpenClearBrowsingDataDialogViaAcceleratorEventId,
+          browser_view);
+    }
   }
 
   chrome::ExecuteCommandWithDisposition(browser, command,
