diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index a82341061c691bf112135d3c84e424183a4ce36c..51debd225b1daa29b7b1f18752f95fe17c70c739 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -27,6 +27,7 @@ static_library("safe_browsing") {
     "//chrome/browser/google",
     "//chrome/browser/optimization_guide",
     "//chrome/browser/password_manager/factories",
+    "//chrome/browser/permissions:permissions_proto",
     "//chrome/browser/profiles",
     "//chrome/browser/profiles:profile",
     "//chrome/browser/signin",
@@ -39,6 +40,7 @@ static_library("safe_browsing") {
     "//components/enterprise/buildflags",
     "//components/enterprise/common:strings",
     "//components/enterprise/obfuscation/core:enterprise_obfuscation",
+    "//components/gcm_driver:gcm_buildflags",
     "//components/keyed_service/content",
     "//components/language/core/browser",
     "//components/no_state_prefetch/browser",
