diff --git chrome/browser/ui/views/frame/browser_native_widget_mac.h chrome/browser/ui/views/frame/browser_native_widget_mac.h
index e5a6aaa3e11cc3a60eef9eb8ace0af6fca30a53a..7a6752e46cb0e01f0d0d8623511c1afac828f4e8 100644
--- chrome/browser/ui/views/frame/browser_native_widget_mac.h
+++ chrome/browser/ui/views/frame/browser_native_widget_mac.h
@@ -15,6 +15,7 @@
 #include "ui/native_theme/native_theme.h"
 #include "ui/views/widget/native_widget_mac.h"
 
+class Browser;
 class BrowserWidget;
 class BrowserView;
 @class BrowserWindowTouchBarController;
@@ -61,6 +62,21 @@ class BrowserNativeWidgetMac : public views::NativeWidgetMac,
   bool ShouldUseInitialVisibleOnAllWorkspaces() const override;
   void AnnounceTextInInProcessWindow(const std::u16string& text) override;
 
+  static void ValidateUserInterfaceItem(
+      Browser* browser,
+      int32_t command,
+      remote_cocoa::mojom::ValidateUserInterfaceItemResult* result);
+  static bool WillExecuteCommand(
+      Browser* browser,
+      int32_t command,
+      WindowOpenDisposition window_open_disposition,
+      bool is_before_first_responder);
+  static bool ExecuteCommand(
+      Browser* browser,
+      int32_t command,
+      WindowOpenDisposition window_open_disposition,
+      bool is_before_first_responder);
+
  protected:
   ~BrowserNativeWidgetMac() override;
 
diff --git chrome/browser/ui/views/frame/browser_native_widget_mac.mm chrome/browser/ui/views/frame/browser_native_widget_mac.mm
index 615e488de7bbce5d35ee06059f983093a468050c..76a20d1190d5e6c7627b15c487d3676c5d58fe84 100644
--- chrome/browser/ui/views/frame/browser_native_widget_mac.mm
+++ chrome/browser/ui/views/frame/browser_native_widget_mac.mm
@@ -234,7 +234,15 @@ void BrowserNativeWidgetMac::OnWidgetDestroyed(views::Widget* widget) {
 void BrowserNativeWidgetMac::ValidateUserInterfaceItem(
     int32_t tag,
     remote_cocoa::mojom::ValidateUserInterfaceItemResult* result) {
-  Browser* const browser = browser_view_ ? browser_view_->browser() : nullptr;
+  return ValidateUserInterfaceItem(
+      browser_view_ ? browser_view_->browser() : nullptr, tag, result);
+}
+
+// static
+void BrowserNativeWidgetMac::ValidateUserInterfaceItem(
+    Browser* browser,
+    int32_t tag,
+    remote_cocoa::mojom::ValidateUserInterfaceItemResult* result) {
   if (!browser || !chrome::SupportsCommand(browser, tag)) {
     result->enable = false;
     return;
@@ -430,9 +438,16 @@ bool BrowserNativeWidgetMac::WillExecuteCommand(
   if (!browser_view_) {
     return false;
   }
+  return WillExecuteCommand(browser_view_->browser(), command,
+                            window_open_disposition, is_before_first_responder);
+}
 
-  Browser* const browser = browser_view_->browser();
-
+// static
+bool BrowserNativeWidgetMac::WillExecuteCommand(
+    Browser* browser,
+    int32_t command,
+    WindowOpenDisposition window_open_disposition,
+    bool is_before_first_responder) {
   if (is_before_first_responder) {
     // The specification for this private extensions API is incredibly vague.
     // For now, we avoid triggering chrome commands prior to giving the
@@ -464,12 +479,23 @@ bool BrowserNativeWidgetMac::ExecuteCommand(
     int32_t command,
     WindowOpenDisposition window_open_disposition,
     bool is_before_first_responder) {
-  if (!WillExecuteCommand(command, window_open_disposition,
-                          is_before_first_responder)) {
+  if (!browser_view_) {
     return false;
   }
+  return ExecuteCommand(browser_view_->browser(), command,
+                        window_open_disposition, is_before_first_responder);
+}
 
-  Browser* browser = browser_view_->browser();
+// static
+bool BrowserNativeWidgetMac::ExecuteCommand(
+    Browser* browser,
+    int32_t command,
+    WindowOpenDisposition window_open_disposition,
+    bool is_before_first_responder) {
+  if (!WillExecuteCommand(browser, command, window_open_disposition,
+                          is_before_first_responder)) {
+    return false;
+  }
 
   if (command == IDC_TOGGLE_VERTICAL_TABS) {
     if (auto* controller =
@@ -479,10 +505,13 @@ bool BrowserNativeWidgetMac::ExecuteCommand(
           is_vertical, tabs::VerticalTabStripEntryPoint::kMacViewMenu);
     }
   } else if (command == IDC_CLEAR_BROWSING_DATA) {
-    views::ElementTrackerViews::GetInstance()->NotifyCustomEvent(
-        browsing_data_important_sites_util::
-            kOpenClearBrowsingDataDialogViaAcceleratorEventId,
-        browser_view_);
+    BrowserView* browser_view = BrowserView::GetBrowserViewForBrowser(browser);
+    if (browser_view) {
+      views::ElementTrackerViews::GetInstance()->NotifyCustomEvent(
+          browsing_data_important_sites_util::
+              kOpenClearBrowsingDataDialogViaAcceleratorEventId,
+          browser_view);
+    }
   }
 
   chrome::ExecuteCommandWithDisposition(browser, command,
