diff --git chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.cc chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.cc
index 78d2b6eef6ca6..ebefa586902c9 100644
--- chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.cc
+++ chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.cc
@@ -94,9 +94,15 @@ WebUIConfigList CreateConfigs() {
 }  // namespace
 
 // static
-void ChromeUntrustedWebUIControllerFactory::RegisterInstance() {
+ChromeUntrustedWebUIControllerFactory*
+ChromeUntrustedWebUIControllerFactory::GetInstance() {
   static base::NoDestructor<ChromeUntrustedWebUIControllerFactory> instance;
-  content::WebUIControllerFactory::RegisterFactory(instance.get());
+  return instance.get();
+}
+
+// static
+void ChromeUntrustedWebUIControllerFactory::RegisterInstance() {
+  content::WebUIControllerFactory::RegisterFactory(GetInstance());
 }
 
 ChromeUntrustedWebUIControllerFactory::ChromeUntrustedWebUIControllerFactory()
diff --git chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.h chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.h
index fd16bca71a3f0..738afa6654e42 100644
--- chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.h
+++ chrome/browser/ui/webui/chrome_untrusted_web_ui_controller_factory.h
@@ -12,6 +12,7 @@ class ChromeUntrustedWebUIControllerFactory
  public:
   // Register the singleton instance of this class.
   static void RegisterInstance();
+  static ChromeUntrustedWebUIControllerFactory* GetInstance();
 
   ChromeUntrustedWebUIControllerFactory();
   ChromeUntrustedWebUIControllerFactory(
