diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 2203610c864eef6042936a83eb2b9797be93a0e0..f7266a73741dbc795b5d89ea2c4b48cc6cc066b5 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -11,6 +11,7 @@ import("//build/config/compiler/pgo/pgo.gni")
 import("//build/config/features.gni")
 import("//build/config/python.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -1649,6 +1650,7 @@ static_library("browser") {
     "//build/config/compiler:compiler_buildflags",
     "//build/config/linux/dbus:buildflags",
     "//cc",
+    "//cef/libcef/features:buildflags",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -2658,6 +2660,10 @@ static_library("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     sources += [
       "after_startup_task_utils_android.cc",
