diff --git chrome/browser/extensions/BUILD.gn chrome/browser/extensions/BUILD.gn
index d3f740db99e1df7bd660fe666922a27b668dfa03..cdba8cadb194c1f2d8615b50f9a94a9efacfcb19 100644
--- chrome/browser/extensions/BUILD.gn
+++ chrome/browser/extensions/BUILD.gn
@@ -3145,10 +3145,10 @@ source_set("browser_tests") {
 
     if (!is_chromeos) {
       # Chrome OS doesn't use background mode or download bubbles.
-      deps += [
-        "//chrome/browser/background/extensions",
-        "//chrome/browser/download/bubble",
-      ]
+      if (enable_background_mode) {
+        deps += [ "//chrome/browser/background/extensions" ]
+      }
+      deps += [ "//chrome/browser/download/bubble" ]
     }
 
     if (use_aura) {
