diff --git chrome/common/features.gni chrome/common/features.gni
index bcc0da3ae5a96..87617098debd3 100644
--- chrome/common/features.gni
+++ chrome/common/features.gni
@@ -7,6 +7,7 @@ import("//build/config/chromeos/ui_mode.gni")
 import("//build/config/compiler/compiler.gni")
 import("//build/config/dcheck_always_on.gni")
 import("//build/config/features.gni")
+import("//cef/libcef/features/features.gni")
 import("//components/compose/features.gni")
 import("//components/feed/features.gni")
 import("//components/nacl/features.gni")
@@ -31,7 +32,7 @@ assert(use_blink, "Chromium without blink shouldn't use anything in //chrome")
 declare_args() {
   # Enables the build to have logging enabled by default.
   # This is intended for use only in developer builds.
-  chrome_enable_logging_by_default = is_debug
+  chrome_enable_logging_by_default = is_debug || enable_cef
 
   # Platforms where Chrome x509 server certificate enterprise policies are
   # supported. This must must match the supported_on/future_on list of the
@@ -95,11 +96,13 @@ declare_args() {
   # optimize_webui was moved to ui/base/ui_features.gni
 }
 
+if (!enable_cef) {
 # Logging must be disabled by default in all official builds (including special
 # DCHECK-enabled builds). Logging is enabled by default for debug builds, and
 # may be selectively enabled by default for release builds.
 assert(!chrome_enable_logging_by_default || !is_official_build,
        "Logging must be disabled by default in Official builds")
+}
 
 # Use brlapi from brltty for braille display support.
 use_brlapi = is_chromeos_ash
