diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 33a6696fd803af080bb0a7d71b66e06d6950f86d..f59633e0da8a13395c7ad95b3f8413be15f70365 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -11,6 +11,7 @@ import("//build/config/compiler/pgo/pgo.gni")
 import("//build/config/features.gni")
 import("//build/config/python.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -2077,6 +2078,7 @@ source_set("browser") {
     ":unexpire_flags",
     "//base",
     "//cc",
+    "//cef/libcef/features:buildflags",
     "//chrome:extra_resources",
     "//chrome/browser/accessibility:impl",
     "//chrome/browser/account_settings",
@@ -2534,6 +2536,10 @@ source_set("browser") {
     ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     deps += [
       "//chrome/browser/android",
