diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index 8d5bdf428063..0aff2229bfa8 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -233,6 +233,7 @@ jumbo_static_library("safe_browsing") {
         "//chrome/common/safe_browsing:download_type_util",
         "//chrome/services/file_util/public/cpp",
         "//components/content_settings/core/browser",
+        "//components/gcm_driver:gcm_buildflags",
         "//components/language/core/common",
         "//components/prefs",
         "//components/safe_browsing/db",
