diff --git components/content_settings/renderer/content_settings_agent_impl.cc components/content_settings/renderer/content_settings_agent_impl.cc
index 0cc2bfa06245f..86655f6746459 100644
--- components/content_settings/renderer/content_settings_agent_impl.cc
+++ components/content_settings/renderer/content_settings_agent_impl.cc
@@ -148,7 +148,7 @@ ContentSetting GetContentSettingFromRules(
       return rule.GetContentSetting();
     }
   }
-  NOTREACHED_IN_MIGRATION();
+  // NOTREACHED_IN_MIGRATION();
   return CONTENT_SETTING_DEFAULT;
 }
 }  // namespace
