diff --git chrome/browser/BUILD.gn chrome/browser/BUILD.gn
index 5d956361e7bae..81e2df3f04e51 100644
--- chrome/browser/BUILD.gn
+++ chrome/browser/BUILD.gn
@@ -12,6 +12,7 @@ import("//build/config/features.gni")
 import("//build/config/linux/gtk/gtk.gni")
 import("//build/config/python.gni")
 import("//build/config/ui.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/browser/buildflags.gni")
 import("//chrome/browser/downgrade/buildflags.gni")
 import("//chrome/common/features.gni")
@@ -1925,6 +1926,7 @@ static_library("browser") {
     "//build:chromeos_buildflags",
     "//build/config/compiler:compiler_buildflags",
     "//cc",
+    "//cef/libcef/features",
     "//chrome:extra_resources",
     "//chrome:resources",
     "//chrome:strings",
@@ -2562,6 +2564,10 @@ static_library("browser") {
     deps += [ "//chrome/browser/ui/webui/connectors_internals:mojo_bindings" ]
   }
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (is_android) {
     sources += [
       "after_startup_task_utils_android.cc",
