diff --git chrome/browser/chrome_content_browser_client.cc chrome/browser/chrome_content_browser_client.cc
index 0b29bc11f07c..9c0ffba4dc20 100644
--- chrome/browser/chrome_content_browser_client.cc
+++ chrome/browser/chrome_content_browser_client.cc
@@ -1067,10 +1067,6 @@ void LaunchURL(
   }
 }
 
-std::string GetProduct() {
-  return version_info::GetProductNameAndVersionForUserAgent();
-}
-
 void MaybeAppendSecureOriginsAllowlistSwitch(base::CommandLine* cmdline) {
   // |allowlist| combines pref/policy + cmdline switch in the browser process.
   // For renderer and utility (e.g. NetworkService) processes the switch is the
@@ -1088,6 +1084,14 @@ void MaybeAppendSecureOriginsAllowlistSwitch(base::CommandLine* cmdline) {
 
 }  // namespace
 
+std::string GetProduct() {
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(switches::kProductVersion))
+    return command_line->GetSwitchValueASCII(switches::kProductVersion);
+
+  return version_info::GetProductNameAndVersionForUserAgent();
+}
+
 std::string GetUserAgent() {
   base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
   if (command_line->HasSwitch(switches::kUserAgent)) {
diff --git chrome/browser/chrome_content_browser_client.h chrome/browser/chrome_content_browser_client.h
index 069f597633a2..e81d7718cfca 100644
--- chrome/browser/chrome_content_browser_client.h
+++ chrome/browser/chrome_content_browser_client.h
@@ -84,7 +84,8 @@ enum class Channel;
 class ChromeHidDelegate;
 class ChromeSerialDelegate;
 
-// Returns the user agent of Chrome.
+// Returns the product and user agent of Chrome.
+std::string GetProduct();
 std::string GetUserAgent();
 
 blink::UserAgentMetadata GetUserAgentMetadata();
