diff --git chrome/renderer/BUILD.gn chrome/renderer/BUILD.gn
index 0903a4520939..be719c54f671 100644
--- chrome/renderer/BUILD.gn
+++ chrome/renderer/BUILD.gn
@@ -3,6 +3,7 @@
 # found in the LICENSE file.
 
 import("//build/config/features.gni")
+import("//cef/libcef/features/features.gni")
 import("//chrome/common/features.gni")
 import("//components/nacl/features.gni")
 import("//components/offline_pages/buildflags/features.gni")
@@ -109,6 +110,7 @@ static_library("renderer") {
   defines = []
 
   deps = [
+    "//cef/libcef/features",
     "//chrome:resources",
     "//chrome:strings",
     "//chrome/common",
@@ -169,6 +171,10 @@ static_library("renderer") {
 
   configs += [ "//build/config/compiler:wexit_time_destructors" ]
 
+  if (enable_cef) {
+    configs += [ "//cef/libcef/features:config" ]
+  }
+
   if (enable_nacl) {
     deps += [
       "//components/nacl/loader",
