diff --git chrome/browser/devtools/chrome_devtools_manager_delegate.cc chrome/browser/devtools/chrome_devtools_manager_delegate.cc
index c9f49cd198cd29a4f4ad3c2321c9b7a874b5e899..e36ba2fd74b2f4ae12265a9ba32d62005eec760b 100644
--- chrome/browser/devtools/chrome_devtools_manager_delegate.cc
+++ chrome/browser/devtools/chrome_devtools_manager_delegate.cc
@@ -14,6 +14,7 @@
 #include "base/strings/utf_string_conversions.h"
 #include "base/values.h"
 #include "build/build_config.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/devtools/chrome_devtools_session.h"
 #include "chrome/browser/devtools/device/android_device_manager.h"
 #include "chrome/browser/devtools/device/tcp_device_provider.h"
@@ -74,6 +75,10 @@
 static_assert(!BUILDFLAG(IS_ANDROID),
               "This file should not be included in Android build");
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/extensions/chrome_extension_util.h"
+#endif
+
 using content::DevToolsAgentHost;
 
 const char ChromeDevToolsManagerDelegate::kTypeApp[] = "app";
@@ -323,6 +328,12 @@ std::string ChromeDevToolsManagerDelegate::GetTargetType(
     return DevToolsAgentHost::kTypePage;
   }
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (cef::IsAlloyContents(web_contents, /*primary_only=*/true)) {
+    return DevToolsAgentHost::kTypePage;
+  }
+#endif
+
   return DevToolsAgentHost::kTypeOther;
 }
 
