diff --git components/performance_manager/graph/frame_node_impl.cc components/performance_manager/graph/frame_node_impl.cc
index c277c918b08ce..8ca8fd2670edd 100644
--- components/performance_manager/graph/frame_node_impl.cc
+++ components/performance_manager/graph/frame_node_impl.cc
@@ -355,8 +355,9 @@ void FrameNodeImpl::SetIntersectsViewport(bool intersects_viewport) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
   // The intersection with the viewport of the outermost main frame or embedder
   // is not tracked.
-  DCHECK(parent_or_outer_document_or_embedder());
-  intersects_viewport_.SetAndMaybeNotify(this, intersects_viewport);
+  if (parent_or_outer_document_or_embedder()) {
+    intersects_viewport_.SetAndMaybeNotify(this, intersects_viewport);
+  }
 }
 
 void FrameNodeImpl::SetInitialVisibility(Visibility visibility) {
