diff --git chrome/browser/ui/views/sad_tab_view.cc chrome/browser/ui/views/sad_tab_view.cc
index 5a3aa5c05907e..3630c7496fb70 100644
--- chrome/browser/ui/views/sad_tab_view.cc
+++ chrome/browser/ui/views/sad_tab_view.cc
@@ -680,6 +680,11 @@ void SadTabView::OnBoundsChanged(const gfx::Rect& previous_bounds) {
   title_->SizeToFit(max_width);
 }
 
+// static
+std::u16string SadTabView::ErrorToString(int error_code) {
+  return ::ErrorToString(error_code);
+}
+
 SadTab* SadTab::Create(content::WebContents* web_contents, SadTabKind kind) {
   return new SadTabView(web_contents, kind);
 }
diff --git chrome/browser/ui/views/sad_tab_view.h chrome/browser/ui/views/sad_tab_view.h
index d6cb0e1a28eb2..59aca7bbb17c3 100644
--- chrome/browser/ui/views/sad_tab_view.h
+++ chrome/browser/ui/views/sad_tab_view.h
@@ -56,6 +56,8 @@ class SadTabView : public SadTab, public views::View {
   // Overridden from views::View:
   void OnBoundsChanged(const gfx::Rect& previous_bounds) override;
 
+  static std::u16string ErrorToString(int error_code);
+
  protected:
   // Overridden from views::View:
   void OnPaint(gfx::Canvas* canvas) override;
