diff --git components/content_settings/renderer/content_settings_agent_impl.cc components/content_settings/renderer/content_settings_agent_impl.cc
index e13faa0b945c9e655a059da24658aa0ea31315dc..20709a9b714ad161c49371b8e5436418141fd49e 100644
--- components/content_settings/renderer/content_settings_agent_impl.cc
+++ components/content_settings/renderer/content_settings_agent_impl.cc
@@ -143,7 +143,7 @@ ContentSetting GetContentSettingFromRules(
       return rule.GetContentSetting();
     }
   }
-  NOTREACHED();
+  return CONTENT_SETTING_DEFAULT;
 }
 }  // namespace
 
