diff --git chrome/browser/safe_browsing/BUILD.gn chrome/browser/safe_browsing/BUILD.gn
index 22fd2393c9c70..e033dbd0d6d10 100644
--- chrome/browser/safe_browsing/BUILD.gn
+++ chrome/browser/safe_browsing/BUILD.gn
@@ -34,6 +34,7 @@ static_library("safe_browsing") {
     "//components/enterprise:enterprise",
     "//components/enterprise/buildflags",
     "//components/enterprise/common:strings",
+    "//components/gcm_driver:gcm_buildflags",
     "//components/keyed_service/content",
     "//components/language/core/browser",
     "//components/no_state_prefetch/browser",
