diff --git components/component_updater/component_updater_service.h components/component_updater/component_updater_service.h
index d8d54fa1b2475717cdb793f63012530fc43ed663..05312ee4215d49fe3d548d4c847b964ad2271b1d 100644
--- components/component_updater/component_updater_service.h
+++ components/component_updater/component_updater_service.h
@@ -19,6 +19,7 @@
 #include "build/build_config.h"
 #include "components/update_client/update_client.h"

+class CefComponentUpdaterImpl;
 class ComponentsHandler;
 class PluginObserver;

@@ -235,6 +236,7 @@ class OnDemandUpdater {
   friend class ash::SmartDimComponentIntegrationTest;
   friend class CrOSComponentInstaller;
 #endif  // BUILDFLAG(IS_CHROMEOS)
+  friend class ::CefComponentUpdaterImpl;
   friend class IwaKeyDistributionComponentInstallerPolicy;

   // Triggers an update check for a component. |id| is a value
