diff --git components/component_updater/component_updater_service.h components/component_updater/component_updater_service.h
index 6c9bd84298bb86e244a4117d22d545e453af870c..9f2722f57335df442d9bfec0d936759d4612ff9c 100644
--- components/component_updater/component_updater_service.h
+++ components/component_updater/component_updater_service.h
@@ -21,6 +21,7 @@
 #include "build/build_config.h"
 #include "components/update_client/update_client.h"
 
+class CefComponentUpdaterImpl;
 class ComponentsHandler;
 class PluginObserver;
 
@@ -243,6 +244,7 @@ class OnDemandUpdater {
   friend class ash::SmartDimComponentIntegrationTest;
   friend class CrOSComponentInstaller;
 #endif  // BUILDFLAG(IS_CHROMEOS)
+  friend class ::CefComponentUpdaterImpl;
   friend class IwaKeyDistributionComponentInstallerPolicy;
   friend class PlatformRuntimeComponentInstallerPolicy;
 
