diff --git components/component_updater/component_updater_service.h components/component_updater/component_updater_service.h
index 39acb4430d9c5cd00c17a83ad82e74adf25dfc01..70f34b1cc6435240fb9285e7239ea3080bb9158c 100644
--- components/component_updater/component_updater_service.h
+++ components/component_updater/component_updater_service.h
@@ -21,6 +21,7 @@
 #include "build/build_config.h"
 #include "components/update_client/update_client.h"
 
+class CefComponentUpdaterImpl;
 class ComponentsHandler;
 class PluginObserver;
 
@@ -243,6 +244,7 @@ class OnDemandUpdater {
   friend class ash::SmartDimComponentIntegrationTest;
   friend class CrOSComponentInstaller;
 #endif  // BUILDFLAG(IS_CHROMEOS)
+  friend class ::CefComponentUpdaterImpl;
   friend class IwaKeyDistributionComponentInstallerPolicy;
   friend class PlatformRuntimeComponentInstallerPolicy;
   friend void UpdateAIEmbeddingsComponentOnDemand(Priority priority,
