diff --git cc/scheduler/compositor_frame_reporting_controller.cc cc/scheduler/compositor_frame_reporting_controller.cc
index f1587ed158d4..1b17021fd229 100644
--- cc/scheduler/compositor_frame_reporting_controller.cc
+++ cc/scheduler/compositor_frame_reporting_controller.cc
@@ -31,8 +31,8 @@ void CompositorFrameReportingController::WillBeginImplFrame() {
 
 void CompositorFrameReportingController::WillBeginMainFrame() {
   DCHECK(reporters_[PipelineStage::kBeginImplFrame]);
-  DCHECK_NE(reporters_[PipelineStage::kBeginMainFrame],
-            reporters_[PipelineStage::kBeginImplFrame]);
+  DCHECK(reporters_[PipelineStage::kBeginMainFrame] !=
+         reporters_[PipelineStage::kBeginImplFrame]);
   reporters_[PipelineStage::kBeginImplFrame]->StartStage(
       "SendBeginMainFrameToCommit");
   AdvanceReporterStage(PipelineStage::kBeginImplFrame,
