diff --git components/component_updater/component_updater_service.h components/component_updater/component_updater_service.h
index 18308f02a6273fe752f8cc41b3b8bdd1293f1972..98301dca212ebf3b658350a844f640aa7c35272e 100644
--- components/component_updater/component_updater_service.h
+++ components/component_updater/component_updater_service.h
@@ -21,6 +21,7 @@
 #include "build/build_config.h"
 #include "components/update_client/update_client.h"
 
+class CefComponentUpdaterImpl;
 class ComponentsHandler;
 class PluginObserver;
 
@@ -243,6 +244,7 @@ class OnDemandUpdater {
   friend class ash::SmartDimComponentIntegrationTest;
   friend class CrOSComponentInstaller;
 #endif  // BUILDFLAG(IS_CHROMEOS)
+  friend class ::CefComponentUpdaterImpl;
   friend class IwaKeyDistributionComponentInstallerPolicy;
   friend class PlatformRuntimeComponentInstallerPolicy;
   friend void UpdateAIEmbeddingsComponentOnDemand(Priority priority,
