diff --git components/component_updater/component_updater_service.h components/component_updater/component_updater_service.h
index c16c8bbd9b59b10a7c281a718d97026d7713fe76..5dbb7a21db7f52790dbd1535c41004ecc7cc714c 100644
--- components/component_updater/component_updater_service.h
+++ components/component_updater/component_updater_service.h
@@ -21,6 +21,7 @@
 #include "build/build_config.h"
 #include "components/update_client/update_client.h"
 
+class CefComponentUpdaterImpl;
 class ComponentsHandler;
 class PluginObserver;
 
@@ -242,6 +243,7 @@ class OnDemandUpdater {
   friend class ash::SmartDimComponentIntegrationTest;
   friend class CrOSComponentInstaller;
 #endif  // BUILDFLAG(IS_CHROMEOS)
+  friend class ::CefComponentUpdaterImpl;
   friend class IwaKeyDistributionComponentInstallerPolicy;
 
   // Triggers an update check for a component. |id| is a value
