diff --git content/browser/renderer_host/navigation_throttle_registry_impl.cc content/browser/renderer_host/navigation_throttle_registry_impl.cc
index 2bbc7d8c31b0ea692f103c48dc5e8dd0c7a0c503..acd3845480558cc877c0251f81fe266e1acb042e 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.cc
+++ content/browser/renderer_host/navigation_throttle_registry_impl.cc
@@ -276,13 +276,18 @@ NavigationHandle& NavigationThrottleRegistryImpl::GetNavigationHandle() {
 }
 
 void NavigationThrottleRegistryImpl::AddThrottle(
-    std::unique_ptr<NavigationThrottle> navigation_throttle) {
+    std::unique_ptr<NavigationThrottle> navigation_throttle,
+    bool first) {
   CHECK(navigation_throttle);
   TRACE_EVENT(TRACE_DISABLED_BY_DEFAULT("navigation"),
               "NavigationThrottleRegistryImpl::AddThrottle",
               "navigation_throttle", navigation_throttle->GetNameForLogging());
   CHECK(!navigation_request_->IsInitialWebUISyncNavigation());
-  throttles_.push_back(std::move(navigation_throttle));
+  if (first) {
+    throttles_.emplace(throttles_.begin(), std::move(navigation_throttle));
+  } else {
+    throttles_.push_back(std::move(navigation_throttle));
+  }
 }
 
 bool NavigationThrottleRegistryImpl::HasThrottle(const std::string& name) {
diff --git content/browser/renderer_host/navigation_throttle_registry_impl.h content/browser/renderer_host/navigation_throttle_registry_impl.h
index fc44a4a0ca10b61ba2414adde737182da0d0bec4..009ae07219c20d5ff43e07388f73513a2c25f4d5 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.h
+++ content/browser/renderer_host/navigation_throttle_registry_impl.h
@@ -135,7 +135,8 @@ class CONTENT_EXPORT NavigationThrottleRegistryImpl
   // Implements NavigationThrottleRegistry:
   NavigationHandle& GetNavigationHandle() override;
   void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) override;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) override;
   bool HasThrottle(const std::string& name) override;
   bool EraseThrottleForTesting(const std::string& name) override;
 
diff --git content/public/browser/navigation_throttle_registry.h content/public/browser/navigation_throttle_registry.h
index d541287beae6717847de91d53f3cf1f5c084f2fa..3132993cf0bd518d3c0c7575676c40ede0ba39da 100644
--- content/public/browser/navigation_throttle_registry.h
+++ content/public/browser/navigation_throttle_registry.h
@@ -28,7 +28,8 @@ class CONTENT_EXPORT NavigationThrottleRegistry {
   // be called on `navigation_throttle`.
   // AddThrottle() disallows `navigation_throttle` to be nullptr.
   virtual void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) = 0;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) = 0;
 
   // Checks if the registry contains a throttle with the given name. Returns
   // true if the throttle is found.
