diff --git content/browser/renderer_host/navigation_throttle_registry_impl.cc content/browser/renderer_host/navigation_throttle_registry_impl.cc
index cc91d46713d9a04a75213ba07bacfebabb468767..2c4430aafbcafae9a0dfb7de71b860fb42f001c2 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.cc
+++ content/browser/renderer_host/navigation_throttle_registry_impl.cc
@@ -277,13 +277,18 @@ NavigationHandle& NavigationThrottleRegistryImpl::GetNavigationHandle() {
 }
 
 void NavigationThrottleRegistryImpl::AddThrottle(
-    std::unique_ptr<NavigationThrottle> navigation_throttle) {
+    std::unique_ptr<NavigationThrottle> navigation_throttle,
+    bool first) {
   CHECK(navigation_throttle);
   TRACE_EVENT(TRACE_DISABLED_BY_DEFAULT("navigation"),
               "NavigationThrottleRegistryImpl::AddThrottle",
               "navigation_throttle", navigation_throttle->GetNameForLogging());
   CHECK(!navigation_request_->IsInitialWebUINavigation());
-  throttles_.push_back(std::move(navigation_throttle));
+  if (first) {
+    throttles_.emplace(throttles_.begin(), std::move(navigation_throttle));
+  } else {
+    throttles_.push_back(std::move(navigation_throttle));
+  }
 }
 
 bool NavigationThrottleRegistryImpl::HasThrottle(const std::string& name) {
diff --git content/browser/renderer_host/navigation_throttle_registry_impl.h content/browser/renderer_host/navigation_throttle_registry_impl.h
index 0970bc9bda0ce10c415cf5f376be1a8db2c62b68..e763e98e6c094f4bffc7ed5be24ed3dff168e48b 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.h
+++ content/browser/renderer_host/navigation_throttle_registry_impl.h
@@ -136,7 +136,8 @@ class CONTENT_EXPORT NavigationThrottleRegistryImpl
   // Implements NavigationThrottleRegistry:
   NavigationHandle& GetNavigationHandle() override;
   void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) override;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) override;
   bool HasThrottle(const std::string& name) override;
   bool EraseThrottleForTesting(const std::string& name) override;
 
diff --git content/public/browser/navigation_throttle_registry.h content/public/browser/navigation_throttle_registry.h
index d541287beae6717847de91d53f3cf1f5c084f2fa..3132993cf0bd518d3c0c7575676c40ede0ba39da 100644
--- content/public/browser/navigation_throttle_registry.h
+++ content/public/browser/navigation_throttle_registry.h
@@ -28,7 +28,8 @@ class CONTENT_EXPORT NavigationThrottleRegistry {
   // be called on `navigation_throttle`.
   // AddThrottle() disallows `navigation_throttle` to be nullptr.
   virtual void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) = 0;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) = 0;
 
   // Checks if the registry contains a throttle with the given name. Returns
   // true if the throttle is found.
