Index: public/renderer/content_renderer_client.cc
===================================================================
--- public/renderer/content_renderer_client.cc	(revision 275973)
+++ public/renderer/content_renderer_client.cc	(working copy)
@@ -92,7 +92,6 @@
   return false;
 }
 
-#ifdef OS_ANDROID
 bool ContentRendererClient::HandleNavigation(
     RenderFrame* render_frame,
     DocumentState* document_state,
@@ -104,7 +103,6 @@
     bool is_redirect) {
   return false;
 }
-#endif
 
 bool ContentRendererClient::ShouldFork(blink::WebFrame* frame,
                                        const GURL& url,
Index: public/renderer/content_renderer_client.h
===================================================================
--- public/renderer/content_renderer_client.h	(revision 275973)
+++ public/renderer/content_renderer_client.h	(working copy)
@@ -176,7 +176,6 @@
   // Returns true if a popup window should be allowed.
   virtual bool AllowPopup();
 
-#ifdef OS_ANDROID
   // TODO(sgurun) This callback is deprecated and will be removed as soon
   // as android webview completes implementation of a resource throttle based
   // shouldoverrideurl implementation. See crbug.com/325351
@@ -191,7 +190,6 @@
                                 blink::WebNavigationType type,
                                 blink::WebNavigationPolicy default_policy,
                                 bool is_redirect);
-#endif
 
   // Returns true if we should fork a new process for the given navigation.
   // If |send_referrer| is set to false (which is the default), no referrer
Index: renderer/render_frame_impl.cc
===================================================================
--- renderer/render_frame_impl.cc	(revision 275973)
+++ renderer/render_frame_impl.cc	(working copy)
@@ -3143,7 +3143,6 @@
     WebNavigationType type,
     WebNavigationPolicy default_policy,
     bool is_redirect) {
-#ifdef OS_ANDROID
   // The handlenavigation API is deprecated and will be removed once
   // crbug.com/325351 is resolved.
   if (request.url() != GURL(kSwappedOutURL) &&
@@ -3158,7 +3157,6 @@
           is_redirect)) {
     return blink::WebNavigationPolicyIgnore;
   }
-#endif
 
   Referrer referrer(RenderViewImpl::GetReferrerFromRequest(frame, request));
 
