diff --git content/browser/renderer_host/navigation_throttle_registry_impl.cc content/browser/renderer_host/navigation_throttle_registry_impl.cc
index 41f3d8bf9e60edd283792e68850a902f7411a1cd..80dba42bb628b150c12bb9c0269baab62c0b7330 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.cc
+++ content/browser/renderer_host/navigation_throttle_registry_impl.cc
@@ -164,11 +164,16 @@ NavigationHandle& NavigationThrottleRegistryImpl::GetNavigationHandle() {
 }
 
 void NavigationThrottleRegistryImpl::AddThrottle(
-    std::unique_ptr<NavigationThrottle> navigation_throttle) {
+    std::unique_ptr<NavigationThrottle> navigation_throttle,
+    bool first) {
   CHECK(navigation_throttle);
   TRACE_EVENT1("navigation", "NavigationThrottleRegistryImpl::AddThrottle",
                "navigation_throttle", navigation_throttle->GetNameForLogging());
-  throttles_.push_back(std::move(navigation_throttle));
+  if (first) {
+    throttles_.emplace(throttles_.begin(), std::move(navigation_throttle));
+  } else {
+    throttles_.push_back(std::move(navigation_throttle));
+  }
 }
 
 bool NavigationThrottleRegistryImpl::HasThrottle(const std::string& name) {
diff --git content/browser/renderer_host/navigation_throttle_registry_impl.h content/browser/renderer_host/navigation_throttle_registry_impl.h
index fb94e73de38ae34bd7bb29667f7c36010fd3a082..fd6a550464dbbb2620487062cce00fe56b9a43d4 100644
--- content/browser/renderer_host/navigation_throttle_registry_impl.h
+++ content/browser/renderer_host/navigation_throttle_registry_impl.h
@@ -84,7 +84,8 @@ class NavigationThrottleRegistryImpl : public NavigationThrottleRegistryBase {
   // Implements NavigationThrottleRegistry:
   NavigationHandle& GetNavigationHandle() override;
   void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) override;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) override;
   bool HasThrottle(const std::string& name) override;
   bool EraseThrottleForTesting(const std::string& name) override;
 
diff --git content/public/browser/navigation_throttle_registry.h content/public/browser/navigation_throttle_registry.h
index d541287beae6717847de91d53f3cf1f5c084f2fa..3132993cf0bd518d3c0c7575676c40ede0ba39da 100644
--- content/public/browser/navigation_throttle_registry.h
+++ content/public/browser/navigation_throttle_registry.h
@@ -28,7 +28,8 @@ class CONTENT_EXPORT NavigationThrottleRegistry {
   // be called on `navigation_throttle`.
   // AddThrottle() disallows `navigation_throttle` to be nullptr.
   virtual void AddThrottle(
-      std::unique_ptr<NavigationThrottle> navigation_throttle) = 0;
+      std::unique_ptr<NavigationThrottle> navigation_throttle,
+      bool first = false) = 0;
 
   // Checks if the registry contains a throttle with the given name. Returns
   // true if the throttle is found.
