diff --git public/renderer/content_renderer_client.cc public/renderer/content_renderer_client.cc
index 3cab7f9..98a151f 100644
--- public/renderer/content_renderer_client.cc
+++ public/renderer/content_renderer_client.cc
@@ -97,7 +97,6 @@ bool ContentRendererClient::AllowPopup() {
   return false;
 }
 
-#if defined(OS_ANDROID)
 bool ContentRendererClient::HandleNavigation(
     RenderFrame* render_frame,
     bool is_content_initiated,
@@ -110,6 +109,7 @@ bool ContentRendererClient::HandleNavigation(
   return false;
 }
 
+#if defined(OS_ANDROID)
 bool ContentRendererClient::ShouldUseMediaPlayerForURL(const GURL& url) {
   return false;
 }
diff --git public/renderer/content_renderer_client.h public/renderer/content_renderer_client.h
index 0764782..6a12636 100644
--- public/renderer/content_renderer_client.h
+++ public/renderer/content_renderer_client.h
@@ -197,7 +197,6 @@ class CONTENT_EXPORT ContentRendererClient {
   // Returns true if a popup window should be allowed.
   virtual bool AllowPopup();
 
-#if defined(OS_ANDROID)
   // TODO(sgurun) This callback is deprecated and will be removed as soon
   // as android webview completes implementation of a resource throttle based
   // shouldoverrideurl implementation. See crbug.com/325351
@@ -213,6 +212,7 @@ class CONTENT_EXPORT ContentRendererClient {
                                 blink::WebNavigationPolicy default_policy,
                                 bool is_redirect);
 
+#if defined(OS_ANDROID)
   // Indicates if the Android MediaPlayer should be used instead of Chrome's
   // built in media player for the given |url|. Defaults to false.
   virtual bool ShouldUseMediaPlayerForURL(const GURL& url);
