diff --git src/dawn/tests/BUILD.gn src/dawn/tests/BUILD.gn
index 2dea894fa5e379aa6e6ef9bcfc3a9d27fd1eec95..b59c79997991a280183df3252a564486f9589bda 100644
--- src/dawn/tests/BUILD.gn
+++ src/dawn/tests/BUILD.gn
@@ -102,8 +102,8 @@ if (!dawn_use_built_dxc && build_with_chromium && is_win) {
   # end-to-end test in Chromium.
   copy("copy_dxc_binary") {
     sources = [
-      "$windows_sdk_path/bin/$windows_sdk_version/$target_cpu/dxcompiler.dll",
-      "$windows_sdk_path/bin/$windows_sdk_version/$target_cpu/dxil.dll",
+      "$windows_sdk_path/Redist/D3D/$target_cpu/dxcompiler.dll",
+      "$windows_sdk_path/Redist/D3D/$target_cpu/dxil.dll",
     ]
 
     outputs = [ "$root_out_dir/{{source_file_part}}" ]
diff --git third_party/directx-shader-compiler/BUILD.gn third_party/directx-shader-compiler/BUILD.gn
index 21dc5de77507673ff1450737e6d4fbf49454d72f..3c85a74aa92336da35a0a33dc61be9fcdbf15fd1 100644
--- third_party/directx-shader-compiler/BUILD.gn
+++ third_party/directx-shader-compiler/BUILD.gn
@@ -2476,8 +2476,7 @@ executable("dxc") {
 if (is_win) {
   import("//build/config/win/visual_studio_version.gni")
   copy("copy_dxil_dll") {
-    sources =
-        [ "$windows_sdk_path/bin/$windows_sdk_version/$target_cpu/dxil.dll" ]
+    sources = [ "$windows_sdk_path/Redist/D3D/$target_cpu/dxil.dll" ]
     outputs = [ "$root_out_dir/{{source_file_part}}" ]
   }
 }
