diff --git src/dawn/tests/BUILD.gn src/dawn/tests/BUILD.gn
index 685133842f2c0e8ab290c75c4d721608ca27d48a..3faeb6b445d2388ccbe15d7b9905916cb9b9a0bb 100644
--- src/dawn/tests/BUILD.gn
+++ src/dawn/tests/BUILD.gn
@@ -104,7 +104,7 @@ if (!dawn_use_built_dxc && build_with_chromium && is_win) {
   # end-to-end test in Chromium.
   copy("copy_dxc_binary") {
     sources = [
-      "$windows_sdk_path/bin/$windows_sdk_version/$target_cpu/dxcompiler.dll",
+      "$windows_sdk_path/Redist/D3D/$target_cpu/dxcompiler.dll",
     ]
 
     outputs = [ "$root_out_dir/{{source_file_part}}" ]
