diff --git content/browser/devtools/devtools_http_handler.cc content/browser/devtools/devtools_http_handler.cc
index cae7c566146e..95b4c23c53c9 100644
--- content/browser/devtools/devtools_http_handler.cc
+++ content/browser/devtools/devtools_http_handler.cc
@@ -564,7 +564,7 @@ void DevToolsHttpHandler::OnJsonRequest(
     version.SetString("Protocol-Version",
                       DevToolsAgentHost::GetProtocolVersion());
     version.SetString("WebKit-Version", GetWebKitVersion());
-    version.SetString("Browser", GetContentClient()->GetProduct());
+    version.SetString("Browser", GetContentClient()->GetChromeProduct());
     version.SetString("User-Agent", GetContentClient()->GetUserAgent());
     version.SetString("V8-Version", V8_VERSION_STRING);
     std::string host = info.headers["host"];
diff --git content/public/common/content_client.h content/public/common/content_client.h
index caf91a31abb9..492d12fbbfda 100644
--- content/public/common/content_client.h
+++ content/public/common/content_client.h
@@ -148,6 +148,10 @@ class CONTENT_EXPORT ContentClient {
   // Used as part of the user agent string.
   virtual std::string GetProduct() const;
 
+  // Returns the Chrome-specific product string. This is used for compatibility
+  // purposes with external tools like Selenium.
+  virtual std::string GetChromeProduct() const { return GetProduct(); }
+
   // Returns the user agent.  Content may cache this value.
   virtual std::string GetUserAgent() const;
 
