diff --git content/browser/devtools/devtools_http_handler.cc content/browser/devtools/devtools_http_handler.cc
index eaf44246e9ec..8b1af37e2fd1 100644
--- content/browser/devtools/devtools_http_handler.cc
+++ content/browser/devtools/devtools_http_handler.cc
@@ -567,7 +567,7 @@ void DevToolsHttpHandler::OnJsonRequest(
     version.SetString("Protocol-Version",
                       DevToolsAgentHost::GetProtocolVersion());
     version.SetString("WebKit-Version", GetWebKitVersion());
-    version.SetString("Browser", GetContentClient()->GetProduct());
+    version.SetString("Browser", GetContentClient()->GetChromeProduct());
     version.SetString("User-Agent", GetContentClient()->GetUserAgent());
     version.SetString("V8-Version", V8_VERSION_STRING);
     std::string host = info.headers["host"];
diff --git content/public/common/content_client.h content/public/common/content_client.h
index 528fd6abf6a6..a45793539166 100644
--- content/public/common/content_client.h
+++ content/public/common/content_client.h
@@ -150,6 +150,10 @@ class CONTENT_EXPORT ContentClient {
   // Used as part of the user agent string.
   virtual std::string GetProduct() const;
 
+  // Returns the Chrome-specific product string. This is used for compatibility
+  // purposes with external tools like Selenium.
+  virtual std::string GetChromeProduct() const { return GetProduct(); }
+
   // Returns the user agent.  Content may cache this value.
   // TODO(yhirano): Move this to ContentBrowserClient.
   virtual std::string GetUserAgent() const;
