diff --git components/embedder_support/user_agent_utils.cc components/embedder_support/user_agent_utils.cc
index 5f72908b2a73a27989f4449d498356045480c469..483657dd42a9206d3b224c1543c2179faab9653d 100644
--- components/embedder_support/user_agent_utils.cc
+++ components/embedder_support/user_agent_utils.cc
@@ -25,6 +25,7 @@
 #include "base/version.h"
 #include "build/branding_buildflags.h"
 #include "build/build_config.h"
+#include "cef/libcef/features/features.h"
 #include "components/embedder_support/pref_names.h"
 #include "components/embedder_support/switches.h"
 #include "components/policy/core/common/policy_pref_names.h"
@@ -53,6 +54,10 @@
 #include <sys/utsname.h>
 #endif
 
+#if BUILDFLAG(ENABLE_CEF)
+constexpr char kUserAgentProductAndVersion[] = "user-agent-product";
+#endif
+
 namespace embedder_support {
 
 namespace {
@@ -455,6 +460,13 @@ std::string BuildOSCpuInfo(
 }  // namespace
 
 std::string GetProductAndVersion() {
+#if BUILDFLAG(ENABLE_CEF)
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kUserAgentProductAndVersion)) {
+    return command_line->GetSwitchValueASCII(kUserAgentProductAndVersion);
+  }
+#endif
+
   return base::FeatureList::IsEnabled(
              blink::features::kReduceUserAgentMinorVersion)
              ? version_info::GetProductNameAndVersionForReducedUserAgent()
