diff --git components/embedder_support/user_agent_utils.cc components/embedder_support/user_agent_utils.cc
index 054a0dc411638ac8e2586ac9a1dd45c821249abb..2b058646e1b37a1cbaf8336f991a2e856ab34d6e 100644
--- components/embedder_support/user_agent_utils.cc
+++ components/embedder_support/user_agent_utils.cc
@@ -27,6 +27,7 @@
 #include "build/branding_buildflags.h"
 #include "build/build_config.h"
 #include "build/util/chromium_git_revision.h"
+#include "cef/libcef/features/features.h"
 #include "components/embedder_support/pref_names.h"
 #include "components/embedder_support/switches.h"
 #include "components/policy/core/common/policy_pref_names.h"
@@ -60,6 +61,10 @@
 #include "device/vr/public/cpp/features.h"
 #endif  // BUILDFLAG(IS_ANDROID) && BUILDFLAG(ENABLE_VR)
 
+#if BUILDFLAG(ENABLE_CEF)
+constexpr char kUserAgentProductAndVersion[] = "user-agent-product";
+#endif
+
 namespace embedder_support {
 
 namespace {
@@ -469,6 +474,13 @@ std::string BuildOSCpuInfo(
 
 std::string GetProductAndVersion(
     UserAgentReductionEnterprisePolicyState user_agent_reduction) {
+#if BUILDFLAG(ENABLE_CEF)
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kUserAgentProductAndVersion)) {
+    return command_line->GetSwitchValueASCII(kUserAgentProductAndVersion);
+  }
+#endif
+
   return ShouldReduceUserAgentMinorVersion(user_agent_reduction)
              ? version_info::GetProductNameAndVersionForReducedUserAgent(
                    blink::features::kUserAgentFrozenBuildVersion.Get())
