diff --git extension_process_policy.h extension_process_policy.h
index 7231c0f..2323483 100644
--- extension_process_policy.h
+++ extension_process_policy.h
@@ -5,12 +5,12 @@
 #ifndef CHROME_COMMON_EXTENSIONS_EXTENSION_PROCESS_POLICY_H_
 #define CHROME_COMMON_EXTENSIONS_EXTENSION_PROCESS_POLICY_H_
 
-class ExtensionSet;
 class GURL;
 
 namespace extensions {
 
 class Extension;
+class ExtensionSet;
 
 // Returns the extension for the given URL.  Excludes extension objects for
 // bookmark apps, which do not use the app process model.
