diff --git gpu/command_buffer/service/feature_info.cc gpu/command_buffer/service/feature_info.cc
index 32e72c6..495ccdd 100644
--- gpu/command_buffer/service/feature_info.cc
+++ gpu/command_buffer/service/feature_info.cc
@@ -1524,8 +1524,6 @@ void FeatureInfo::InitializeFloatAndHalfFloatFeatures(
     glBindFramebufferEXT(GL_FRAMEBUFFER, static_cast<GLuint>(fb_binding));
     glBindTexture(GL_TEXTURE_2D, static_cast<GLuint>(tex_binding));
 
-    DCHECK_EQ(glGetError(), static_cast<GLuint>(GL_NO_ERROR));
-
     if (status_rgba == GL_FRAMEBUFFER_COMPLETE) {
       feature_flags_.chromium_color_buffer_float_rgba = true;
       if (!disallowed_features_.chromium_color_buffer_float_rgba)
