diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 972382ea6c9043e06779330a11de90f87e2a11cd..d15d496c773985a2c0da4148f5500f7b54465c5a 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1557,6 +1557,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
