diff --git tools/gritsettings/resource_ids tools/gritsettings/resource_ids
index e823c9b9edfd..2402613c8d44 100644
--- tools/gritsettings/resource_ids
+++ tools/gritsettings/resource_ids
@@ -431,4 +431,11 @@
   # Please read the header and find the right section above instead.
 
   # Resource ids starting at 31000 are reserved for projects built on Chromium.
+
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [31500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [32000],
+  },
 }
