diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 5ac909144d0a82914e6f0e89cc885dd8bac05d45..11130e971b97cab871f8ec1d41d655a21d1ffb0d 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1648,6 +1648,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
