diff --git resource_ids resource_ids
index c500512..0ac0e98 100644
--- resource_ids
+++ resource_ids
@@ -358,5 +358,12 @@
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [31500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [32000],
+  },
+
   # Resource ids starting at 31000 are reserved for projects built on Chromium.
 }
