diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 2f4ab49cbe968..5e78796e1820b 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1532,11 +1532,18 @@
 
   "third_party/search_engines_data/resources/search_engines_scaled_resources.grd": {
     "structures": [10100],
-  }
+  },
 
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
