diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index aeb9f8b8eaf68c076348278f6a8b394cae2ef827..a2363427c3b66615f1ef410cbb0be7c449d388b3 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1600,6 +1600,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
