diff --git tools/gritsettings/resource_ids tools/gritsettings/resource_ids
index dc660db144df..6fc116b2c8ff 100644
--- tools/gritsettings/resource_ids
+++ tools/gritsettings/resource_ids
@@ -434,4 +434,11 @@
   # Please read the header and find the right section above instead.
 
   # Resource ids starting at 31000 are reserved for projects built on Chromium.
+
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [31500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [32000],
+  },
 }
