diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 8749c14d44f1b32abd54c21a87087f958680167d..f7906e3fa1d29fd034fd42c07eedc93df74bb4b5 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1807,6 +1807,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
