diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index bba1bed7b351171434d9b6d8170f6b015f5acb6c..0106d6079281a2f4a418468a4ccc17c46b0ccbd6 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1578,6 +1578,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
