diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index f78fa04ff6bfc43d2554ea833945e89b08e8ec14..8f1699ef615f1f2459c5cfd268ac70102b07cf0c 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1544,6 +1544,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
