diff --git tools/gritsettings/resource_ids tools/gritsettings/resource_ids
index b41d8cbc5150..dbd98d1c4058 100644
--- tools/gritsettings/resource_ids
+++ tools/gritsettings/resource_ids
@@ -432,4 +432,11 @@
   # Please read the header and find the right section above instead.
 
   # Resource ids starting at 31000 are reserved for projects built on Chromium.
+
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [31500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [32000],
+  },
 }
