diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index bb7dc253bbc1a9e8b194bc55de59bd4e6b7aecc4..000fc0c8f9547a57d679af23ba01ed8e15b64c0c 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1660,6 +1660,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
