diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index aaf7050604d4968c8eb1b9a42decdadfec527e20..0d7fc4a9209e510946cc25c4f62051b0cb5827e8 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1632,6 +1632,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
