diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index b2295b7374c9055fbaab4c153b061aa03ae46487..1a71a21038b3b98f2aa2902f6850cb24a387ea63 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1739,6 +1739,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
