diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index e1f3030f0c0cc1b870cbe75395539d37ca1451dd..958a9425d7552dd7fbfffb582f6f33882a7aa488 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1810,6 +1810,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
