diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 9d8b4ade5aff13eb00892614ad7b309805f00ec2..650c13eb37c6df4fc31d58e8cea149b130c79cd2 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1791,6 +1791,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
