diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 24ed276c40a30a015cd97cfb9f40f3bc485e2fd7..8217cb0d5dea9d84642a76822c184df6c1d86c46 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1806,6 +1806,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
