diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 01f27f9eb13bd..fdf9ad6b62e9a 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1035,6 +1035,15 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "META": {"align": 31500},
+    "includes": [31500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "META": {"align": 32000},
+    "messages": [32000],
+  },
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
