diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index ee45675b7b852eb79ef112f6c564d39d9abfd5ba..af465ed9c3e65309280a5b0992ee5dab01afa5be 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1632,6 +1632,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
