diff --git tools/gritsettings/resource_ids.spec tools/gritsettings/resource_ids.spec
index 7a9ffcdf31d89d0bf51ce31bfd1371d09bcea9c6..de649fd1c94f34f6e8343a7fe19c9367cafda10f 100644
--- tools/gritsettings/resource_ids.spec
+++ tools/gritsettings/resource_ids.spec
@@ -1664,6 +1664,13 @@
   # END "everything else" section.
   # Everything but chrome/, components/, content/, and ios/
 
+  "cef/libcef/resources/cef_resources.grd": {
+    "includes": [12500],
+  },
+  "cef/libcef/resources/cef_strings.grd": {
+    "messages": [12510],
+  }
+
   # Thinking about appending to the end?
   # Please read the header and find the right section above instead.
 }
