diff --git ui/gtk/gtk_compat.cc ui/gtk/gtk_compat.cc
index a4f93deaa7c42..de236923186cc 100644
--- ui/gtk/gtk_compat.cc
+++ ui/gtk/gtk_compat.cc
@@ -127,7 +127,7 @@ bool LoadGtkImpl() {
   }
   auto env = base::Environment::Create();
   const auto desktop = base::nix::GetDesktopEnvironment(env.get());
-  if (desktop == base::nix::DESKTOP_ENVIRONMENT_GNOME) {
+  if (!gtk_version && desktop == base::nix::DESKTOP_ENVIRONMENT_GNOME) {
     // GNOME is currently the only desktop to support GTK4 starting with version
     // 42+. Try to match the loaded GTK version with the GNOME GTK version.
     // Checking the GNOME version is not necessary since GTK4 is available iff
