diff --git chrome_elf/chrome_elf.gyp chrome_elf/chrome_elf.gyp
index c9a412b..95a7fa6 100644
--- chrome_elf/chrome_elf.gyp
+++ chrome_elf/chrome_elf.gyp
@@ -65,12 +65,6 @@
           ],
           # Set /SUBSYSTEM:WINDOWS.
           'SubSystem': '2',
-          'AdditionalDependencies!': [
-            'user32.lib',
-          ],
-          'IgnoreDefaultLibraryNames': [
-            'user32.lib',
-          ],
         },
       },
     },
