diff --git chrome_elf/chrome_elf.gyp chrome_elf/chrome_elf.gyp
index 7a49fe8..b4be214 100644
--- chrome_elf/chrome_elf.gyp
+++ chrome_elf/chrome_elf.gyp
@@ -59,12 +59,6 @@
           ],
           # Set /SUBSYSTEM:WINDOWS.
           'SubSystem': '2',
-          'AdditionalDependencies!': [
-            'user32.lib',
-          ],
-          'IgnoreDefaultLibraryNames': [
-            'user32.lib',
-          ],
         },
       },
     },
