diff --git input_method_win.cc input_method_win.cc
index f6bc4c0..9c382fe 100644
--- input_method_win.cc
+++ input_method_win.cc
@@ -25,7 +25,7 @@ static const size_t kExtraNumberOfChars = 20;
 
 InputMethodWin::InputMethodWin(internal::InputMethodDelegate* delegate,
                                HWND toplevel_window_handle)
-    : toplevel_window_handle_(toplevel_window_handle),
+    : toplevel_window_handle_(::GetAncestor(toplevel_window_handle, GA_ROOT)),
       pending_requested_direction_(base::i18n::UNKNOWN_DIRECTION),
       accept_carriage_return_(false),
       active_(false),
