diff --git src/libANGLE/renderer/d3d/d3d11/Clear11.cpp src/libANGLE/renderer/d3d/d3d11/Clear11.cpp
index 2b960fe..4755132 100644
--- src/libANGLE/renderer/d3d/d3d11/Clear11.cpp
+++ src/libANGLE/renderer/d3d/d3d11/Clear11.cpp
@@ -370,8 +370,8 @@ gl::Error Clear11::clearFramebuffer(const ClearParameters &clearParams,
     {
         const gl::FramebufferAttachment &attachment = colorAttachments[colorAttachmentIndex];
 
-        if (clearParams.clearColor[colorAttachmentIndex] && attachment.isAttached() &&
-            drawBufferStates[colorAttachmentIndex] != GL_NONE)
+        if (!clearParams.clearColor[colorAttachmentIndex] || !attachment.isAttached() ||
+            drawBufferStates[colorAttachmentIndex] == GL_NONE)
         {
             continue;
         }
