diff --git third_party/libxml/BUILD.gn third_party/libxml/BUILD.gn
index 890afc0e83a926d1c9c08c34da54556699110691..a470d9a527d9421db90d4cd60af9d310dd5536e2 100644
--- third_party/libxml/BUILD.gn
+++ third_party/libxml/BUILD.gn
@@ -143,6 +143,7 @@ static_library("libxml") {
     ":libxml_utils",
     ":xml_reader",
     ":xml_writer",
+    "//cef:*",
     "//chromecast/internal",
     "//testing/libfuzzer/*",
     "//third_party/blink/renderer/*",
