diff --git third_party/libxml/BUILD.gn third_party/libxml/BUILD.gn
index 43bdff9ce7a007ecaa8fe8db48f1331d36d3f87c..75083059da72ccafad013cbc40f80f3575c40a2a 100644
--- third_party/libxml/BUILD.gn
+++ third_party/libxml/BUILD.gn
@@ -143,6 +143,7 @@ static_library("libxml") {
     ":libxml_utils",
     ":xml_reader",
     ":xml_writer",
+    "//cef:*",
     "//chromecast/internal",
     "//testing/libfuzzer/*",
     "//third_party/blink/renderer/*",
