diff --git third_party/libxml/BUILD.gn third_party/libxml/BUILD.gn
index a155534f66d2..7a29f910be79 100644
--- third_party/libxml/BUILD.gn
+++ third_party/libxml/BUILD.gn
@@ -142,6 +142,7 @@ static_library("libxml") {
     "//third_party/blink/renderer/*",
     "//third_party/fontconfig",
     "//third_party/libxslt",
+    "//cef:*",
   ]
   if (is_ios) {
     foreach(tgt, ios_libxml_visibility_additions) {
