diff --git third_party/libxml/BUILD.gn third_party/libxml/BUILD.gn
index 3fe4dfd2990b..64848dfd33d7 100644
--- third_party/libxml/BUILD.gn
+++ third_party/libxml/BUILD.gn
@@ -141,6 +141,7 @@ static_library("libxml") {
     "//third_party/blink/renderer/*",
     "//third_party/fontconfig",
     "//third_party/libxslt",
+    "//cef:*",
   ]
   if (is_ios) {
     foreach(tgt, ios_libxml_visibility_additions) {
