diff --git third_party/libxml/src/tree.c third_party/libxml/src/tree.c
index 8077348a980a1..46c0b6eddb486 100644
--- third_party/libxml/src/tree.c
+++ third_party/libxml/src/tree.c
@@ -4361,7 +4361,8 @@ xmlStaticCopyNode(xmlNodePtr node, xmlDocPtr doc, xmlNodePtr parent,
             }
             insert->last = copy;
 
-            if (cur->children != NULL) {
+            if ((cur->type != XML_ENTITY_REF_NODE) &&
+                (cur->children != NULL)) {
                 cur = cur->children;
                 insert = copy;
                 continue;
