diff --git third_party/libxml/BUILD.gn third_party/libxml/BUILD.gn
index 5780cb8336be31a37baf365b655dc142158d6b05..a00143646bf6e70784244dc8fba7eb53c576ca41 100644
--- third_party/libxml/BUILD.gn
+++ third_party/libxml/BUILD.gn
@@ -143,6 +143,7 @@ static_library("libxml") {
     ":libxml_utils",
     ":xml_reader",
     ":xml_writer",
+    "//cef:*",
     "//chromecast/internal",
     "//testing/libfuzzer/*",
     "//third_party/blink/renderer/*",
