diff --git base/win/dark_mode_support.cc base/win/dark_mode_support.cc
index daecebb5070705c1eb5281a3a7df4ace6b89c119..aed51c2e6b06e3527bd6c2003e4b6a82ee4469e5 100644
--- base/win/dark_mode_support.cc
+++ base/win/dark_mode_support.cc
@@ -7,6 +7,7 @@
 #include <windows.h>
 
 #include "base/check.h"
+#include "base/command_line.h"
 #include "base/native_library.h"
 #include "base/win/win_util.h"
 #include "base/win/windows_version.h"
@@ -85,11 +86,20 @@ const DarkModeSupport& GetDarkModeSupport() {
   return dark_mode_support;
 }
 
+bool IsForcedLightMode() {
+  static bool kIsForcedLightMode =
+      base::CommandLine::ForCurrentProcess()->HasSwitch(
+          "force-light-mode");
+  return kIsForcedLightMode;
+}
+
 }  // namespace
 
 namespace base::win {
 
 bool IsDarkModeAvailable() {
+  if (IsForcedLightMode())
+    return false;
   auto& dark_mode_support = GetDarkModeSupport();
   return (dark_mode_support.allow_dark_mode_for_app ||
           dark_mode_support.set_preferred_app_mode) &&
diff --git chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
index 8eca0f15ad724bfe7b3f83b361984f581f79c347..214eb82df050d5b41efffb8aef44d8ba71e785ae 100644
--- chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
+++ chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
@@ -58,7 +58,10 @@ void ChromeBrowserMainExtraPartsViewsLinux::ToolkitInitialized() {
     ui::CursorFactory::GetInstance()->ObserveThemeChanges();
   }
 #if BUILDFLAG(USE_DBUS)
-  dark_mode_manager_ = std::make_unique<ui::DarkModeManagerLinux>();
+  if (!ui::NativeTheme::IsForcedDarkMode() &&
+      !ui::NativeTheme::IsForcedLightMode()) {
+    dark_mode_manager_ = std::make_unique<ui::DarkModeManagerLinux>();
+  }
 #endif
 }
 
diff --git ui/gtk/native_theme_gtk.cc ui/gtk/native_theme_gtk.cc
index 9693faf3e397d98807146c320975bca241445afb..774c43fd1a67ef68ad90b5eac4e0827b7b74a05e 100644
--- ui/gtk/native_theme_gtk.cc
+++ ui/gtk/native_theme_gtk.cc
@@ -156,9 +156,11 @@ void NativeThemeGtk::OnThemeChanged(GtkSettings* settings,
   // have a light variant and aren't affected by the setting.  Because of this,
   // experimentally check if the theme is dark by checking if the window
   // background color is dark.
-  const SkColor window_bg_color = GetBgColor("");
-  set_use_dark_colors(IsForcedDarkMode() ||
-                      color_utils::IsDark(window_bg_color));
+  if (!IsForcedLightMode()) {
+    const SkColor window_bg_color = GetBgColor("");
+    set_use_dark_colors(IsForcedDarkMode() ||
+                        color_utils::IsDark(window_bg_color));
+  }
   set_preferred_color_scheme(CalculatePreferredColorScheme());
 
   // GTK doesn't have a native high contrast setting.  Rather, it's implied by
diff --git ui/native_theme/native_theme.cc ui/native_theme/native_theme.cc
index e761050d9afbbc6d1c25dd8ec6a2a90d548c1195..23bd4923520a688bc66ea621c1a5da206c23a80a 100644
--- ui/native_theme/native_theme.cc
+++ ui/native_theme/native_theme.cc
@@ -139,6 +139,7 @@ void NativeTheme::NotifyOnNativeThemeUpdated() {
   color_provider_manager.ResetColorProviderCache();
   native_theme_observers_.Notify(&NativeThemeObserver::OnNativeThemeUpdated,
                                  this);
+  color_provider_manager.AfterNativeThemeUpdated();
 
   RecordNumColorProvidersInitializedDuringOnNativeThemeUpdated(
       color_provider_manager.num_providers_initialized() -
@@ -318,6 +319,13 @@ bool NativeTheme::IsForcedDarkMode() {
   return kIsForcedDarkMode;
 }
 
+bool NativeTheme::IsForcedLightMode() {
+  static bool kIsForcedLightMode =
+      base::CommandLine::ForCurrentProcess()->HasSwitch(
+          "force-light-mode");
+  return kIsForcedLightMode;
+}
+
 bool NativeTheme::IsForcedHighContrast() {
   static bool kIsForcedHighContrast =
       base::CommandLine::ForCurrentProcess()->HasSwitch(
diff --git ui/native_theme/native_theme.h ui/native_theme/native_theme.h
index cbe76bbc68692228072fa69a8562c290e571505e..80f047de2f3ba8c48880dd7de5562701ead2a00e 100644
--- ui/native_theme/native_theme.h
+++ ui/native_theme/native_theme.h
@@ -622,6 +622,9 @@ class COMPONENT_EXPORT(NATIVE_THEME) NativeTheme {
   // Calculates and returns the use overlay scrollbar setting.
   static bool CalculateUseOverlayScrollbar();
 
+  // Whether light mode is forced via command-line flag.
+  static bool IsForcedLightMode();
+
  protected:
   explicit NativeTheme(
       bool should_only_use_dark_colors,
diff --git ui/native_theme/native_theme_mac.mm ui/native_theme/native_theme_mac.mm
index e12701fe87bcd15a0db1c51b0885e5128cc84d71..e98f3d25e022ec50208d6694e08fa3a07711be4a 100644
--- ui/native_theme/native_theme_mac.mm
+++ ui/native_theme/native_theme_mac.mm
@@ -626,11 +626,15 @@ void NativeThemeMac::PaintSelectedMenuItem(
 
 void NativeThemeMac::InitializeDarkModeStateAndObserver() {
   __block auto theme = this;
-  set_use_dark_colors(IsDarkMode());
+  if (!IsForcedLightMode()) {
+    set_use_dark_colors(IsForcedDarkMode() || IsDarkMode());
+  }
   set_preferred_color_scheme(CalculatePreferredColorScheme());
   appearance_observer_ =
       [[NativeThemeEffectiveAppearanceObserver alloc] initWithHandler:^{
-        theme->set_use_dark_colors(IsDarkMode());
+        if (!IsForcedLightMode()) {
+          theme->set_use_dark_colors(IsForcedDarkMode() || IsDarkMode());
+        }
         theme->set_preferred_color_scheme(CalculatePreferredColorScheme());
         theme->NotifyOnNativeThemeUpdated();
       }];
diff --git ui/native_theme/native_theme_win.cc ui/native_theme/native_theme_win.cc
index 0ca923fb1b402ce7a5b112600b86e79b02d96168..660cbd9359d24b3ed4538808c84512a6769421d5 100644
--- ui/native_theme/native_theme_win.cc
+++ ui/native_theme/native_theme_win.cc
@@ -675,7 +675,10 @@ bool NativeThemeWin::ShouldUseDarkColors() const {
   // Windows high contrast modes are entirely different themes,
   // so let them take priority over dark mode.
   // ...unless --force-dark-mode was specified in which case caveat emptor.
-  if (InForcedColorsMode() && !IsForcedDarkMode()) {
+  if (supports_windows_dark_mode_ && IsForcedDarkMode()) {
+    return true;
+  }
+  if (InForcedColorsMode()) {
     return false;
   }
   return NativeTheme::ShouldUseDarkColors();
@@ -683,7 +686,7 @@ bool NativeThemeWin::ShouldUseDarkColors() const {
 
 NativeTheme::PreferredColorScheme
 NativeThemeWin::CalculatePreferredColorScheme() const {
-  if (!InForcedColorsMode()) {
+  if (!InForcedColorsMode() || !supports_windows_dark_mode_) {
     return NativeTheme::CalculatePreferredColorScheme();
   }
 
@@ -1664,8 +1667,9 @@ void NativeThemeWin::RegisterColorFilteringRegkeyObserver() {
 }
 
 void NativeThemeWin::UpdateDarkModeStatus() {
-  bool dark_mode_enabled = false;
-  if (hkcu_themes_regkey_.Valid()) {
+  bool dark_mode_enabled = ShouldUseDarkColors();
+  if (supports_windows_dark_mode_ && !IsForcedDarkMode() &&
+      hkcu_themes_regkey_.Valid()) {
     DWORD apps_use_light_theme = 1;
     hkcu_themes_regkey_.ReadValueDW(L"AppsUseLightTheme",
                                     &apps_use_light_theme);
