diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 30440d1b5b..c16bda68af 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -355,6 +355,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index 95887af4fe..906bf15c12 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -175,4 +175,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
