diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 040207453f9f95974e8199c31f2d3cf31bb0e29f..f7a6828321241f29392e8e24625b6c7bc3c6679e 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -362,6 +362,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index afb08ff72d342edcbdb3a00c7a8cafc21fd906a1..2c5ebcab15422f65eac7d65cd4735f6515899fb8 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -196,6 +196,10 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
 
 if (use_gtk) {
