diff --git chrome/browser/ui/libgtkui/gtk_ui.cc chrome/browser/ui/libgtkui/gtk_ui.cc
index f45f2b1..dbabf1b 100644
--- chrome/browser/ui/libgtkui/gtk_ui.cc
+++ chrome/browser/ui/libgtkui/gtk_ui.cc
@@ -92,6 +92,7 @@ namespace libgtkui {
 
 namespace {
 
+#if GTK_MAJOR_VERSION == 3
 // We would like this to be a feature flag, but GtkUi gets initialized
 // earlier than the feature flag registry, so just use a simple bool.
 // The reason for wanting a flag is so that we can release the GTK3
@@ -100,6 +101,7 @@ namespace {
 // Since this was never really intended to be toggled by users, this
 // is fine for now.
 const bool kUseGtkNavButtonLayoutManager = true;
+#endif
 
 const double kDefaultDPI = 96;
 
