diff --git build/config/linux/atk/BUILD.gn build/config/linux/atk/BUILD.gn
index 3931a5a897c948baa2af8ce1ec6a6731b58d5a39..a05e937a9e48eed6c9d7fc62c79e537183a12a31 100644
--- build/config/linux/atk/BUILD.gn
+++ build/config/linux/atk/BUILD.gn
@@ -10,7 +10,7 @@ import("//build/config/ui.gni")
 assert(!is_chromeos)
 
 # These packages should _only_ be expected when building for a target.
-assert(current_toolchain == default_toolchain)
+# assert(current_toolchain == default_toolchain)
 
 if (use_atk) {
   assert(use_glib, "use_atk=true requires that use_glib=true")
diff --git build/config/linux/atspi2/BUILD.gn build/config/linux/atspi2/BUILD.gn
index 30bc77f9c23fbdd4cf46f3ae022266c75f0e9ddf..778d4bc35003684dc5a1da63abfa66806e3b1704 100644
--- build/config/linux/atspi2/BUILD.gn
+++ build/config/linux/atspi2/BUILD.gn
@@ -6,7 +6,7 @@ import("//build/config/linux/pkg_config.gni")
 import("//build/config/ui.gni")
 
 # These packages should _only_ be expected when building for a target.
-assert(current_toolchain == default_toolchain)
+# assert(current_toolchain == default_toolchain)
 
 if (use_atk) {
   pkg_config("atspi2") {
