diff --git build/config/linux/atk/BUILD.gn build/config/linux/atk/BUILD.gn
index aa8723b55c6a0b7375435e2d175868163d523d78..0dbb3405d35cec123a6f816f023866d8d6548706 100644
--- build/config/linux/atk/BUILD.gn
+++ build/config/linux/atk/BUILD.gn
@@ -9,7 +9,7 @@ import("//build/config/ui.gni")
 assert(!is_chromeos)
 
 # These packages should _only_ be expected when building for a target.
-assert(current_toolchain == default_toolchain)
+# assert(current_toolchain == default_toolchain)
 
 if (use_atk) {
   assert(use_glib, "use_atk=true requires that use_glib=true")
diff --git build/config/linux/atspi2/BUILD.gn build/config/linux/atspi2/BUILD.gn
index 89a6049446ce1fe184cb981bcb8d36c6b4930c58..5d3af28e4acdfa777576a1520c44bbcb73ccf3ae 100644
--- build/config/linux/atspi2/BUILD.gn
+++ build/config/linux/atspi2/BUILD.gn
@@ -6,7 +6,7 @@ import("//build/config/linux/pkg_config.gni")
 import("//build/config/ui.gni")
 
 # These packages should _only_ be expected when building for a target.
-assert(current_toolchain == default_toolchain)
+# assert(current_toolchain == default_toolchain)
 
 if (use_atk) {
   pkg_config("atspi2") {
