diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 79a01b007924230edf8c0b32945d3fb4e6f2445d..fbf082eb10f47161291498fd2e44d7a033ab24f0 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -358,6 +358,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index 799ce76a32dcd4438d3d4057e26f3636842f318e..7a2ebca0f9857777be6bf5ac4e24604d54a40c2d 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -181,4 +181,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
