diff --git third_party/tflite/features.gni third_party/tflite/features.gni
index 2466577542552..d94d80cda02d6 100644
--- third_party/tflite/features.gni
+++ third_party/tflite/features.gni
@@ -7,9 +7,9 @@ import("//build/config/chrome_build.gni")
 declare_args() {
   # This enables building TFLite with XNNPACK. Currently only available for
   # Linux, macOS and Windows arm64/x64/x86 targets and ChromeOS non-ARM targets.
-  build_tflite_with_xnnpack =
-      is_win || is_mac || is_linux ||
-      (is_chromeos && target_cpu != "arm64" && target_cpu != "arm")
+  build_tflite_with_xnnpack = (is_win || is_mac || is_linux ||
+      (is_chromeos && current_cpu != "arm64" && current_cpu != "arm")) &&
+      (current_cpu == "arm64" || current_cpu == "x64" || current_cpu == "x86")
 
   # Turns on TFLITE_WITH_RUY, using ruy as the gemm backend instead of gemmlowp.
   build_tflite_with_ruy = true
