diff --git device/bluetooth/BUILD.gn device/bluetooth/BUILD.gn
index facb18d9bc855..77edc67f54653 100644
--- device/bluetooth/BUILD.gn
+++ device/bluetooth/BUILD.gn
@@ -45,13 +45,6 @@ source_set("deprecated_experimental_mojo") {
     "socket.h",
   ]
 
-  if (is_chromeos || is_linux) {
-    sources += [
-      "bluez/metrics_recorder.cc",
-      "bluez/metrics_recorder.h",
-    ]
-  }
-
   deps = [
     ":bluetooth",
     "//device/bluetooth/public/mojom:deprecated_experimental_interfaces",
