diff --git chrome/browser/ui/page_info/chrome_page_info_client.cc chrome/browser/ui/page_info/chrome_page_info_client.cc
index e2972da9e548d..2f356616ed97c 100644
--- chrome/browser/ui/page_info/chrome_page_info_client.cc
+++ chrome/browser/ui/page_info/chrome_page_info_client.cc
@@ -17,5 +17,9 @@ std::unique_ptr<PageInfoDelegate> ChromePageInfoClient::CreatePageInfoDelegate(
 }
 
 int ChromePageInfoClient::GetJavaResourceId(int native_resource_id) {
+#if defined(OS_ANDROID)
   return ResourceMapper::MapToJavaDrawableId(native_resource_id);
+#else
+  return 0;
+#endif
 }
