diff --git printing/printing_context_linux.cc printing/printing_context_linux.cc
index 5c555d8768e2a1d9e6ac86fc9392420a42f5193d..9d343a47c7896569b3ba4de67fbed50e3007be0c 100644
--- printing/printing_context_linux.cc
+++ printing/printing_context_linux.cc
@@ -88,8 +88,8 @@ mojom::ResultCode PrintingContextLinux::UseDefaultSettings() {
 }
 
 gfx::Size PrintingContextLinux::GetPdfPaperSizeDeviceUnits() {
-  if (ui::LinuxUi::instance()) {
-    return ui::LinuxUi::instance()->GetPdfPaperSize(this);
+  if (ui::PrintingContextLinuxDelegate::instance()) {
+    return ui::PrintingContextLinuxDelegate::instance()->GetPdfPaperSize(this);
   }
 
   return gfx::Size();
diff --git ui/linux/BUILD.gn ui/linux/BUILD.gn
index 267de0acbaf5a01f0b9aa5c456e90e32721a31cd..ce4f0836692a292fb9c768124a63ff468ced225b 100644
--- ui/linux/BUILD.gn
+++ ui/linux/BUILD.gn
@@ -3,6 +3,7 @@
 # found in the LICENSE file.
 
 import("//build/config/linux/gtk/gtk.gni")
+import("//printing/buildflags/buildflags.gni")
 import("//ui/qt/qt.gni")
 
 assert(is_linux)
@@ -35,6 +36,11 @@ component("linux_ui") {
     "//ui/display/types",
     "//ui/gfx/geometry",
   ]
+
+  if (enable_printing) {
+    deps += [ "//printing/mojom" ]
+  }
+
   public_deps = [ "//printing/buildflags" ]
 }
 
diff --git ui/linux/linux_ui.cc ui/linux/linux_ui.cc
index 4c86d688230bdca3611fd1a1454390b30cb57dbe..88b706bffb1605835710fb54375aa752839ff7d6 100644
--- ui/linux/linux_ui.cc
+++ ui/linux/linux_ui.cc
@@ -21,11 +21,29 @@ namespace ui {
 namespace {
 
 LinuxUi* g_linux_ui = nullptr;
+static PrintingContextLinuxDelegate* g_delegate = nullptr;
 
 }  // namespace
 
+// static
+PrintingContextLinuxDelegate* PrintingContextLinuxDelegate::SetInstance(
+    PrintingContextLinuxDelegate* delegate) {
+  auto old_delegate = g_delegate;
+  g_delegate = delegate;
+  return old_delegate;
+}
+
+// static
+PrintingContextLinuxDelegate* PrintingContextLinuxDelegate::instance() {
+  return g_delegate;
+}
+
 // static
 LinuxUi* LinuxUi::SetInstance(LinuxUi* instance) {
+#if BUILDFLAG(IS_LINUX) && BUILDFLAG(ENABLE_PRINTING)
+  PrintingContextLinuxDelegate::SetInstance(instance);
+#endif
+
   return std::exchange(g_linux_ui, instance);
 }
 
diff --git ui/linux/linux_ui.h ui/linux/linux_ui.h
index 737739af638a50fef2a4e431847a2fb02cd46663..04fdba8d30263aa5071925c55cadb2729a646fef 100644
--- ui/linux/linux_ui.h
+++ ui/linux/linux_ui.h
@@ -20,6 +20,10 @@
 #include "printing/buildflags/buildflags.h"
 #include "ui/display/types/display_config.h"
 
+#if BUILDFLAG(ENABLE_PRINTING)
+#include "printing/printing_context_linux.h"  // nogncheck
+#endif
+
 // The main entrypoint into Linux toolkit specific code. GTK/QT code should only
 // be executed behind this interface.
 
@@ -59,9 +63,27 @@ class WindowButtonOrderObserver;
 class WindowFrameProvider;
 enum class TextEditCommand;
 
+class COMPONENT_EXPORT(LINUX_UI) PrintingContextLinuxDelegate {
+ public:
+  virtual ~PrintingContextLinuxDelegate() = default;
+
+  virtual std::unique_ptr<printing::PrintDialogLinuxInterface> CreatePrintDialog(
+      printing::PrintingContextLinux* context) = 0;
+
+  virtual gfx::Size GetPdfPaperSize(printing::PrintingContextLinux* context) = 0;
+
+  static PrintingContextLinuxDelegate* SetInstance(
+      PrintingContextLinuxDelegate* delegate);
+  static PrintingContextLinuxDelegate* instance();
+};
+
 // Adapter class with targets to render like different toolkits. Set by any
 // project that wants to do linux desktop native rendering.
-class COMPONENT_EXPORT(LINUX_UI) LinuxUi {
+class COMPONENT_EXPORT(LINUX_UI) LinuxUi
+#if BUILDFLAG(ENABLE_PRINTING)
+    : public PrintingContextLinuxDelegate
+#endif
+ {
  public:
   // 8px is the default value in GTK for how far the mouse needs to move before
   // a drag begins.
@@ -154,14 +176,6 @@ class COMPONENT_EXPORT(LINUX_UI) LinuxUi {
   // Returns a map of KeyboardEvent code to KeyboardEvent key values.
   virtual base::flat_map<std::string, std::string> GetKeyboardLayoutMap() = 0;
 
-#if BUILDFLAG(ENABLE_PRINTING)
-  virtual std::unique_ptr<printing::PrintDialogLinuxInterface>
-  CreatePrintDialog(printing::PrintingContextLinux* context) = 0;
-
-  virtual gfx::Size GetPdfPaperSize(
-      printing::PrintingContextLinux* context) = 0;
-#endif
-
   // Returns a native file selection dialog.  `listener` is of type
   // SelectFileDialog::Listener.  TODO(thomasanderson): Move
   // SelectFileDialog::Listener to SelectFileDialogListener so that it can be
