diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 2d982e757d94976ec68863726db8e8a85b1034e5..3417747a6e87a9e5d3381918efa72829c03cb928 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -358,6 +358,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index 94d30cc6d300e6bdb963980d3962df4999eefc31..e15e4d214f242aeef8b5aeac087e681b45afd855 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -181,4 +181,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
