diff --git device/bluetooth/BUILD.gn device/bluetooth/BUILD.gn
index 05cce11f233e0..91db8a1344d6a 100644
--- device/bluetooth/BUILD.gn
+++ device/bluetooth/BUILD.gn
@@ -50,10 +50,12 @@ source_set("deprecated_experimental_mojo") {
   ]
 
   if (is_chromeos || is_linux) {
-    sources += [
-      "bluez/metrics_recorder.cc",
-      "bluez/metrics_recorder.h",
-    ]
+    if (is_component_build) {
+      sources += [
+        "bluez/metrics_recorder.cc",
+        "bluez/metrics_recorder.h",
+      ]
+    }
   }
 
   deps = [
