diff --git chrome/browser/ui/libgtkui/gtk_ui.cc chrome/browser/ui/libgtkui/gtk_ui.cc
index 7df1d0c511c1..3dd32090eb8d 100644
--- chrome/browser/ui/libgtkui/gtk_ui.cc
+++ chrome/browser/ui/libgtkui/gtk_ui.cc
@@ -320,8 +320,12 @@ float GetRawDeviceScaleFactor() {
     return display::Display::GetForcedDeviceScaleFactor();
 
   GdkScreen* screen = gdk_screen_get_default();
+#if GTK_MAJOR_VERSION == 3
   gint scale = gdk_screen_get_monitor_scale_factor(
       screen, gdk_screen_get_primary_monitor(screen));
+#else
+  gint scale = 1;
+#endif
   gdouble resolution = gdk_screen_get_resolution(screen);
   return resolution <= 0 ? scale : resolution * scale / kDefaultDPI;
 }
