diff --git device/bluetooth/BUILD.gn device/bluetooth/BUILD.gn
index 82d4f2bf563f6..87652664f9e56 100644
--- device/bluetooth/BUILD.gn
+++ device/bluetooth/BUILD.gn
@@ -57,10 +57,12 @@ source_set("deprecated_experimental_mojo") {
   ]
 
   if (is_chromeos || is_linux) {
-    sources += [
-      "bluez/metrics_recorder.cc",
-      "bluez/metrics_recorder.h",
-    ]
+    if (is_component_build) {
+      sources += [
+        "bluez/metrics_recorder.cc",
+        "bluez/metrics_recorder.h",
+      ]
+    }
   }
 
   deps = [
