diff --git build/config/compiler/BUILD.gn build/config/compiler/BUILD.gn
index 148e91397142..ea325faec030 100644
--- build/config/compiler/BUILD.gn
+++ build/config/compiler/BUILD.gn
@@ -163,7 +163,7 @@ declare_args() {
             !use_clang_coverage && !(is_android && use_order_profiling) &&
             (use_lld ||
              (use_gold &&
-              ((!is_android && linux_use_bundled_binutils) || is_chromeos ||
+              ((!is_android && linux_use_bundled_binutils && current_cpu != "x86") || is_chromeos ||
                !(current_cpu == "x86" || current_cpu == "x64"))))
 }
 
