diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 54d483d6b3f1a56573e21b1f873b8dee8fc25a0f..c8d7e4e2ae9a6bb5e9b0a3d81b089274708442b0 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -359,6 +359,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index ec1e229d4974c8985d5e13d60be7fd6f680f1f84..faf4cc3eb01e345b702f1a24d116dba4445b9cdf 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -184,4 +184,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
