diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 09b5f1a0165ca..441bcedb3ac8b 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -357,6 +357,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index dc98ce8281066..31a1eac703aba 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -175,4 +175,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
