diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 54d483d6b3f1a56573e21b1f873b8dee8fc25a0f..c8d7e4e2ae9a6bb5e9b0a3d81b089274708442b0 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -359,6 +359,10 @@ component("platform") {
       if (use_gio) {
         configs += [ "//build/linux:gio_config" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index 799ce76a32dcd4438d3d4057e26f3636842f318e..7a2ebca0f9857777be6bf5ac4e24604d54a40c2d 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -181,4 +181,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
