diff --git ui/accessibility/platform/BUILD.gn ui/accessibility/platform/BUILD.gn
index 4db17696494ad..c4a56a3237d58 100644
--- ui/accessibility/platform/BUILD.gn
+++ ui/accessibility/platform/BUILD.gn
@@ -272,6 +272,10 @@ source_set("platform") {
       if (use_glib) {
         configs += [ "//build/config/linux:glib" ]
       }
+
+      if (is_clang) {
+        cflags = [ "-Wno-deprecated-volatile" ]
+      }
     }
   }
 }
diff --git ui/gtk/BUILD.gn ui/gtk/BUILD.gn
index 4fc6b6d617832..77f17ecf4caf0 100644
--- ui/gtk/BUILD.gn
+++ ui/gtk/BUILD.gn
@@ -176,4 +176,8 @@ component("gtk") {
 
   # TODO: This should be removed.
   deps += [ "//ui/ozone" ]
+
+  if (is_clang) {
+    cflags = [ "-Wno-deprecated-volatile" ]
+  }
 }
