diff --git content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
index 1b458cd3bfc444fd380b4c426bb7844d5ac1b47a..4a82829c4d0ddb7f21d9815164115853bb7b3f66 100644
--- content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
+++ content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
@@ -172,6 +172,13 @@ void ExtractUnderlines(NSAttributedString* string,
 
 // RenderWidgetHostViewCocoa ---------------------------------------------------
 
+@interface NSWindow (CefCustomMethods)
+- (int)acceptsFirstMouse;
+@end
+
+constexpr int kStateEnabled = 1;
+constexpr int kStateDisabled = 2;
+
 // Private methods:
 @interface RenderWidgetHostViewCocoa ()
 
@@ -783,6 +790,15 @@ void ExtractUnderlines(NSAttributedString* string,
 }
 
 - (BOOL)acceptsFirstMouse:(NSEvent*)theEvent {
+  if ([self.window respondsToSelector:@selector(acceptsFirstMouse)]) {
+    const auto mode = [self.window acceptsFirstMouse];
+    if (mode == kStateEnabled) {
+      return YES;
+    } else if (mode == kStateDisabled) {
+      return NO;
+    }
+  }
+
   // Enable "click-through" if mouse clicks are accepted in inactive windows
   return [self acceptsMouseEventsOption] > kAcceptMouseEventsInActiveWindow;
 }
