diff --git third_party/widevine/cdm/BUILD.gn third_party/widevine/cdm/BUILD.gn
index f35d81872096..d14e71e0319b 100644
--- third_party/widevine/cdm/BUILD.gn
+++ third_party/widevine/cdm/BUILD.gn
@@ -110,7 +110,9 @@ if (widevine_cdm_binary_files != []) {
       "//build/config:exe_and_shlib_deps",
     ]
 
-    if (is_posix && !is_mac) {
+    if (is_mac) {
+      ldflags = [ "-Wl,-install_name,@loader_path/libwidevinecdm.dylib" ]
+    } else if (is_posix) {
       cflags = [ "-fvisibility=hidden" ]
     }
   }
