diff --git content/browser/scheduler/responsiveness/native_event_observer_mac.mm content/browser/scheduler/responsiveness/native_event_observer_mac.mm
index 38740d438e263..7657fa28bfcb9 100644
--- content/browser/scheduler/responsiveness/native_event_observer_mac.mm
+++ content/browser/scheduler/responsiveness/native_event_observer_mac.mm
@@ -15,13 +15,15 @@
 namespace content::responsiveness {
 
 void NativeEventObserver::RegisterObserver() {
-  DCHECK([NSApp conformsToProtocol:@protocol(NativeEventProcessor)]);
+  if (![NSApp conformsToProtocol:@protocol(NativeEventProcessor)])
+    return;
   id<NativeEventProcessor> processor =
       static_cast<id<NativeEventProcessor>>(NSApp);
   [processor addNativeEventProcessorObserver:this];
 }
 void NativeEventObserver::DeregisterObserver() {
-  DCHECK([NSApp conformsToProtocol:@protocol(NativeEventProcessor)]);
+  if (![NSApp conformsToProtocol:@protocol(NativeEventProcessor)])
+    return;
   id<NativeEventProcessor> processor =
       static_cast<id<NativeEventProcessor>>(NSApp);
   [processor removeNativeEventProcessorObserver:this];
