diff --git ui/gl/init/gl_initializer_mac.cc ui/gl/init/gl_initializer_mac.cc
index 380f2935952d6..dcee9c5aeb849 100644
--- ui/gl/init/gl_initializer_mac.cc
+++ ui/gl/init/gl_initializer_mac.cc
@@ -46,11 +46,8 @@ bool InitializeOneOffForSandbox() {
   // GPU-related stuff is very slow without this, probably because
   // the sandbox prevents loading graphics drivers or some such.
   std::vector<CGLPixelFormatAttribute> attribs;
-  if (GLContext::SwitchableGPUsSupported()) {
-    // Avoid switching to the discrete GPU just for this pixel
-    // format selection.
-    attribs.push_back(kCGLPFAAllowOfflineRenderers);
-  }
+  // Avoid switching to the discrete GPU just for this pixel format selection.
+  attribs.push_back(kCGLPFAAllowOfflineRenderers);
   attribs.push_back(static_cast<CGLPixelFormatAttribute>(0));
 
   CGLPixelFormatObj format;
