diff --git ui/events/keycodes/keyboard_code_conversion_mac.mm ui/events/keycodes/keyboard_code_conversion_mac.mm
index ab8df8aa5c3fd69d453a0a37f1ceceeda5a300a5..ac6184c85bbe0affacac60495e21e629c4f9fc5f 100644
--- ui/events/keycodes/keyboard_code_conversion_mac.mm
+++ ui/events/keycodes/keyboard_code_conversion_mac.mm
@@ -892,7 +892,7 @@ DomKey DomKeyFromNSEvent(NSEvent* event) {
       return DomKeyFromKeyCode(event.keyCode);
     }
     default:
-      NOTREACHED();
+      return ui::DomKey::NONE;
   }
 }
 
