diff --git content/browser/renderer_host/input/fling_scheduler_mac.mm content/browser/renderer_host/input/fling_scheduler_mac.mm
index 50ed39df38044..7839bdaf7b7e5 100644
--- content/browser/renderer_host/input/fling_scheduler_mac.mm
+++ content/browser/renderer_host/input/fling_scheduler_mac.mm
@@ -26,6 +26,10 @@ ui::Compositor* FlingSchedulerMac::GetCompositor() {
       return nullptr;
   }
 
+  // For CEF this will always be false when running in OSR mode.
+  if (!view->GetNativeView())
+    return nullptr;
+
   RenderWidgetHostViewMac* mac_view =
       static_cast<RenderWidgetHostViewMac*>(view);
   if (mac_view->BrowserCompositor())
