diff --git ui/events/keycodes/keyboard_code_conversion_mac.mm ui/events/keycodes/keyboard_code_conversion_mac.mm
index 23a5e234f7030..0bcc9e3fb2eb1 100644
--- ui/events/keycodes/keyboard_code_conversion_mac.mm
+++ ui/events/keycodes/keyboard_code_conversion_mac.mm
@@ -935,7 +935,7 @@ DomKey DomKeyFromNSEvent(NSEvent* event) {
       return DomKeyFromKeyCode(event.keyCode);
     }
     default:
-      NOTREACHED_NORETURN();
+      return ui::DomKey::NONE;
   }
 }
 
