diff --git content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
index 2e30fbae5a125d4916c8d2cda5ad565c61933da8..1306fe575702365db989b0983d51f35a5fbe794e 100644
--- content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
+++ content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
@@ -192,6 +192,13 @@ void ExtractUnderlines(NSAttributedString* string,
 
 // RenderWidgetHostViewCocoa ---------------------------------------------------
 
+@interface NSWindow (CefCustomMethods)
+- (int)acceptsFirstMouse;
+@end
+
+constexpr int kStateEnabled = 1;
+constexpr int kStateDisabled = 2;
+
 // Private methods:
 @interface RenderWidgetHostViewCocoa ()
 
@@ -850,6 +857,15 @@ static NSWindow* __weak _deferredResignKeyWindow;
 }
 
 - (BOOL)acceptsFirstMouse:(NSEvent*)theEvent {
+  if ([self.window respondsToSelector:@selector(acceptsFirstMouse)]) {
+    const auto mode = [self.window acceptsFirstMouse];
+    if (mode == kStateEnabled) {
+      return YES;
+    } else if (mode == kStateDisabled) {
+      return NO;
+    }
+  }
+
   // Enable "click-through" if mouse clicks are accepted in inactive windows.
   return
       [self acceptsMouseEventsOption] > AcceptMouseEvents::kWhenInActiveWindow;
