diff --git webrtc_internals.cc webrtc_internals.cc
index 92ca83c..51c19cb 100644
--- webrtc_internals.cc
+++ webrtc_internals.cc
@@ -61,7 +61,8 @@ WebRTCInternals::~WebRTCInternals() {
 }
 
 WebRTCInternals* WebRTCInternals::GetInstance() {
-  return Singleton<WebRTCInternals>::get();
+  return Singleton<WebRTCInternals,
+                   LeakySingletonTraits<WebRTCInternals> >::get();
 }
 
 void WebRTCInternals::OnAddPeerConnection(int render_process_id,
