Index: webrtc_internals.cc
===================================================================
--- webrtc_internals.cc	(revision 290621)
+++ webrtc_internals.cc	(working copy)
@@ -59,7 +59,8 @@
 }
 
 WebRTCInternals* WebRTCInternals::GetInstance() {
-  return Singleton<WebRTCInternals>::get();
+  return Singleton<WebRTCInternals,
+                   LeakySingletonTraits<WebRTCInternals> >::get();
 }
 
 void WebRTCInternals::OnAddPeerConnection(int render_process_id,
