diff --git base/message_loop/message_loop.cc base/message_loop/message_loop.cc
index f72f17d9317c..40c421ad4dcb 100644
--- base/message_loop/message_loop.cc
+++ base/message_loop/message_loop.cc
@@ -177,6 +177,9 @@ MessageLoopBase* MessageLoop::GetMessageLoopBase() {
 
 //------------------------------------------------------------------------------
 
+MessageLoopForUI::MessageLoopForUI(std::unique_ptr<MessagePump> pump)
+    : MessageLoop(TYPE_UI, BindOnce(&ReturnPump, std::move(pump))) {}
+
 // static
 std::unique_ptr<MessageLoop> MessageLoop::CreateUnbound(
     Type type,
diff --git base/message_loop/message_loop.h base/message_loop/message_loop.h
index 1adfd4cd1904..acfc0886003c 100644
--- base/message_loop/message_loop.h
+++ base/message_loop/message_loop.h
@@ -188,6 +188,8 @@ class BASE_EXPORT MessageLoopBase {
   // can post other tasks when destructed.
   virtual void DeletePendingTasks() = 0;
 
+  virtual void ReleasePump() = 0;
+
  protected:
   friend class MessageLoop;
   friend class MessageLoopForUI;
@@ -294,6 +296,9 @@ class BASE_EXPORT MessageLoop {
     SEQUENCE_MANAGER,
   };
 
+  // Called from Thread::CleanUp() to release resources.
+  void ReleasePump() { pump_ = nullptr; backend_->ReleasePump(); }
+
   //----------------------------------------------------------------------------
  protected:
   using MessagePumpFactoryCallback =
@@ -402,6 +407,7 @@ class BASE_EXPORT MessageLoop {
 class BASE_EXPORT MessageLoopForUI : public MessageLoop {
  public:
   explicit MessageLoopForUI(Type type = TYPE_UI);
+  explicit MessageLoopForUI(std::unique_ptr<MessagePump> pump);
 
 #if defined(OS_IOS)
   // On iOS, the main message loop cannot be Run().  Instead call Attach(),
diff --git base/message_loop/message_loop_current.cc base/message_loop/message_loop_current.cc
index dca465455e36..bc2dc7c33110 100644
--- base/message_loop/message_loop_current.cc
+++ base/message_loop/message_loop_current.cc
@@ -49,6 +49,8 @@ void MessageLoopCurrent::AddDestructionObserver(
 
 void MessageLoopCurrent::RemoveDestructionObserver(
     DestructionObserver* destruction_observer) {
+  if (!current_)
+    return;
   DCHECK(current_->IsBoundToCurrentThread());
   current_->RemoveDestructionObserver(destruction_observer);
 }
diff --git base/message_loop/message_loop_current.h base/message_loop/message_loop_current.h
index 56c147d1e739..4afd37ba5b10 100644
--- base/message_loop/message_loop_current.h
+++ base/message_loop/message_loop_current.h
@@ -134,6 +134,16 @@ class BASE_EXPORT MessageLoopCurrent {
   // posted tasks.
   void SetAddQueueTimeToTasks(bool enable);
 
+#if defined(OS_WIN)
+  void set_os_modal_loop(bool os_modal_loop) {
+    os_modal_loop_ = os_modal_loop;
+  }
+
+  bool os_modal_loop() const {
+    return os_modal_loop_;
+  }
+#endif  // OS_WIN
+
   // Enables or disables the recursive task processing. This happens in the case
   // of recursive message loops. Some unwanted message loops may occur when
   // using common controls or printer functions. By default, recursive task
@@ -187,7 +197,6 @@ class BASE_EXPORT MessageLoopCurrent {
   // level.
   bool IsIdleForTesting();
 
- protected:
   // Binds |current| to the current thread. It will from then on be the
   // MessageLoop driven by MessageLoopCurrent on this thread. This is only meant
   // to be invoked by the MessageLoop itself.
@@ -198,6 +207,7 @@ class BASE_EXPORT MessageLoopCurrent {
   // meant to be invoked by the MessageLoop itself.
   static void UnbindFromCurrentThreadInternal(MessageLoopBase* current);
 
+ protected:
   explicit MessageLoopCurrent(MessageLoopBase* current) : current_(current) {}
 
   friend class MessageLoopImpl;
@@ -215,6 +225,13 @@ class BASE_EXPORT MessageLoopCurrent {
   MessageLoopBase* ToMessageLoopBaseDeprecated() const { return current_; }
 
   MessageLoopBase* current_;
+
+#if defined(OS_WIN)
+ private:
+  // Should be set to true before calling Windows APIs like TrackPopupMenu, etc.
+  // which enter a modal message loop.
+  bool os_modal_loop_ = false;
+#endif
 };
 
 #if !defined(OS_NACL)
diff --git base/message_loop/message_loop_impl.h base/message_loop/message_loop_impl.h
index cefa94ff52d6..13acc90d10dc 100644
--- base/message_loop/message_loop_impl.h
+++ base/message_loop/message_loop_impl.h
@@ -77,6 +77,8 @@ class BASE_EXPORT MessageLoopImpl : public MessageLoopBase,
   void DeletePendingTasks() override;
   bool HasTasks() override;
 
+  void ReleasePump() override { pump_ = nullptr; }
+
   // Gets the TaskRunner associated with this message loop.
   const scoped_refptr<SingleThreadTaskRunner>& task_runner() const {
     return task_runner_;
diff --git base/message_loop/message_pump_win.cc base/message_loop/message_pump_win.cc
index 1d6748e0e88a..1598fb65484e 100644
--- base/message_loop/message_pump_win.cc
+++ base/message_loop/message_pump_win.cc
@@ -11,6 +11,7 @@
 
 #include "base/debug/alias.h"
 #include "base/memory/ptr_util.h"
+#include "base/message_loop/message_loop_current.h"
 #include "base/metrics/histogram_macros.h"
 #include "base/strings/stringprintf.h"
 #include "base/trace_event/trace_event.h"
@@ -384,20 +385,28 @@ bool MessagePumpForUI::ProcessMessageHelper(const MSG& msg) {
 }
 
 bool MessagePumpForUI::ProcessPumpReplacementMessage() {
-  // When we encounter a kMsgHaveWork message, this method is called to peek and
-  // process a replacement message. The goal is to make the kMsgHaveWork as non-
-  // intrusive as possible, even though a continuous stream of such messages are
-  // posted. This method carefully peeks a message while there is no chance for
-  // a kMsgHaveWork to be pending, then resets the |have_work_| flag (allowing a
-  // replacement kMsgHaveWork to possibly be posted), and finally dispatches
-  // that peeked replacement. Note that the re-post of kMsgHaveWork may be
-  // asynchronous to this thread!!
-
+  // When we encounter a kMsgHaveWork message, this method is called to peek
+  // and process a replacement message, such as a WM_PAINT or WM_TIMER.  The
+  // goal is to make the kMsgHaveWork as non-intrusive as possible, even though
+  // a continuous stream of such messages are posted.  This method carefully
+  // peeks a message while there is no chance for a kMsgHaveWork to be pending,
+  // then resets the have_work_ flag (allowing a replacement kMsgHaveWork to
+  // possibly be posted), and finally dispatches that peeked replacement.  Note
+  // that the re-post of kMsgHaveWork may be asynchronous to this thread!!
+
+  bool have_message = false;
   MSG msg;
-  const bool have_message =
-      ::PeekMessage(&msg, nullptr, 0, 0, PM_REMOVE) != FALSE;
+  // We should not process all window messages if we are in the context of an
+  // OS modal loop, i.e. in the context of a windows API call like MessageBox.
+  // This is to ensure that these messages are peeked out by the OS modal loop.
+  if (MessageLoopCurrent::Get()->os_modal_loop()) {
+    // We only peek out WM_PAINT and WM_TIMER here for reasons mentioned above.
+    have_message = PeekMessage(&msg, NULL, WM_PAINT, WM_PAINT, PM_REMOVE) ||
+                   PeekMessage(&msg, NULL, WM_TIMER, WM_TIMER, PM_REMOVE);
+  } else {
+    have_message = PeekMessage(&msg, NULL, 0, 0, PM_REMOVE) != FALSE;
+  }
 
-  // Expect no message or a message different than kMsgHaveWork.
   DCHECK(!have_message || kMsgHaveWork != msg.message ||
          msg.hwnd != message_window_.hwnd());
 
diff --git base/task/sequence_manager/sequence_manager_impl.h base/task/sequence_manager/sequence_manager_impl.h
index 8dafb175bb5e..291cdc9c1cd3 100644
--- base/task/sequence_manager/sequence_manager_impl.h
+++ base/task/sequence_manager/sequence_manager_impl.h
@@ -164,6 +164,8 @@ class BASE_EXPORT SequenceManagerImpl
   void BindToCurrentThread(std::unique_ptr<MessagePump> pump) override;
   void DeletePendingTasks() override;
   bool HasTasks() override;
+  
+  void ReleasePump() override { }
 
   // Requests that a task to process work is posted on the main task runner.
   // These tasks are de-duplicated in two buckets: main-thread and all other
