diff --git base/message_loop/message_pump_mac.mm base/message_loop/message_pump_mac.mm
index 1d0ac66a9174..4ddaca449e8b 100644
--- base/message_loop/message_pump_mac.mm
+++ base/message_loop/message_pump_mac.mm
@@ -644,7 +644,8 @@ void MessagePumpUIApplication::Detach() {
 #else
 
 ScopedPumpMessagesInPrivateModes::ScopedPumpMessagesInPrivateModes() {
-  DCHECK(g_app_pump);
+  if (!g_app_pump)
+    return;
   DCHECK_EQ(kNSApplicationModalSafeModeMask, g_app_pump->GetModeMask());
   // Pumping events in private runloop modes is known to interact badly with
   // app modal windows like NSAlert.
@@ -654,7 +655,8 @@ ScopedPumpMessagesInPrivateModes::ScopedPumpMessagesInPrivateModes() {
 }
 
 ScopedPumpMessagesInPrivateModes::~ScopedPumpMessagesInPrivateModes() {
-  DCHECK(g_app_pump);
+  if (!g_app_pump)
+    return;
   g_app_pump->SetModeMask(kNSApplicationModalSafeModeMask);
 }
 
