diff --git components/metrics/persistent_system_profile.cc components/metrics/persistent_system_profile.cc
index 5ffc5ec97f90a..d322d2a26264a 100644
--- components/metrics/persistent_system_profile.cc
+++ components/metrics/persistent_system_profile.cc
@@ -415,6 +415,10 @@ bool PersistentSystemProfile::GetSystemProfile(
   return true;
 }
 
+void PersistentSystemProfile::DetachFromCurrentThread() {
+  DETACH_FROM_THREAD(thread_checker_);
+}
+
 // static
 void PersistentSystemProfile::MergeUpdateRecords(
     const base::PersistentMemoryAllocator& memory_allocator,
diff --git components/metrics/persistent_system_profile.h components/metrics/persistent_system_profile.h
index 426ed30d15be2..088fffc40f8a0 100644
--- components/metrics/persistent_system_profile.h
+++ components/metrics/persistent_system_profile.h
@@ -64,6 +64,10 @@ class PersistentSystemProfile {
       const base::PersistentMemoryAllocator& memory_allocator,
       SystemProfileProto* system_profile);
 
+  // Detaches this object from the current thread in preparation for a move to
+  // a different thread.
+  void DetachFromCurrentThread();
+
  private:
   friend class PersistentSystemProfileTest;
 
