diff --git extensions/browser/guest_view/mime_handler_view/mime_handler_view_embedder.cc extensions/browser/guest_view/mime_handler_view/mime_handler_view_embedder.cc
index 045cc6e50916..205945f06910 100644
--- extensions/browser/guest_view/mime_handler_view/mime_handler_view_embedder.cc
+++ extensions/browser/guest_view/mime_handler_view/mime_handler_view_embedder.cc
@@ -110,7 +110,7 @@ void MimeHandlerViewEmbedder::DidFinishNavigation(
     return;
   // We should've deleted the MimeHandlerViewEmbedder at this point if the frame
   // is sandboxed.
-  DCHECK(!render_frame_host_->IsSandboxed(
+  DCHECK(!render_frame_host_ || !render_frame_host_->IsSandboxed(
       network::mojom::WebSandboxFlags::kPlugins));
 }
 
