diff --git net/base/load_flags_list.h net/base/load_flags_list.h
index c10df92f30b1292a8243f376025e81afcefba78d..59e6208b3954635076022807edfc9a305c8c22af 100644
--- net/base/load_flags_list.h
+++ net/base/load_flags_list.h
@@ -123,4 +123,7 @@ LOAD_FLAG(SHOULD_BYPASS_HSTS, 1 << 18)
 // compromised renderer concern is acceptable.
 LOAD_FLAG(IS_MAIN_FRAME_ORIGIN_RECENTLY_ACCESSED, 1 << 19)
 
+// This load will not send any cookies. For CEF usage.
+LOAD_FLAG(DO_NOT_SEND_COOKIES, 1 << 19)
+
 // See note at top of file about why adding LoadFlags is often a bad idea.
diff --git net/url_request/url_request_http_job.cc net/url_request/url_request_http_job.cc
index 6537c74f5812f0d24bf8b31b7093822201b82a70..2ace801880fb99ce6b638d3b835f7b44a1bea889 100644
--- net/url_request/url_request_http_job.cc
+++ net/url_request/url_request_http_job.cc
@@ -2185,7 +2185,8 @@ bool URLRequestHttpJob::ShouldAddCookieHeader() const {
   // Read cookies whenever allow_credentials() is true, even if the PrivacyMode
   // is being overridden by NetworkDelegate and will eventually block them, as
   // blocked cookies still need to be logged in that case.
-  return request_->context()->cookie_store() && request_->allow_credentials();
+  return request_->context()->cookie_store() && request_->allow_credentials() &&
+         !(request_info_.load_flags & LOAD_DO_NOT_SEND_COOKIES);
 }
 
 bool URLRequestHttpJob::ShouldRecordPartitionedCookieUsage() const {
diff --git services/network/cors/cors_url_loader_factory.cc services/network/cors/cors_url_loader_factory.cc
index 0fa4c9ccd9c596c46fef491d0693e61211e44422..797115b0711feac887f31969ed5714e03066af59 100644
--- services/network/cors/cors_url_loader_factory.cc
+++ services/network/cors/cors_url_loader_factory.cc
@@ -660,7 +660,8 @@ bool CorsURLLoaderFactory::IsValidRequest(const ResourceRequest& request,
           net::LOAD_DISABLE_CACHE | net::LOAD_PREFETCH |
           net::LOAD_IGNORE_LIMITS | net::LOAD_DO_NOT_USE_EMBEDDED_IDENTITY |
           net::LOAD_SUPPORT_ASYNC_REVALIDATION |
-          net::LOAD_RESTRICTED_PREFETCH_FOR_MAIN_FRAME)) {
+          net::LOAD_RESTRICTED_PREFETCH_FOR_MAIN_FRAME |
+          net::LOAD_DO_NOT_SEND_COOKIES | net::LOAD_DO_NOT_SAVE_COOKIES)) {
       mojo::ReportBadMessage(
           "CorsURLLoaderFactory: Untrusted caller using restricted load flag");
       return false;
diff --git services/network/public/cpp/resource_request.cc services/network/public/cpp/resource_request.cc
index dd0ebad1fdcadab06b25df192a85153d1e65e141..cb7b39406e974ea3e271520351f99cbc996b486a 100644
--- services/network/public/cpp/resource_request.cc
+++ services/network/public/cpp/resource_request.cc
@@ -374,7 +374,8 @@ bool ResourceRequest::EqualsForTesting(const ResourceRequest& request) const {
 }
 
 bool ResourceRequest::SendsCookies() const {
-  return credentials_mode == network::mojom::CredentialsMode::kInclude;
+  return credentials_mode == network::mojom::CredentialsMode::kInclude &&
+         !(load_flags & net::LOAD_DO_NOT_SEND_COOKIES);
 }
 
 bool ResourceRequest::SavesCookies() const {
