diff --git net/url_request/url_request.h net/url_request/url_request.h
index 8551e92a205d..0bc5443c9f2e 100644
--- net/url_request/url_request.h
+++ net/url_request/url_request.h
@@ -724,10 +724,10 @@ class NET_EXPORT URLRequest : public base::SupportsUserData {
   }
   bool upgrade_if_insecure() const { return upgrade_if_insecure_; }
 
- protected:
   // Allow the URLRequestJob class to control the is_pending() flag.
   void set_is_pending(bool value) { is_pending_ = value; }
 
+ protected:
   // Allow the URLRequestJob class to set our status too.
   void set_status(URLRequestStatus status);
 
