diff --git net/url_request/url_request.h net/url_request/url_request.h
index b264ca045f24..1746e0588926 100644
--- net/url_request/url_request.h
+++ net/url_request/url_request.h
@@ -748,10 +748,10 @@ class NET_EXPORT URLRequest : public base::SupportsUserData {
 
   base::WeakPtr<URLRequest> GetWeakPtr();
 
- protected:
   // Allow the URLRequestJob class to control the is_pending() flag.
   void set_is_pending(bool value) { is_pending_ = value; }
 
+ protected:
   // Allow the URLRequestJob class to set our status too.
   void set_status(URLRequestStatus status);
 
