diff --git net/test/embedded_test_server/embedded_test_server.cc net/test/embedded_test_server/embedded_test_server.cc
index 988b223d61af0c83050b3474d21c2546de2dd124..6c6a2f0eb2a6536714300235ca7f719a37aec5b6 100644
--- net/test/embedded_test_server/embedded_test_server.cc
+++ net/test/embedded_test_server/embedded_test_server.cc
@@ -1225,7 +1225,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWait(base::OnceClosure closure) {
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   if (!io_thread_->task_runner()->PostTaskAndReply(
           FROM_HERE, std::move(closure), run_loop.QuitClosure())) {
     return false;
@@ -1252,7 +1252,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWaitWithResult(
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   bool task_result = false;
   if (!io_thread_->task_runner()->PostTaskAndReplyWithResult(
           FROM_HERE, std::move(task),
