diff --git net/url_request/url_request_job.cc net/url_request/url_request_job.cc
index f65b7e6eccd2ef33c9addb5797aab62e75fd3803..1391b7a3a040f2bf8338080ccaaf6fbbf0de5d01 100644
--- net/url_request/url_request_job.cc
+++ net/url_request/url_request_job.cc
@@ -39,6 +39,7 @@
 #include "net/url_request/redirect_info.h"
 #include "net/url_request/redirect_util.h"
 #include "net/url_request/url_request_context.h"
+#include "url/url_util.h"
 
 namespace net {
 
@@ -59,6 +60,16 @@ const scoped_refptr<base::SingleThreadTaskRunner>& TaskRunner(
   return base::SingleThreadTaskRunner::GetCurrentDefault();
 }
 
+bool IsSecureScheme(const GURL& url) {
+  if (!url.has_scheme()) {
+    return false;
+  }
+  if (GURL::SchemeIsCryptographic(url.scheme())) {
+    return true;
+  }
+  return base::Contains(url::GetSecureSchemes(), url.scheme());
+}
+
 }  // namespace
 
 // Each SourceStreams own the previous SourceStream in the chain, but the
@@ -344,8 +355,7 @@ GURL URLRequestJob::ComputeReferrerForPolicy(
   }
 
   bool secure_referrer_but_insecure_destination =
-      original_referrer.SchemeIsCryptographic() &&
-      !destination.SchemeIsCryptographic();
+      IsSecureScheme(original_referrer) && !IsSecureScheme(destination);
 
   switch (policy) {
     case ReferrerPolicy::CLEAR_ON_TRANSITION_FROM_SECURE_TO_INSECURE:
