diff --git net/test/embedded_test_server/embedded_test_server.cc net/test/embedded_test_server/embedded_test_server.cc
index 731594d6ec78701bc6e39cf9b987a08dcaa3c34f..a066840be2637dff1ebb5040ed2ce1451357a0b8 100644
--- net/test/embedded_test_server/embedded_test_server.cc
+++ net/test/embedded_test_server/embedded_test_server.cc
@@ -1283,7 +1283,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWait(base::OnceClosure closure) {
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   if (!io_thread_->task_runner()->PostTaskAndReply(
           FROM_HERE, std::move(closure), run_loop.QuitClosure())) {
     return false;
@@ -1310,7 +1310,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWaitWithResult(
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   bool task_result = false;
   if (!io_thread_->task_runner()->PostTaskAndReplyWithResult(
           FROM_HERE, std::move(task),
