diff --git net/test/embedded_test_server/embedded_test_server.cc net/test/embedded_test_server/embedded_test_server.cc
index 80d0f7c615308eafdf1735fcffbb74ad90574722..0a599a62318219c67bcca863ab5bbeae5d86b5b9 100644
--- net/test/embedded_test_server/embedded_test_server.cc
+++ net/test/embedded_test_server/embedded_test_server.cc
@@ -1156,7 +1156,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWait(base::OnceClosure closure) {
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   if (!io_thread_->task_runner()->PostTaskAndReply(
           FROM_HERE, std::move(closure), run_loop.QuitClosure())) {
     return false;
@@ -1183,7 +1183,7 @@ bool EmbeddedTestServer::PostTaskToIOThreadAndWaitWithResult(
   if (!base::CurrentThread::Get())
     temporary_loop = std::make_unique<base::SingleThreadTaskExecutor>();
 
-  base::RunLoop run_loop;
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
   bool task_result = false;
   if (!io_thread_->task_runner()->PostTaskAndReplyWithResult(
           FROM_HERE, std::move(task),
