diff --git net/url_request/url_request.h net/url_request/url_request.h
index a64498d1786c..bf440a5b5f6b 100644
--- net/url_request/url_request.h
+++ net/url_request/url_request.h
@@ -737,10 +737,10 @@ class NET_EXPORT URLRequest : public base::SupportsUserData {
 
   base::WeakPtr<URLRequest> GetWeakPtr();
 
- protected:
   // Allow the URLRequestJob class to control the is_pending() flag.
   void set_is_pending(bool value) { is_pending_ = value; }
 
+ protected:
   // Allow the URLRequestJob class to set our status too.
   void set_status(URLRequestStatus status);
 
