diff --git net/url_request/url_request.h net/url_request/url_request.h
index 95a26676dcae..68d7bfb95914 100644
--- net/url_request/url_request.h
+++ net/url_request/url_request.h
@@ -657,10 +657,10 @@ class NET_EXPORT URLRequest : NON_EXPORTED_BASE(public base::NonThreadSafe),
     return traffic_annotation_;
   }
 
- protected:
   // Allow the URLRequestJob class to control the is_pending() flag.
   void set_is_pending(bool value) { is_pending_ = value; }
 
+ protected:
   // Allow the URLRequestJob class to set our status too.
   void set_status(URLRequestStatus status);
 
