diff --git net/url_request/url_request.h net/url_request/url_request.h
index 13b7326d6d50..d9b1dac19cd8 100644
--- net/url_request/url_request.h
+++ net/url_request/url_request.h
@@ -723,10 +723,10 @@ class NET_EXPORT URLRequest : public base::SupportsUserData {
   }
   bool upgrade_if_insecure() const { return upgrade_if_insecure_; }
 
- protected:
   // Allow the URLRequestJob class to control the is_pending() flag.
   void set_is_pending(bool value) { is_pending_ = value; }
 
+ protected:
   // Allow the URLRequestJob class to set our status too.
   void set_status(URLRequestStatus status);
 
