diff --git components/input/fling_scheduler_base.h components/input/fling_scheduler_base.h
index 6e331ee24bddf1b6f4b0f99d40c94504707dbba8..ab6c61c2e57ac8a93d7f1413b2627c3b78699777 100644
--- components/input/fling_scheduler_base.h
+++ components/input/fling_scheduler_base.h
@@ -5,9 +5,14 @@
 #ifndef COMPONENTS_INPUT_FLING_SCHEDULER_BASE_H_
 #define COMPONENTS_INPUT_FLING_SCHEDULER_BASE_H_
 
+#include "base/memory/raw_ptr.h"
 #include "components/input/fling_controller.h"
 #include "components/viz/common/frame_sinks/begin_frame_source.h"
 
+namespace ui {
+class Compositor;
+}
+
 namespace input {
 
 class FlingSchedulerBase : public FlingControllerSchedulerClient {
@@ -19,6 +24,13 @@ class FlingSchedulerBase : public FlingControllerSchedulerClient {
   // in Viz currently, but in future we would want to migrate browser's
   // implementations to use this to progress flings.
   virtual void SetBeginFrameSource(viz::BeginFrameSource* begin_frame_source) {}
+  
+  void SetCompositor(ui::Compositor* compositor) {
+    compositor_ = compositor;
+  }
+
+ protected:
+  raw_ptr<ui::Compositor> compositor_ = nullptr;
 };
 
 }  // namespace input
diff --git components/input/render_input_router.h components/input/render_input_router.h
index a5be83281a411282346656f88feb6b9eed15a9bc..8d947d1054c307500a8e6957f3efabc0c7f10136 100644
--- components/input/render_input_router.h
+++ components/input/render_input_router.h
@@ -67,6 +67,7 @@ class COMPONENT_EXPORT(INPUT) RenderInputRouter
 
   InputRouter* input_router() { return input_router_.get(); }
   RenderInputRouterDelegate* delegate() { return delegate_; }
+  input::FlingSchedulerBase* fling_scheduler() { return fling_scheduler_.get(); }
 
   void SetForceEnableZoom(bool);
   void SetDeviceScaleFactor(float device_scale_factor);
diff --git content/browser/renderer_host/input/fling_scheduler.cc content/browser/renderer_host/input/fling_scheduler.cc
index 79df9203ec3b63fbfb30ceb0c402d7cd4595c628..60abc7b2ac845bc02d03b343719b2ce5ddd3be88 100644
--- content/browser/renderer_host/input/fling_scheduler.cc
+++ content/browser/renderer_host/input/fling_scheduler.cc
@@ -98,6 +98,9 @@ void FlingScheduler::ProgressFlingOnBeginFrameIfneeded(
 }
 
 ui::Compositor* FlingScheduler::GetCompositor() {
+  if (compositor_) {
+    return compositor_;
+  }
   if (!host_->GetView()) {
     return nullptr;
   }
diff --git content/browser/renderer_host/render_widget_host_impl.cc content/browser/renderer_host/render_widget_host_impl.cc
index f56703978d6d416ecbd378742e2b23ae7f69c64f..daad99a315ea903531795429a68495e4ee489a5b 100644
--- content/browser/renderer_host/render_widget_host_impl.cc
+++ content/browser/renderer_host/render_widget_host_impl.cc
@@ -3826,6 +3826,11 @@ void RenderWidgetHostImpl::StopFling() {
   GetRenderInputRouter()->StopFling();
 }
 
+void RenderWidgetHostImpl::SetCompositorForFlingScheduler(
+    ui::Compositor* compositor) {
+  GetRenderInputRouter()->fling_scheduler()->SetCompositor(compositor);
+}
+
 void RenderWidgetHostImpl::SetScreenOrientationForTesting(
     uint16_t angle,
     display::mojom::ScreenOrientation type) {
diff --git content/browser/renderer_host/render_widget_host_impl.h content/browser/renderer_host/render_widget_host_impl.h
index 614390e29fa84af4c141102d86f3ebdb7293b813..1c9947272fb8dc05f0776d1e22d5c82994cc5815 100644
--- content/browser/renderer_host/render_widget_host_impl.h
+++ content/browser/renderer_host/render_widget_host_impl.h
@@ -859,6 +859,7 @@ class CONTENT_EXPORT RenderWidgetHostImpl
 
   void ProgressFlingIfNeeded(base::TimeTicks current_time);
   void StopFling();
+  void SetCompositorForFlingScheduler(ui::Compositor* compositor);
 
   RenderWidgetHostViewBase* GetRenderWidgetHostViewBase();
 
