diff --git BUILD.gn BUILD.gn
index 52c22d6a1..42888ca9f 100644
--- BUILD.gn
+++ BUILD.gn
@@ -209,6 +209,10 @@ jumbo_static_library("pdfium") {
     complete_static_lib = true
     configs -= [ "//build/config/compiler:thin_archive" ]
   }
+
+  include_dirs = [
+    "//v8/include",
+  ]
 }
 
 # Targets below this are only visible within this file (and to the
diff --git fpdfsdk/fpdf_view.cpp fpdfsdk/fpdf_view.cpp
index e46c3dbe3..acb963c7b 100644
--- fpdfsdk/fpdf_view.cpp
+++ fpdfsdk/fpdf_view.cpp
@@ -38,6 +38,7 @@
 #include "fpdfsdk/cpdfsdk_helpers.h"
 #include "fpdfsdk/cpdfsdk_pageview.h"
 #include "fpdfsdk/ipdfsdk_pauseadapter.h"
+#include "fxjs/cfxjs_engine.h"
 #include "fxjs/ijs_runtime.h"
 #include "public/fpdf_formfill.h"
 #include "third_party/base/ptr_util.h"
@@ -183,6 +184,7 @@ FPDF_EXPORT void FPDF_CALLCONV FPDF_DestroyLibrary() {
 
   CPDF_ModuleMgr::Destroy();
   CFX_GEModule::Destroy();
+  FXJS_Release();
 
   IJS_Runtime::Destroy();
 
