diff --git BUILD.gn BUILD.gn
index de7f21b..cb3dcb1 100644
--- BUILD.gn
+++ BUILD.gn
@@ -208,6 +208,10 @@ static_library("pdfium") {
   } else {
     libs += [ "freetype" ]
   }
+
+  include_dirs = [
+    "//v8/include",
+  ]
 }
 
 static_library("test_support") {
diff --git fpdfsdk/fpdfview.cpp fpdfsdk/fpdfview.cpp
index 1e7a651..9523c7e 100644
--- fpdfsdk/fpdfview.cpp
+++ fpdfsdk/fpdfview.cpp
@@ -31,6 +31,7 @@
 #include "fpdfsdk/fsdk_define.h"
 #include "fpdfsdk/fsdk_pauseadapter.h"
 #include "fpdfsdk/javascript/ijs_runtime.h"
+#include "fxjs/fxjs_v8.h"
 #include "public/fpdf_ext.h"
 #include "public/fpdf_progressive.h"
 #include "third_party/base/numerics/safe_conversions_impl.h"
@@ -419,6 +420,7 @@ DLLEXPORT void STDCALL FPDF_DestroyLibrary() {
 
   CPDF_ModuleMgr::Destroy();
   CFX_GEModule::Destroy();
+  FXJS_Release();
 
   delete g_pCodecModule;
   g_pCodecModule = nullptr;
