diff --git BUILD.gn BUILD.gn
index 5802752af..94c247e04 100644
--- BUILD.gn
+++ BUILD.gn
@@ -222,6 +222,10 @@ static_library("pdfium") {
   if (pdf_is_complete_lib) {
     complete_static_lib = true
   }
+
+  include_dirs = [
+    "//v8/include",
+  ]
 }
 
 static_library("test_support") {
diff --git fpdfsdk/fpdfview.cpp fpdfsdk/fpdfview.cpp
index f20e8ab9f..bc34b858d 100644
--- fpdfsdk/fpdfview.cpp
+++ fpdfsdk/fpdfview.cpp
@@ -35,6 +35,7 @@
 #include "fpdfsdk/fsdk_define.h"
 #include "fpdfsdk/fsdk_pauseadapter.h"
 #include "fpdfsdk/javascript/ijs_runtime.h"
+#include "fxjs/fxjs_v8.h"
 #include "public/fpdf_ext.h"
 #include "public/fpdf_progressive.h"
 #include "third_party/base/allocator/partition_allocator/partition_alloc.h"
@@ -382,6 +383,7 @@ DLLEXPORT void STDCALL FPDF_DestroyLibrary() {
 
   CPDF_ModuleMgr::Destroy();
   CFX_GEModule::Destroy();
+  FXJS_Release();
 
   IJS_Runtime::Destroy();
 
