diff --git BUILD.gn BUILD.gn
index 80a258e..7f343f1 100644
--- BUILD.gn
+++ BUILD.gn
@@ -227,6 +227,10 @@ static_library("pdfium") {
   if (pdf_is_complete_lib) {
     complete_static_lib = true
   }
+
+  include_dirs = [
+    "//v8/include",
+  ]
 }
 
 static_library("test_support") {
diff --git fpdfsdk/fpdfview.cpp fpdfsdk/fpdfview.cpp
index 7dbaf7f..3b41421 100644
--- fpdfsdk/fpdfview.cpp
+++ fpdfsdk/fpdfview.cpp
@@ -37,6 +37,7 @@
 #include "fpdfsdk/fsdk_define.h"
 #include "fpdfsdk/fsdk_pauseadapter.h"
 #include "fpdfsdk/javascript/ijs_runtime.h"
+#include "fxjs/fxjs_v8.h"
 #include "public/fpdf_edit.h"
 #include "public/fpdf_ext.h"
 #include "public/fpdf_progressive.h"
@@ -419,6 +420,7 @@ DLLEXPORT void STDCALL FPDF_DestroyLibrary() {
 
   CPDF_ModuleMgr::Destroy();
   CFX_GEModule::Destroy();
+  FXJS_Release();
 
   IJS_Runtime::Destroy();
 
