Index: power_monitor_device_source_win.cc
===================================================================
--- power_monitor_device_source_win.cc	(revision 227628)
+++ power_monitor_device_source_win.cc	(working copy)
@@ -35,8 +35,8 @@
   if (MessageLoop::current()->type() != MessageLoop::TYPE_UI) {
     // Creating this window in (e.g.) a renderer inhibits shutdown on Windows.
     // See http://crbug.com/230122. TODO(vandebo): http://crbug.com/236031
-    DLOG(ERROR)
-        << "Cannot create windows on non-UI thread, power monitor disabled!";
+    // DLOG(ERROR)
+    //     << "Cannot create windows on non-UI thread, power monitor disabled!";
     return;
   }
   WNDCLASSEX window_class;
