diff --git content/public/common/common_param_traits_macros.h content/public/common/common_param_traits_macros.h
index 39ff86491246..163eaec35230 100644
--- content/public/common/common_param_traits_macros.h
+++ content/public/common/common_param_traits_macros.h
@@ -186,6 +186,7 @@ IPC_STRUCT_TRAITS_BEGIN(content::WebPreferences)
   IPC_STRUCT_TRAITS_MEMBER(main_frame_resizes_are_orientation_changes)
   IPC_STRUCT_TRAITS_MEMBER(initialize_at_minimum_page_scale)
   IPC_STRUCT_TRAITS_MEMBER(smart_insert_delete_enabled)
+  IPC_STRUCT_TRAITS_MEMBER(base_background_color)
   IPC_STRUCT_TRAITS_MEMBER(cookie_enabled)
   IPC_STRUCT_TRAITS_MEMBER(navigate_on_drag_drop)
   IPC_STRUCT_TRAITS_MEMBER(spatial_navigation_enabled)
diff --git content/public/common/web_preferences.cc content/public/common/web_preferences.cc
index b76f327dccca..eecf13a609ca 100644
--- content/public/common/web_preferences.cc
+++ content/public/common/web_preferences.cc
@@ -168,6 +168,7 @@ WebPreferences::WebPreferences()
       spatial_navigation_enabled(false),
       use_solid_color_scrollbars(false),
       navigate_on_drag_drop(true),
+      base_background_color(0xFFFFFFFF),  // Color::white
       v8_cache_options(blink::mojom::V8CacheOptions::kDefault),
       record_whole_document(false),
       cookie_enabled(true),
diff --git content/public/common/web_preferences.h content/public/common/web_preferences.h
index 9052d1d6f4db..acb7027bb637 100644
--- content/public/common/web_preferences.h
+++ content/public/common/web_preferences.h
@@ -183,6 +183,7 @@ struct CONTENT_EXPORT WebPreferences {
   bool spatial_navigation_enabled;
   bool use_solid_color_scrollbars;
   bool navigate_on_drag_drop;
+  uint32_t base_background_color;
   blink::mojom::V8CacheOptions v8_cache_options;
   bool record_whole_document;
 
diff --git content/renderer/render_view_impl.cc content/renderer/render_view_impl.cc
index c748edbf5a32..0d549ad30f4c 100644
--- content/renderer/render_view_impl.cc
+++ content/renderer/render_view_impl.cc
@@ -1028,6 +1028,8 @@ void RenderView::ApplyWebPreferences(const WebPreferences& prefs,
 #endif
 
   WebRuntimeFeatures::EnableTranslateService(prefs.translate_service_available);
+
+  web_view->SetBaseBackgroundColor(prefs.base_background_color);
 }
 
 /*static*/
