diff --git content/public/common/common_param_traits_macros.h content/public/common/common_param_traits_macros.h
index 63a73ad504d5..9b8cde03ecf4 100644
--- content/public/common/common_param_traits_macros.h
+++ content/public/common/common_param_traits_macros.h
@@ -195,6 +195,7 @@ IPC_STRUCT_TRAITS_BEGIN(content::WebPreferences)
   IPC_STRUCT_TRAITS_MEMBER(main_frame_resizes_are_orientation_changes)
   IPC_STRUCT_TRAITS_MEMBER(initialize_at_minimum_page_scale)
   IPC_STRUCT_TRAITS_MEMBER(smart_insert_delete_enabled)
+  IPC_STRUCT_TRAITS_MEMBER(base_background_color)
   IPC_STRUCT_TRAITS_MEMBER(cookie_enabled)
   IPC_STRUCT_TRAITS_MEMBER(navigate_on_drag_drop)
   IPC_STRUCT_TRAITS_MEMBER(spatial_navigation_enabled)
diff --git content/public/common/web_preferences.cc content/public/common/web_preferences.cc
index cc20936125f4..a1cde3a9a2b2 100644
--- content/public/common/web_preferences.cc
+++ content/public/common/web_preferences.cc
@@ -178,6 +178,7 @@ WebPreferences::WebPreferences()
       spatial_navigation_enabled(false),
       use_solid_color_scrollbars(false),
       navigate_on_drag_drop(true),
+      base_background_color(0xFFFFFFFF),  // Color::white
       v8_cache_options(V8_CACHE_OPTIONS_DEFAULT),
       record_whole_document(false),
       save_previous_document_resources(SavePreviousDocumentResources::NEVER),
diff --git content/public/common/web_preferences.h content/public/common/web_preferences.h
index 361505babff0..341873d000bc 100644
--- content/public/common/web_preferences.h
+++ content/public/common/web_preferences.h
@@ -197,6 +197,7 @@ struct CONTENT_EXPORT WebPreferences {
   bool spatial_navigation_enabled;
   bool use_solid_color_scrollbars;
   bool navigate_on_drag_drop;
+  uint32_t base_background_color;
   V8CacheOptions v8_cache_options;
   bool record_whole_document;
   SavePreviousDocumentResources save_previous_document_resources;
diff --git content/renderer/render_view_impl.cc content/renderer/render_view_impl.cc
index 0f5f1d8e1649..ae5c98646619 100644
--- content/renderer/render_view_impl.cc
+++ content/renderer/render_view_impl.cc
@@ -1250,6 +1250,8 @@ void RenderViewImpl::ApplyWebPreferencesInternal(
     blink::WebView* web_view,
     CompositorDependencies* compositor_deps) {
   ApplyWebPreferences(prefs, web_view);
+  if (GetWebFrameWidget())
+    GetWebFrameWidget()->SetBaseBackgroundColor(prefs.base_background_color);
 }
 
 void RenderViewImpl::OnForceRedraw(const ui::LatencyInfo& latency_info) {
