diff --git public/common/common_param_traits_macros.h public/common/common_param_traits_macros.h
index 119fc1d..c7e8699 100644
--- public/common/common_param_traits_macros.h
+++ public/common/common_param_traits_macros.h
@@ -196,6 +196,7 @@ IPC_STRUCT_TRAITS_BEGIN(content::WebPreferences)
   IPC_STRUCT_TRAITS_MEMBER(main_frame_resizes_are_orientation_changes)
   IPC_STRUCT_TRAITS_MEMBER(initialize_at_minimum_page_scale)
   IPC_STRUCT_TRAITS_MEMBER(smart_insert_delete_enabled)
+  IPC_STRUCT_TRAITS_MEMBER(base_background_color)
   IPC_STRUCT_TRAITS_MEMBER(cookie_enabled)
   IPC_STRUCT_TRAITS_MEMBER(navigate_on_drag_drop)
   IPC_STRUCT_TRAITS_MEMBER(spatial_navigation_enabled)
diff --git public/common/web_preferences.cc public/common/web_preferences.cc
index 11803fa..efe32ff 100644
--- public/common/web_preferences.cc
+++ public/common/web_preferences.cc
@@ -171,6 +171,7 @@ WebPreferences::WebPreferences()
       pinch_overlay_scrollbar_thickness(0),
       use_solid_color_scrollbars(false),
       navigate_on_drag_drop(true),
+      base_background_color(0xFFFFFFFF),  // Color::white
       v8_cache_options(V8_CACHE_OPTIONS_DEFAULT),
       slimming_paint_enabled(false),
       slimming_paint_v2_enabled(false),
diff --git public/common/web_preferences.h public/common/web_preferences.h
index bfb67a1..b7170ef 100644
--- public/common/web_preferences.h
+++ public/common/web_preferences.h
@@ -173,6 +173,7 @@ struct CONTENT_EXPORT WebPreferences {
   int pinch_overlay_scrollbar_thickness;
   bool use_solid_color_scrollbars;
   bool navigate_on_drag_drop;
+  uint32_t base_background_color;
   V8CacheOptions v8_cache_options;
   bool slimming_paint_enabled;
   bool slimming_paint_v2_enabled;
diff --git renderer/render_view_impl.cc renderer/render_view_impl.cc
index 5f1e6d8..09e81e2 100644
--- renderer/render_view_impl.cc
+++ renderer/render_view_impl.cc
@@ -970,6 +970,8 @@ void RenderView::ApplyWebPreferences(const WebPreferences& prefs,
 
   settings->setJavaEnabled(prefs.java_enabled);
 
+  web_view->setBaseBackgroundColor(prefs.base_background_color);
+
   // By default, allow_universal_access_from_file_urls is set to false and thus
   // we mitigate attacks from local HTML files by not granting file:// URLs
   // universal access. Only test shell will enable this.
