Index: common/webpreferences.cc
===================================================================
--- common/webpreferences.cc	(revision 272007)
+++ common/webpreferences.cc	(working copy)
@@ -114,6 +114,7 @@
       use_solid_color_scrollbars(false),
       compositor_touch_hit_testing(true),
       navigate_on_drag_drop(true),
+      base_background_color(0xFFFFFFFF),  // Color::white
       cookie_enabled(true),
       pepper_accelerated_video_decode_enabled(false)
 #if defined(OS_ANDROID)
Index: common/webpreferences.h
===================================================================
--- common/webpreferences.h	(revision 272007)
+++ common/webpreferences.h	(working copy)
@@ -149,6 +149,7 @@
   bool use_solid_color_scrollbars;
   bool compositor_touch_hit_testing;
   bool navigate_on_drag_drop;
+  uint32_t base_background_color;
 
   // This flags corresponds to a Page's Settings' setCookieEnabled state. It
   // only controls whether or not the "document.cookie" field is properly
