Index: content_browser_client.cc
===================================================================
--- content_browser_client.cc	(revision 275973)
+++ content_browser_client.cc	(working copy)
@@ -252,6 +252,10 @@
   return true;
 }
 
+SkColor ContentBrowserClient::GetBaseBackgroundColor(RenderViewHost* rvh) {
+  return SK_ColorWHITE;
+}
+
 base::FilePath ContentBrowserClient::GetDefaultDownloadDirectory() {
   return base::FilePath();
 }
Index: content_browser_client.h
===================================================================
--- content_browser_client.h	(revision 275973)
+++ content_browser_client.h	(working copy)
@@ -23,6 +23,7 @@
 #include "net/cookies/canonical_cookie.h"
 #include "net/url_request/url_request_interceptor.h"
 #include "net/url_request/url_request_job_factory.h"
+#include "third_party/skia/include/core/SkColor.h"
 #include "third_party/WebKit/public/web/WebNotificationPresenter.h"
 #include "ui/base/window_open_disposition.h"
 #include "webkit/browser/fileapi/file_system_context.h"
@@ -533,6 +534,9 @@
   // Clears browser cookies.
   virtual void ClearCookies(RenderViewHost* rvh) {}
 
+  // Returns the base background color.
+  virtual SkColor GetBaseBackgroundColor(RenderViewHost* rvh);
+
   // Returns the default download directory.
   // This can be called on any thread.
   virtual base::FilePath GetDefaultDownloadDirectory();
Index: web_contents.cc
===================================================================
--- web_contents.cc	(revision 275973)
+++ web_contents.cc	(working copy)
@@ -17,7 +17,9 @@
       main_frame_routing_id(MSG_ROUTING_NONE),
       initially_hidden(false),
       guest_instance_id(0),
-      context(NULL) {}
+      context(NULL),
+      view(NULL),
+      delegate_view(NULL) {}
 
 WebContents::CreateParams::CreateParams(
     BrowserContext* context, SiteInstance* site)
@@ -29,7 +31,9 @@
       main_frame_routing_id(MSG_ROUTING_NONE),
       initially_hidden(false),
       guest_instance_id(0),
-      context(NULL) {}
+      context(NULL),
+      view(NULL),
+      delegate_view(NULL) {}
 
 WebContents::CreateParams::~CreateParams() {
 }
Index: web_contents.h
===================================================================
--- web_contents.h	(revision 275973)
+++ web_contents.h	(working copy)
@@ -50,9 +50,11 @@
 class RenderFrameHost;
 class RenderProcessHost;
 class RenderViewHost;
+class RenderViewHostDelegateView;
 class RenderWidgetHostView;
 class SiteInstance;
 class WebContentsDelegate;
+class WebContentsView;
 struct CustomContextMenuContext;
 struct DropData;
 struct RendererPreferences;
@@ -120,6 +122,10 @@
     // Used to specify the location context which display the new view should
     // belong. This can be NULL if not needed.
     gfx::NativeView context;
+    
+    // Optionally specify the view and delegate view.
+    content::WebContentsView* view;
+    content::RenderViewHostDelegateView* delegate_view;
   };
 
   // Creates a new WebContents.
Index: web_contents_delegate.cc
===================================================================
--- web_contents_delegate.cc	(revision 275973)
+++ web_contents_delegate.cc	(working copy)
@@ -136,7 +136,9 @@
     const base::string16& frame_name,
     const GURL& target_url,
     const std::string& partition_id,
-    SessionStorageNamespace* session_storage_namespace) {
+    SessionStorageNamespace* session_storage_namespace,
+    content::WebContentsView** view,
+    content::RenderViewHostDelegateView** delegate_view) {
   return true;
 }
 
Index: web_contents_delegate.h
===================================================================
--- web_contents_delegate.h	(revision 275973)
+++ web_contents_delegate.h	(working copy)
@@ -36,9 +36,11 @@
 class JavaScriptDialogManager;
 class PageState;
 class RenderViewHost;
+class RenderViewHostDelegateView;
 class SessionStorageNamespace;
 class WebContents;
 class WebContentsImpl;
+class WebContentsView;
 struct ColorSuggestion;
 struct ContextMenuParams;
 struct DropData;
@@ -312,7 +314,9 @@
       const base::string16& frame_name,
       const GURL& target_url,
       const std::string& partition_id,
-      SessionStorageNamespace* session_storage_namespace);
+      SessionStorageNamespace* session_storage_namespace,
+      content::WebContentsView** view,
+      content::RenderViewHostDelegateView** delegate_view);
 
   // Notifies the delegate about the creation of a new WebContents. This
   // typically happens when popups are created.
