diff --git net/base/directory_lister.cc net/base/directory_lister.cc
index 58cce4935385662099e84fadd0c734b0b112647e..33b00235918c34e62422a6b12df4a7f8abf3c111 100644
--- net/base/directory_lister.cc
+++ net/base/directory_lister.cc
@@ -200,7 +200,7 @@ void DirectoryLister::OnListFile(const DirectoryListerData& data) {
 }
 
 void DirectoryLister::OnListDone(int error) {
-  delegate_->OnListDone(error);
+  delegate_.ExtractAsDangling()->OnListDone(error);
 }
 
 }  // namespace net
diff --git net/base/directory_lister.h net/base/directory_lister.h
index c600b1204bf63a7e910ea468e1c2770a7b2a0d65..b1cc437f8e64f3ef5211a6c1e166f58d292ea2a0 100644
--- net/base/directory_lister.h
+++ net/base/directory_lister.h
@@ -133,7 +133,7 @@ class NET_EXPORT DirectoryLister  {
   void OnListDone(int error);
 
   scoped_refptr<Core> core_;
-  const raw_ptr<DirectoryListerDelegate> delegate_;
+  raw_ptr<DirectoryListerDelegate> delegate_;
 };
 
 }  // namespace net
