diff --git net/base/directory_lister.cc net/base/directory_lister.cc
index 58cce4935385662099e84fadd0c734b0b112647e..33b00235918c34e62422a6b12df4a7f8abf3c111 100644
--- net/base/directory_lister.cc
+++ net/base/directory_lister.cc
@@ -200,7 +200,7 @@ void DirectoryLister::OnListFile(const DirectoryListerData& data) {
 }
 
 void DirectoryLister::OnListDone(int error) {
-  delegate_->OnListDone(error);
+  delegate_.ExtractAsDangling()->OnListDone(error);
 }
 
 }  // namespace net
diff --git net/base/directory_lister.h net/base/directory_lister.h
index 7dfb7f451e9365d2c5154d516edadb43b149c409..8d0d35e500fdc9b0f7ecfc49522a0bb2383c356f 100644
--- net/base/directory_lister.h
+++ net/base/directory_lister.h
@@ -133,7 +133,7 @@ class NET_EXPORT DirectoryLister  {
   void OnListDone(int error);
 
   scoped_refptr<Core> core_;
-  const raw_ptr<DirectoryListerDelegate> delegate_;
+  raw_ptr<DirectoryListerDelegate> delegate_;
 };
 
 }  // namespace net
