diff --git build/config/compiler/compiler.gni build/config/compiler/compiler.gni
index c351d6babb8d..db5b4f0b973a 100644
--- build/config/compiler/compiler.gni
+++ build/config/compiler/compiler.gni
@@ -248,18 +248,6 @@ if (symbol_level == -1) {
   }
 }
 
-# Non-component debug builds with symbol_level = 2 are an undesirable (very slow
-# build times) and unsupported (some test binaries will fail with > 4 GB PDBs)
-# combination. This is only checked when current_toolchain == default_toolchain
-# because the is_component_build flag is set to false in various components of
-# the build (like nacl) and we don't want to assert on those.
-# iOS does not support component builds so add an exception for this platform.
-if (build_with_chromium) {
-  assert(symbol_level != 2 || current_toolchain != default_toolchain ||
-             is_component_build || !is_debug || is_ios,
-         "Can't do non-component debug builds at symbol_level=2")
-}
-
 # Assert that the configuration isn't going to hit https://crbug.com/648948.
 # An exception is made when target_os == "chromeos" as we only use the Android
 # toolchain there to build relatively small binaries.
