diff --git render_thread_impl.cc render_thread_impl.cc
index 3842bca..caaa3e3 100644
--- render_thread_impl.cc
+++ render_thread_impl.cc
@@ -834,9 +834,7 @@ void RenderThreadImpl::EnsureWebKitInitialized() {
 
   webkit_platform_support_.reset(new RendererWebKitPlatformSupportImpl);
   blink::initialize(webkit_platform_support_.get());
-  main_thread_compositor_task_runner_ =
-      make_scoped_refptr(new SchedulerProxyTaskRunner<
-          &blink::WebSchedulerProxy::postCompositorTask>());
+  main_thread_compositor_task_runner_ = base::MessageLoopProxy::current();
 
   v8::Isolate* isolate = blink::mainThreadIsolate();
 
