diff --git content/browser/renderer_host/input/render_widget_host_latency_tracker.cc content/browser/renderer_host/input/render_widget_host_latency_tracker.cc
index 02d6fe5..6d26a26 100644
--- content/browser/renderer_host/input/render_widget_host_latency_tracker.cc
+++ content/browser/renderer_host/input/render_widget_host_latency_tracker.cc
@@ -99,7 +99,6 @@ void UpdateLatencyCoordinates(const WebInputEvent& event,
 #define CONFIRM_VALID_TIMING(start, end)     \
   DCHECK(!start.first_event_time.is_null()); \
   DCHECK(!end.last_event_time.is_null());    \
-  DCHECK_GE(end.last_event_time, start.first_event_time);
 
 // Touch/wheel to scroll latency using Rappor.
 #define RAPPOR_TOUCH_WHEEL_TO_SCROLL_LATENCY(delegate, name, start, end) \
