diff --git content/browser/renderer_host/render_view_host_impl.cc content/browser/renderer_host/render_view_host_impl.cc
index 950fbd61f3a95..27064eb0a26bf 100644
--- content/browser/renderer_host/render_view_host_impl.cc
+++ content/browser/renderer_host/render_view_host_impl.cc
@@ -643,6 +643,8 @@ bool RenderViewHostImpl::IsRenderViewLive() {
 }
 
 void RenderViewHostImpl::SetBackgroundOpaque(bool opaque) {
+  if (!GetWidget()->GetAssociatedFrameWidget().is_bound())
+    return;
   GetWidget()->GetAssociatedFrameWidget()->SetBackgroundOpaque(opaque);
 }
 
