diff --git content/browser/renderer_host/render_view_host_impl.cc content/browser/renderer_host/render_view_host_impl.cc
index 4f914119d21b..5be76fbbda36 100644
--- content/browser/renderer_host/render_view_host_impl.cc
+++ content/browser/renderer_host/render_view_host_impl.cc
@@ -522,6 +522,8 @@ bool RenderViewHostImpl::IsRenderViewLive() {
 }
 
 void RenderViewHostImpl::SetBackgroundOpaque(bool opaque) {
+  if (!GetWidget()->GetAssociatedFrameWidget().is_bound())
+    return;
   GetWidget()->GetAssociatedFrameWidget()->SetBackgroundOpaque(opaque);
 }
 
