diff --git content/browser/renderer_host/render_widget_host_view_aura.cc content/browser/renderer_host/render_widget_host_view_aura.cc
index 3324954f298227456904c4d3cfe9e1e42dcc61af..c7db8755f9f25ad1a56ce8b002fe85c260d9157a 100644
--- content/browser/renderer_host/render_widget_host_view_aura.cc
+++ content/browser/renderer_host/render_widget_host_view_aura.cc
@@ -8,6 +8,7 @@
 #include <memory>
 #include <set>
 #include <string_view>
+#include <tuple>
 #include <utility>
 
 #include "base/auto_reset.h"
@@ -56,6 +57,7 @@
 #include "content/public/browser/content_browser_client.h"
 #include "content/public/browser/device_service.h"
 #include "content/public/browser/render_view_host.h"
+#include "content/public/browser/web_contents.h"
 #include "content/public/common/page_visibility_state.h"
 #include "services/service_manager/public/cpp/interface_provider.h"
 #include "third_party/blink/public/common/input/web_input_event.h"
@@ -868,6 +870,9 @@ gfx::Rect RenderWidgetHostViewAura::GetViewBoundsWithoutTransform() {
 void RenderWidgetHostViewAura::UpdateBackgroundColor() {
   CHECK(GetBackgroundColor());
 
+  if (!window_) {
+    return;
+  }
   SkColor4f background_color =
       SkColor4f::FromColor(GetBackgroundColor().value());
   window_->layer()->SetFillsBoundsOpaquely(background_color.isOpaque());
@@ -1253,6 +1258,12 @@ void RenderWidgetHostViewAura::TransformPointToRootSurface(gfx::PointF* point) {
 }
 
 gfx::Rect RenderWidgetHostViewAura::GetBoundsInScreen() {
+  if (!root_window_bounds_callback_.is_null()) {
+    if (auto bounds = root_window_bounds_callback_.Run()) {
+      return *bounds;
+    }
+  }
+
   aura::Window* top_level = window_->GetToplevelWindow();
   gfx::Rect bounds(top_level->GetBoundsInScreen());
 
@@ -2986,6 +2997,16 @@ void RenderWidgetHostViewAura::CreateAuraWindow(aura::client::WindowType type) {
   window_->layer()->SetFillsBoundsOpaquely(background_color.isOpaque());
   window_->layer()->AsSurface()->SetFallbackBackgroundColor(background_color);
   UpdateFrameSinkIdRegistration();
+
+  // Do this after |window_| is created to avoid crashes on Win10.
+  // See https://crbug.com/761389.
+  auto* web_contents =
+      WebContents::FromRenderViewHost(RenderViewHost::From(host()));
+  if (web_contents) {
+    // TODO(mostynb): actually use prefs.  Landing this as a separate CL
+    // first to rebaseline some unreliable layout tests.
+    std::ignore = web_contents->GetOrCreateWebPreferences();
+  }
 }
 
 void RenderWidgetHostViewAura::UpdateFrameSinkIdRegistration() {
diff --git content/browser/renderer_host/render_widget_host_view_aura.h content/browser/renderer_host/render_widget_host_view_aura.h
index 21422e7e2d44aa4766d081da2e42e7ba5ebd82f4..ffea7823346063f5307927377987df84c18e95ed 100644
--- content/browser/renderer_host/render_widget_host_view_aura.h
+++ content/browser/renderer_host/render_widget_host_view_aura.h
@@ -482,6 +482,12 @@ class CONTENT_EXPORT RenderWidgetHostViewAura
 
   bool ShouldUseDefaultDeadlineOnResize() const;
 
+  using RootWindowBoundsCallback =
+      base::RepeatingCallback<std::optional<gfx::Rect>()>;
+  void SetRootWindowBoundsCallback(const RootWindowBoundsCallback& callback) {
+    root_window_bounds_callback_ = callback;
+  }
+
  protected:
   ~RenderWidgetHostViewAura() override;
 
@@ -909,6 +915,8 @@ class CONTENT_EXPORT RenderWidgetHostViewAura
 
   bool use_default_deadline_on_resize_ = false;
 
+  RootWindowBoundsCallback root_window_bounds_callback_;
+
   base::WeakPtrFactory<RenderWidgetHostViewAura> weak_ptr_factory_{this};
 };
 
