diff --git ui/base/resource/resource_bundle.cc ui/base/resource/resource_bundle.cc
index fbdb93989f2332c9ed193575153fe0463ce13306..f10317f58de6092f6fad450bf935c260875c5f4e 100644
--- ui/base/resource/resource_bundle.cc
+++ ui/base/resource/resource_bundle.cc
@@ -1019,6 +1019,12 @@ ResourceBundle::ResourceBundle(Delegate* delegate)
       locale_resources_data_lock_(new base::Lock),
       resource_handles_lock_(new base::Lock),
       max_scale_factor_(k100Percent) {
+  // With CEF's multi-threaded mode the ResourceBundle may be created on the
+  // main thread and then accessed on the UI thread. Allow the SequenceChecker
+  // to re-bind on the UI thread when CalledOnValidSequence() is called for the
+  // first time.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+
   mangle_localized_strings_ = base::CommandLine::ForCurrentProcess()->HasSwitch(
       switches::kMangleLocalizedStrings);
 }
@@ -1028,6 +1034,11 @@ ResourceBundle::~ResourceBundle() {
   UnloadLocaleResources();
 }
 
+void ResourceBundle::CleanupOnUIThread() {
+  FreeImages();
+  font_cache_.clear();
+}
+
 // static
 void ResourceBundle::InitSharedInstance(Delegate* delegate) {
   DCHECK(g_shared_instance_ == nullptr) << "ResourceBundle initialized twice";
diff --git ui/base/resource/resource_bundle.h ui/base/resource/resource_bundle.h
index 874d77bd9543a4dfcea0e930bc8baabcc92c6a69..4f7c5d52b4eb93771dbea564d65b9b035ab11155 100644
--- ui/base/resource/resource_bundle.h
+++ ui/base/resource/resource_bundle.h
@@ -266,6 +266,11 @@ class COMPONENT_EXPORT(UI_BASE) ResourceBundle {
   ResourceBundle(const ResourceBundle&) = delete;
   ResourceBundle& operator=(const ResourceBundle&) = delete;
 
+  // With CEF's multi-threaded mode the ResourceBundle may be created/destroyed
+  // on the main thread but accessed on the UI thread. Call this method on the
+  // UI thread to clean up resources before destruction.
+  void CleanupOnUIThread();
+
   // Loads a secondary locale data pack using the given file region.
   void LoadAdditionalLocaleDataWithPakFileRegion(
       base::File pak_file,
