diff --git ui/base/resource/resource_bundle.cc ui/base/resource/resource_bundle.cc
index 432a14f39c27a17b6e82e1c77b863c85642f0e5e..052e94e293e55456d027d15225841f493a19cebe 100644
--- ui/base/resource/resource_bundle.cc
+++ ui/base/resource/resource_bundle.cc
@@ -1018,6 +1018,12 @@ ResourceBundle::ResourceBundle(Delegate* delegate)
       locale_resources_data_lock_(new base::Lock),
       resource_handles_lock_(new base::Lock),
       max_scale_factor_(k100Percent) {
+  // With CEF's multi-threaded mode the ResourceBundle may be created on the
+  // main thread and then accessed on the UI thread. Allow the SequenceChecker
+  // to re-bind on the UI thread when CalledOnValidSequence() is called for the
+  // first time.
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+
   mangle_localized_strings_ = base::CommandLine::ForCurrentProcess()->HasSwitch(
       switches::kMangleLocalizedStrings);
 }
@@ -1027,6 +1033,11 @@ ResourceBundle::~ResourceBundle() {
   UnloadLocaleResources();
 }
 
+void ResourceBundle::CleanupOnUIThread() {
+  FreeImages();
+  font_cache_.clear();
+}
+
 // static
 void ResourceBundle::InitSharedInstance(Delegate* delegate) {
   DCHECK(g_shared_instance_ == nullptr) << "ResourceBundle initialized twice";
diff --git ui/base/resource/resource_bundle.h ui/base/resource/resource_bundle.h
index c742164e29ab9898a73c3070bf83fe474ca3ba5f..bf9848fcffa7711b9acd248b02c293e62b5ee16d 100644
--- ui/base/resource/resource_bundle.h
+++ ui/base/resource/resource_bundle.h
@@ -263,6 +263,11 @@ class COMPONENT_EXPORT(UI_BASE) ResourceBundle {
   ResourceBundle(const ResourceBundle&) = delete;
   ResourceBundle& operator=(const ResourceBundle&) = delete;
 
+  // With CEF's multi-threaded mode the ResourceBundle may be created/destroyed
+  // on the main thread but accessed on the UI thread. Call this method on the
+  // UI thread to clean up resources before destruction.
+  void CleanupOnUIThread();
+
   // Loads a secondary locale data pack using the given file region.
   void LoadAdditionalLocaleDataWithPakFileRegion(
       base::File pak_file,
