diff --git content/browser/renderer_host/render_frame_host_impl.cc content/browser/renderer_host/render_frame_host_impl.cc
index ff42c4bc7d776dbb4cb27d5c2c7b1a78dba9c750..3f5682dffa3952beb9074b91a6d58f93ae59cc4a 100644
--- content/browser/renderer_host/render_frame_host_impl.cc
+++ content/browser/renderer_host/render_frame_host_impl.cc
@@ -9770,6 +9770,16 @@ void RenderFrameHostImpl::CreateNewWindow(
     return;
   }
 
+  callback = base::BindOnce(
+      [](RenderFrameHostImpl* self,
+         CreateNewWindowCallback callback,
+         mojom::CreateNewWindowStatus status,
+         mojom::CreateNewWindowReplyPtr reply) {
+        GetContentClient()->browser()->CreateWindowResult(
+            self, status == mojom::CreateNewWindowStatus::kSuccess);
+        std::move(callback).Run(status, std::move(reply));
+      }, base::Unretained(this), std::move(callback));
+
   // Otherwise, consume user activation before we proceed. In particular, it is
   // important to do this before we return from the |opener_suppressed| case
   // below.
@@ -12192,6 +12202,7 @@ void RenderFrameHostImpl::CommitNavigation(
       ProcessLock::FromSiteInfo(GetSiteInstance()->GetSiteInfo());
   auto browser_calc_origin_to_commit = navigation_request->GetOriginToCommit();
   if (!process_lock.is_error_page() && !is_mhtml_subframe &&
+      common_params->url.IsStandard() &&
       !policy->CanAccessOrigin(
           GetProcess()->GetDeprecatedID(),
           browser_calc_origin_to_commit.value(),
