diff --git content/browser/renderer_host/render_frame_host_impl.cc content/browser/renderer_host/render_frame_host_impl.cc
index 08bd4bc6c74145c8469ae34e25a3044dab86a941..1541a97ff8d19ae6c94b80ad4d1f06b0641d1efb 100644
--- content/browser/renderer_host/render_frame_host_impl.cc
+++ content/browser/renderer_host/render_frame_host_impl.cc
@@ -9885,6 +9885,16 @@ void RenderFrameHostImpl::CreateNewWindow(
     return;
   }
 
+  callback = base::BindOnce(
+      [](RenderFrameHostImpl* self,
+         CreateNewWindowCallback callback,
+         mojom::CreateNewWindowStatus status,
+         mojom::CreateNewWindowReplyPtr reply) {
+        GetContentClient()->browser()->CreateWindowResult(
+            self, status == mojom::CreateNewWindowStatus::kSuccess);
+        std::move(callback).Run(status, std::move(reply));
+      }, base::Unretained(this), std::move(callback));
+
   // Otherwise, consume user activation before we proceed. In particular, it is
   // important to do this before we return from the |opener_suppressed| case
   // below.
@@ -12431,6 +12441,7 @@ void RenderFrameHostImpl::CommitNavigation(
       ProcessLock::FromSiteInfo(GetSiteInstance()->GetSiteInfo());
   auto browser_calc_origin_to_commit = navigation_request->GetOriginToCommit();
   if (!process_lock.is_error_page() && !is_mhtml_subframe &&
+      common_params->url.IsStandard() &&
       !policy->CanAccessOrigin(
           GetProcess()->GetDeprecatedID(),
           browser_calc_origin_to_commit.value(),
