diff --git ui/display/screen.cc ui/display/screen.cc
index b918c494ed0c664f456718c635efdcb173c7baec..8cbcd615b585032ff0a7a08dd490734896c0b7a4 100644
--- ui/display/screen.cc
+++ ui/display/screen.cc
@@ -104,13 +104,13 @@ base::TimeDelta Screen::CalculateIdleTime() const {
 gfx::Rect Screen::ScreenToDIPRectInWindow(gfx::NativeWindow window,
                                           const gfx::Rect& screen_rect) const {
   float scale = GetDisplayNearestWindow(window).device_scale_factor();
-  return ScaleToEnclosingRect(screen_rect, 1.0f / scale);
+  return ScaleToRoundedRect(screen_rect, 1.0f / scale);
 }
 
 gfx::Rect Screen::DIPToScreenRectInWindow(gfx::NativeWindow window,
                                           const gfx::Rect& dip_rect) const {
   float scale = GetDisplayNearestWindow(window).device_scale_factor();
-  return ScaleToEnclosingRect(dip_rect, scale);
+  return ScaleToRoundedRect(dip_rect, scale);
 }
 
 bool Screen::GetDisplayWithDisplayId(int64_t display_id,
diff --git ui/display/win/screen_win.cc ui/display/win/screen_win.cc
index bfc40f8073decb0394afc7f4f589549760a6219a..dc7a02fef54709776556fcf844c30afb62b691aa 100644
--- ui/display/win/screen_win.cc
+++ ui/display/win/screen_win.cc
@@ -749,7 +749,7 @@ gfx::Rect ScreenWin::ScreenToDIPRect(HWND hwnd,
                 gfx::PointF(pixel_bounds.origin()), screen_win_display))
           : ToFlooredPoint(display::win::ScreenToDIPPoint(
                 gfx::PointF(pixel_bounds.origin()), screen_win_display));
-  return {origin, ScaleToEnclosingRect(pixel_bounds, scale_factor).size()};
+  return {origin, ScaleToRoundedRect(pixel_bounds, scale_factor).size()};
 }
 
 gfx::Rect ScreenWin::DIPToScreenRect(HWND hwnd,
@@ -764,7 +764,7 @@ gfx::Rect ScreenWin::DIPToScreenRect(HWND hwnd,
   const gfx::Point origin =
       display::win::DIPToScreenPoint(dip_bounds.origin(), screen_win_display);
   const float scale_factor = screen_win_display.display().device_scale_factor();
-  return {origin, ScaleToEnclosingRect(dip_bounds, scale_factor).size()};
+  return {origin, ScaleToRoundedRect(dip_bounds, scale_factor).size()};
 }
 
 gfx::Rect ScreenWin::ClientToDIPRect(HWND hwnd,
