diff --git ui/display/screen.cc ui/display/screen.cc
index 8208c0df3b8405d4e3f5d01b3664c72048b00699..8450effc2afc17abdeba28f3b772e2a72953fa28 100644
--- ui/display/screen.cc
+++ ui/display/screen.cc
@@ -104,13 +104,13 @@ base::TimeDelta Screen::CalculateIdleTime() const {
 gfx::Rect Screen::ScreenToDIPRectInWindow(gfx::NativeWindow window,
                                           const gfx::Rect& screen_rect) const {
   float scale = GetDisplayNearestWindow(window).device_scale_factor();
-  return ScaleToEnclosingRect(screen_rect, 1.0f / scale);
+  return ScaleToRoundedRect(screen_rect, 1.0f / scale);
 }
 
 gfx::Rect Screen::DIPToScreenRectInWindow(gfx::NativeWindow window,
                                           const gfx::Rect& dip_rect) const {
   float scale = GetDisplayNearestWindow(window).device_scale_factor();
-  return ScaleToEnclosingRect(dip_rect, scale);
+  return ScaleToRoundedRect(dip_rect, scale);
 }
 
 bool Screen::GetDisplayWithDisplayId(int64_t display_id,
diff --git ui/display/win/screen_win.cc ui/display/win/screen_win.cc
index a3b287d975909f860901b2a8b22efcd8eb4e0155..80b2ce772b83586473ac40ebfc49277c22ee2cfc 100644
--- ui/display/win/screen_win.cc
+++ ui/display/win/screen_win.cc
@@ -701,7 +701,7 @@ gfx::Rect ScreenWin::ScreenToDIPRect(HWND hwnd,
       gfx::PointF(pixel_bounds.origin()), screen_win_display));
   const float scale_factor =
       1.0f / screen_win_display.display().device_scale_factor();
-  return {origin, ScaleToEnclosingRect(pixel_bounds, scale_factor).size()};
+  return {origin, ScaleToRoundedRect(pixel_bounds, scale_factor).size()};
 }
 
 gfx::Rect ScreenWin::DIPToScreenRect(HWND hwnd,
@@ -716,7 +716,7 @@ gfx::Rect ScreenWin::DIPToScreenRect(HWND hwnd,
   const gfx::Point origin =
       display::win::DIPToScreenPoint(dip_bounds.origin(), screen_win_display);
   const float scale_factor = screen_win_display.display().device_scale_factor();
-  return {origin, ScaleToEnclosingRect(dip_bounds, scale_factor).size()};
+  return {origin, ScaleToRoundedRect(dip_bounds, scale_factor).size()};
 }
 
 gfx::Rect ScreenWin::ClientToDIPRect(HWND hwnd,
