diff --git content/browser/storage_partition_impl.cc content/browser/storage_partition_impl.cc
index 89d3396a4b71ad2502a37bf6dc2b2c9ae201e9b6..69aa2de49739f30325e3424e3ef6d33b3faa3e07 100644
--- content/browser/storage_partition_impl.cc
+++ content/browser/storage_partition_impl.cc
@@ -3583,9 +3583,12 @@ void StoragePartitionImpl::InitNetworkContext() {
   cert_verifier::mojom::CertVerifierCreationParamsPtr
       cert_verifier_creation_params =
           cert_verifier::mojom::CertVerifierCreationParams::New();
-  GetContentClient()->browser()->ConfigureNetworkContextParams(
+  if (!GetContentClient()->browser()->ConfigureNetworkContextParams(
       browser_context_, is_in_memory(), relative_partition_path_,
-      context_params.get(), cert_verifier_creation_params.get());
+      context_params.get(), cert_verifier_creation_params.get())) {
+    // Don't re-initialize the network context during shutdown.
+    return;
+  }
   // Should be initialized with existing per-profile CORS access lists.
   DCHECK(context_params->cors_origin_access_list.empty())
       << "NetworkContextParams::cors_origin_access_list should be populated "
