diff --git ui/views/controls/webview/webview.cc ui/views/controls/webview/webview.cc
index 41afec12a8b37454b6e4606aeb3f1121c9f195f2..17f6cfb2d63d4d16c5275120e5a7004585b50d18 100644
--- ui/views/controls/webview/webview.cc
+++ ui/views/controls/webview/webview.cc
@@ -223,6 +223,10 @@ const gfx::Size& WebView::AutoResizeMaxSize() const {
   return max_size_;
 }
 
+void WebView::SetResizeBackgroundColor(SkColor resize_background_color) {
+  holder_->SetBackgroundColorWhenClipped(resize_background_color);
+}
+
 void WebView::TakeCrashedOverlayViewImpl(
     std::unique_ptr<View> crashed_overlay_view,
     ReturnCrashOverlayToOwnerCallback return_to_owner) {
diff --git ui/views/controls/webview/webview.h ui/views/controls/webview/webview.h
index 28aaa94abd83d47fdc5cb3c2082c0da3f1a28725..b38320797d00b6edd00bf5d013cf74854ac56803 100644
--- ui/views/controls/webview/webview.h
+++ ui/views/controls/webview/webview.h
@@ -137,6 +137,10 @@ class WEBVIEW_EXPORT WebView : public View,
   const gfx::Size& AutoResizeMinSize() const;
   const gfx::Size& AutoResizeMaxSize() const;
 
+  // Set the background color to use while resizing with a clip. This is white
+  // by default.
+  void SetResizeBackgroundColor(SkColor resize_background_color);
+
   // A scoped object that disconnects the webview from the accessibility tree.
   // When destroyed, it restores the previous accessibility state.
   class WEBVIEW_EXPORT ScopedAxDisconnectLock {
