diff --git ui/views/controls/webview/webview.cc ui/views/controls/webview/webview.cc
index 30564a6a03ffc7e1f201c7f0942b9814788f38b1..f15ce67ebabc8d05bbec41ea24b8b3cb8e8ed460 100644
--- ui/views/controls/webview/webview.cc
+++ ui/views/controls/webview/webview.cc
@@ -174,6 +174,10 @@ void WebView::EnableSizingFromWebContents(const gfx::Size& min_size,
   }
 }
 
+void WebView::SetResizeBackgroundColor(SkColor resize_background_color) {
+  holder_->SetBackgroundColorWhenClipped(resize_background_color);
+}
+
 void WebView::SetCrashedOverlayView(View* crashed_overlay_view) {
   if (crashed_overlay_view_.view() == crashed_overlay_view) {
     return;
diff --git ui/views/controls/webview/webview.h ui/views/controls/webview/webview.h
index 50fb7d2d5769dece09f9e2845b70c251ae2326e1..9983d6f8bffff9d35f1663dd325a4b0267d8a5e3 100644
--- ui/views/controls/webview/webview.h
+++ ui/views/controls/webview/webview.h
@@ -110,6 +110,10 @@ class WEBVIEW_EXPORT WebView : public View,
   void EnableSizingFromWebContents(const gfx::Size& min_size,
                                    const gfx::Size& max_size);
 
+  // Set the background color to use while resizing with a clip. This is white
+  // by default.
+  void SetResizeBackgroundColor(SkColor resize_background_color);
+
   // If provided, this View will be shown in place of the web contents
   // when the web contents is in a crashed state. This is cleared automatically
   // if the web contents is changed. The passed-in overlay view must be owned by
