diff --git ui/views/controls/webview/webview.cc ui/views/controls/webview/webview.cc
index bb046b60656d9a2cb521cf39ec4e8b358b080d02..e12f0d68330c2f3e4d82d03f57e6dcc8ffe97ae0 100644
--- ui/views/controls/webview/webview.cc
+++ ui/views/controls/webview/webview.cc
@@ -175,6 +175,10 @@ void WebView::EnableSizingFromWebContents(const gfx::Size& min_size,
   }
 }
 
+void WebView::SetResizeBackgroundColor(SkColor resize_background_color) {
+  holder_->SetBackgroundColorWhenClipped(resize_background_color);
+}
+
 void WebView::SetCrashedOverlayView(View* crashed_overlay_view) {
   if (crashed_overlay_view_.view() == crashed_overlay_view) {
     return;
diff --git ui/views/controls/webview/webview.h ui/views/controls/webview/webview.h
index 78a7e12e14794896fc57434bece1579d20f0a3a1..71775ac2eb8fd17d809b9557874a2c917b7b376e 100644
--- ui/views/controls/webview/webview.h
+++ ui/views/controls/webview/webview.h
@@ -125,6 +125,10 @@ class WEBVIEW_EXPORT WebView : public View,
   void EnableSizingFromWebContents(const gfx::Size& min_size,
                                    const gfx::Size& max_size);
 
+  // Set the background color to use while resizing with a clip. This is white
+  // by default.
+  void SetResizeBackgroundColor(SkColor resize_background_color);
+
   // If provided, this View will be shown in place of the web contents
   // when the web contents is in a crashed state. This is cleared automatically
   // if the web contents is changed. The passed-in overlay view must be owned by
