diff --git ui/views/controls/webview/webview.cc ui/views/controls/webview/webview.cc
index 823fabedf1f065a11a8c6d4c1e8a975663955418..f30cc1adb2e9185793d8bf44624220071ced778d 100644
--- ui/views/controls/webview/webview.cc
+++ ui/views/controls/webview/webview.cc
@@ -211,6 +211,10 @@ void WebView::EnableSizingFromWebContents(const gfx::Size& min_size,
   }
 }
 
+void WebView::SetResizeBackgroundColor(SkColor resize_background_color) {
+  holder_->SetBackgroundColorWhenClipped(resize_background_color);
+}
+
 void WebView::TakeCrashedOverlayViewImpl(
     std::unique_ptr<View> crashed_overlay_view,
     ReturnCrashOverlayToOwnerCallback return_to_owner) {
diff --git ui/views/controls/webview/webview.h ui/views/controls/webview/webview.h
index 867f5f4cd43c47bc078d0e40d5363c84032250f2..a16a23743b6c5858fd66dbaad4973290d6d39f46 100644
--- ui/views/controls/webview/webview.h
+++ ui/views/controls/webview/webview.h
@@ -135,6 +135,10 @@ class WEBVIEW_EXPORT WebView : public View,
   void EnableSizingFromWebContents(const gfx::Size& min_size,
                                    const gfx::Size& max_size);
 
+  // Set the background color to use while resizing with a clip. This is white
+  // by default.
+  void SetResizeBackgroundColor(SkColor resize_background_color);
+
   // A scoped object that disconnects the webview from the accessibility tree.
   // When destroyed, it restores the previous accessibility state.
   class WEBVIEW_EXPORT ScopedAxDisconnectLock {
