Index: src/GrGpuGL.cpp
===================================================================
--- src/GrGpuGL.cpp	(revision 1763)
+++ src/GrGpuGL.cpp	(working copy)
@@ -707,7 +707,6 @@
     static const StencilFormat esStencilFormats[] = {
         {GR_GL_STENCIL_INDEX8,     8,   false},
         {GR_GL_DEPTH24_STENCIL8,   8,   true },
-        {GR_GL_STENCIL_INDEX4,     4,   false},
     };
 
     if (GR_GL_SUPPORT_DESKTOP) {
