Index: frame/FrameView.cpp
===================================================================
--- frame/FrameView.cpp	(revision 160715)
+++ frame/FrameView.cpp	(working copy)
@@ -200,8 +200,10 @@
     if (!isMainFrame())
         return;
 
+#if 0
     ScrollableArea::setVerticalScrollElasticity(ScrollElasticityAllowed);
     ScrollableArea::setHorizontalScrollElasticity(ScrollElasticityAllowed);
+#endif
 }
 
 PassRefPtr<FrameView> FrameView::create(Frame* frame)
Index: platform/mac/NSScrollerImpDetails.mm
===================================================================
--- platform/mac/NSScrollerImpDetails.mm	(revision 160715)
+++ platform/mac/NSScrollerImpDetails.mm	(working copy)
@@ -73,10 +73,14 @@
 
 bool isScrollbarOverlayAPIAvailable()
 {
+#if 0
     static bool apiAvailable =
         [NSClassFromString(@"NSScrollerImp") respondsToSelector:@selector(scrollerImpWithStyle:controlSize:horizontal:replacingScrollerImp:)]
         && [NSClassFromString(@"NSScrollerImpPair") instancesRespondToSelector:@selector(scrollerStyle)];
     return apiAvailable;
+#else
+    return false;
+#endif
 }
 
 NSScrollerStyle recommendedScrollerStyle()
