Index: core/frame/FrameView.cpp
===================================================================
--- core/frame/FrameView.cpp	(revision 166298)
+++ core/frame/FrameView.cpp	(working copy)
@@ -168,8 +168,10 @@
     if (!isMainFrame())
         return;
 
+#if 0
     ScrollableArea::setVerticalScrollElasticity(ScrollElasticityAllowed);
     ScrollableArea::setHorizontalScrollElasticity(ScrollElasticityAllowed);
+#endif
 }
 
 PassRefPtr<FrameView> FrameView::create(Frame* frame)
Index: platform/scroll/ScrollbarThemeMacCommon.mm
===================================================================
--- platform/scroll/ScrollbarThemeMacCommon.mm	(revision 166298)
+++ platform/scroll/ScrollbarThemeMacCommon.mm	(working copy)
@@ -369,10 +369,14 @@
 // static
 bool ScrollbarThemeMacCommon::isOverlayAPIAvailable()
 {
+#if 0
     static bool apiAvailable =
         [NSClassFromString(@"NSScrollerImp") respondsToSelector:@selector(scrollerImpWithStyle:controlSize:horizontal:replacingScrollerImp:)]
         && [NSClassFromString(@"NSScrollerImpPair") instancesRespondToSelector:@selector(scrollerStyle)];
     return apiAvailable;
+#else
+    return false;
+#endif
 }
 
 } // namespace WebCore
