Index: page/FrameView.cpp
===================================================================
--- page/FrameView.cpp	(revision 109991)
+++ page/FrameView.cpp	(working copy)
@@ -157,10 +157,12 @@
             m_page = page;
             m_page->addScrollableArea(this);
 
+#if 0
             if (m_frame == m_page->mainFrame()) {
                 ScrollableArea::setVerticalScrollElasticity(ScrollElasticityAllowed);
                 ScrollableArea::setHorizontalScrollElasticity(ScrollElasticityAllowed);
             }
+#endif
         }
     }
 }
Index: platform/mac/NSScrollerImpDetails.mm
===================================================================
--- platform/mac/NSScrollerImpDetails.mm	(revision 109991)
+++ platform/mac/NSScrollerImpDetails.mm	(working copy)
@@ -33,6 +33,7 @@
 #if PLATFORM(CHROMIUM)
 bool isScrollbarOverlayAPIAvailable()
 {
+#if 0
     static bool apiAvailable;
     static bool shouldInitialize = true;
     if (shouldInitialize) {
@@ -43,6 +44,9 @@
                        && [scrollerImpPairClass instancesRespondToSelector:@selector(scrollerStyle)];
     }
     return apiAvailable;
+#else
+    return false;
+#endif
 }
 #endif
 
