diff --git stream_context.cc stream_context.cc
index 3782205..5d97643 100644
--- stream_context.cc
+++ stream_context.cc
@@ -21,6 +21,10 @@ namespace content {
 
 StreamContext::StreamContext() {}
 
+const void* StreamContext::GetUserDataKey() {
+  return kStreamContextKeyName;
+}
+
 StreamContext* StreamContext::GetFor(BrowserContext* context) {
   if (!context->GetUserData(kStreamContextKeyName)) {
     scoped_refptr<StreamContext> stream = new StreamContext();
diff --git stream_context.h stream_context.h
index 075ae3e..57fb5fd 100644
--- stream_context.h
+++ stream_context.h
@@ -29,6 +29,7 @@ class StreamContext
  public:
   StreamContext();
 
+  CONTENT_EXPORT static const void* GetUserDataKey();
   CONTENT_EXPORT static StreamContext* GetFor(BrowserContext* browser_context);
 
   void InitializeOnIOThread();
