diff --git gclient.py gclient.py
index 933a4be3531e5275ecdd498229b6733af4fc632b..f5e86d9e579e15fc79214f6ec371f3a19df10640 100755
--- gclient.py
+++ gclient.py
@@ -1002,7 +1002,8 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
         if self.git_dependencies_state == gclient_eval.SUBMODULES:
             deps.update(self.ParseGitSubmodules())
 
-        if self.git_dependencies_state != gclient_eval.DEPS:
+        if not self.get_vars().get('source_tarball', False) and \
+           self.git_dependencies_state != gclient_eval.DEPS:
             # Git submodules are used - get their state.
             self._known_dependency_diff = self.CreateSCM().GetSubmoduleDiff()
             self._dependency_index_state = self.CreateSCM(
@@ -1220,7 +1221,7 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
             self.FuzzyMatchUrl(revision_overrides), None)
         if not revision_override and not self.managed:
             revision_override = 'unmanaged'
-        if run_scm and self.url:
+        if not self.get_vars().get('source_tarball', False) and run_scm and self.url:
             # Create a shallow copy to mutate revision.
             options = copy.copy(options)
             options.revision = revision_override
@@ -2968,7 +2969,12 @@ class CipdDependency(Dependency):
         if dep_value.get('version_file'):
             fp = os.path.join(parent.root.root_dir, parent.name,
                               dep_value['version_file'])
-            version = gclient_utils.FileRead(fp).strip()
+            if os.path.isfile(fp):
+                version = gclient_utils.FileRead(fp).strip()
+            elif not should_process:
+                version = 'unknown'
+            else:
+                raise FileNotFoundError(fp)
         else:
             version = dep_value['version']
         url = urllib.parse.urljoin(cipd_root.service_url,
