diff --git gclient.py gclient.py
index 9f19223375a6e7beff929bafdf10c72dc6c434fb..07cc34a73a67481d6485d06e337a2fb1ac63bd12 100755
--- gclient.py
+++ gclient.py
@@ -1123,7 +1123,8 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
         if self.git_dependencies_state == gclient_eval.SUBMODULES:
             deps.update(self.ParseGitSubmodules())
 
-        if self.git_dependencies_state != gclient_eval.DEPS:
+        if not self.get_vars().get('source_tarball', False) and \
+           self.git_dependencies_state != gclient_eval.DEPS:
             # Git submodules are used - get their state.
             self._known_dependency_diff = self.CreateSCM().GetSubmoduleDiff()
             self._dependency_index_state = (
@@ -1363,7 +1364,7 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
         )
         if not revision_override and not self.managed:
             revision_override = "unmanaged"
-        if run_scm and self.url:
+        if not self.get_vars().get('source_tarball', False) and run_scm and self.url:
             # Create a shallow copy to mutate revision.
             options = copy.copy(options)
             options.revision = revision_override
@@ -3330,7 +3331,12 @@ class CipdDependency(Dependency):
             fp = os.path.join(
                 parent.root.root_dir, parent.name, dep_value["version_file"]
             )
-            version = gclient_utils.FileRead(fp).strip()
+            if os.path.isfile(fp):
+                version = gclient_utils.FileRead(fp).strip()
+            elif not should_process:
+                version = 'unknown'
+            else:
+                raise FileNotFoundError(fp)
         else:
             version = dep_value["version"]
         url = urllib.parse.urljoin(
