diff --git gclient.py gclient.py
index ee205ec257651e71132f3c7261a5a3d656b23606..895b950d0bfa5f8df2694ef883d8bad96fc8b6ff 100755
--- gclient.py
+++ gclient.py
@@ -1001,7 +1001,8 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
         if self.git_dependencies_state == gclient_eval.SUBMODULES:
             deps.update(self.ParseGitSubmodules())
 
-        if self.git_dependencies_state != gclient_eval.DEPS:
+        if not self.get_vars().get('source_tarball', False) and \
+           self.git_dependencies_state != gclient_eval.DEPS:
             # Git submodules are used - get their state.
             self._known_dependency_diff = self.CreateSCM().GetSubmoduleDiff()
             self._dependency_index_state = self.CreateSCM(
@@ -1219,7 +1220,7 @@ class Dependency(gclient_utils.WorkItem, DependencySettings):
             self.FuzzyMatchUrl(revision_overrides), None)
         if not revision_override and not self.managed:
             revision_override = 'unmanaged'
-        if run_scm and self.url:
+        if not self.get_vars().get('source_tarball', False) and run_scm and self.url:
             # Create a shallow copy to mutate revision.
             options = copy.copy(options)
             options.revision = revision_override
