diff --git DEPS DEPS
index 353b1d9422a396852273e08bf6ff0eeb7767b1b9..de4b83576a22660cd2b65ea734ce64f18261f7c2 100644
--- DEPS
+++ DEPS
@@ -4191,6 +4191,7 @@ hooks = [
     # Update LASTCHANGE.
     'name': 'lastchange',
     'pattern': '.',
+    'condition': 'not source_tarball',
     'action': ['python3', 'src/build/util/lastchange.py',
                '-o', 'src/build/util/LASTCHANGE'],
   },
@@ -4198,7 +4199,7 @@ hooks = [
     # Update lastchange_commit_position.h (only for CrOS).
     'name': 'lastchange_commit_position_cros',
     'pattern': '.',
-    'condition': 'checkout_chromeos',
+    'condition': 'checkout_chromeos and not source_tarball',
     'action': ['python3', 'src/build/util/lastchange.py',
                '-m', 'CHROMIUM',
                '--commit-position-header',
@@ -4208,6 +4209,7 @@ hooks = [
     # Update GPU lists version string (for gpu/config).
     'name': 'gpu_lists_version',
     'pattern': '.',
+    'condition': 'not source_tarball',
     'action': ['python3', 'src/build/util/lastchange.py',
                '-m', 'GPU_LISTS_VERSION',
                '--revision-id-only',
@@ -4217,6 +4219,7 @@ hooks = [
     # Update skia_commit_hash.h.
     'name': 'lastchange_skia',
     'pattern': '.',
+    'condition': 'not source_tarball',
     'action': ['python3', 'src/build/util/lastchange.py',
                '-m', 'SKIA_COMMIT_HASH',
                '-s', 'src/third_party/skia',
@@ -4226,6 +4229,7 @@ hooks = [
     # Update dawn_version.h.
     'name': 'lastchange_dawn',
     'pattern': '.',
+    'condition': 'not source_tarball',
     'action': ['python3', 'src/build/util/lastchange.py',
                '-s', 'src/third_party/dawn',
                '--revision', 'src/gpu/webgpu/DAWN_VERSION'],
