Index: pylib/gyp/input.py
===================================================================
--- pylib/gyp/input.py	(revision 1323)
+++ pylib/gyp/input.py	(working copy)
@@ -666,7 +666,8 @@
             # that don't load quickly, this can be faster than
             # <!(python modulename param eters). Do this in |build_file_dir|.
             oldwd = os.getcwd()  # Python doesn't like os.open('.'): no fchdir.
-            os.chdir(build_file_dir)
+            if not build_file_dir is None:
+              os.chdir(build_file_dir)
 
             parsed_contents = shlex.split(contents)
             py_module = __import__(parsed_contents[0])
Index: pylib/gyp/mac_tool.py
===================================================================
--- pylib/gyp/mac_tool.py	(revision 1323)
+++ pylib/gyp/mac_tool.py	(working copy)
@@ -54,16 +54,8 @@
       return self._CopyXIBFile(source, dest)
     elif extension == '.strings':
       self._CopyStringsFile(source, dest)
-    # TODO: Given that files with arbitrary extensions can be copied to the
-    # bundle, we will want to get rid of this whitelist eventually.
-    elif extension in [
-        '.icns', '.manifest', '.pak', '.pdf', '.png', '.sb', '.sh',
-        '.ttf', '.sdef']:
-      shutil.copyfile(source, dest)
     else:
-      raise NotImplementedError(
-          "Don't know how to copy bundle resources of type %s while copying "
-          "%s to %s)" % (extension, source, dest))
+      shutil.copyfile(source, dest)
 
   def _CopyXIBFile(self, source, dest):
     """Compiles a XIB file with ibtool into a binary plist in the bundle."""
