diff --git base/trace_event/builtin_categories.h base/trace_event/builtin_categories.h
index 60bba6a85d393..70eee12a53157 100644
--- base/trace_event/builtin_categories.h
+++ base/trace_event/builtin_categories.h
@@ -61,6 +61,8 @@
   X("cc")                                                                \
   X("cc.debug")                                                          \
   X("cdp.perf")                                                          \
+  X("cef")                                                               \
+  X("cef.client")                                                        \
   X("chromeos")                                                          \
   X("cma")                                                               \
   X("compositor")                                                        \
