diff --git base/trace_event/builtin_categories.h base/trace_event/builtin_categories.h
index 56408734b26428933a1fb8c04cee317c659feea7..7eb9a8f2d15f928894c3c8077223dc1a55553eab 100644
--- base/trace_event/builtin_categories.h
+++ base/trace_event/builtin_categories.h
@@ -106,6 +106,8 @@ PERFETTO_DEFINE_CATEGORIES_IN_NAMESPACE_WITH_ATTRS(
     perfetto::Category("cc").SetTags("rendering"),
     perfetto::Category("cc.debug").SetTags("debug"),
     perfetto::Category("cdp.perf"),
+    perfetto::Category("cef"),
+    perfetto::Category("cef.client"),
     perfetto::Category("chromeos"),
     perfetto::Category("cma"),
     perfetto::Category("compositor"),
