diff --git base/trace_event/builtin_categories.h base/trace_event/builtin_categories.h
index a96b2338f39ad..48f1ef2d96045 100644
--- base/trace_event/builtin_categories.h
+++ base/trace_event/builtin_categories.h
@@ -64,6 +64,8 @@
   X("cc")                                                                \
   X("cc.debug")                                                          \
   X("cdp.perf")                                                          \
+  X("cef")                                                               \
+  X("cef.client")                                                        \
   X("chromeos")                                                          \
   X("cma")                                                               \
   X("compositor")                                                        \
