diff --git base/trace_event/builtin_categories.h base/trace_event/builtin_categories.h
index 8649b24d17fc9d8acf988f44422134ecc3ed0203..478442388e04ed92e95c675bc4dd2fb3ea4a72ef 100644
--- base/trace_event/builtin_categories.h
+++ base/trace_event/builtin_categories.h
@@ -106,6 +106,8 @@ PERFETTO_DEFINE_CATEGORIES_IN_NAMESPACE_WITH_ATTRS(
     perfetto::Category("cc").SetTags("rendering"),
     perfetto::Category("cc.debug").SetTags("debug"),
     perfetto::Category("cdp.perf"),
+    perfetto::Category("cef"),
+    perfetto::Category("cef.client"),
     perfetto::Category("chromeos"),
     perfetto::Category("cma"),
     perfetto::Category("compositor"),
