diff --git base/trace_event/builtin_categories.h base/trace_event/builtin_categories.h
index ad3e86fb45ed..61c3db8b59d1 100644
--- base/trace_event/builtin_categories.h
+++ base/trace_event/builtin_categories.h
@@ -58,6 +58,8 @@
   X("cc")                                                                \
   X("cc.debug")                                                          \
   X("cdp.perf")                                                          \
+  X("cef")                                                               \
+  X("cef.client")                                                        \
   X("chromeos")                                                          \
   X("cma")                                                               \
   X("compositor")                                                        \
