diff --git ui/base/x/x11_util.h ui/base/x/x11_util.h
index d45396238fc802444b5db1bb3d56c9aab6309f05..327f1d2c7da171479afab4643b570384a9e69b08 100644
--- ui/base/x/x11_util.h
+++ ui/base/x/x11_util.h
@@ -246,8 +246,9 @@ x11::Future<void> SendClientMessage(
     x11::Window target,
     x11::Atom type,
     const std::array<uint32_t, 5> data,
-    x11::EventMask event_mask = x11::EventMask::SubstructureNotify |
-                                x11::EventMask::SubstructureRedirect);
+    x11::EventMask event_mask = static_cast<x11::EventMask>(
+        static_cast<int>(x11::EventMask::SubstructureNotify) |
+        static_cast<int>(x11::EventMask::SubstructureRedirect)));
 
 // Return true if VulkanSurface is supported.
 COMPONENT_EXPORT(UI_BASE_X) bool IsVulkanSurfaceSupported();
