diff --git ui/base/x/x11_os_exchange_data_provider.cc ui/base/x/x11_os_exchange_data_provider.cc
index 2f1430d55f0e61ec450a13c4fc1db870fa95ea9c..b9bdd71d62f4739e1126541636880d954526552d 100644
--- ui/base/x/x11_os_exchange_data_provider.cc
+++ ui/base/x/x11_os_exchange_data_provider.cc
@@ -166,7 +166,8 @@ void XOSExchangeDataProvider::SetURL(const GURL& url,
     format_map_.Insert(x11::GetAtom(kMimeTypeMozillaUrl), mem);
 
     // Set a string fallback as well.
-    SetString(spec);
+    if (!HasString())
+      SetString(spec);
 
     // Return early if this drag already contains file contents (this implies
     // that file contents must be populated before URLs). Nautilus (and possibly
