diff --git ui/menus/simple_menu_model.cc ui/menus/simple_menu_model.cc
index 9f56505d9502c685f66ab082b60eaae4011e76b7..509d93ae1fa22c0fe578d74b8528af4a01250240 100644
--- ui/menus/simple_menu_model.cc
+++ ui/menus/simple_menu_model.cc
@@ -11,6 +11,7 @@
 #include "base/functional/bind.h"
 #include "base/location.h"
 #include "base/task/single_thread_task_runner.h"
+#include "cef/libcef/features/features.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/models/image_model.h"
 #include "ui/gfx/image/image.h"
@@ -171,8 +172,10 @@ void SimpleMenuModel::AddSeparator(MenuSeparatorType separator_type) {
     }
 
     if (items_.at(last_visible_item).type == TYPE_SEPARATOR) {
+#if !BUILDFLAG(ENABLE_CEF)
       DCHECK_EQ(NORMAL_SEPARATOR, separator_type);
       DCHECK_EQ(NORMAL_SEPARATOR, items_.at(last_visible_item).separator_type);
+#endif
       // The last item is already a separator. Don't add another.
       return;
     }
