diff --git ui/menus/simple_menu_model.cc ui/menus/simple_menu_model.cc
index 156e5068000ceb89a096f51d358ae69a3f0cf5d8..bf8ef99561ed03fbd4a23525d48b40825fd29706 100644
--- ui/menus/simple_menu_model.cc
+++ ui/menus/simple_menu_model.cc
@@ -13,6 +13,7 @@
 #include "base/location.h"
 #include "base/notimplemented.h"
 #include "base/task/single_thread_task_runner.h"
+#include "cef/libcef/features/features.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/models/image_model.h"
 #include "ui/base/models/menu_model.h"
@@ -174,8 +175,10 @@ void SimpleMenuModel::AddSeparator(MenuSeparatorType separator_type) {
     }
 
     if (items_.at(last_visible_item).type == TYPE_SEPARATOR) {
+#if !BUILDFLAG(ENABLE_CEF)
       DCHECK_EQ(NORMAL_SEPARATOR, separator_type);
       DCHECK_EQ(NORMAL_SEPARATOR, items_.at(last_visible_item).separator_type);
+#endif
       // The last item is already a separator. Don't add another.
       return;
     }
