diff --git chrome/browser/ui/views/chrome_views_delegate_linux.cc chrome/browser/ui/views/chrome_views_delegate_linux.cc
index 8e068b986b08a93c2d8415e34d517e16d8ca01b6..3c68c31fd1038ba011f45c385ebde45f371d8052 100644
--- chrome/browser/ui/views/chrome_views_delegate_linux.cc
+++ chrome/browser/ui/views/chrome_views_delegate_linux.cc
@@ -70,6 +70,8 @@ NativeWidgetType GetNativeWidgetTypeForInitParams(
   }
 
   return (params.parent &&
+          (params.child ||
+           params.type == views::Widget::InitParams::TYPE_BUBBLE) &&
           params.type != views::Widget::InitParams::TYPE_MENU &&
           params.type != views::Widget::InitParams::TYPE_TOOLTIP)
              ? NativeWidgetType::kNativeWidgetAura
diff --git ui/views/test/desktop_test_views_delegate_aura.cc ui/views/test/desktop_test_views_delegate_aura.cc
index 18281d9a3002647786ca99b2b2bd7bef104154a0..5e86ab38b9e35044a879475846ed8d600352a1e5 100644
--- ui/views/test/desktop_test_views_delegate_aura.cc
+++ ui/views/test/desktop_test_views_delegate_aura.cc
@@ -27,7 +27,8 @@ void DesktopTestViewsDelegate::OnBeforeWidgetInit(
     return;
   }
 
-  if (params->parent && params->type != views::Widget::InitParams::TYPE_MENU &&
+  if (params->parent && params->child &&
+      params->type != views::Widget::InitParams::TYPE_MENU &&
       params->type != views::Widget::InitParams::TYPE_TOOLTIP) {
     params->native_widget = new views::NativeWidgetAura(delegate);
   } else {
