diff --git BUILD.gn BUILD.gn
index c5d180039d2..0068199adb1 100644
--- BUILD.gn
+++ BUILD.gn
@@ -11,6 +11,7 @@ import("//build/config/mips.gni")
 import("//build/config/riscv.gni")
 import("//build/config/sanitizers/sanitizers.gni")
 import("//build_overrides/build.gni")
+import("//cef/libcef/features/features.gni")
 import("//third_party/icu/config.gni")
 
 import("gni/snapshot_toolchain.gni")
@@ -454,6 +455,9 @@ declare_args() {
   # Some fuzzers depend on fuzzing functionality linked into the v8 library.
   # For binary size reasons this functionality is not always available.
   v8_wasm_random_fuzzers = ""
+
+  # Set to true if V8 will be used in a shared library.
+  v8_used_in_shared_library = enable_cef
 }
 
 # Derived defaults.
@@ -805,6 +809,10 @@ config("internal_config") {
     defines += [ "BUILDING_V8_SHARED" ]
   }
 
+  if (v8_used_in_shared_library) {
+    defines += [ "V8_TLS_USED_IN_LIBRARY" ]
+  }
+
   if (v8_current_cpu == "riscv64" || v8_current_cpu == "riscv32") {
     if (!is_clang) {
       libs = [ "atomic" ]
