diff --git BUILD.gn BUILD.gn
index 8dda88f8ad8c20256bafb94f221e4a6bde86a98b..848a0ca092b657f77f5bbd417d1fc000191e2421 100644
--- BUILD.gn
+++ BUILD.gn
@@ -12,6 +12,7 @@ import("//build/config/riscv.gni")
 import("//build/config/rust.gni")
 import("//build/config/sanitizers/sanitizers.gni")
 import("//build_overrides/build.gni")
+import("//cef/libcef/features/features.gni")
 import("//third_party/icu/config.gni")
 
 import("gni/snapshot_toolchain.gni")
@@ -471,6 +472,9 @@ declare_args() {
   # Experimental testing mode where various limits are artificially set lower.
   v8_lower_limits_mode = false
 
+  # Set to true if V8 will be used in a shared library.
+  v8_used_in_shared_library = enable_cef
+
   # Enables the use of partition_alloc as the default allocator for standalone
   # V8. This should be used for benchmarking and testing purposes to more
   # closely mimic in-browser behavior.
@@ -851,6 +855,10 @@ config("internal_config") {
     defines += [ "BUILDING_V8_SHARED" ]
   }
 
+  if (v8_used_in_shared_library) {
+    defines += [ "V8_TLS_USED_IN_LIBRARY" ]
+  }
+
   if (v8_current_cpu == "riscv64" || v8_current_cpu == "riscv32") {
     if (!is_clang) {
       libs = [ "atomic" ]
