diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index 5405f62d9292c37a3611de36625e7dff6804ce97..731e0a3c0b82c97fd24596696ceb5181e75c1790 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3333,6 +3333,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   if (GetWidget()) {
