diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index e80a6344ab662592a71f0870da2f226e8bf65614..7565d733b1aeb93e6549f1b506307fbb5ab02b34 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3133,6 +3133,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
