diff --git ui/views/accessibility/ax_virtual_view.cc ui/views/accessibility/ax_virtual_view.cc
index f2ad0e75..0660d4d 100644
--- ui/views/accessibility/ax_virtual_view.cc
+++ ui/views/accessibility/ax_virtual_view.cc
@@ -19,13 +19,6 @@
 
 namespace views {
 
-// GetClassName will be mapped to GetClassNameW if windows.h has been included
-// and the UNICODE macro has been defined. We need to undef it to use it in this
-// file.
-#ifdef GetClassName
-#undef GetClassName
-#endif
-
 // static
 const char AXVirtualView::kViewClassName[] = "AXVirtualView";
 
@@ -34,7 +27,7 @@
   ax_platform_node_ = ui::AXPlatformNode::Create(this);
   DCHECK(ax_platform_node_);
   custom_data_.AddStringAttribute(ax::mojom::StringAttribute::kClassName,
-                                  GetClassName());
+                                  GetViewClassName());
 }
 
 AXVirtualView::~AXVirtualView() {
@@ -148,7 +141,7 @@
                                  : -1;
 }
 
-const char* AXVirtualView::GetClassName() const {
+const char* AXVirtualView::GetViewClassName() const {
   return kViewClassName;
 }
 
diff --git ui/views/accessibility/ax_virtual_view.h ui/views/accessibility/ax_virtual_view.h
index c18eab3..c7e08d3 100644
--- ui/views/accessibility/ax_virtual_view.h
+++ ui/views/accessibility/ax_virtual_view.h
@@ -98,7 +98,7 @@
   // Other methods.
   //
 
-  const char* GetClassName() const;
+  const char* GetViewClassName() const;
   gfx::NativeViewAccessible GetNativeObject() const;
   void NotifyAccessibilityEvent(ax::mojom::Event event_type);
   // Allows clients to modify the AXNodeData for this virtual view.
