diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index acf09b04faa30ef3dbd6bf55214570cfe9eda3ee..8417863f690c689ed311f44e3173cba37d19faca 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3160,6 +3160,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
