diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index e672ba29b448148e15e2eb42ede9e18501d904dc..0eb719a0cc73fb45c8c3aef62802d4bd4f476866 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3302,6 +3302,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
