diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index 7c94559eda8c8799c7366399b995863d84b80a87..cb032ed925b7850e608900168de29cb3883c8ea1 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3133,6 +3133,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
