diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index c0b7fb9eb01667097add9d10021219bdbd1b940e..cebe8368a7c12f5c645870246c2e5f6c1aff9385 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3150,6 +3150,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
