diff --git ui/views/controls/textfield/textfield.cc ui/views/controls/textfield/textfield.cc
index 9510441335854bd14307fb51a2e48fd37bb13259..73f4d1850981b351c64ffa50b9288e3f28fdf402 100644
--- ui/views/controls/textfield/textfield.cc
+++ ui/views/controls/textfield/textfield.cc
@@ -3142,6 +3142,10 @@ void Textfield::OnEnabledChanged() {
   if (GetInputMethod()) {
     GetInputMethod()->OnTextInputTypeChanged(this);
   }
+  if (GetWidget()) {
+    SetColor(GetTextColor());
+    UpdateBackgroundColor();
+  }
   UpdateDefaultBorder();
 
   // Only expose readonly if enabled. Don't overwrite the disabled restriction.
