diff --git components/viz/service/frame_sinks/video_capture/frame_sink_video_capturer_impl.cc components/viz/service/frame_sinks/video_capture/frame_sink_video_capturer_impl.cc
index 371498e19a268deacf1afa6fa6b4dc3497b13ec7..a4576f9363b920d68db330ff13c6fbee2493a87e 100644
--- components/viz/service/frame_sinks/video_capture/frame_sink_video_capturer_impl.cc
+++ components/viz/service/frame_sinks/video_capture/frame_sink_video_capturer_impl.cc
@@ -1139,12 +1139,15 @@ void FrameSinkVideoCapturerImpl::MaybeCaptureFrame(
   // TODO(crbug.com/346799708): The condition to check `pixel_format_` shouldn't
   // be necessary but video capture is started with I420+mappable SI in tests.
   // That still captures software I420 frames and not textures.
-  const bool capture_texture_results =
+  const bool prefer_shared_image =
+      buffer_format_preference_ ==
+          mojom::BufferFormatPreference::kPreferMappableSharedImage ||
       buffer_format_preference_ ==
-          mojom::BufferFormatPreference::kPreferMappableSharedImage &&
-      (pixel_format_ == media::PIXEL_FORMAT_NV12 ||
-       pixel_format_ == media::PIXEL_FORMAT_ARGB ||
-       pixel_format_ == media::PIXEL_FORMAT_RGBAF16);
+          mojom::BufferFormatPreference::kPreferSharedImageWithNativeHandle;
+  const bool capture_texture_results =
+      prefer_shared_image && (pixel_format_ == media::PIXEL_FORMAT_NV12 ||
+                              pixel_format_ == media::PIXEL_FORMAT_ARGB ||
+                              pixel_format_ == media::PIXEL_FORMAT_RGBAF16);
 
   std::optional<BlitRequest> blit_request;
   if (capture_texture_results) {
@@ -1154,10 +1157,12 @@ void FrameSinkVideoCapturerImpl::MaybeCaptureFrame(
 
     // TODO(crbug.com/41350322): change the capturer to only request the
     // parts of the frame that have changed whenever possible.
-    blit_request =
-        BlitRequest(content_rect.origin(), LetterboxingBehavior::kLetterbox,
-                    frame_capture.frame->shared_image(), sync_token,
-                    /*populates_mappable_shared_image=*/true);
+    blit_request = BlitRequest(
+        content_rect.origin(), LetterboxingBehavior::kLetterbox,
+        frame_capture.frame->shared_image(), sync_token,
+        /*populates_mappable_shared_image=*/
+        buffer_format_preference_ ==
+            mojom::BufferFormatPreference::kPreferMappableSharedImage);
 
     // We haven't captured the frame yet, but let's pretend that we did for
     // the sake of blend information computation. We will be asking for an
@@ -1367,8 +1372,10 @@ void FrameSinkVideoCapturerImpl::DidCopyFrame(
         frame_capture.CaptureFailed(CaptureResult::kARGBReadbackFailed);
       }
     } else {
-      CHECK_EQ(buffer_format_preference_,
-               mojom::BufferFormatPreference::kPreferMappableSharedImage);
+      CHECK(buffer_format_preference_ ==
+                mojom::BufferFormatPreference::kPreferMappableSharedImage ||
+            buffer_format_preference_ == mojom::BufferFormatPreference::
+                                            kPreferSharedImageWithNativeHandle);
       // MappableSI ARGB results are written to the existing pool texture.
       if (result->IsEmpty()) {
         frame_capture.CaptureFailed(
