diff --git content/renderer/loader/web_url_loader_impl.h content/renderer/loader/web_url_loader_impl.h
index d73a7818cf67..8acffdd5f8e0 100644
--- content/renderer/loader/web_url_loader_impl.h
+++ content/renderer/loader/web_url_loader_impl.h
@@ -99,6 +99,7 @@ class CONTENT_EXPORT WebURLLoaderImpl : public blink::WebURLLoader {
       std::unique_ptr<blink::ResourceLoadInfoNotifierWrapper>
           resource_load_info_notifier_wrapper,
       blink::WebURLLoaderClient* client) override;
+  void Cancel() override;
   void SetDefersLoading(DeferType value) override;
   void DidChangePriority(blink::WebURLRequest::Priority new_priority,
                          int intra_priority_value) override;
@@ -109,8 +110,6 @@ class CONTENT_EXPORT WebURLLoaderImpl : public blink::WebURLLoader {
   class Context;
   class RequestPeerImpl;
 
-  void Cancel();
-
   scoped_refptr<Context> context_;
 
   DISALLOW_COPY_AND_ASSIGN(WebURLLoaderImpl);
diff --git third_party/blink/public/platform/web_url_loader.h third_party/blink/public/platform/web_url_loader.h
index 2db7b64b7948..a9c468d24c21 100644
--- third_party/blink/public/platform/web_url_loader.h
+++ third_party/blink/public/platform/web_url_loader.h
@@ -93,6 +93,10 @@ class WebURLLoader {
       bool no_mime_sniffing,
       std::unique_ptr<ResourceLoadInfoNotifierWrapper>,
       WebURLLoaderClient*) = 0;
+ 
+  // Cancels an asynchronous load.  This will appear as a load error to
+  // the client.
+  virtual void Cancel() {}
 
   // |kDeferred| is when an asynchronous load is suspended.
   // |kDeferredWithBackForwardCache| is when an asynchronous load is suspended
