diff --git content/renderer/loader/web_url_loader_impl.h content/renderer/loader/web_url_loader_impl.h
index 71ddc4cd98aa..714d1612e4eb 100644
--- content/renderer/loader/web_url_loader_impl.h
+++ content/renderer/loader/web_url_loader_impl.h
@@ -99,6 +99,7 @@ class CONTENT_EXPORT WebURLLoaderImpl : public blink::WebURLLoader {
       std::unique_ptr<blink::ResourceLoadInfoNotifierWrapper>
           resource_load_info_notifier_wrapper,
       blink::WebURLLoaderClient* client) override;
+  void Cancel() override;
   void SetDefersLoading(bool value) override;
   void DidChangePriority(blink::WebURLRequest::Priority new_priority,
                          int intra_priority_value) override;
@@ -109,8 +110,6 @@ class CONTENT_EXPORT WebURLLoaderImpl : public blink::WebURLLoader {
   class Context;
   class RequestPeerImpl;
 
-  void Cancel();
-
   scoped_refptr<Context> context_;
 
   DISALLOW_COPY_AND_ASSIGN(WebURLLoaderImpl);
diff --git third_party/blink/public/platform/web_url_loader.h third_party/blink/public/platform/web_url_loader.h
index 18f605fdf283..760e690ac3d2 100644
--- third_party/blink/public/platform/web_url_loader.h
+++ third_party/blink/public/platform/web_url_loader.h
@@ -94,6 +94,10 @@ class WebURLLoader {
       std::unique_ptr<ResourceLoadInfoNotifierWrapper>,
       WebURLLoaderClient*) = 0;
 
+  // Cancels an asynchronous load.  This will appear as a load error to
+  // the client.
+  virtual void Cancel() {}
+
   // Suspends/resumes an asynchronous load.
   virtual void SetDefersLoading(bool) = 0;
 
