diff --git third_party/blink/public/platform/web_url_loader.h third_party/blink/public/platform/web_url_loader.h
index 48b95a412fe0..569e8c750e0a 100644
--- third_party/blink/public/platform/web_url_loader.h
+++ third_party/blink/public/platform/web_url_loader.h
@@ -161,12 +161,14 @@ class BLINK_PLATFORM_EXPORT WebURLLoader {
   void SetResourceRequestSenderForTesting(
       std::unique_ptr<WebResourceRequestSender> resource_request_sender);
 
+  // Cancels an asynchronous load.  This will appear as a load error to
+  // the client.
+  void Cancel();
+
  private:
   class Context;
   class RequestPeerImpl;
 
-  void Cancel();
-
   scoped_refptr<Context> context_;
 
   DISALLOW_COPY_AND_ASSIGN(WebURLLoader);
