Index: V8DOMWindowCustom.cpp
===================================================================
--- V8DOMWindowCustom.cpp	(revision 102901)
+++ V8DOMWindowCustom.cpp	(working copy)
@@ -297,8 +297,11 @@
 static v8::Handle<v8::Value> handlePostMessageCallback(const v8::Arguments& args, bool extendedTransfer)
 {
     DOMWindow* window = V8DOMWindow::toNative(args.Holder());
+    Frame* frame = V8Proxy::retrieveFrameForCallingContext();
+    if (!frame)
+        return v8::Undefined();
 
-    DOMWindow* source = V8Proxy::retrieveFrameForCallingContext()->domWindow();
+    DOMWindow* source = frame->domWindow();
     ASSERT(source->frame());
 
     // This function has variable arguments and can be:
