Index: public/WebView.h
===================================================================
--- public/WebView.h	(revision 151551)
+++ public/WebView.h	(working copy)
@@ -417,6 +417,7 @@
 
     // Sets whether select popup menus should be rendered by the browser.
     WEBKIT_EXPORT static void setUseExternalPopupMenus(bool);
+    virtual void setUseExternalPopupMenusThisInstance(bool) = 0;
 
 
     // Visited link state --------------------------------------------------
Index: src/ChromeClientImpl.cpp
===================================================================
--- src/ChromeClientImpl.cpp	(revision 151551)
+++ src/ChromeClientImpl.cpp	(working copy)
@@ -894,7 +894,7 @@
 
 PassRefPtr<PopupMenu> ChromeClientImpl::createPopupMenu(Frame& frame, PopupMenuClient* client) const
 {
-    if (WebViewImpl::useExternalPopupMenus())
+    if (m_webView->useExternalPopupMenus())
         return adoptRef(new ExternalPopupMenu(frame, client, m_webView->client()));
 
     return adoptRef(new PopupMenuChromium(frame, client));
Index: src/WebViewImpl.cpp
===================================================================
--- src/WebViewImpl.cpp	(revision 151551)
+++ src/WebViewImpl.cpp	(working copy)
@@ -385,6 +385,7 @@
     , m_fakeDoubleTapPageScaleFactor(0)
     , m_fakeDoubleTapUseAnchor(false)
     , m_contextMenuAllowed(false)
+    , m_shouldUseExternalPopupMenus(shouldUseExternalPopupMenus)
     , m_doingDragAndDrop(false)
     , m_ignoreInputEvents(false)
     , m_suppressNextKeypressEvent(false)
@@ -3634,9 +3635,14 @@
     updateLayerTreeViewport();
 }
 
+void WebViewImpl::setUseExternalPopupMenusThisInstance(bool useExternalPopupMenus)
+{
+    m_shouldUseExternalPopupMenus = useExternalPopupMenus;
+}
+
 bool WebViewImpl::useExternalPopupMenus()
 {
-    return shouldUseExternalPopupMenus;
+    return m_shouldUseExternalPopupMenus;
 }
 
 void WebViewImpl::setEmulatedTextZoomFactor(float textZoomFactor)
Index: src/WebViewImpl.h
===================================================================
--- src/WebViewImpl.h	(revision 151551)
+++ src/WebViewImpl.h	(working copy)
@@ -417,7 +417,8 @@
 
     // Returns true if popup menus should be rendered by the browser, false if
     // they should be rendered by WebKit (which is the default).
-    static bool useExternalPopupMenus();
+    void setUseExternalPopupMenusThisInstance(bool);
+    bool useExternalPopupMenus();
 
     bool contextMenuAllowed() const
     {
@@ -730,6 +731,8 @@
 
     bool m_contextMenuAllowed;
 
+    bool m_shouldUseExternalPopupMenus;
+
     bool m_doingDragAndDrop;
 
     bool m_ignoreInputEvents;
