Index: public/web/WebView.h
===================================================================
--- public/web/WebView.h	(revision 160715)
+++ public/web/WebView.h	(working copy)
@@ -441,6 +441,7 @@
 
     // Sets whether select popup menus should be rendered by the browser.
     BLINK_EXPORT static void setUseExternalPopupMenus(bool);
+    virtual void setUseExternalPopupMenusThisInstance(bool) = 0;
 
 
     // Visited link state --------------------------------------------------
Index: Source/web/ChromeClientImpl.cpp
===================================================================
--- Source/web/ChromeClientImpl.cpp	(revision 160715)
+++ Source/web/ChromeClientImpl.cpp	(working copy)
@@ -871,7 +871,7 @@
 
 PassRefPtr<PopupMenu> ChromeClientImpl::createPopupMenu(Frame& frame, PopupMenuClient* client) const
 {
-    if (WebViewImpl::useExternalPopupMenus())
+    if (m_webView->useExternalPopupMenus())
         return adoptRef(new ExternalPopupMenu(frame, client, m_webView->client()));
 
     return adoptRef(new PopupMenuChromium(frame, client));
Index: Source/web/WebViewImpl.cpp
===================================================================
--- Source/web/WebViewImpl.cpp	(revision 160715)
+++ Source/web/WebViewImpl.cpp	(working copy)
@@ -402,6 +402,7 @@
     , m_fakePageScaleAnimationPageScaleFactor(0)
     , m_fakePageScaleAnimationUseAnchor(false)
     , m_contextMenuAllowed(false)
+    , m_shouldUseExternalPopupMenus(shouldUseExternalPopupMenus)
     , m_doingDragAndDrop(false)
     , m_ignoreInputEvents(false)
     , m_compositorDeviceScaleFactorOverride(0)
@@ -3643,9 +3644,14 @@
     updateLayerTreeViewport();
 }
 
+void WebViewImpl::setUseExternalPopupMenusThisInstance(bool useExternalPopupMenus)
+{
+    m_shouldUseExternalPopupMenus = useExternalPopupMenus;
+}
+
 bool WebViewImpl::useExternalPopupMenus()
 {
-    return shouldUseExternalPopupMenus;
+    return m_shouldUseExternalPopupMenus;
 }
 
 void WebViewImpl::startDragging(Frame* frame,
Index: Source/web/WebViewImpl.h
===================================================================
--- Source/web/WebViewImpl.h	(revision 160715)
+++ Source/web/WebViewImpl.h	(working copy)
@@ -409,7 +409,8 @@
 
     // Returns true if popup menus should be rendered by the browser, false if
     // they should be rendered by WebKit (which is the default).
-    static bool useExternalPopupMenus();
+    void setUseExternalPopupMenusThisInstance(bool);
+    bool useExternalPopupMenus();
 
     bool contextMenuAllowed() const
     {
@@ -702,6 +703,8 @@
 
     bool m_contextMenuAllowed;
 
+    bool m_shouldUseExternalPopupMenus;
+
     bool m_doingDragAndDrop;
 
     bool m_ignoreInputEvents;
