Index: public/WebFrame.h
===================================================================
--- public/WebFrame.h	(revision 23266)
+++ public/WebFrame.h	(working copy)
@@ -332,6 +332,11 @@
         // given page size.
         virtual int printBegin(const WebSize& pageSize) = 0;
 
+        // Returns the page shrinking factor calculated by webkit (usually
+        // between 1/1.25 and 1/2). Returns 0 if the page number is invalid or
+        // not in printing mode.
+        virtual float getPrintPageShrink(int page) = 0;
+
         // Prints one page, and returns the calculated page shrinking factor
         // (usually between 1/1.25 and 1/2).  Returns 0 if the page number is
         // invalid or not in printing mode.
