Index: features.gypi
===================================================================
--- features.gypi	(revision 143980)
+++ features.gypi	(working copy)
@@ -190,7 +190,7 @@
           'ENABLE_CALENDAR_PICKER=1',
           'ENABLE_DATALIST_ELEMENT=1',
           'ENABLE_INPUT_SPEECH=1',
-          'ENABLE_INPUT_TYPE_COLOR=1',
+          'ENABLE_INPUT_TYPE_COLOR=0',
           'ENABLE_INPUT_TYPE_WEEK=1',
           'ENABLE_INPUT_MULTIPLE_FIELDS_UI=1',
           'ENABLE_LEGACY_NOTIFICATIONS=1',
@@ -199,7 +199,7 @@
           'ENABLE_NOTIFICATIONS=1',
           'ENABLE_ORIENTATION_EVENTS=0',
           'ENABLE_PAGE_POPUP=1',
-          'ENABLE_PRINTING=1',
+          'ENABLE_PRINTING=0',
           'ENABLE_SHARED_WORKERS=1',
           'ENABLE_WEB_AUDIO=1',
         ],
