diff --git third_party/WebKit/Source/web/ChromeClientImpl.cpp third_party/WebKit/Source/web/ChromeClientImpl.cpp
index ba81ee3..648d7ce 100644
--- third_party/WebKit/Source/web/ChromeClientImpl.cpp
+++ third_party/WebKit/Source/web/ChromeClientImpl.cpp
@@ -908,7 +908,7 @@ bool ChromeClientImpl::HasOpenedPopup() const {
 PopupMenu* ChromeClientImpl::OpenPopupMenu(LocalFrame& frame,
                                            HTMLSelectElement& select) {
   NotifyPopupOpeningObservers();
-  if (WebViewImpl::UseExternalPopupMenus())
+  if (web_view_->UseExternalPopupMenus())
     return new ExternalPopupMenu(frame, select, *web_view_);
 
   DCHECK(RuntimeEnabledFeatures::pagePopupEnabled());
diff --git third_party/WebKit/Source/web/WebViewImpl.cpp third_party/WebKit/Source/web/WebViewImpl.cpp
index 16ac66b..7053b54 100644
--- third_party/WebKit/Source/web/WebViewImpl.cpp
+++ third_party/WebKit/Source/web/WebViewImpl.cpp
@@ -355,6 +355,7 @@ WebViewImpl::WebViewImpl(WebViewClient* client,
       enable_fake_page_scale_animation_for_testing_(false),
       fake_page_scale_animation_page_scale_factor_(0),
       fake_page_scale_animation_use_anchor_(false),
+      should_use_external_popup_menus_(g_should_use_external_popup_menus),
       compositor_device_scale_factor_override_(0),
       suppress_next_keypress_event_(false),
       ime_accept_events_(true),
@@ -3743,8 +3744,13 @@ void WebViewImpl::MainFrameScrollOffsetChanged() {
   dev_tools_emulator_->MainFrameScrollOrScaleChanged();
 }
 
+void WebViewImpl::SetUseExternalPopupMenusThisInstance(
+    bool useExternalPopupMenus) {
+  should_use_external_popup_menus_ = useExternalPopupMenus;
+}
+
 bool WebViewImpl::UseExternalPopupMenus() {
-  return g_should_use_external_popup_menus;
+  return should_use_external_popup_menus_;
 }
 
 void WebViewImpl::SetBackgroundColorOverride(WebColor color) {
diff --git third_party/WebKit/Source/web/WebViewImpl.h third_party/WebKit/Source/web/WebViewImpl.h
index a22ef37..33dbab4 100644
--- third_party/WebKit/Source/web/WebViewImpl.h
+++ third_party/WebKit/Source/web/WebViewImpl.h
@@ -362,7 +362,8 @@ class WEB_EXPORT WebViewImpl final
 
   // Returns true if popup menus should be rendered by the browser, false if
   // they should be rendered by WebKit (which is the default).
-  static bool UseExternalPopupMenus();
+  void SetUseExternalPopupMenusThisInstance(bool);
+  bool UseExternalPopupMenus();
 
   bool ShouldAutoResize() const { return should_auto_resize_; }
 
@@ -647,6 +648,8 @@ class WEB_EXPORT WebViewImpl final
   float fake_page_scale_animation_page_scale_factor_;
   bool fake_page_scale_animation_use_anchor_;
 
+  bool should_use_external_popup_menus_;
+
   float compositor_device_scale_factor_override_;
   TransformationMatrix device_emulation_transform_;
 
diff --git third_party/WebKit/public/web/WebView.h third_party/WebKit/public/web/WebView.h
index e811392..d6d5bf8 100644
--- third_party/WebKit/public/web/WebView.h
+++ third_party/WebKit/public/web/WebView.h
@@ -403,6 +403,7 @@ class WebView : protected WebWidget {
 
   // Sets whether select popup menus should be rendered by the browser.
   BLINK_EXPORT static void SetUseExternalPopupMenus(bool);
+  virtual void SetUseExternalPopupMenusThisInstance(bool) = 0;
 
   // Hides any popup (suggestions, selects...) that might be showing.
   virtual void HidePopups() = 0;
