diff --git third_party/blink/renderer/build/scripts/templates/runtime_enabled_features.h.tmpl third_party/blink/renderer/build/scripts/templates/runtime_enabled_features.h.tmpl
index e065ef718e980..5c94887561a95 100644
--- third_party/blink/renderer/build/scripts/templates/runtime_enabled_features.h.tmpl
+++ third_party/blink/renderer/build/scripts/templates/runtime_enabled_features.h.tmpl
@@ -13,6 +13,8 @@
 #include "third_party/blink/renderer/platform/platform_export.h"
 #include "third_party/blink/renderer/platform/wtf/allocator/allocator.h"
 
+class AlloyContentRendererClient;
+
 #define ASSERT_ORIGIN_TRIAL(feature) \
   static_assert(std::is_same<decltype(::blink::RuntimeEnabledFeatures::     \
                                           feature##EnabledByRuntimeFlag()), \
@@ -117,6 +119,7 @@ class PLATFORM_EXPORT RuntimeEnabledFeatures : public RuntimeEnabledFeaturesBase
   // protected section of RuntimeEnabledFeaturesBase. Normally, unit tests
   // should use the ScopedFeatureNameForTest classes defined in
   // platform/testing/runtime_enabled_features_test_helpers.h.
+  friend class ::AlloyContentRendererClient;
   friend class DevToolsEmulator;
   friend class InternalRuntimeFlags;
   friend class V8ContextSnapshotImpl;
