Index: public/web/WebView.h
===================================================================
--- public/web/WebView.h	(revision 152953)
+++ public/web/WebView.h	(working copy)
@@ -421,6 +421,7 @@
 
     // Sets whether select popup menus should be rendered by the browser.
     WEBKIT_EXPORT static void setUseExternalPopupMenus(bool);
+    virtual void setUseExternalPopupMenusThisInstance(bool) = 0;
 
 
     // Visited link state --------------------------------------------------
Index: Source/WebKit/chromium/src/ChromeClientImpl.cpp
===================================================================
--- Source/WebKit/chromium/src/ChromeClientImpl.cpp	(revision 152953)
+++ Source/WebKit/chromium/src/ChromeClientImpl.cpp	(working copy)
@@ -895,7 +895,7 @@
 
 PassRefPtr<PopupMenu> ChromeClientImpl::createPopupMenu(Frame& frame, PopupMenuClient* client) const
 {
-    if (WebViewImpl::useExternalPopupMenus())
+    if (m_webView->useExternalPopupMenus())
         return adoptRef(new ExternalPopupMenu(frame, client, m_webView->client()));
 
     return adoptRef(new PopupMenuChromium(frame, client));
Index: Source/WebKit/chromium/src/WebViewImpl.cpp
===================================================================
--- Source/WebKit/chromium/src/WebViewImpl.cpp	(revision 152953)
+++ Source/WebKit/chromium/src/WebViewImpl.cpp	(working copy)
@@ -393,6 +393,7 @@
     , m_fakeDoubleTapPageScaleFactor(0)
     , m_fakeDoubleTapUseAnchor(false)
     , m_contextMenuAllowed(false)
+    , m_shouldUseExternalPopupMenus(shouldUseExternalPopupMenus)
     , m_doingDragAndDrop(false)
     , m_ignoreInputEvents(false)
     , m_suppressNextKeypressEvent(false)
@@ -3676,9 +3677,14 @@
     updateLayerTreeViewport();
 }
 
+void WebViewImpl::setUseExternalPopupMenusThisInstance(bool useExternalPopupMenus)
+{
+    m_shouldUseExternalPopupMenus = useExternalPopupMenus;
+}
+
 bool WebViewImpl::useExternalPopupMenus()
 {
-    return shouldUseExternalPopupMenus;
+    return m_shouldUseExternalPopupMenus;
 }
 
 void WebViewImpl::setEmulatedTextZoomFactor(float textZoomFactor)
Index: Source/WebKit/chromium/src/WebViewImpl.h
===================================================================
--- Source/WebKit/chromium/src/WebViewImpl.h	(revision 152953)
+++ Source/WebKit/chromium/src/WebViewImpl.h	(working copy)
@@ -415,7 +415,8 @@
 
     // Returns true if popup menus should be rendered by the browser, false if
     // they should be rendered by WebKit (which is the default).
-    static bool useExternalPopupMenus();
+    void setUseExternalPopupMenusThisInstance(bool);
+    bool useExternalPopupMenus();
 
     bool contextMenuAllowed() const
     {
@@ -733,6 +734,8 @@
 
     bool m_contextMenuAllowed;
 
+    bool m_shouldUseExternalPopupMenus;
+
     bool m_doingDragAndDrop;
 
     bool m_ignoreInputEvents;
