diff --git third_party/blink/renderer/platform/widget/widget_base.cc third_party/blink/renderer/platform/widget/widget_base.cc
index 048460f4dd81..59a70580ded0 100644
--- third_party/blink/renderer/platform/widget/widget_base.cc
+++ third_party/blink/renderer/platform/widget/widget_base.cc
@@ -1281,8 +1281,8 @@ gfx::Point WidgetBase::BlinkSpaceToFlooredDIPs(const gfx::Point& point) {
     return point;
   // TODO(danakj): Should this be GetScreenInfo() so it changes under emulation?
   // TODO(dtapuska): Determine if this should be a floor vs rounded.
-  return gfx::ScaleToFlooredPoint(
-      point, client_->GetOriginalScreenInfo().device_scale_factor);
+  float reverse = 1 / client_->GetOriginalScreenInfo().device_scale_factor;
+  return gfx::ScaleToFlooredPoint(point, reverse);
 }
 
 gfx::Size WidgetBase::DIPsToCeiledBlinkSpace(const gfx::Size& size) {
