diff --git third_party/blink/renderer/core/input/pointer_event_manager.cc third_party/blink/renderer/core/input/pointer_event_manager.cc
index e48a89c9b8d8..512cf3aff54b 100644
--- third_party/blink/renderer/core/input/pointer_event_manager.cc
+++ third_party/blink/renderer/core/input/pointer_event_manager.cc
@@ -284,7 +284,7 @@ void PointerEventManager::HandlePointerInterruption(
   for (auto pointer_event : canceled_pointer_events) {
     // If we are sending a pointercancel we have sent the pointerevent to some
     // target before.
-    DCHECK(node_under_pointer_.Contains(pointer_event->pointerId()));
+    // DCHECK(node_under_pointer_.Contains(pointer_event->pointerId()));
     EventTarget* target =
         node_under_pointer_.at(pointer_event->pointerId()).target;
 
