Index: webframe_impl.cc
===================================================================
--- webframe_impl.cc	(revision 23266)
+++ webframe_impl.cc	(working copy)
@@ -1045,6 +1045,16 @@
   return print_context_->pageCount();
 }
 
+float WebFrameImpl::getPrintPageShrink(int page) {
+  // Ensure correct state.
+  if (!print_context_.get() || page < 0) {
+    NOTREACHED();
+    return 0;
+  }
+
+  return print_context_->getPageShrink(page);
+}
+
 float WebFrameImpl::printPage(int page, WebCanvas* canvas) {
   // Ensure correct state.
   if (!print_context_.get() || page < 0 || !frame() || !frame()->document()) {
Index: webframe_impl.h
===================================================================
--- webframe_impl.h	(revision 23266)
+++ webframe_impl.h	(working copy)
@@ -147,6 +147,7 @@
   virtual WebKit::WebString selectionAsMarkup() const;
   virtual int printBegin(const WebKit::WebSize& page_size);
   virtual float printPage(int page_to_print, WebKit::WebCanvas* canvas);
+  virtual float getPrintPageShrink(int page);
   virtual void printEnd();
   virtual bool find(
       int identifier, const WebKit::WebString& search_text,
