diff --git chrome/browser/picture_in_picture/picture_in_picture_window_manager.cc chrome/browser/picture_in_picture/picture_in_picture_window_manager.cc
index e8a525444092a20624616b280b99f9b218180410..7874d8d2e58d9564841ce2b661abed713bf8afba 100644
--- chrome/browser/picture_in_picture/picture_in_picture_window_manager.cc
+++ chrome/browser/picture_in_picture/picture_in_picture_window_manager.cc
@@ -443,7 +443,8 @@ gfx::Rect PictureInPictureWindowManager::CalculateOuterWindowBounds(
     // Ignore the result if we're asked to do so.  Note that we still have to
     // ask the cache, so that it's set up to accept position updates later for
     // this request.
-    if (cached_window_bounds && !pip_options.prefer_initial_window_placement) {
+    if (cached_window_bounds && !pip_options.prefer_initial_window_placement &&
+        !pip_options.initial_position.has_value()) {
       // Cache hit!  Just return it as the window bounds.
       return *cached_window_bounds;
     }
@@ -497,6 +498,11 @@ gfx::Rect PictureInPictureWindowManager::CalculateOuterWindowBounds(
   }
 #endif
   // Position the window.
+  if (pip_options.initial_position.has_value()) {
+    window_bounds.set_origin(pip_options.initial_position.value());
+    return window_bounds;
+  }
+
   int window_diff_width = work_area.right() - window_bounds.width();
   int window_diff_height = work_area.bottom() - window_bounds.height();
 
diff --git content/renderer/render_frame_impl.cc content/renderer/render_frame_impl.cc
index cd9a5bb5b14a00b555f86bee503fc3f8d51db80b..267e8a5747e19c8479ca53cbb2ffeb13e93e438d 100644
--- content/renderer/render_frame_impl.cc
+++ content/renderer/render_frame_impl.cc
@@ -6831,6 +6831,8 @@ WebView* RenderFrameImpl::CreateNewWindow(
         pip_options->disallow_return_to_opener;
     pip_mojom_opts->prefer_initial_window_placement =
         pip_options->prefer_initial_window_placement;
+    pip_mojom_opts->initial_position = pip_options->initial_position;
+    pip_mojom_opts->resizable = pip_options->resizable;
     params->pip_options = std::move(pip_mojom_opts);
   }
 
diff --git third_party/blink/public/mojom/picture_in_picture_window_options/picture_in_picture_window_options.mojom third_party/blink/public/mojom/picture_in_picture_window_options/picture_in_picture_window_options.mojom
index 0a10dd95413ad7fc333eec91c866d73339dd8c11..72495326c304f46d91de3e7f911a2e2a2467e8e1 100644
--- third_party/blink/public/mojom/picture_in_picture_window_options/picture_in_picture_window_options.mojom
+++ third_party/blink/public/mojom/picture_in_picture_window_options/picture_in_picture_window_options.mojom
@@ -4,6 +4,8 @@
 
 module blink.mojom;
 
+import "ui/gfx/geometry/mojom/geometry.mojom";
+
 struct PictureInPictureWindowOptions {
   // The requested width for the picture-in-picture window to be initially sized
   // to.
@@ -21,4 +23,8 @@ struct PictureInPictureWindowOptions {
   // When true, don't try to re-use the previous window bounds from the
   // bounds cache.  Instead, just use the default position and size.
   bool prefer_initial_window_placement;
+
+  // CEF extension
+  gfx.mojom.Point? initial_position;
+  bool resizable = true;
 };
diff --git third_party/blink/public/web/web_picture_in_picture_window_options.h third_party/blink/public/web/web_picture_in_picture_window_options.h
index f3a5b90b3bb119558da1d9c56a73675c9f32cefd..53c5c690bcf2b6af49d44945e93df58bf3b9f5b4 100644
--- third_party/blink/public/web/web_picture_in_picture_window_options.h
+++ third_party/blink/public/web/web_picture_in_picture_window_options.h
@@ -6,6 +6,9 @@
 #define THIRD_PARTY_BLINK_PUBLIC_WEB_WEB_PICTURE_IN_PICTURE_WINDOW_OPTIONS_H_
 
 #include <cstdint>
+#include <optional>
+
+#include "ui/gfx/geometry/point.h"
 
 namespace blink {
 
@@ -14,6 +17,9 @@ struct WebPictureInPictureWindowOptions {
   uint64_t height = 0;
   bool disallow_return_to_opener = false;
   bool prefer_initial_window_placement = false;
+  // CEF extension
+  std::optional<gfx::Point> initial_position;
+  bool resizable = true;
 };
 
 }  // namespace blink
diff --git third_party/blink/public/web/web_view.h third_party/blink/public/web/web_view.h
index 932658273154ef2e022358e493a8e7c00c86e732..6a1b5bbe94772a947ac8055e4e0971f7b3e166e1 100644
--- third_party/blink/public/web/web_view.h
+++ third_party/blink/public/web/web_view.h
@@ -465,6 +465,16 @@ class BLINK_EXPORT WebView {
   virtual void SetPageAttributionSupport(
       network::mojom::AttributionSupport support) = 0;
 
+  // Sets whether to allow the use of JavaScript moveTo/By() and resizeTo/By()
+  // (without user activation) with Document picture-in-picture popups.
+  virtual void SetMovePictureInPictureEnabled(bool enabled) = 0;
+  virtual bool MovePictureInPictureEnabled() const = 0;
+
+  // Sets whether to allow opening Document picture-in-picture windows without
+  // user activation. Defaults to false (user activation required).
+  virtual void SetAllowPictureInPictureWithoutUserActivation(bool allow) = 0;
+  virtual bool AllowPictureInPictureWithoutUserActivation() const = 0;
+
  protected:
   ~WebView() = default;
 };
diff --git third_party/blink/renderer/bindings/generated_in_modules.gni third_party/blink/renderer/bindings/generated_in_modules.gni
index ca5996cfa12f9cb6792a50af5e28d50a1c91d312..3aa1ae6a8388c1334101ccc0d597ecbd3feeae44 100644
--- third_party/blink/renderer/bindings/generated_in_modules.gni
+++ third_party/blink/renderer/bindings/generated_in_modules.gni
@@ -392,6 +392,10 @@ generated_dictionary_sources_in_modules = [
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_directory_picker_options.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_display_media_stream_options.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_display_media_stream_options.h",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_cef_options.cc",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_cef_options.h",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_position.cc",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_position.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_event_init.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_event_init.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_options.cc",
diff --git third_party/blink/renderer/core/exported/web_view_impl.h third_party/blink/renderer/core/exported/web_view_impl.h
index db7b6217c87a02023635ab2ed078ddc5f7688c7d..382c5eef1335c8b4438d5eb1d54faf9891ff03d7 100644
--- third_party/blink/renderer/core/exported/web_view_impl.h
+++ third_party/blink/renderer/core/exported/web_view_impl.h
@@ -321,6 +321,20 @@ class CORE_EXPORT WebViewImpl final : public WebView,
       const ColorProviderColorMaps& color_provider_colors) override;
   void SetSupportsDraggableRegions(bool supports_draggable_regions) override;
 
+  void SetMovePictureInPictureEnabled(bool enabled) override {
+    move_pip_enabled_ = enabled;
+  }
+  bool MovePictureInPictureEnabled() const override {
+    return move_pip_enabled_;
+  }
+
+  void SetAllowPictureInPictureWithoutUserActivation(bool allow) override {
+    allow_pip_without_user_activation_ = allow;
+  }
+  bool AllowPictureInPictureWithoutUserActivation() const override {
+    return allow_pip_without_user_activation_;
+  }
+
   void DispatchPersistedPageshow(base::TimeTicks navigation_start);
   void DispatchPagehide(mojom::blink::PagehideDispatch pagehide_dispatch);
   void HookBackForwardCacheEviction(bool hook);
@@ -1027,6 +1041,9 @@ class CORE_EXPORT WebViewImpl final : public WebView,
   // CSS property.
   bool supports_draggable_regions_ = false;
 
+  bool move_pip_enabled_ = false;
+  bool allow_pip_without_user_activation_ = false;
+
   // All the registered observers.
   base::ObserverList<WebViewObserver> observers_;
 #if !BUILDFLAG(IS_ANDROID) && !BUILDFLAG(IS_IOS)
diff --git third_party/blink/renderer/core/frame/local_dom_window.cc third_party/blink/renderer/core/frame/local_dom_window.cc
index c425735d53a8a79c9ce20c049506659c17b51be3..ba115ee7c5e4634dc059b54be1d4b4f0896c6c2c 100644
--- third_party/blink/renderer/core/frame/local_dom_window.cc
+++ third_party/blink/renderer/core/frame/local_dom_window.cc
@@ -51,6 +51,7 @@
 #include "third_party/blink/public/platform/task_type.h"
 #include "third_party/blink/public/platform/web_string.h"
 #include "third_party/blink/public/web/web_picture_in_picture_window_options.h"
+#include "third_party/blink/public/web/web_view.h"
 #include "third_party/blink/renderer/bindings/core/v8/binding_security.h"
 #include "third_party/blink/renderer/bindings/core/v8/capture_source_location.h"
 #include "third_party/blink/renderer/bindings/core/v8/isolated_world_csp.h"
@@ -110,6 +111,7 @@
 #include "third_party/blink/renderer/core/frame/settings.h"
 #include "third_party/blink/renderer/core/frame/viewport_data.h"
 #include "third_party/blink/renderer/core/frame/visual_viewport.h"
+#include "third_party/blink/renderer/core/frame/web_local_frame_impl.h"
 #include "third_party/blink/renderer/core/html/custom/custom_element_registry.h"
 #include "third_party/blink/renderer/core/html/fenced_frame/fence.h"
 #include "third_party/blink/renderer/core/html/forms/form_controller.h"
@@ -1977,8 +1979,9 @@ void LocalDOMWindow::moveBy(int x, int y) const {
     return;
   }
 
-  if (IsPictureInPictureWindow())
+  if (IsPictureInPictureWindow() && !MovePictureInPictureEnabled()) {
     return;
+  }
 
   LocalFrame* frame = GetFrame();
   Page* page = frame->GetPage();
@@ -1998,8 +2001,9 @@ void LocalDOMWindow::moveTo(int x, int y) const {
     return;
   }
 
-  if (IsPictureInPictureWindow())
+  if (IsPictureInPictureWindow() && !MovePictureInPictureEnabled()) {
     return;
+  }
 
   LocalFrame* frame = GetFrame();
   Page* page = frame->GetPage();
@@ -2022,7 +2026,8 @@ void LocalDOMWindow::resizeBy(int x,
   }
 
   if (IsPictureInPictureWindow()) {
-    if (!LocalFrame::ConsumeTransientUserActivation(GetFrame())) {
+    if (!MovePictureInPictureEnabled() &&
+        !LocalFrame::ConsumeTransientUserActivation(GetFrame())) {
       exception_state.ThrowDOMException(
           DOMExceptionCode::kNotAllowedError,
           "resizeBy() requires user activation in document picture-in-picture");
@@ -2050,7 +2055,8 @@ void LocalDOMWindow::resizeTo(int width,
   }
 
   if (IsPictureInPictureWindow()) {
-    if (!LocalFrame::ConsumeTransientUserActivation(GetFrame())) {
+    if (!MovePictureInPictureEnabled() &&
+        !LocalFrame::ConsumeTransientUserActivation(GetFrame())) {
       exception_state.ThrowDOMException(
           DOMExceptionCode::kNotAllowedError,
           "resizeTo() requires user activation in document picture-in-picture");
@@ -2487,6 +2493,12 @@ DOMWindow* LocalDOMWindow::openPictureInPictureWindow(
       To<LocalDOMWindow>(result.frame->DomWindow());
   pip_dom_window->SetIsPictureInPictureWindow();
 
+  if (WebLocalFrameImpl::FromFrame(entered_window->GetFrame())
+          ->View()
+          ->MovePictureInPictureEnabled()) {
+    pip_dom_window->SetMovePictureInPictureEnabled(true);
+  }
+
   // Ensure that we're using the same compatibility mode as the opener document.
   pip_dom_window->document()->SetCompatibilityMode(
       entered_window->document()->GetCompatibilityMode());
diff --git third_party/blink/renderer/core/frame/local_dom_window.h third_party/blink/renderer/core/frame/local_dom_window.h
index 3d8f63c8beb03ce680b50eb7b6010d736156e29b..902d6429fb66237617e278109cbde911504d42d3 100644
--- third_party/blink/renderer/core/frame/local_dom_window.h
+++ third_party/blink/renderer/core/frame/local_dom_window.h
@@ -604,6 +604,11 @@ class CORE_EXPORT LocalDOMWindow final : public DOMWindow,
 
   void SetIsPictureInPictureWindow();
 
+  void SetMovePictureInPictureEnabled(bool enabled) {
+    move_pip_enabled_ = enabled;
+  }
+  bool MovePictureInPictureEnabled() const { return move_pip_enabled_; }
+
   // Return the viewport size including scrollbars.
   gfx::Size GetViewportSize() const;
 
@@ -710,6 +715,8 @@ class CORE_EXPORT LocalDOMWindow final : public DOMWindow,
   // https://wicg.github.io/document-picture-in-picture/
   bool is_picture_in_picture_window_ = false;
 
+  bool move_pip_enabled_ = false;
+
   // Records this window's Storage Access API status. It cannot be downgraded.
   net::StorageAccessApiStatus storage_access_api_status_ =
       net::StorageAccessApiStatus::kNone;
diff --git third_party/blink/renderer/modules/document_picture_in_picture/document_picture_in_picture_options.idl third_party/blink/renderer/modules/document_picture_in_picture/document_picture_in_picture_options.idl
index de4f0294a2f7ea20517c71cad3d17fb0f1975211..8fa3befaf1d90d20ffe819ed09667511be1b8cf9 100644
--- third_party/blink/renderer/modules/document_picture_in_picture/document_picture_in_picture_options.idl
+++ third_party/blink/renderer/modules/document_picture_in_picture/document_picture_in_picture_options.idl
@@ -4,9 +4,20 @@
 
 // https://wicg.github.io/document-picture-in-picture/#dictdef-documentpictureinpictureoptions
 
+dictionary DocumentPictureInPicturePosition {
+  required long x;
+  required long y;
+};
+
+dictionary DocumentPictureInPictureCefOptions {
+  DocumentPictureInPicturePosition position;
+  boolean resizable = true;
+};
+
 dictionary DocumentPictureInPictureOptions {
   [EnforceRange] unsigned long long width = 0;
   [EnforceRange] unsigned long long height = 0;
   boolean disallowReturnToOpener = false;
   [RuntimeEnabled=DocumentPictureInPicturePreferInitialPlacement] boolean preferInitialWindowPlacement = false;
+  [RuntimeEnabled=DocumentPictureInPictureCefOptions] DocumentPictureInPictureCefOptions cefOptions;
 };
diff --git third_party/blink/renderer/modules/document_picture_in_picture/picture_in_picture_controller_impl.cc third_party/blink/renderer/modules/document_picture_in_picture/picture_in_picture_controller_impl.cc
index 12d793161d71deb8ba4997ff214a3a4291e67ec2..858583165eccea97e28e2167b24ac2ef7acb1d99 100644
--- third_party/blink/renderer/modules/document_picture_in_picture/picture_in_picture_controller_impl.cc
+++ third_party/blink/renderer/modules/document_picture_in_picture/picture_in_picture_controller_impl.cc
@@ -15,16 +15,20 @@
 #include "third_party/blink/public/platform/browser_interface_broker_proxy.h"
 #include "third_party/blink/public/platform/web_media_player.h"
 #include "third_party/blink/public/web/web_picture_in_picture_window_options.h"
+#include "third_party/blink/public/web/web_view.h"
 #include "third_party/blink/renderer/bindings/core/v8/script_controller.h"
 #include "third_party/blink/renderer/bindings/core/v8/script_promise_resolver.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_binding_for_core.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_throw_dom_exception.h"
+#include "third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_cef_options.h"
 #include "third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_options.h"
+#include "third_party/blink/renderer/bindings/modules/v8/v8_document_picture_in_picture_position.h"
 #include "third_party/blink/renderer/core/dom/document.h"
 #include "third_party/blink/renderer/core/dom/dom_exception.h"
 #include "third_party/blink/renderer/core/dom/events/event.h"
 #include "third_party/blink/renderer/core/frame/local_frame.h"
 #include "third_party/blink/renderer/core/frame/settings.h"
+#include "third_party/blink/renderer/core/frame/web_local_frame_impl.h"
 #include "third_party/blink/renderer/core/fullscreen/fullscreen.h"
 #include "third_party/blink/renderer/core/html/media/html_media_element.h"
 #include "third_party/blink/renderer/core/html/media/html_video_element.h"
@@ -37,6 +41,7 @@
 #include "third_party/blink/renderer/platform/weborigin/scheme_registry.h"
 #include "third_party/blink/renderer/platform/widget/frame_widget.h"
 #include "third_party/blink/renderer/platform/wtf/functional.h"
+#include "ui/gfx/geometry/point.h"
 
 namespace blink {
 
@@ -380,7 +385,12 @@ void PictureInPictureControllerImpl::CreateDocumentPictureInPictureWindow(
     LocalDOMWindow& opener,
     DocumentPictureInPictureOptions* options,
     ScriptPromiseResolver<DOMWindow>* resolver) {
-  if (!LocalFrame::ConsumeTransientUserActivation(opener.GetFrame())) {
+  auto* opener_frame_view =
+      WebLocalFrameImpl::FromFrame(opener.GetFrame())->View();
+  bool allow_without_activation =
+      opener_frame_view->AllowPictureInPictureWithoutUserActivation();
+  if (!allow_without_activation &&
+      !LocalFrame::ConsumeTransientUserActivation(opener.GetFrame())) {
     resolver->RejectWithDOMException(DOMExceptionCode::kNotAllowedError,
                                      "Document PiP requires user activation");
     return;
@@ -393,6 +403,17 @@ void PictureInPictureControllerImpl::CreateDocumentPictureInPictureWindow(
   web_options.prefer_initial_window_placement =
       options->preferInitialWindowPlacement();
 
+  if (options->hasCefOptions()) {
+    const auto* cef_options = options->cefOptions();
+    if (cef_options->hasPosition()) {
+      const auto* position = cef_options->position();
+      web_options.initial_position = gfx::Point(position->x(), position->y());
+    }
+    if (cef_options->hasResizable()) {
+      web_options.resizable = cef_options->resizable();
+    }
+  }
+
   // If either width or height is specified, then both must be specified.
   if (web_options.width > 0 && web_options.height == 0) {
     resolver->RejectWithRangeError(
diff --git third_party/blink/renderer/platform/runtime_enabled_features.override.json5 third_party/blink/renderer/platform/runtime_enabled_features.override.json5
index 43cfe5a54118d60898fb4f6bfbcae7ec3a96c53d..2d80c0463e2774ce5a1189348814289e5a837d16 100644
--- third_party/blink/renderer/platform/runtime_enabled_features.override.json5
+++ third_party/blink/renderer/platform/runtime_enabled_features.override.json5
@@ -46,5 +46,13 @@
 
   data: [
     // Entries go here.
+    // Enables cefOptions for Document PiP to specify initial window position
+    {
+      name: "DocumentPictureInPictureCefOptions",
+      status: {
+        "Android": "",
+        "default": "",
+      },
+    },
   ]
 }
\ No newline at end of file
