Index: features.gypi
===================================================================
--- features.gypi	(revision 151551)
+++ features.gypi	(working copy)
@@ -78,7 +78,7 @@
         'feature_defines': [
           'ENABLE_CALENDAR_PICKER=1',
           'ENABLE_INPUT_SPEECH=1',
-          'ENABLE_INPUT_TYPE_COLOR=1',
+          'ENABLE_INPUT_TYPE_COLOR=0',
           'ENABLE_INPUT_MULTIPLE_FIELDS_UI=1',
           'ENABLE_LEGACY_NOTIFICATIONS=1',
           'ENABLE_MEDIA_CAPTURE=0',
