diff --git content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
index 1cc7bf44645f..7ec28c21ec1d 100644
--- content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
+++ content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
@@ -48,7 +48,7 @@ bool ScreenlockMonitorDeviceSource::SessionMessageWindow::OnWndProc(
   if (message == WM_WTSSESSION_CHANGE) {
     ProcessWTSSessionLockMessage(wparam);
   }
-  return true;
+  return false;
 }
 
 void ScreenlockMonitorDeviceSource::SessionMessageWindow::
