diff --git ui/accessibility/platform/ax_platform_node_win.cc ui/accessibility/platform/ax_platform_node_win.cc
index 992b06daefd6a84acb3d8902f3502529650acdad..9c1d900cad9082df23aacb114d22702e873484fa 100644
--- ui/accessibility/platform/ax_platform_node_win.cc
+++ ui/accessibility/platform/ax_platform_node_win.cc
@@ -1683,6 +1683,10 @@ IFACEMETHODIMP AXPlatformNodeWin::accLocation(LONG* physical_pixel_left,
   AXPlatform::GetInstance().SetMsaaActive();
   AXPlatform::GetInstance().OnMinimalPropertiesUsed();
 
+  if (!target || target->IsDestroyed()) {
+    return S_FALSE;
+  }
+
   gfx::Rect bounds = target->GetDelegate()->GetBoundsRect(
       AXCoordinateSystem::kScreenPhysicalPixels,
       AXClippingBehavior::kUnclipped);
