diff --git services/service_manager/sandbox/win/sandbox_win.cc services/service_manager/sandbox/win/sandbox_win.cc
index 6c7356eacc42..4804ce497e1b 100644
--- services/service_manager/sandbox/win/sandbox_win.cc
+++ services/service_manager/sandbox/win/sandbox_win.cc
@@ -865,8 +865,11 @@ sandbox::ResultCode SandboxWin::StartSandboxedProcess(
 #endif
 
   // Post-startup mitigations.
-  mitigations = sandbox::MITIGATION_STRICT_HANDLE_CHECKS |
-                sandbox::MITIGATION_DLL_SEARCH_ORDER;
+  mitigations = sandbox::MITIGATION_DLL_SEARCH_ORDER;
+  if (!launcher_process_command_line.HasSwitch("win-rt-app")) {
+    // Don't enable this mitigation in WinRT apps. See issue #2274.
+    mitigations |= sandbox::MITIGATION_STRICT_HANDLE_CHECKS;
+  }
   if (base::FeatureList::IsEnabled(
           service_manager::features::kWinSboxForceMsSigned) &&
       !cmd_line->HasSwitch(switches::kAllowThirdPartyModules)) {
