diff --git chrome/browser/taskbar/taskbar_decorator_win.cc chrome/browser/taskbar/taskbar_decorator_win.cc
index 90969f7bb07a642d37a348d33d7220ec035d7a3f..a91b4587514165c5925e8d8632651acd0bf93a25 100644
--- chrome/browser/taskbar/taskbar_decorator_win.cc
+++ chrome/browser/taskbar/taskbar_decorator_win.cc
@@ -20,12 +20,14 @@
 #include "base/task/single_thread_task_runner.h"
 #include "base/task/thread_pool.h"
 #include "base/win/scoped_gdi_object.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/profiles/avatar_menu.h"
 #include "chrome/browser/profiles/profile.h"
 #include "chrome/browser/profiles/profile_attributes_storage.h"
 #include "chrome/browser/profiles/profile_avatar_icon_util.h"
 #include "chrome/browser/profiles/profile_manager.h"
+#include "chrome/browser/ui/views/frame/browser_view.h"
 #include "skia/ext/font_utils.h"
 #include "skia/ext/image_operations.h"
 #include "skia/ext/legacy_display_globals.h"
@@ -42,6 +44,10 @@
 #include "ui/gfx/win/icon_util.h"
 #include "ui/views/win/hwnd_util.h"
 
+#if BUILDFLAG(ENABLE_CEF)
+#include "cef/libcef/browser/chrome/browser_delegate.h"
+#endif
+
 namespace taskbar {
 
 namespace {
@@ -203,6 +209,24 @@ void UpdateTaskbarDecoration(Profile* profile, gfx::NativeWindow window) {
     return;
   }
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (auto browser_view = BrowserView::GetBrowserViewForNativeWindow(window)) {
+    if (auto browser = browser_view->browser()) {
+      if (browser->cef_delegate() &&
+          (browser->cef_delegate()->IsViewsHosted() ||
+           ((browser->GetType() ==
+                 BrowserWindowInterface::Type::TYPE_PICTURE_IN_PICTURE ||
+             browser->GetType() ==
+                 BrowserWindowInterface::Type::TYPE_DEVTOOLS) &&
+            browser->cef_delegate()->HasViewsHostedOpener()))) {
+        // Don't add taskbar decoration.
+        taskbar::DrawTaskbarDecoration(window, nullptr);
+        return;
+      }
+    }
+  }
+#endif  // BUILDFLAG(ENABLE_CEF)
+
   // We need to draw the taskbar decoration. Even though we have an icon on the
   // window's relaunch details, we draw over it because the user may have
   // pinned the badge-less Chrome shortcut which will cause Windows to ignore
