diff --git content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
index 444e19ccb601..6d696b28f417 100644
--- content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
+++ content/browser/screenlock_monitor/screenlock_monitor_device_source_win.cc
@@ -49,7 +49,7 @@ bool ScreenlockMonitorDeviceSource::SessionMessageWindow::OnWndProc(
   if (message == WM_WTSSESSION_CHANGE) {
     ProcessWTSSessionLockMessage(wparam);
   }
-  return true;
+  return false;
 }
 
 void ScreenlockMonitorDeviceSource::SessionMessageWindow::
