diff --git base/third_party/double_conversion/BUILD.gn base/third_party/double_conversion/BUILD.gn
index 8380f58d995bc..c0755bf9870b4 100644
--- base/third_party/double_conversion/BUILD.gn
+++ base/third_party/double_conversion/BUILD.gn
@@ -9,6 +9,11 @@ config("config") {
     "-Wno-unused-const-variable",
     "-Wno-unused-function",
   ]
+
+  # Build as C++17 to avoid export of templates that should be inlined.
+  if (is_win) {
+    cflags_cc = [ "/std:c++17" ]
+  }
 }
 
 static_library("double_conversion") {
diff --git base/win/BUILD.gn base/win/BUILD.gn
index d915a9320750c..5661652c9b2da 100644
--- base/win/BUILD.gn
+++ base/win/BUILD.gn
@@ -33,4 +33,7 @@ static_library("pe_image") {
     "pe_image.cc",
     "pe_image.h",
   ]
+
+  # Build as C++17 to avoid export of templates that should be inlined.
+  cflags_cc = [ "/std:c++17" ]
 }
