diff --git chrome/browser/ui/views/frame/browser_window_property_manager_win.cc chrome/browser/ui/views/frame/browser_window_property_manager_win.cc
index a2b4f28e7ec965c95c7e52972bdeb22330f4e0a1..f1fac633b4dcd2e43aaca30b9043e228d9087453 100644
--- chrome/browser/ui/views/frame/browser_window_property_manager_win.cc
+++ chrome/browser/ui/views/frame/browser_window_property_manager_win.cc
@@ -8,6 +8,7 @@
 #include "base/functional/bind.h"
 #include "base/strings/utf_string_conversions.h"
 #include "base/win/windows_version.h"
+#include "cef/libcef/features/features.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/profiles/profile.h"
 #include "chrome/browser/profiles/profile_manager.h"
@@ -50,6 +51,16 @@ void BrowserWindowPropertyManager::UpdateWindowProperties() {
   const Browser* browser = view_->browser();
   Profile* profile = const_cast<Profile*>(browser->GetProfile());
 
+#if BUILDFLAG(ENABLE_CEF)
+  if (browser->cef_delegate() &&
+        (browser->is_type_picture_in_picture() ||
+         browser->is_type_devtools()) &&
+        browser->cef_delegate()->HasViewsHostedOpener()) {
+    // Don't create a separate taskbar group.
+    return;
+  }
+#endif
+
   // Set the app user model id for this application to that of the application
   // name. See http://crbug.com/41308099.
   std::wstring app_id =
