diff --git ui/views/win/fullscreen_handler.cc ui/views/win/fullscreen_handler.cc
index bf14eb24f08b1..1d3ce1e151849 100644
--- ui/views/win/fullscreen_handler.cc
+++ ui/views/win/fullscreen_handler.cc
@@ -109,7 +109,7 @@ void FullscreenHandler::ProcessFullscreen(bool fullscreen,
     // Restore the window style and bounds saved prior to entering fullscreen.
     // Making multiple window adjustments here is ugly, but if SetWindowPos()
     // doesn't redraw, the taskbar won't be repainted.
-    SetWindowLong(hwnd_, GWL_STYLE, saved_window_info_.style);
+    SetWindowLong(hwnd_, GWL_STYLE, saved_window_info_.style | WS_VISIBLE);
     SetWindowLong(hwnd_, GWL_EXSTYLE, saved_window_info_.ex_style);
 
     gfx::Rect window_rect(saved_window_info_.rect);
