diff --git sandbox/policy/win/sandbox_win.cc sandbox/policy/win/sandbox_win.cc
index d63d942c7e3e..a100fb4d685f 100644
--- sandbox/policy/win/sandbox_win.cc
+++ sandbox/policy/win/sandbox_win.cc
@@ -953,8 +953,11 @@ ResultCode SandboxWin::StartSandboxedProcess(
   }
   // TODO(wfh): Relax strict handle checks for network process until root cause
   // for this crash can be resolved. See https://crbug.com/939590.
-  if (sandbox_type != SandboxType::kNetwork)
+  if (!launcher_process_command_line.HasSwitch("win-rt-app") &&
+      sandbox_type != SandboxType::kNetwork) {
+    // Don't enable this mitigation in WinRT apps. See issue #2274.
     mitigations |= MITIGATION_STRICT_HANDLE_CHECKS;
+  }
 
   result = policy->SetDelayedProcessMitigations(mitigations);
   if (result != SBOX_ALL_OK)
