diff --git services/service_manager/sandbox/win/sandbox_win.cc services/service_manager/sandbox/win/sandbox_win.cc
index d4370c2797c0..7509eb4d734d 100644
--- services/service_manager/sandbox/win/sandbox_win.cc
+++ services/service_manager/sandbox/win/sandbox_win.cc
@@ -934,8 +934,11 @@ sandbox::ResultCode SandboxWin::StartSandboxedProcess(
   }
   // TODO(wfh): Relax strict handle checks for network process until root cause
   // for this crash can be resolved. See https://crbug.com/939590.
-  if (sandbox_type != SANDBOX_TYPE_NETWORK)
+  if (!launcher_process_command_line.HasSwitch("win-rt-app") &&
+      sandbox_type != SANDBOX_TYPE_NETWORK) {
+    // Don't enable this mitigation in WinRT apps. See issue #2274.
     mitigations |= sandbox::MITIGATION_STRICT_HANDLE_CHECKS;
+  }
 
   result = policy->SetDelayedProcessMitigations(mitigations);
   if (result != sandbox::SBOX_ALL_OK)
