diff --git ui/accessibility/platform/ax_platform_node_win.cc ui/accessibility/platform/ax_platform_node_win.cc
index 1f68b3bf64265e089fc3d53e06f8e6f94aff8d7c..76aaf983bcc610c8dd842fd01666d8d89a62cbf7 100644
--- ui/accessibility/platform/ax_platform_node_win.cc
+++ ui/accessibility/platform/ax_platform_node_win.cc
@@ -1681,6 +1681,10 @@ IFACEMETHODIMP AXPlatformNodeWin::accLocation(LONG* physical_pixel_left,
   AXPlatform::GetInstance().SetMsaaActive();
   AXPlatform::GetInstance().OnMinimalPropertiesUsed();
 
+  if (!target || target->IsDestroyed()) {
+    return S_FALSE;
+  }
+
   gfx::Rect bounds = target->GetDelegate()->GetBoundsRect(
       AXCoordinateSystem::kScreenPhysicalPixels,
       AXClippingBehavior::kUnclipped);
