Index: Vector.h
===================================================================
--- Vector.h	(revision 175829)
+++ Vector.h	(working copy)
@@ -224,7 +224,10 @@
     {
         static bool compare(const T* a, const T* b, size_t size)
         {
-            return std::equal(a, a + size, b);
+            for (size_t i = 0; i < size; ++i)
+                if (!(a[i] == b[i]))
+                    return false;
+            return true;
         }
     };
 
